// tds-ui/esm2020/pagination/pagination-options.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { toNumber } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/select";
import * as i3 from "@angular/forms";
import * as i4 from "tds-ui/form-field";
import * as i5 from "tds-ui/tds-input";
const _c0 = ["tds-pagination-options", ""];
function TDSPaginationOptionsComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 2)(1, "tds-select", 3);
    i0.ɵɵlistener("ngModelChange", function TDSPaginationOptionsComponent_div_0_Template_tds_select_ngModelChange_1_listener($event) { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.onPageSizeChange($event); });
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(2, "div", 4);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("disabled", ctx_r0.disabled)("size", ctx_r0.size)("ngModel", ctx_r0.pageSize)("data", ctx_r0.listOfPageSizeOption);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r0.locale.items_per_page, " ");
} }
function TDSPaginationOptionsComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtext(1);
    i0.ɵɵelementStart(2, "tds-form-field", 6)(3, "input", 7);
    i0.ɵɵlistener("keydown.enter", function TDSPaginationOptionsComponent_div_1_Template_input_keydown_enter_3_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return ctx_r4.jumpToPageViaInput($event); });
    i0.ɵɵelementEnd()();
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r1.locale.jump_to, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("disabled", ctx_r1.disabled)("size", ctx_r1.size);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r1.locale.page, " ");
} }
export class TDSPaginationOptionsComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.size = 'md';
        this.disabled = false;
        this.showSizeChanger = false;
        this.showQuickJumper = false;
        this.total = 0;
        this.pageIndex = 1;
        this.pageSize = 10;
        this.pageSizeOptions = [];
        this.pageIndexChange = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.listOfPageSizeOption = [];
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-pagination-options');
    }
    onPageSizeChange(size) {
        if (this.pageSize !== size) {
            this.pageSizeChange.next(size);
        }
    }
    jumpToPageViaInput($event) {
        const target = $event.target;
        const index = Math.floor(toNumber(target.value, this.pageIndex));
        this.pageIndexChange.next(index);
        target.value = '';
    }
    trackByOption(_, option) {
        return option.value;
    }
    ngOnChanges(changes) {
        const { pageSize, pageSizeOptions, locale } = changes;
        if (pageSize || pageSizeOptions || locale) {
            this.listOfPageSizeOption = [...new Set([...this.pageSizeOptions, this.pageSize])].map(item => ({
                value: item,
                label: `${item}`
            }));
        }
    }
}
TDSPaginationOptionsComponent.ɵfac = function TDSPaginationOptionsComponent_Factory(t) { return new (t || TDSPaginationOptionsComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSPaginationOptionsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationOptionsComponent, selectors: [["div", "tds-pagination-options", ""]], inputs: { size: "size", disabled: "disabled", showSizeChanger: "showSizeChanger", showQuickJumper: "showQuickJumper", locale: "locale", total: "total", pageIndex: "pageIndex", pageSize: "pageSize", pageSizeOptions: "pageSizeOptions" }, outputs: { pageIndexChange: "pageIndexChange", pageSizeChange: "pageSizeChange" }, features: [i0.ɵɵNgOnChangesFeature], attrs: _c0, decls: 2, vars: 2, consts: [["class", "inline-flex items-center", 4, "ngIf"], ["class", "tds-pagination-options-quick-jumper flex items-center ", 4, "ngIf"], [1, "inline-flex", "items-center"], ["textField", "label", "valueField", "value", 1, "tds-pagination-options-size-changer", 3, "disabled", "size", "ngModel", "data", "ngModelChange"], [1, "ml-2", "min-w-max"], [1, "tds-pagination-options-quick-jumper", "flex", "items-center"], [1, "w-20", "mx-3"], ["autocomplete", "off", "tdsInput", "", "autocomplete", "off", 3, "disabled", "size", "keydown.enter"]], template: function TDSPaginationOptionsComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSPaginationOptionsComponent_div_0_Template, 4, 5, "div", 0);
        i0.ɵɵtemplate(1, TDSPaginationOptionsComponent_div_1_Template, 5, 4, "div", 1);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.showSizeChanger);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.showQuickJumper);
    } }, directives: [i1.NgIf, i2.TDSSelect, i3.NgControlStatus, i3.NgModel, i4.TDSFormField, i5.TDSInput], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationOptionsComponent, [{
        type: Component,
        args: [{
                selector: 'div[tds-pagination-options]',
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <div class='inline-flex items-center' *ngIf="showSizeChanger">
    <tds-select
    class="tds-pagination-options-size-changer"
    
    [disabled]="disabled"
    [size]="size"
    [ngModel]="pageSize"
    (ngModelChange)="onPageSizeChange($event)"
    [data]="listOfPageSizeOption"
    textField="label"
    valueField="value"
  >     
  </tds-select>
      <div class='ml-2 min-w-max'>
        {{locale.items_per_page}}
      </div>
    </div>
   
    <div class="tds-pagination-options-quick-jumper flex items-center " *ngIf="showQuickJumper">
      {{ locale.jump_to }}
      <tds-form-field class="w-20 mx-3">     
        <input autocomplete="off" tdsInput autocomplete="off" 
         [disabled]="disabled" [size]="size" (keydown.enter)="jumpToPageViaInput($event)" />    
      </tds-form-field>
      {{ locale.page }}
    </div>
  `
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { size: [{
            type: Input
        }], disabled: [{
            type: Input
        }], showSizeChanger: [{
            type: Input
        }], showQuickJumper: [{
            type: Input
        }], locale: [{
            type: Input
        }], total: [{
            type: Input
        }], pageIndex: [{
            type: Input
        }], pageSize: [{
            type: Input
        }], pageSizeOptions: [{
            type: Input
        }], pageIndexChange: [{
            type: Output
        }], pageSizeChange: [{
            type: Output
        }] }); })();
