// tds-ui/esm2020/pagination/pagination.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSFormFieldModule } from 'tds-ui/form-field';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSSelectModule } from 'tds-ui/select';
import { TDSInputModule } from 'tds-ui/tds-input';
import { TDSPaginationDefaultComponent } from './pagination-default.component';
import { TDSPaginationItemComponent } from './pagination-item.component';
import { TDSPaginationOptionsComponent } from './pagination-options.component';
import { TDSPaginationSimpleComponent } from './pagination-simple.component';
import { TDSPaginationComponent } from './pagination.component';
import * as i0 from "@angular/core";
export class TDSPaginationModule {
}
TDSPaginationModule.ɵfac = function TDSPaginationModule_Factory(t) { return new (t || TDSPaginationModule)(); };
TDSPaginationModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSPaginationModule });
TDSPaginationModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, FormsModule, TDSSelectModule, TDSI18nModule, TDSFormFieldModule, TDSInputModule, TDSButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSPaginationComponent,
                    TDSPaginationSimpleComponent,
                    TDSPaginationOptionsComponent,
                    TDSPaginationItemComponent,
                    TDSPaginationDefaultComponent
                ],
                exports: [TDSPaginationComponent],
                imports: [BidiModule, CommonModule, FormsModule, TDSSelectModule, TDSI18nModule, TDSFormFieldModule, TDSInputModule, TDSButtonModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSPaginationModule, { declarations: [TDSPaginationComponent,
        TDSPaginationSimpleComponent,
        TDSPaginationOptionsComponent,
        TDSPaginationItemComponent,
        TDSPaginationDefaultComponent], imports: [BidiModule, CommonModule, FormsModule, TDSSelectModule, TDSI18nModule, TDSFormFieldModule, TDSInputModule, TDSButtonModule], exports: [TDSPaginationComponent] }); })();
