// tds-ui/esm2020/pagination/pagination-item.component.mjs
/* eslint-disable */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["tds-pagination-item", ""];
function TDSPaginationItemComponent_ng_template_0_a_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 6);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const page_r4 = i0.ɵɵnextContext().page;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(page_r4);
} }
const _c1 = function (a0, a1) { return { "cursor-pointer": a0, "cursor-not-allowed": a1 }; };
const _c2 = function (a0, a1) { return { "text-neutral-1-100 dark:text-d-neutral-1-100": a0, "text-neutral-1-500 dark:text-d-neutral-1-500": a1 }; };
function TDSPaginationItemComponent_ng_template_0_button_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "button", 7);
    i0.ɵɵelement(1, "i", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("disabled", ctx_r6.disabled)("ngClass", i0.ɵɵpureFunction2(3, _c1, !ctx_r6.disabled, ctx_r6.disabled));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(6, _c2, ctx_r6.disabled, !ctx_r6.disabled));
} }
function TDSPaginationItemComponent_ng_template_0_button_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "button", 7);
    i0.ɵɵelement(1, "i", 9);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("disabled", ctx_r7.disabled)("ngClass", i0.ɵɵpureFunction2(3, _c1, ctx_r7.disabled, ctx_r7.disabled));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(6, _c2, ctx_r7.disabled, !ctx_r7.disabled));
} }
function TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0, 2);
    i0.ɵɵelement(1, "span", 15);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("ngSwitch", ctx_r11.direction);
} }
function TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0, 2);
    i0.ɵɵelement(1, "span", 16);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("ngSwitch", ctx_r12.direction);
} }
function TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 12);
    i0.ɵɵelementContainerStart(1, 2);
    i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_2_Template, 2, 1, "ng-container", 13);
    i0.ɵɵtemplate(3, TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_3_Template, 2, 1, "ng-container", 13);
    i0.ɵɵelementContainerEnd();
    i0.ɵɵelement(4, "span", 14);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const type_r3 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitch", type_r3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "prev_5");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "next_5");
} }
function TDSPaginationItemComponent_ng_template_0_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "a", 10);
    i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_Template, 5, 3, "div", 11);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const type_r3 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitch", type_r3);
} }
function TDSPaginationItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0, 2);
    i0.ɵɵtemplate(1, TDSPaginationItemComponent_ng_template_0_a_1_Template, 2, 1, "a", 3);
    i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_0_button_2_Template, 2, 9, "button", 4);
    i0.ɵɵtemplate(3, TDSPaginationItemComponent_ng_template_0_button_3_Template, 2, 9, "button", 4);
    i0.ɵɵtemplate(4, TDSPaginationItemComponent_ng_template_0_ng_container_4_Template, 3, 1, "ng-container", 5);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const type_r3 = ctx.$implicit;
    i0.ɵɵproperty("ngSwitch", type_r3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "page");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "prev");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "next");
} }
function TDSPaginationItemComponent_ng_template_2_Template(rf, ctx) { }
const _c3 = function (a0, a1) { return { $implicit: a0, page: a1 }; };
export class TDSPaginationItemComponent {
    constructor() {
        this.active = false;
        this.index = null;
        this.disabled = false;
        this.direction = 'ltr';
        this.type = null;
        this.itemRender = null;
        this.diffIndex = new EventEmitter();
        this.gotoIndex = new EventEmitter();
        this.title = null;
    }
    clickItem() {
        if (!this.disabled) {
            if (this.type === 'page') {
                this.gotoIndex.emit(this.index);
            }
            else {
                this.diffIndex.emit({
                    next: 1,
                    prev: -1,
                    prev_5: -5,
                    next_5: 5
                }[this.type]);
            }
        }
    }
    ngOnChanges(changes) {
        const { locale, index, type } = changes;
        if (locale || index || type) {
            this.title = {
                page: `${this.index}`,
                next: this.locale?.next_page,
                prev: this.locale?.prev_page,
                prev_5: this.locale?.prev_5,
                next_5: this.locale?.next_5
            }[this.type];
        }
    }
}
TDSPaginationItemComponent.ɵfac = function TDSPaginationItemComponent_Factory(t) { return new (t || TDSPaginationItemComponent)(); };
TDSPaginationItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationItemComponent, selectors: [["li", "tds-pagination-item", ""]], hostVars: 19, hostBindings: function TDSPaginationItemComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function TDSPaginationItemComponent_click_HostBindingHandler() { return ctx.clickItem(); });
    } if (rf & 2) {
        i0.ɵɵattribute("title", ctx.title);
        i0.ɵɵclassProp("tds-pagination-prev", ctx.type === "prev")("tds-pagination-next", ctx.type === "next")("tds-pagination-item", ctx.type === "page")("tds-pagination-jump-prev", ctx.type === "prev_5")("tds-pagination-jump-prev-custom-icon", ctx.type === "prev_5")("tds-pagination-jump-next", ctx.type === "next_5")("tds-pagination-jump-next-custom-icon", ctx.type === "next_5")("tds-pagination-disabled", ctx.disabled)("tds-pagination-item-active", ctx.active);
    } }, inputs: { active: "active", locale: "locale", index: "index", disabled: "disabled", direction: "direction", type: "type", itemRender: "itemRender" }, outputs: { diffIndex: "diffIndex", gotoIndex: "gotoIndex" }, features: [i0.ɵɵNgOnChangesFeature], attrs: _c0, decls: 3, vars: 5, consts: [["renderItemTemplate", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [3, "ngSwitch"], ["class", "px-0.5", 4, "ngSwitchCase"], ["type", "button", "class", "tds-pagination-item-link flex", 3, "disabled", "ngClass", 4, "ngSwitchCase"], [4, "ngSwitchDefault"], [1, "px-0.5"], ["type", "button", 1, "tds-pagination-item-link", "flex", 3, "disabled", "ngClass"], [1, "tdsi-arrow-left-fill", "leading-4", "text-base", 3, "ngClass"], [1, "tdsi-arrow-right-fill", "leading-4", "text-base", 3, "ngClass"], [1, "tds-pagination-item-link", "flex", 3, "ngSwitch"], ["class", "tds-pagination-item-container flex", 4, "ngSwitchDefault"], [1, "tds-pagination-item-container", "flex"], [3, "ngSwitch", 4, "ngSwitchCase"], [1, "tds-pagination-item-ellipsis", "tdsi-three-dots-horizon-fill", "group-hover:hidden", "leading-4", "text-base"], [1, "tds-pagination-item-link-icon", "tdsi-chevron-previous-line", "group-hover:inline", "hidden", "leading-4", "text-base"], [1, "tds-pagination-item-link-icon", "tdsi-chevron-next-line", "group-hover:inline", "hidden", "leading-4", "text-base"]], template: function TDSPaginationItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSPaginationItemComponent_ng_template_0_Template, 5, 4, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_2_Template, 0, 0, "ng-template", 1);
    } if (rf & 2) {
        const _r0 = i0.ɵɵreference(1);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.itemRender || _r0)("ngTemplateOutletContext", i0.ɵɵpureFunction2(2, _c3, ctx.type, ctx.index));
    } }, directives: [i1.NgSwitch, i1.NgSwitchCase, i1.NgClass, i1.NgSwitchDefault, i1.NgTemplateOutlet], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationItemComponent, [{
        type: Component,
        args: [{ selector: 'li[tds-pagination-item]', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                    '[class.tds-pagination-prev]': `type === 'prev'`,
                    '[class.tds-pagination-next]': `type === 'next'`,
                    '[class.tds-pagination-item]': `type === 'page'`,
                    '[class.tds-pagination-jump-prev]': `type === 'prev_5'`,
                    '[class.tds-pagination-jump-prev-custom-icon]': `type === 'prev_5'`,
                    '[class.tds-pagination-jump-next]': `type === 'next_5'`,
                    '[class.tds-pagination-jump-next-custom-icon]': `type === 'next_5'`,
                    '[class.tds-pagination-disabled]': 'disabled',
                    '[class.tds-pagination-item-active]': 'active',
                    '[attr.title]': 'title',
                    '(click)': 'clickItem()'
                }, template: "<ng-template #renderItemTemplate let-type let-page=\"page\">\r\n  <ng-container [ngSwitch]=\"type\">\r\n    <a class=\"px-0.5\" *ngSwitchCase=\"'page'\">{{ page }}</a>\r\n    <button type=\"button\" [disabled]=\"disabled\" class=\"tds-pagination-item-link flex\"\r\n      [ngClass]=\"{'cursor-pointer':!disabled,'cursor-not-allowed':disabled}\" *ngSwitchCase=\"'prev'\">\r\n      <i class=\"tdsi-arrow-left-fill leading-4 text-base\" [ngClass]=\"{'text-neutral-1-100 dark:text-d-neutral-1-100':disabled,'text-neutral-1-500 dark:text-d-neutral-1-500':!disabled}\"></i>\r\n\r\n    </button>\r\n    <button type=\"button\" [disabled]=\"disabled\" [ngClass]=\"{'cursor-pointer':disabled,'cursor-not-allowed':disabled}\"\r\n      class=\"tds-pagination-item-link flex\" *ngSwitchCase=\"'next'\">\r\n      <i class=\"tdsi-arrow-right-fill leading-4 text-base\" [ngClass]=\"{'text-neutral-1-100 dark:text-d-neutral-1-100':disabled,'text-neutral-1-500 dark:text-d-neutral-1-500':!disabled}\"></i>\r\n\r\n    </button>\r\n    <ng-container *ngSwitchDefault>\r\n      <a class=\"tds-pagination-item-link flex\" [ngSwitch]=\"type\">\r\n        <div class=\"tds-pagination-item-container flex\" *ngSwitchDefault>\r\n          <ng-container [ngSwitch]=\"type\">\r\n            <ng-container *ngSwitchCase=\"'prev_5'\" [ngSwitch]=\"direction\">\r\n\r\n              <span class=\"tds-pagination-item-link-icon tdsi-chevron-previous-line group-hover:inline hidden leading-4 text-base\"></span>\r\n            </ng-container>\r\n            <ng-container *ngSwitchCase=\"'next_5'\" [ngSwitch]=\"direction\">\r\n\r\n              <span class=\"tds-pagination-item-link-icon tdsi-chevron-next-line group-hover:inline hidden leading-4 text-base\"></span>\r\n            </ng-container>\r\n          </ng-container>\r\n          <span class=\"tds-pagination-item-ellipsis tdsi-three-dots-horizon-fill group-hover:hidden leading-4 text-base\"></span>\r\n        </div>\r\n      </a>\r\n    </ng-container>\r\n  </ng-container>\r\n</ng-template>\r\n<ng-template [ngTemplateOutlet]=\"itemRender || renderItemTemplate\"\r\n  [ngTemplateOutletContext]=\"{ $implicit: type, page: index }\"></ng-template>" }]
    }], null, { active: [{
            type: Input
        }], locale: [{
            type: Input
        }], index: [{
            type: Input
        }], disabled: [{
            type: Input
        }], direction: [{
            type: Input
        }], type: [{
            type: Input
        }], itemRender: [{
            type: Input
        }], diffIndex: [{
            type: Output
        }], gotoIndex: [{
            type: Output
        }] }); })();
