// tds-ui/esm2020/pagination/pagination-default.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Optional, Output, ViewChild, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/bidi";
import * as i2 from "@angular/common";
import * as i3 from "./pagination-options.component";
import * as i4 from "tds-ui/button";
import * as i5 from "./pagination-item.component";
const _c0 = ["containerTemplate"];
const _c1 = function (a0, a1, a2, a3) { return { $implicit: a0, locale: a1, bordered: a2, dir: a3 }; };
function TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementContainer(1, 13);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const page_r10 = i0.ɵɵnextContext().$implicit;
    const ctx_r11 = i0.ɵɵnextContext(2);
    const _r2 = i0.ɵɵreference(3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", i0.ɵɵpureFunction4(2, _c1, page_r10, ctx_r11.locale, ctx_r11.bordered, ctx_r11.dir));
} }
function TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementContainer(1, 13);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const page_r10 = i0.ɵɵnextContext().$implicit;
    const ctx_r12 = i0.ɵɵnextContext(2);
    const _r4 = i0.ɵɵreference(5);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r4)("ngTemplateOutletContext", i0.ɵɵpureFunction4(2, _c1, page_r10, ctx_r12.locale, ctx_r12.bordered, ctx_r12.dir));
} }
function TDSPaginationDefaultComponent_ng_template_0_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0, 10);
    i0.ɵɵtemplate(1, TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_1_Template, 2, 7, "ng-container", 11);
    i0.ɵɵtemplate(2, TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_2_Template, 2, 7, "ng-container", 12);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const page_r10 = ctx.$implicit;
    i0.ɵɵproperty("ngSwitch", page_r10.type);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "page");
} }
function TDSPaginationDefaultComponent_ng_template_0_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r16 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 14);
    i0.ɵɵlistener("pageIndexChange", function TDSPaginationDefaultComponent_ng_template_0_div_3_Template_div_pageIndexChange_0_listener($event) { i0.ɵɵrestoreView(_r16); const ctx_r15 = i0.ɵɵnextContext(2); return ctx_r15.onPageIndexChange($event); })("pageSizeChange", function TDSPaginationDefaultComponent_ng_template_0_div_3_Template_div_pageSizeChange_0_listener($event) { i0.ɵɵrestoreView(_r16); const ctx_r17 = i0.ɵɵnextContext(2); return ctx_r17.onPageSizeChange($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("total", ctx_r7.total)("locale", ctx_r7.locale)("disabled", ctx_r7.disabled)("size", ctx_r7.size)("showSizeChanger", ctx_r7.showSizeChanger)("showQuickJumper", ctx_r7.showQuickJumper)("pageIndex", ctx_r7.pageIndex)("pageSize", ctx_r7.pageSize)("pageSizeOptions", ctx_r7.pageSizeOptions);
} }
function TDSPaginationDefaultComponent_ng_template_0_div_5_ng_template_1_Template(rf, ctx) { }
const _c2 = function (a0, a1) { return { $implicit: a0, range: a1 }; };
function TDSPaginationDefaultComponent_ng_template_0_div_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 15);
    i0.ɵɵtemplate(1, TDSPaginationDefaultComponent_ng_template_0_div_5_ng_template_1_Template, 0, 0, "ng-template", 13);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r8.showTotal)("ngTemplateOutletContext", i0.ɵɵpureFunction2(2, _c2, ctx_r8.total, ctx_r8.ranges));
} }
function TDSPaginationDefaultComponent_ng_template_0_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r20 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 16);
    i0.ɵɵlistener("click", function TDSPaginationDefaultComponent_ng_template_0_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r20); const ctx_r19 = i0.ɵɵnextContext(2); return ctx_r19.onRefresh($event); });
    i0.ɵɵelementStart(1, "span", 4);
    i0.ɵɵelement(2, "i", 17);
    i0.ɵɵelementEnd()();
} }
function TDSPaginationDefaultComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 3)(1, "ul", 4);
    i0.ɵɵtemplate(2, TDSPaginationDefaultComponent_ng_template_0_ng_template_2_Template, 3, 2, "ng-template", 5);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(3, TDSPaginationDefaultComponent_ng_template_0_div_3_Template, 1, 9, "div", 6);
    i0.ɵɵelementStart(4, "div", 7);
    i0.ɵɵtemplate(5, TDSPaginationDefaultComponent_ng_template_0_div_5_Template, 2, 5, "div", 8);
    i0.ɵɵtemplate(6, TDSPaginationDefaultComponent_ng_template_0_button_6_Template, 3, 0, "button", 9);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r1.listOfPageItem)("ngForTrackBy", ctx_r1.trackByPageItem);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.showQuickJumper || ctx_r1.showSizeChanger);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.showTotal);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.showRefresh);
} }
const _c3 = function (a0, a1, a2, a3, a4, a5, a6) { return { "bg-primary-1 text-white": a0, "text-neutral-1-600 dark:text-d-neutral-1-600 hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300": a1, "cursor-pointer": a2, "cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500": a3, "border-transparent": a4, " border-primary-1": a5, "border-neutral-2-200 dark:border-d-neutral-2-200": a6 }; };
function TDSPaginationDefaultComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    const _r26 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "li", 18);
    i0.ɵɵlistener("gotoIndex", function TDSPaginationDefaultComponent_ng_template_2_Template_li_gotoIndex_0_listener($event) { i0.ɵɵrestoreView(_r26); const ctx_r25 = i0.ɵɵnextContext(); return ctx_r25.jumpPage($event); })("diffIndex", function TDSPaginationDefaultComponent_ng_template_2_Template_li_diffIndex_0_listener($event) { i0.ɵɵrestoreView(_r26); const ctx_r27 = i0.ɵɵnextContext(); return ctx_r27.jumpDiff($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const page_r21 = ctx.$implicit;
    const dir_r23 = ctx.dir;
    const bordered_r24 = ctx.bordered;
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵproperty("locale", ctx_r3.locale)("ngClass", i0.ɵɵpureFunction7(8, _c3, ctx_r3.pageIndex === page_r21.index, ctx_r3.pageIndex !== page_r21.index, !page_r21.disabled, !!page_r21.disabled, !bordered_r24, ctx_r3.pageIndex === page_r21.index && bordered_r24, ctx_r3.pageIndex !== page_r21.index && bordered_r24))("type", page_r21.type)("index", page_r21.index)("disabled", !!page_r21.disabled)("itemRender", ctx_r3.itemRender)("active", ctx_r3.pageIndex === page_r21.index)("direction", dir_r23);
} }
const _c4 = function (a0, a1) { return { "cursor-pointer": a0, "cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500": a1 }; };
function TDSPaginationDefaultComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    const _r33 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "li", 19);
    i0.ɵɵlistener("gotoIndex", function TDSPaginationDefaultComponent_ng_template_4_Template_li_gotoIndex_0_listener($event) { i0.ɵɵrestoreView(_r33); const ctx_r32 = i0.ɵɵnextContext(); return ctx_r32.jumpPage($event); })("diffIndex", function TDSPaginationDefaultComponent_ng_template_4_Template_li_diffIndex_0_listener($event) { i0.ɵɵrestoreView(_r33); const ctx_r34 = i0.ɵɵnextContext(); return ctx_r34.jumpDiff($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const page_r28 = ctx.$implicit;
    const dir_r30 = ctx.dir;
    const ctx_r5 = i0.ɵɵnextContext();
    i0.ɵɵproperty("locale", ctx_r5.locale)("ngClass", i0.ɵɵpureFunction2(8, _c4, !page_r28.disabled, !!page_r28.disabled))("type", page_r28.type)("index", page_r28.index)("disabled", !!page_r28.disabled)("itemRender", ctx_r5.itemRender)("active", ctx_r5.pageIndex === page_r28.index)("direction", dir_r30);
} }
export class TDSPaginationDefaultComponent {
    constructor(cdr, renderer, elementRef, directionality) {
        this.cdr = cdr;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.directionality = directionality;
        this.size = 'md';
        this.itemRender = null;
        this.showTotal = null;
        this.disabled = false;
        this.showSizeChanger = false;
        this.showQuickJumper = false;
        this.total = 0;
        this.pageIndex = 1;
        this.pageSize = 10;
        this.pageSizeOptions = [10, 20, 30, 40];
        this.showRefresh = false;
        this.bordered = true;
        this.pageIndexChange = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.clickRefresh = new EventEmitter();
        this.ranges = [0, 0];
        this.listOfPageItem = [];
        this.dir = 'ltr';
        this.destroy$ = new Subject();
        renderer.removeChild(renderer.parentNode(elementRef.nativeElement), elementRef.nativeElement);
    }
    ngOnInit() {
        this.directionality.change?.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.updateRtlStyle();
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
        this.updateRtlStyle();
    }
    updateRtlStyle() {
        if (this.dir === 'rtl') {
            this.renderer.addClass(this.elementRef.nativeElement, 'tds-pagination-rtl');
        }
        else {
            this.renderer.removeClass(this.elementRef.nativeElement, 'tds-pagination-rtl');
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    jumpPage(index) {
        this.onPageIndexChange(index);
    }
    jumpDiff(diff) {
        this.jumpPage(this.pageIndex + diff);
    }
    trackByPageItem(_, value) {
        return `${value.type}-${value.index}`;
    }
    onPageIndexChange(index) {
        this.pageIndexChange.next(index);
    }
    onPageSizeChange(size) {
        this.pageSizeChange.next(size);
    }
    getLastIndex(total, pageSize) {
        return Math.ceil(total / pageSize);
    }
    buildIndexes() {
        const lastIndex = this.getLastIndex(this.total, this.pageSize);
        this.listOfPageItem = this.getListOfPageItem(this.pageIndex, lastIndex);
    }
    getListOfPageItem(pageIndex, lastIndex) {
        const concatWithPrevNext = (listOfPage) => {
            const prevItem = {
                type: 'prev',
                disabled: pageIndex === 1
            };
            const nextItem = {
                type: 'next',
                disabled: pageIndex === lastIndex
            };
            return [prevItem, ...listOfPage, nextItem];
        };
        const generatePage = (start, end) => {
            const list = [];
            for (let i = start; i <= end; i++) {
                list.push({
                    index: i,
                    type: 'page'
                });
            }
            return list;
        };
        if (lastIndex <= 9) {
            return concatWithPrevNext(generatePage(1, lastIndex));
        }
        else {
            const generateRangeItem = (selected, last) => {
                let listOfRange = [];
                const prevFiveItem = {
                    type: 'prev_5'
                };
                const nextFiveItem = {
                    type: 'next_5'
                };
                const firstPageItem = generatePage(1, 1);
                const lastPageItem = generatePage(lastIndex, lastIndex);
                if (selected < 4) {
                    listOfRange = [...generatePage(2, 5), nextFiveItem];
                }
                else if (selected < last - 3) {
                    listOfRange = [prevFiveItem, ...generatePage(selected - 2, selected + 2), nextFiveItem];
                }
                else {
                    listOfRange = [prevFiveItem, ...generatePage(last - 4, last - 1)];
                }
                return [...firstPageItem, ...listOfRange, ...lastPageItem];
            };
            return concatWithPrevNext(generateRangeItem(pageIndex, lastIndex));
        }
    }
    ngOnChanges(changes) {
        const { pageIndex, pageSize, total } = changes;
        if (pageIndex || pageSize || total) {
            this.ranges = [(this.pageIndex - 1) * this.pageSize + 1, Math.min(this.pageIndex * this.pageSize, this.total)];
            this.buildIndexes();
        }
    }
    onRefresh(e) {
        this.clickRefresh.emit(e);
    }
}
TDSPaginationDefaultComponent.ɵfac = function TDSPaginationDefaultComponent_Factory(t) { return new (t || TDSPaginationDefaultComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.Directionality, 8)); };
TDSPaginationDefaultComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationDefaultComponent, selectors: [["tds-pagination-default"]], viewQuery: function TDSPaginationDefaultComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { size: "size", itemRender: "itemRender", showTotal: "showTotal", disabled: "disabled", locale: "locale", showSizeChanger: "showSizeChanger", showQuickJumper: "showQuickJumper", total: "total", pageIndex: "pageIndex", pageSize: "pageSize", pageSizeOptions: "pageSizeOptions", showRefresh: "showRefresh", bordered: "bordered" }, outputs: { pageIndexChange: "pageIndexChange", pageSizeChange: "pageSizeChange", clickRefresh: "clickRefresh" }, features: [i0.ɵɵNgOnChangesFeature], decls: 6, vars: 0, consts: [["containerTemplate", ""], ["pageTemplate", ""], ["nextPrevTemplate", ""], [1, "flex", "items-center", "w-full", "select-none"], [1, "flex", "items-center"], ["ngFor", "", 3, "ngForOf", "ngForTrackBy"], ["class", "ml-10", "tds-pagination-options", "", 3, "total", "locale", "disabled", "size", "showSizeChanger", "showQuickJumper", "pageIndex", "pageSize", "pageSizeOptions", "pageIndexChange", "pageSizeChange", 4, "ngIf"], [1, "ml-auto", "flex", "items-center"], ["class", "tds-pagination-total-text", 4, "ngIf"], ["class", "mr-2.5 ml-3 ", "tds-button-icon", "", "color", "secondary", "size", "sm", 3, "click", 4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["tds-pagination-options", "", 1, "ml-10", 3, "total", "locale", "disabled", "size", "showSizeChanger", "showQuickJumper", "pageIndex", "pageSize", "pageSizeOptions", "pageIndexChange", "pageSizeChange"], [1, "tds-pagination-total-text"], ["tds-button-icon", "", "color", "secondary", "size", "sm", 1, "mr-2.5", "ml-3", 3, "click"], [1, "tdsi-loading-fill"], ["tds-pagination-item", "", 1, "min-w-7", "min-h-7", "text-sm", "text-center", "align-middle", "mx-0.5", "leading-7", "box-content", "select-auto", "group", "border", "rounded-md", 3, "locale", "ngClass", "type", "index", "disabled", "itemRender", "active", "direction", "gotoIndex", "diffIndex"], ["tds-pagination-item", "", 1, "min-w-7", "min-h-7", "text-sm", "text-center", "align-middle", "mx-0.5", "leading-7", "box-content", "select-auto", "group", "border", "border-transparent", "hover:bg-neutral-3-50", "dark:hover:bg-d-neutral-3-300", "rounded-md", "flex", "items-center", "justify-center", 3, "locale", "ngClass", "type", "index", "disabled", "itemRender", "active", "direction", "gotoIndex", "diffIndex"]], template: function TDSPaginationDefaultComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSPaginationDefaultComponent_ng_template_0_Template, 7, 5, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(2, TDSPaginationDefaultComponent_ng_template_2_Template, 1, 16, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(4, TDSPaginationDefaultComponent_ng_template_4_Template, 1, 11, "ng-template", null, 2, i0.ɵɵtemplateRefExtractor);
    } }, directives: [i2.NgForOf, i2.NgSwitch, i2.NgSwitchCase, i2.NgTemplateOutlet, i2.NgSwitchDefault, i2.NgIf, i3.TDSPaginationOptionsComponent, i4.TDSButtonIconComponent, i5.TDSPaginationItemComponent, i2.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationDefaultComponent, [{
        type: Component,
        args: [{ selector: 'tds-pagination-default', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-template #containerTemplate>\r\n    <div class=\"flex items-center w-full select-none\">\r\n        <ul class=\"flex items-center\">\r\n            <ng-template ngFor let-page [ngForOf]=\"listOfPageItem\" [ngForTrackBy]=\"trackByPageItem\">\r\n                <ng-container [ngSwitch]=\"page.type\">\r\n                    <ng-container *ngSwitchCase=\"'page'\">\r\n                        <ng-container [ngTemplateOutlet]=\"pageTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale,bordered:bordered, dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container>\r\n                    <!-- <ng-container *ngSwitchCase=\"'prev'\">\r\n                        <ng-container [ngTemplateOutlet]=\"nextPrevTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale, dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container>\r\n                    <ng-container *ngSwitchCase=\"'next'\">\r\n                        <ng-container [ngTemplateOutlet]=\"nextPrevTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale, dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container> -->\r\n                    <ng-container *ngSwitchDefault>\r\n                        <ng-container [ngTemplateOutlet]=\"nextPrevTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale, bordered:bordered ,  dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container>\r\n\r\n                </ng-container>\r\n\r\n\r\n            </ng-template>\r\n\r\n        </ul>\r\n        <div class=\"ml-10\" tds-pagination-options *ngIf=\"showQuickJumper || showSizeChanger\" [total]=\"total\"\r\n            [locale]=\"locale\" [disabled]=\"disabled\" [size]=\"size\" [showSizeChanger]=\"showSizeChanger\"\r\n            [showQuickJumper]=\"showQuickJumper\" [pageIndex]=\"pageIndex\" [pageSize]=\"pageSize\"\r\n            [pageSizeOptions]=\"pageSizeOptions\" (pageIndexChange)=\"onPageIndexChange($event)\"\r\n            (pageSizeChange)=\"onPageSizeChange($event)\"></div>\r\n\r\n        <div class=\"ml-auto flex items-center\">\r\n            <div class=\"tds-pagination-total-text\" *ngIf=\"showTotal\">\r\n                <ng-template [ngTemplateOutlet]=\"showTotal\"\r\n                    [ngTemplateOutletContext]=\"{ $implicit: total, range: ranges }\">\r\n                </ng-template>\r\n            </div>\r\n            <button class=\"mr-2.5 ml-3 \" (click)=\"onRefresh($event)\" tds-button-icon color=\"secondary\" *ngIf=\"showRefresh\"\r\n                size=\"sm\">\r\n                <span class=\"flex items-center\">\r\n                    <i class=\"tdsi-loading-fill\"></i>\r\n                </span>\r\n            </button>\r\n        </div>\r\n    </div>\r\n\r\n\r\n</ng-template>\r\n<ng-template #pageTemplate let-page let-local=\"local\" let-dir=\"dir\" let-bordered=\"bordered\">\r\n    <li tds-pagination-item [locale]=\"locale\"\r\n        class=\"min-w-7 min-h-7 text-sm  text-center align-middle mx-0.5 leading-7 box-content select-auto group border rounded-md\"\r\n        [ngClass]=\"{'bg-primary-1 text-white':pageIndex === page.index,\r\n     'text-neutral-1-600 dark:text-d-neutral-1-600 hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300':pageIndex !== page.index,\r\n        'cursor-pointer':!page.disabled,'cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500':!!page.disabled,\r\n        'border-transparent':!bordered,\r\n        ' border-primary-1':pageIndex === page.index && bordered,\r\n        'border-neutral-2-200 dark:border-d-neutral-2-200':pageIndex !== page.index && bordered\r\n    }\" [type]=\"page.type\" [index]=\"page.index\" [disabled]=\"!!page.disabled\" [itemRender]=\"itemRender\"\r\n        [active]=\"pageIndex === page.index\" (gotoIndex)=\"jumpPage($event)\" (diffIndex)=\"jumpDiff($event)\"\r\n        [direction]=\"dir\"></li>\r\n</ng-template>\r\n<ng-template #nextPrevTemplate let-page let-local=\"local\" let-dir=\"dir\" let-bordered=\"bordered\">\r\n    <li tds-pagination-item [locale]=\"locale\"\r\n        class=\"min-w-7 min-h-7 text-sm text-center align-middle mx-0.5 leading-7 box-content select-auto group border border-transparent hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300 rounded-md flex items-center justify-center\"\r\n        [ngClass]=\"{'cursor-pointer':!page.disabled,'cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500':!!page.disabled}\"\r\n        [type]=\"page.type\" [index]=\"page.index\" [disabled]=\"!!page.disabled\" [itemRender]=\"itemRender\"\r\n        [active]=\"pageIndex === page.index\" (gotoIndex)=\"jumpPage($event)\" (diffIndex)=\"jumpDiff($event)\"\r\n        [direction]=\"dir\"></li>\r\n</ng-template>" }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i1.Directionality, decorators: [{
                type: Optional
            }] }]; }, { template: [{
            type: ViewChild,
            args: ['containerTemplate', { static: true }]
        }], size: [{
            type: Input
        }], itemRender: [{
            type: Input
        }], showTotal: [{
            type: Input
        }], disabled: [{
            type: Input
        }], locale: [{
            type: Input
        }], showSizeChanger: [{
            type: Input
        }], showQuickJumper: [{
            type: Input
        }], total: [{
            type: Input
        }], pageIndex: [{
            type: Input
        }], pageSize: [{
            type: Input
        }], pageSizeOptions: [{
            type: Input
        }], showRefresh: [{
            type: Input
        }], bordered: [{
            type: Input
        }], pageIndexChange: [{
            type: Output
        }], pageSizeChange: [{
            type: Output
        }], clickRefresh: [{
            type: Output
        }] }); })();
