// tds-ui/esm2020/date-picker/lib/decade-header.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { AbstractPanelHeader } from './abstract-panel-header';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/select";
import * as i3 from "@angular/forms";
function DecadeHeaderComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r4 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 2);
    i0.ɵɵlistener("click", function DecadeHeaderComponent_button_1_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r4); const ctx_r3 = i0.ɵɵnextContext(); return ctx_r3.superPrevious(); });
    i0.ɵɵelement(1, "span", 6);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassMapInterpolate1("", ctx_r0.prefixCls, "-super-prev-btn text-xl text-neutral-1-500 dark:text-d-neutral-1-500 p-1.5 flex items-center justify-center border border-transparent");
    i0.ɵɵstyleProp("visibility", ctx_r0.showSuperPreBtn ? "visible" : "hidden");
    i0.ɵɵpropertyInterpolate("title", ctx_r0.superPreviousTitle());
} }
function DecadeHeaderComponent_ng_container_5_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r11 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 9);
    i0.ɵɵlistener("click", function DecadeHeaderComponent_ng_container_5_button_1_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r11); const selector_r5 = i0.ɵɵnextContext().$implicit; return selector_r5.onClick($event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const selector_r5 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMap(selector_r5.className);
    i0.ɵɵpropertyInterpolate("title", selector_r5.title || null);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", selector_r5.label, " \u00A0 ");
} }
function DecadeHeaderComponent_ng_container_5_div_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0);
} if (rf & 2) {
    const item_r15 = ctx.item;
    i0.ɵɵtextInterpolate1(" ", item_r15.name, " ");
} }
const _c0 = function (a0, a1) { return { "w-[85px]": a0, "flex-auto": a1 }; };
function DecadeHeaderComponent_ng_container_5_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r20 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 10)(1, "tds-select", 11);
    i0.ɵɵlistener("ngModelChange", function DecadeHeaderComponent_ng_container_5_div_2_Template_tds_select_ngModelChange_1_listener($event) { i0.ɵɵrestoreView(_r20); const selector_r5 = i0.ɵɵnextContext().$implicit; return selector_r5.onClick($event); });
    i0.ɵɵtemplate(2, DecadeHeaderComponent_ng_container_5_div_2_ng_template_2_Template, 1, 1, "ng-template", 12);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r21 = i0.ɵɵnextContext();
    const indx_r6 = ctx_r21.index;
    const selector_r5 = ctx_r21.$implicit;
    const ctx_r8 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(i0.ɵɵpureFunction2(6, _c0, ctx_r8.selectors.length == 1, ctx_r8.selectors.length > 1 && indx_r6 == 0));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("placeholder", selector_r5.title)("ngModel", selector_r5.value)("data", selector_r5.data)("allowSearch", false);
} }
function DecadeHeaderComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, DecadeHeaderComponent_ng_container_5_button_1_Template, 2, 5, "button", 7);
    i0.ɵɵtemplate(2, DecadeHeaderComponent_ng_container_5_div_2_Template, 3, 9, "div", 8);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const selector_r5 = ctx.$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !selector_r5.data);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", selector_r5.data);
} }
function DecadeHeaderComponent_button_8_Template(rf, ctx) { if (rf & 1) {
    const _r23 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 2);
    i0.ɵɵlistener("click", function DecadeHeaderComponent_button_8_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r23); const ctx_r22 = i0.ɵɵnextContext(); return ctx_r22.superNext(); });
    i0.ɵɵelement(1, "span", 13);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵclassMapInterpolate1("", ctx_r2.prefixCls, "-super-next-btn text-xl text-neutral-1-500 dark:text-d-neutral-1-500 p-1.5 flex items-center justify-center border border-transparent");
    i0.ɵɵstyleProp("visibility", ctx_r2.showSuperNextBtn ? "visible" : "hidden");
    i0.ɵɵpropertyInterpolate("title", ctx_r2.superNextTitle());
} }
export class DecadeHeaderComponent extends AbstractPanelHeader {
    previous() { }
    next() { }
    get startYear() {
        return parseInt(`${this.value.getYear() / 100}`, 10) * 100;
    }
    get endYear() {
        return this.startYear + 99;
    }
    superPrevious() {
        this.changeValue(this.value.addYears(-100));
    }
    superNext() {
        this.changeValue(this.value.addYears(100));
    }
    getSelectors() {
        return [
            {
                className: `${this.prefixCls}-decade-btn`,
                title: '',
                onClick: () => {
                    // noop
                },
                label: `${this.startYear}-${this.endYear}`
            }
        ];
    }
}
DecadeHeaderComponent.ɵfac = /*@__PURE__*/ function () { let ɵDecadeHeaderComponent_BaseFactory; return function DecadeHeaderComponent_Factory(t) { return (ɵDecadeHeaderComponent_BaseFactory || (ɵDecadeHeaderComponent_BaseFactory = i0.ɵɵgetInheritedFactory(DecadeHeaderComponent)))(t || DecadeHeaderComponent); }; }();
DecadeHeaderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: DecadeHeaderComponent, selectors: [["decade-header"]], exportAs: ["decadeHeader"], features: [i0.ɵɵInheritDefinitionFeature], decls: 9, vars: 19, consts: [[1, "flex", "py-2", "px-0.5", "w-full", 3, "ngClass"], ["role", "button", "tabindex", "-1", 3, "visibility", "class", "title", "click", 4, "ngIf"], ["role", "button", "tabindex", "-1", 3, "title", "click"], [1, "tdsi-arrow-left-line"], [4, "ngFor", "ngForOf"], [1, "tdsi-arrow-right-line"], [1, "tdsi-arrow-previous-fill"], ["role", "button", "type", "button", 3, "class", "title", "click", 4, "ngIf"], ["class", "px-0.5", 3, "class", 4, "ngIf"], ["role", "button", "type", "button", 3, "title", "click"], [1, "px-0.5"], ["valueField", "id", "textField", "name", 3, "placeholder", "ngModel", "data", "allowSearch", "ngModelChange"], ["tds-option-tmp", ""], [1, "tdsi-arrow-next-fill"]], template: function DecadeHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, DecadeHeaderComponent_button_1_Template, 2, 6, "button", 1);
        i0.ɵɵelementStart(2, "button", 2);
        i0.ɵɵlistener("click", function DecadeHeaderComponent_Template_button_click_2_listener() { return ctx.previous(); });
        i0.ɵɵelement(3, "span", 3);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(4, "div");
        i0.ɵɵtemplate(5, DecadeHeaderComponent_ng_container_5_Template, 3, 2, "ng-container", 4);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(6, "button", 2);
        i0.ɵɵlistener("click", function DecadeHeaderComponent_Template_button_click_6_listener() { return ctx.next(); });
        i0.ɵɵelement(7, "span", 5);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(8, DecadeHeaderComponent_button_8_Template, 2, 6, "button", 1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵproperty("ngClass", ctx.prefixCls);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.showSuperPreBtn);
        i0.ɵɵadvance(1);
        i0.ɵɵclassMapInterpolate1("", ctx.prefixCls, "-prev-btn p-1.5 text-xl text-neutral-1-500 dark:text-d-neutral-1-500 flex items-center justify-center border border-transparent");
        i0.ɵɵstyleProp("visibility", ctx.showPreBtn ? "visible" : "hidden");
        i0.ɵɵpropertyInterpolate("title", ctx.previousTitle());
        i0.ɵɵadvance(2);
        i0.ɵɵclassMapInterpolate1("", ctx.prefixCls, "-view flex-auto flex items-center justify-center font-semibold text-title-1 text-neutral-1-900 dark:text-d-neutral-1-900");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngForOf", ctx.selectors);
        i0.ɵɵadvance(1);
        i0.ɵɵclassMapInterpolate1("", ctx.prefixCls, "-next-btn text-xl text-neutral-1-500 dark:text-d-neutral-1-500 p-1.5 flex items-center justify-center border border-transparent");
        i0.ɵɵstyleProp("visibility", ctx.showNextBtn ? "visible" : "hidden");
        i0.ɵɵpropertyInterpolate("title", ctx.nextTitle());
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.showSuperNextBtn);
    } }, directives: [i1.NgClass, i1.NgIf, i1.NgForOf, i2.TDSSelect, i3.NgControlStatus, i3.NgModel, i2.TDSOptionTemplateDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DecadeHeaderComponent, [{
        type: Component,
        args: [{ encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, selector: 'decade-header', exportAs: 'decadeHeader', template: "<div class=\"flex py-2 px-0.5 w-full\" [ngClass]=\"prefixCls\">\r\n  <button *ngIf=\"showSuperPreBtn\" [style.visibility]=\"showSuperPreBtn ? 'visible' : 'hidden'\"\r\n    class=\"{{ prefixCls }}-super-prev-btn text-xl text-neutral-1-500 dark:text-d-neutral-1-500 p-1.5 flex items-center justify-center border border-transparent\"\r\n    role=\"button\" tabindex=\"-1\" title=\"{{ superPreviousTitle() }}\" (click)=\"superPrevious()\">\r\n    <span class=\"tdsi-arrow-previous-fill\"></span>\r\n  </button>\r\n  <button [style.visibility]=\"showPreBtn ? 'visible' : 'hidden'\"\r\n    class=\"{{ prefixCls }}-prev-btn p-1.5 text-xl text-neutral-1-500 dark:text-d-neutral-1-500 flex items-center justify-center border border-transparent\"\r\n    role=\"button\" title=\"{{ previousTitle() }}\" tabindex=\"-1\" (click)=\"previous()\">\r\n    <span class=\"tdsi-arrow-left-line\"></span>\r\n  </button>\r\n\r\n  <div\r\n    class=\"{{ prefixCls }}-view flex-auto flex items-center justify-center font-semibold text-title-1 text-neutral-1-900 dark:text-d-neutral-1-900\">\r\n    <ng-container *ngFor=\"let selector of selectors; let indx= index\">\r\n      <button *ngIf=\"!selector.data\" class=\"{{ selector.className }}\" role=\"button\" type=\"button\"\r\n        title=\"{{ selector.title || null }}\" (click)=\"selector.onClick($event)\">\r\n        {{ selector.label }} &nbsp;\r\n      </button>\r\n      <div *ngIf=\"selector.data\" class=\"px-0.5\"\r\n        [class]=\"{'w-[85px]':selectors.length == 1,'flex-auto':selectors.length > 1 && indx == 0}\">\r\n        <tds-select valueField=\"id\" textField=\"name\" [placeholder]='selector.title!' [ngModel]=\"selector.value\"\r\n          (ngModelChange)=\"selector.onClick($event)\" [data]=\"selector.data\" [allowSearch]=\"false\" >\r\n          <ng-template tds-option-tmp let-label='label' let-item=\"item\" let-value='value' let-selected=\"selected\">\r\n            {{item.name}}\r\n          </ng-template>\r\n        </tds-select>\r\n      </div>\r\n    </ng-container>\r\n  </div>\r\n  <button [style.visibility]=\"showNextBtn ? 'visible' : 'hidden'\"\r\n    class=\"{{ prefixCls }}-next-btn text-xl text-neutral-1-500 dark:text-d-neutral-1-500 p-1.5 flex items-center justify-center border border-transparent\"\r\n    role=\"button\" tabindex=\"-1\" title=\"{{ nextTitle() }}\" (click)=\"next()\">\r\n    <span class=\"tdsi-arrow-right-line\"></span>\r\n  </button>\r\n  <button *ngIf=\"showSuperNextBtn\" [style.visibility]=\"showSuperNextBtn ? 'visible' : 'hidden'\"\r\n    class=\"{{ prefixCls }}-super-next-btn text-xl text-neutral-1-500 dark:text-d-neutral-1-500 p-1.5 flex items-center justify-center border border-transparent\"\r\n    role=\"button\" tabindex=\"-1\" title=\"{{ superNextTitle() }}\" (click)=\"superNext()\">\r\n    <span class=\"tdsi-arrow-next-fill\"></span>\r\n  </button>\r\n</div>" }]
    }], null, null); })();
