// tds-ui/esm2020/popover/popover.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSOverlayModule } from 'tds-ui/core/overlay';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSPopoverComponent, TDSPopoverDirective } from './popover';
import * as i0 from "@angular/core";
export class TDSPopoverModule {
}
TDSPopoverModule.ɵfac = function TDSPopoverModule_Factory(t) { return new (t || TDSPopoverModule)(); };
TDSPopoverModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSPopoverModule });
TDSPopoverModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSOverlayModule, TDSToolTipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPopoverModule, [{
        type: NgModule,
        args: [{
                exports: [TDSPopoverDirective, TDSPopoverComponent],
                declarations: [TDSPopoverDirective, TDSPopoverComponent],
                imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSOverlayModule, TDSToolTipModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSPopoverModule, { declarations: [TDSPopoverDirective, TDSPopoverComponent], imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSOverlayModule, TDSToolTipModule], exports: [TDSPopoverDirective, TDSPopoverComponent] }); })();
