// tds-ui/esm2020/cascader/cascader-li.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/progress-spinner";
import * as i3 from "tds-ui/core/outlet";
import * as i4 from "tds-ui/core/highlight";
const _c0 = ["tds-cascader-option", ""];
function TDSCascaderOptionComponent_ng_container_0_ng_template_1_Template(rf, ctx) { }
const _c1 = function (a0, a1) { return { $implicit: a0, index: a1 }; };
function TDSCascaderOptionComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSCascaderOptionComponent_ng_container_0_ng_template_1_Template, 0, 0, "ng-template", 3);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r0.optionTemplate)("ngTemplateOutletContext", i0.ɵɵpureFunction2(2, _c1, ctx_r0.option, ctx_r0.columnIndex));
} }
function TDSCascaderOptionComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 4);
    i0.ɵɵpipe(1, "tdsHighlight");
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("innerHTML", i0.ɵɵpipeBind4(1, 1, ctx_r2.optionLabel, ctx_r2.highlightText, "g", "tds-cascader-menu-item-keyword"), i0.ɵɵsanitizeHtml);
} }
function TDSCascaderOptionComponent_div_3_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵelement(1, "tds-spinner", 7);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("color", "primary");
} }
function TDSCascaderOptionComponent_div_3_ng_template_2_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "span");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵclassMap(ctx_r8.expandIcon);
} }
function TDSCascaderOptionComponent_div_3_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSCascaderOptionComponent_div_3_ng_template_2_ng_container_0_Template, 2, 2, "ng-container", 8);
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r7.expandIcon);
} }
function TDSCascaderOptionComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtemplate(1, TDSCascaderOptionComponent_div_3_span_1_Template, 2, 1, "span", 0);
    i0.ɵɵtemplate(2, TDSCascaderOptionComponent_div_3_ng_template_2_Template, 1, 1, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const _r6 = i0.ɵɵreference(3);
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r3.option.loading)("ngIfElse", _r6);
} }
export class TDSCascaderOptionComponent {
    constructor(cdr, elementRef, renderer) {
        this.cdr = cdr;
        this.optionTemplate = null;
        this.activated = false;
        this.tdsLabelProperty = 'label';
        this.expandIcon = '';
        this.dir = 'ltr';
        renderer.addClass(elementRef.nativeElement, 'tds-cascader-menu-item');
        renderer.addClass(elementRef.nativeElement, 'tds-cascader-menu-item-expanded');
        this.nativeElement = elementRef.nativeElement;
    }
    ngOnInit() {
        if (this.expandIcon === '' && this.dir === 'rtl') {
            this.expandIcon = 'tdsi-arrow-left-line';
        }
        else if (this.expandIcon === '') {
            this.expandIcon = 'tdsi-arrow-right-line';
        }
    }
    get optionLabel() {
        return this.option[this.tdsLabelProperty];
    }
    markForCheck() {
        this.cdr.markForCheck();
    }
}
TDSCascaderOptionComponent.ɵfac = function TDSCascaderOptionComponent_Factory(t) { return new (t || TDSCascaderOptionComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Renderer2)); };
TDSCascaderOptionComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCascaderOptionComponent, selectors: [["", "tds-cascader-option", ""]], hostVars: 7, hostBindings: function TDSCascaderOptionComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("title", ctx.option.title || ctx.optionLabel);
        i0.ɵɵclassProp("tds-cascader-menu-item-active", ctx.activated)("tds-cascader-menu-item-expand", !ctx.option.isLeaf)("tds-cascader-menu-item-disabled", ctx.option.disabled);
    } }, inputs: { optionTemplate: "optionTemplate", option: "option", activated: "activated", highlightText: "highlightText", tdsLabelProperty: "tdsLabelProperty", columnIndex: "columnIndex", expandIcon: "expandIcon", dir: "dir" }, exportAs: ["tdsCascaderOption"], attrs: _c0, decls: 4, vars: 3, consts: [[4, "ngIf", "ngIfElse"], ["defaultOptionTemplate", ""], ["class", "tds-cascader-menu-item-expand-icon text-base leading-4 text-neutral-600 dark:text-d-neutral-1-600", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "tds-cascader-menu-item-content", 3, "innerHTML"], [1, "tds-cascader-menu-item-expand-icon", "text-base", "leading-4", "text-neutral-600", "dark:text-d-neutral-1-600"], ["icon", ""], [1, "h-4", "w-4", 3, "color"], [4, "tdsStringTemplateOutlet"]], template: function TDSCascaderOptionComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSCascaderOptionComponent_ng_container_0_Template, 2, 5, "ng-container", 0);
        i0.ɵɵtemplate(1, TDSCascaderOptionComponent_ng_template_1_Template, 2, 6, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(3, TDSCascaderOptionComponent_div_3_Template, 4, 2, "div", 2);
    } if (rf & 2) {
        const _r1 = i0.ɵɵreference(2);
        i0.ɵɵproperty("ngIf", ctx.optionTemplate)("ngIfElse", _r1);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", !ctx.option.isLeaf || (ctx.option.children == null ? null : ctx.option.children.length) || ctx.option.loading);
    } }, directives: [i1.NgIf, i1.NgTemplateOutlet, i2.TDSProgressSpinnerComponent, i3.TDSStringTemplateOutletDirective], pipes: [i4.TDSHighlightPipe], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCascaderOptionComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: '[tds-cascader-option]',
                exportAs: 'tdsCascaderOption',
                template: `
    <ng-container *ngIf="optionTemplate; else defaultOptionTemplate">
      <ng-template
        [ngTemplateOutlet]="optionTemplate"
        [ngTemplateOutletContext]="{ $implicit: option, index: columnIndex }"
      ></ng-template>
    </ng-container>
    <ng-template #defaultOptionTemplate>
      <div
        class="tds-cascader-menu-item-content"
        [innerHTML]="optionLabel | tdsHighlight: highlightText:'g':'tds-cascader-menu-item-keyword'"
      ></div>
    </ng-template>
    <div *ngIf="!option.isLeaf || option.children?.length || option.loading" class="tds-cascader-menu-item-expand-icon text-base leading-4 text-neutral-600 dark:text-d-neutral-1-600">
      <span *ngIf="option.loading; else icon">
        <tds-spinner [color]="'primary'" class="h-4 w-4"></tds-spinner>
      </span>      
      <ng-template #icon>
      <ng-container *tdsStringTemplateOutlet="expandIcon">
      <span [class]="expandIcon"></span>
  </ng-container>
      </ng-template>
    </div>
  `,
                host: {
                    '[attr.title]': 'option.title || optionLabel',
                    '[class.tds-cascader-menu-item-active]': 'activated',
                    '[class.tds-cascader-menu-item-expand]': '!option.isLeaf',
                    '[class.tds-cascader-menu-item-disabled]': 'option.disabled',
                }
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i0.ElementRef }, { type: i0.Renderer2 }]; }, { optionTemplate: [{
            type: Input
        }], option: [{
            type: Input
        }], activated: [{
            type: Input
        }], highlightText: [{
            type: Input
        }], tdsLabelProperty: [{
            type: Input
        }], columnIndex: [{
            type: Input
        }], expandIcon: [{
            type: Input
        }], dir: [{
            type: Input
        }] }); })();
