// tds-ui/esm2020/cascader/cascader.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSHighlightModule } from 'tds-ui/core/highlight';
import { TDSNoAnimationModule } from 'tds-ui/core/no-animation';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSOverlayModule } from 'tds-ui/core/overlay';
import { TDSEmptyModule } from 'tds-ui/empty';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSInputModule } from 'tds-ui/tds-input';
import { TDSCascaderOptionComponent } from './cascader-li.component';
import { TDSCascaderComponent } from './cascader.component';
import * as i0 from "@angular/core";
export class TDSCascaderModule {
}
TDSCascaderModule.ɵfac = function TDSCascaderModule_Factory(t) { return new (t || TDSCascaderModule)(); };
TDSCascaderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCascaderModule });
TDSCascaderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            FormsModule,
            OverlayModule,
            TDSOutletModule,
            TDSEmptyModule,
            TDSHighlightModule,
            TDSInputModule,
            TDSNoAnimationModule,
            TDSOverlayModule,
            TDSSpinnerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCascaderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    FormsModule,
                    OverlayModule,
                    TDSOutletModule,
                    TDSEmptyModule,
                    TDSHighlightModule,
                    TDSInputModule,
                    TDSNoAnimationModule,
                    TDSOverlayModule,
                    TDSSpinnerModule
                ],
                declarations: [TDSCascaderComponent, TDSCascaderOptionComponent],
                exports: [TDSCascaderComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCascaderModule, { declarations: [TDSCascaderComponent, TDSCascaderOptionComponent], imports: [BidiModule,
        CommonModule,
        FormsModule,
        OverlayModule,
        TDSOutletModule,
        TDSEmptyModule,
        TDSHighlightModule,
        TDSInputModule,
        TDSNoAnimationModule,
        TDSOverlayModule,
        TDSSpinnerModule], exports: [TDSCascaderComponent] }); })();
