// tds-ui/esm2020/progress-spinner/progress-spinner.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSProgressSpinnerComponent } from './progress-spinner.component';
import { TDSSpinComponent } from './spin.component';
import * as i0 from "@angular/core";
export class TDSSpinnerModule {
}
TDSSpinnerModule.ɵfac = function TDSSpinnerModule_Factory(t) { return new (t || TDSSpinnerModule)(); };
TDSSpinnerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSSpinnerModule });
TDSSpinnerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSpinnerModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSProgressSpinnerComponent,
                    TDSSpinComponent
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    TDSProgressSpinnerComponent,
                    TDSSpinComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSSpinnerModule, { declarations: [TDSProgressSpinnerComponent,
        TDSSpinComponent], imports: [CommonModule], exports: [TDSProgressSpinnerComponent,
        TDSSpinComponent] }); })();
