// tds-ui/esm2020/page-header/page-header-cells.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSPageHeaderTitleDirective {
}
TDSPageHeaderTitleDirective.ɵfac = function TDSPageHeaderTitleDirective_Factory(t) { return new (t || TDSPageHeaderTitleDirective)(); };
TDSPageHeaderTitleDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderTitleDirective, selectors: [["tds-page-header-title"], ["", "tds-page-header-title", ""]], hostAttrs: [1, "tds-page-header-heading-title", "text-neutral-1-900", "dark:text-d-neutral-1-900", "text-header-1", "font-semibold", "pr-3"], exportAs: ["tdsPageHeaderTitle"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderTitleDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-page-header-title, [tds-page-header-title]',
                exportAs: 'tdsPageHeaderTitle',
                host: {
                    class: 'tds-page-header-heading-title text-neutral-1-900 dark:text-d-neutral-1-900 text-header-1 font-semibold pr-3'
                }
            }]
    }], null, null); })();
export class TDSPageHeaderSubtitleDirective {
}
TDSPageHeaderSubtitleDirective.ɵfac = function TDSPageHeaderSubtitleDirective_Factory(t) { return new (t || TDSPageHeaderSubtitleDirective)(); };
TDSPageHeaderSubtitleDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderSubtitleDirective, selectors: [["tds-page-header-subtitle"], ["", "tds-page-header-subtitle", ""]], hostAttrs: [1, "tds-page-header-heading-sub-title", "text-caption-1", "font-regular", "text-neutral-1-400", "dark:text-d-neutral-1-400", "pr-3"], exportAs: ["tdsPageHeaderSubtitle"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderSubtitleDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-page-header-subtitle, [tds-page-header-subtitle]',
                exportAs: 'tdsPageHeaderSubtitle',
                host: {
                    class: 'tds-page-header-heading-sub-title text-caption-1 font-regular text-neutral-1-400 dark:text-d-neutral-1-400 pr-3'
                }
            }]
    }], null, null); })();
export class TDSPageHeaderContentDirective {
}
TDSPageHeaderContentDirective.ɵfac = function TDSPageHeaderContentDirective_Factory(t) { return new (t || TDSPageHeaderContentDirective)(); };
TDSPageHeaderContentDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderContentDirective, selectors: [["tds-page-header-content"], ["", "tds-page-header-content", ""]], hostAttrs: [1, "tds-page-header-content", "px-4", "py-2.5", "w-full", "block"], exportAs: ["tdsPageHeaderContent"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderContentDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-page-header-content, [tds-page-header-content]',
                exportAs: 'tdsPageHeaderContent',
                host: {
                    class: 'tds-page-header-content  px-4 py-2.5 w-full block'
                }
            }]
    }], null, null); })();
export class TDSPageHeaderTagDirective {
}
TDSPageHeaderTagDirective.ɵfac = function TDSPageHeaderTagDirective_Factory(t) { return new (t || TDSPageHeaderTagDirective)(); };
TDSPageHeaderTagDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderTagDirective, selectors: [["tds-page-header-tags"], ["", "tds-page-header-tags", ""]], hostAttrs: [1, "tds-page-header-heading-tags"], exportAs: ["tdsPageHeaderTags"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderTagDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-page-header-tags, [tds-page-header-tags]',
                exportAs: 'tdsPageHeaderTags',
                host: {
                    class: 'tds-page-header-heading-tags'
                }
            }]
    }], null, null); })();
export class TDSPageHeaderExtraDirective {
}
TDSPageHeaderExtraDirective.ɵfac = function TDSPageHeaderExtraDirective_Factory(t) { return new (t || TDSPageHeaderExtraDirective)(); };
TDSPageHeaderExtraDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderExtraDirective, selectors: [["tds-page-header-extra"], ["", "tds-page-header-extra", ""]], hostAttrs: [1, "tds-page-header-heading-extra", "flex"], exportAs: ["tdsPageHeaderExtra"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderExtraDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-page-header-extra, [tds-page-header-extra]',
                exportAs: 'tdsPageHeaderExtra',
                host: {
                    class: 'tds-page-header-heading-extra flex '
                }
            }]
    }], null, null); })();
export class TDSPageHeaderFooterDirective {
}
TDSPageHeaderFooterDirective.ɵfac = function TDSPageHeaderFooterDirective_Factory(t) { return new (t || TDSPageHeaderFooterDirective)(); };
TDSPageHeaderFooterDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderFooterDirective, selectors: [["tds-page-header-footer"], ["", "tds-page-header-footer", ""]], hostAttrs: [1, "tds-page-header-footer", "border-t", "border-neutral-2-200", "dark:border-d-neutral-2-200", "px-4", "py-2.5", "w-full", "block"], exportAs: ["tdsPageHeaderFooter"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderFooterDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-page-header-footer, [tds-page-header-footer]',
                exportAs: 'tdsPageHeaderFooter',
                host: {
                    class: 'tds-page-header-footer  border-t border-neutral-2-200 dark:border-d-neutral-2-200 px-4 py-2.5 w-full block'
                }
            }]
    }], null, null); })();
export class TDSPageHeaderBreadcrumbDirective {
}
TDSPageHeaderBreadcrumbDirective.ɵfac = function TDSPageHeaderBreadcrumbDirective_Factory(t) { return new (t || TDSPageHeaderBreadcrumbDirective)(); };
TDSPageHeaderBreadcrumbDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderBreadcrumbDirective, selectors: [["tds-breadcrumb", "tds-page-header-breadcrumb", ""]], hostAttrs: [1, "px-4", "pt-2.5", "pb-1"], exportAs: ["tdsPageHeaderBreadcrumb"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderBreadcrumbDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-breadcrumb[tds-page-header-breadcrumb]',
                exportAs: 'tdsPageHeaderBreadcrumb',
                host: {
                    class: 'px-4 pt-2.5 pb-1'
                }
            }]
    }], null, null); })();
export class TDSPageHeaderAvatarDirective {
}
TDSPageHeaderAvatarDirective.ɵfac = function TDSPageHeaderAvatarDirective_Factory(t) { return new (t || TDSPageHeaderAvatarDirective)(); };
TDSPageHeaderAvatarDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPageHeaderAvatarDirective, selectors: [["tds-avatar", "tds-page-header-avatar", ""]], exportAs: ["tdsPageHeaderAvatar"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderAvatarDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-avatar[tds-page-header-avatar]',
                exportAs: 'tdsPageHeaderAvatar'
            }]
    }], null, null); })();
