// tds-ui/esm2020/page-header/page-header.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSPageHeaderAvatarDirective, TDSPageHeaderBreadcrumbDirective, TDSPageHeaderContentDirective, TDSPageHeaderExtraDirective, TDSPageHeaderFooterDirective, TDSPageHeaderSubtitleDirective, TDSPageHeaderTagDirective, TDSPageHeaderTitleDirective } from './page-header-cells';
import { TDSPageHeaderComponent } from './page-header.component';
import * as i0 from "@angular/core";
const TDSPageHeaderCells = [
    TDSPageHeaderTitleDirective,
    TDSPageHeaderSubtitleDirective,
    TDSPageHeaderContentDirective,
    TDSPageHeaderTagDirective,
    TDSPageHeaderExtraDirective,
    TDSPageHeaderFooterDirective,
    TDSPageHeaderBreadcrumbDirective,
    TDSPageHeaderAvatarDirective
];
export class TDSPageHeaderModule {
}
TDSPageHeaderModule.ɵfac = function TDSPageHeaderModule_Factory(t) { return new (t || TDSPageHeaderModule)(); };
TDSPageHeaderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSPageHeaderModule });
TDSPageHeaderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, TDSOutletModule],
                exports: [TDSPageHeaderComponent, TDSPageHeaderCells],
                declarations: [TDSPageHeaderComponent, TDSPageHeaderCells]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSPageHeaderModule, { declarations: [TDSPageHeaderComponent, TDSPageHeaderTitleDirective,
        TDSPageHeaderSubtitleDirective,
        TDSPageHeaderContentDirective,
        TDSPageHeaderTagDirective,
        TDSPageHeaderExtraDirective,
        TDSPageHeaderFooterDirective,
        TDSPageHeaderBreadcrumbDirective,
        TDSPageHeaderAvatarDirective], imports: [BidiModule, CommonModule, TDSOutletModule], exports: [TDSPageHeaderComponent, TDSPageHeaderTitleDirective,
        TDSPageHeaderSubtitleDirective,
        TDSPageHeaderContentDirective,
        TDSPageHeaderTagDirective,
        TDSPageHeaderExtraDirective,
        TDSPageHeaderFooterDirective,
        TDSPageHeaderBreadcrumbDirective,
        TDSPageHeaderAvatarDirective] }); })();
