// tds-ui/esm2020/page-header/page-header.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Input, Optional, Output, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { map, takeUntil } from 'rxjs/operators';
import { TDSPageHeaderBreadcrumbDirective, TDSPageHeaderFooterDirective } from './page-header-cells';
import { WithConfig } from 'tds-ui/core/config';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/core/config";
import * as i3 from "tds-ui/core/resize-observers";
import * as i4 from "@angular/cdk/bidi";
import * as i5 from "tds-ui/core/outlet";
function TDSPageHeaderComponent_div_3_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "span", 9);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const backIcon_r6 = ctx.$implicit;
    const ctx_r5 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", backIcon_r6 || ctx_r5.getBackIcon());
} }
function TDSPageHeaderComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r8 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 6);
    i0.ɵɵlistener("click", function TDSPageHeaderComponent_div_3_Template_div_click_0_listener() { i0.ɵɵrestoreView(_r8); const ctx_r7 = i0.ɵɵnextContext(); return ctx_r7._onBack(); });
    i0.ɵɵelementStart(1, "div", 7);
    i0.ɵɵtemplate(2, TDSPageHeaderComponent_div_3_ng_container_2_Template, 2, 1, "ng-container", 8);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r0.backIcon);
} }
function TDSPageHeaderComponent_span_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r9.title);
} }
function TDSPageHeaderComponent_span_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 10);
    i0.ɵɵtemplate(1, TDSPageHeaderComponent_span_5_ng_container_1_Template, 2, 1, "ng-container", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r1.title);
} }
function TDSPageHeaderComponent_ng_content_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0, 6, ["*ngIf", "!title"]);
} }
function TDSPageHeaderComponent_span_7_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r10.subtitle);
} }
function TDSPageHeaderComponent_span_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 11);
    i0.ɵɵtemplate(1, TDSPageHeaderComponent_span_7_ng_container_1_Template, 2, 1, "ng-container", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r3.subtitle);
} }
function TDSPageHeaderComponent_ng_content_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0, 7, ["*ngIf", "!subtitle"]);
} }
const _c0 = [[["tds-breadcrumb", "tds-page-header-breadcrumb", ""]], [["tds-avatar", "tds-page-header-avatar", ""]], [["tds-page-header-tags"], ["", "tds-page-header-tags", ""]], [["tds-page-header-extra"], ["", "tds-page-header-extra", ""]], [["tds-page-header-content"], ["", "tds-page-header-content", ""]], [["tds-page-header-footer"], ["", "tds-page-header-footer", ""]], [["tds-page-header-title"], ["", "tds-page-header-title", ""]], [["tds-page-header-subtitle"], ["", "tds-page-header-subtitle", ""]]];
const _c1 = ["tds-breadcrumb[tds-page-header-breadcrumb]", "tds-avatar[tds-page-header-avatar]", "tds-page-header-tags, [tds-page-header-tags]", "tds-page-header-extra, [tds-page-header-extra]", "tds-page-header-content, [tds-page-header-content]", "tds-page-header-footer, [tds-page-header-footer]", "tds-page-header-title, [tds-page-header-title]", "tds-page-header-subtitle, [tds-page-header-subtitle]"];
// const TDS_CONFIG_MODULE_NAME: TDSConfigKey = 'pageHeader';
export class TDSPageHeaderComponent {
    constructor(location, tdsConfigService, elementRef, tdsResizeObserver, cdr, directionality) {
        this.location = location;
        this.tdsConfigService = tdsConfigService;
        this.elementRef = elementRef;
        this.tdsResizeObserver = tdsResizeObserver;
        this.cdr = cdr;
        this.directionality = directionality;
        this.backIcon = null;
        this.ghost = true;
        this.back = new EventEmitter();
        this.compact = false;
        this.destroy$ = new Subject();
        this.dir = 'ltr';
    }
    ngOnInit() {
        this.directionality.change?.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
    }
    ngAfterViewInit() {
        this.tdsResizeObserver
            .observe(this.elementRef)
            .pipe(map(([entry]) => entry.contentRect.width), takeUntil(this.destroy$))
            .subscribe((width) => {
            this.compact = width < 768;
            this.cdr.markForCheck();
        });
    }
    _onBack() {
        if (this.back.observers.length) {
            this.back.emit();
        }
        else {
            if (!this.location) {
                throw new Error(`[TDS] you should import 'RouterModule' or register 'Location' if you want to use 'back' default event!`);
            }
            this.location.back();
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    getBackIcon() {
        if (this.dir === 'rtl') {
            return 'arrow-right';
        }
        return 'arrow-left';
    }
}
TDSPageHeaderComponent.ɵfac = function TDSPageHeaderComponent_Factory(t) { return new (t || TDSPageHeaderComponent)(i0.ɵɵdirectiveInject(i1.Location, 8), i0.ɵɵdirectiveInject(i2.TDSConfigService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i3.TDSResizeObserver), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i4.Directionality, 8)); };
TDSPageHeaderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPageHeaderComponent, selectors: [["tds-page-header"]], contentQueries: function TDSPageHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSPageHeaderFooterDirective, 5);
        i0.ɵɵcontentQuery(dirIndex, TDSPageHeaderBreadcrumbDirective, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.tdsPageHeaderFooter = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.tdsPageHeaderBreadcrumb = _t.first);
    } }, hostAttrs: [1, "tds-page-header", "w-full", "block"], hostVars: 10, hostBindings: function TDSPageHeaderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("has-footer", ctx.tdsPageHeaderFooter)("tds-page-header-ghost", ctx.ghost)("has-breadcrumb", ctx.tdsPageHeaderBreadcrumb)("tds-page-header-compact", ctx.compact)("tds-page-header-rtl", ctx.dir === "rtl");
    } }, inputs: { backIcon: "backIcon", title: "title", subtitle: "subtitle", ghost: "ghost" }, outputs: { back: "back" }, exportAs: ["tdsPageHeader"], ngContentSelectors: _c1, decls: 13, vars: 5, consts: [[1, "tds-page-header-heading", "flex", "justify-between", "pl-4", "pr-4", "pb-2.5"], [1, "tds-page-header-heading-left", "flex", "items-center"], ["class", "tds-page-header-back", 3, "click", 4, "ngIf"], ["class", "tds-page-header-heading-title text-neutral-1-900 dark:text-d-neutral-1-900 text-header-1 font-semibold pr-3", 4, "ngIf"], [4, "ngIf"], ["class", "tds-page-header-heading-sub-title text-caption-1 font-regular text-neutral-1-400 dark:text-d-neutral-1-400 pr-3", 4, "ngIf"], [1, "tds-page-header-back", 3, "click"], ["role", "button", "tabindex", "0", 1, "tds-page-header-back-button"], [4, "tdsStringTemplateOutlet"], [3, "ngClass"], [1, "tds-page-header-heading-title", "text-neutral-1-900", "dark:text-d-neutral-1-900", "text-header-1", "font-semibold", "pr-3"], [1, "tds-page-header-heading-sub-title", "text-caption-1", "font-regular", "text-neutral-1-400", "dark:text-d-neutral-1-400", "pr-3"]], template: function TDSPageHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c0);
        i0.ɵɵprojection(0);
        i0.ɵɵelementStart(1, "div", 0)(2, "div", 1);
        i0.ɵɵtemplate(3, TDSPageHeaderComponent_div_3_Template, 3, 1, "div", 2);
        i0.ɵɵprojection(4, 1);
        i0.ɵɵtemplate(5, TDSPageHeaderComponent_span_5_Template, 2, 1, "span", 3);
        i0.ɵɵtemplate(6, TDSPageHeaderComponent_ng_content_6_Template, 1, 0, "ng-content", 4);
        i0.ɵɵtemplate(7, TDSPageHeaderComponent_span_7_Template, 2, 1, "span", 5);
        i0.ɵɵtemplate(8, TDSPageHeaderComponent_ng_content_8_Template, 1, 0, "ng-content", 4);
        i0.ɵɵprojection(9, 2);
        i0.ɵɵelementEnd();
        i0.ɵɵprojection(10, 3);
        i0.ɵɵelementEnd();
        i0.ɵɵprojection(11, 4);
        i0.ɵɵprojection(12, 5);
    } if (rf & 2) {
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.backIcon !== null);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.title);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.title);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.subtitle);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.subtitle);
    } }, directives: [i1.NgIf, i5.TDSStringTemplateOutletDirective, i1.NgClass], encapsulation: 2, changeDetection: 0 });
__decorate([
    WithConfig()
], TDSPageHeaderComponent.prototype, "ghost", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPageHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'tds-page-header',
                exportAs: 'tdsPageHeader',
                template: `
    <ng-content select="tds-breadcrumb[tds-page-header-breadcrumb]"></ng-content>

    <div class="tds-page-header-heading flex justify-between pl-4 pr-4 pb-2.5">
      <div class="tds-page-header-heading-left flex items-center  ">
        <!--back-->
        <div *ngIf="backIcon !== null" (click)="_onBack()" class="tds-page-header-back">
          <div role="button" tabindex="0" class="tds-page-header-back-button">
            <ng-container *tdsStringTemplateOutlet="backIcon; let backIcon">
              <span [ngClass]="backIcon || getBackIcon()"></span>             
            </ng-container>
          </div>
        </div>
        <!--avatar-->
        <ng-content select="tds-avatar[tds-page-header-avatar]"></ng-content>
        <!--title-->
        <span class="tds-page-header-heading-title text-neutral-1-900 dark:text-d-neutral-1-900 text-header-1 font-semibold pr-3" *ngIf="title">
          <ng-container *tdsStringTemplateOutlet="title">{{ title }}</ng-container>
        </span>
        <ng-content *ngIf="!title" select="tds-page-header-title, [tds-page-header-title]"></ng-content>
        <!--subtitle-->
        <span class="tds-page-header-heading-sub-title text-caption-1 font-regular text-neutral-1-400 dark:text-d-neutral-1-400 pr-3" *ngIf="subtitle">
          <ng-container *tdsStringTemplateOutlet="subtitle">{{ subtitle }}</ng-container>
        </span>
        <ng-content *ngIf="!subtitle" select="tds-page-header-subtitle, [tds-page-header-subtitle]"></ng-content>
        <ng-content select="tds-page-header-tags, [tds-page-header-tags]"></ng-content>
      </div>

      <ng-content select="tds-page-header-extra, [tds-page-header-extra]"></ng-content>
    </div>

    <ng-content select="tds-page-header-content, [tds-page-header-content]"></ng-content>
    <ng-content select="tds-page-header-footer, [tds-page-header-footer]"></ng-content>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                host: {
                    class: 'tds-page-header w-full block',
                    '[class.has-footer]': 'tdsPageHeaderFooter',
                    '[class.tds-page-header-ghost]': 'ghost',
                    '[class.has-breadcrumb]': 'tdsPageHeaderBreadcrumb',
                    '[class.tds-page-header-compact]': 'compact',
                    '[class.tds-page-header-rtl]': `dir === 'rtl'`,
                }
            }]
    }], function () { return [{ type: i1.Location, decorators: [{
                type: Optional
            }] }, { type: i2.TDSConfigService }, { type: i0.ElementRef }, { type: i3.TDSResizeObserver }, { type: i0.ChangeDetectorRef }, { type: i4.Directionality, decorators: [{
                type: Optional
            }] }]; }, { backIcon: [{
            type: Input
        }], title: [{
            type: Input
        }], subtitle: [{
            type: Input
        }], ghost: [{
            type: Input
        }], back: [{
            type: Output
        }], tdsPageHeaderFooter: [{
            type: ContentChild,
            args: [TDSPageHeaderFooterDirective, { static: false }]
        }], tdsPageHeaderBreadcrumb: [{
            type: ContentChild,
            args: [TDSPageHeaderBreadcrumbDirective, { static: false }]
        }] }); })();
