// tds-ui/esm2020/notification/notification.component.mjs
import { Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { notificationMotion } from 'tds-ui/core/animation';
import { TDSMNComponent } from 'tds-ui/message';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/core/outlet";
import * as i3 from "tds-ui/button";
function TDSNotificationComponent_div_1_i_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 18);
} }
function TDSNotificationComponent_div_1_i_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 19);
} }
function TDSNotificationComponent_div_1_i_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 20);
} }
function TDSNotificationComponent_div_1_i_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 21);
} }
const _c0 = function (a0) { return { "tds-notification-notice-with-icon": a0 }; };
const _c1 = function (a0, a1) { return { "tds-notification-notice-with-icon mr-4": a0, "hidden": a1 }; };
function TDSNotificationComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 7)(1, "div", 8)(2, "div", 9);
    i0.ɵɵelementContainerStart(3, 10);
    i0.ɵɵtemplate(4, TDSNotificationComponent_div_1_i_4_Template, 1, 0, "i", 11);
    i0.ɵɵtemplate(5, TDSNotificationComponent_div_1_i_5_Template, 1, 0, "i", 12);
    i0.ɵɵtemplate(6, TDSNotificationComponent_div_1_i_6_Template, 1, 0, "i", 13);
    i0.ɵɵtemplate(7, TDSNotificationComponent_div_1_i_7_Template, 1, 0, "i", 14);
    i0.ɵɵelementContainerEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(8, "div", 15);
    i0.ɵɵelement(9, "div", 16)(10, "div", 17);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(9, _c0, ctx_r0.instance.type !== "blank"));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(11, _c1, ctx_r0.instance.type !== "blank", ctx_r0.instance.type == "blank"));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitch", ctx_r0.instance.type);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "success");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "info");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "warning");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "error");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("innerHTML", ctx_r0.instance.title, i0.ɵɵsanitizeHtml);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("innerHTML", ctx_r0.instance.content, i0.ɵɵsanitizeHtml);
} }
function TDSNotificationComponent_div_2_ng_template_1_Template(rf, ctx) { }
const _c2 = function (a0, a1) { return { $implicit: a0, data: a1 }; };
function TDSNotificationComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 22);
    i0.ɵɵtemplate(1, TDSNotificationComponent_div_2_ng_template_1_Template, 0, 0, "ng-template", 23);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.instance.template)("ngTemplateOutlet", ctx_r1.instance.template)("ngTemplateOutletContext", i0.ɵɵpureFunction2(3, _c2, ctx_r1, ctx_r1.instance.options.data));
} }
function TDSNotificationComponent_ng_container_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "button", 25);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", ctx_r10.instance.options.closeIcon);
} }
function TDSNotificationComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSNotificationComponent_ng_container_5_ng_container_1_Template, 2, 1, "ng-container", 24);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r2.instance.options == null ? null : ctx_r2.instance.options.closeIcon);
} }
function TDSNotificationComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "button", 26);
} }
export class TDSNotificationComponent extends TDSMNComponent {
    constructor(cdr) {
        super(cdr);
        this.destroyed = new EventEmitter();
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        this.instance.onClick.complete();
    }
    onClick(event) {
        this.instance.onClick.next(event);
    }
    close() {
        this.destroy(true);
    }
    get state() {
        if (this.instance.state === 'enter') {
            if (this.placement === 'topLeft' || this.placement === 'bottomLeft') {
                return 'enterLeft';
            }
            else {
                return 'enterRight';
            }
        }
        else {
            return this.instance.state;
        }
    }
}
TDSNotificationComponent.ɵfac = function TDSNotificationComponent_Factory(t) { return new (t || TDSNotificationComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSNotificationComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSNotificationComponent, selectors: [["tds-notification"]], hostAttrs: [1, "mb-2", "block"], inputs: { instance: "instance", placement: "placement", index: "index" }, outputs: { destroyed: "destroyed" }, exportAs: ["tdsNotification"], features: [i0.ɵɵInheritDefinitionFeature], decls: 8, vars: 7, consts: [[1, "tds-notification-notice", "tds-notification-notice-closable", "relative", "mr-4", 3, "ngStyle", "ngClass", "click", "mouseenter", "mouseleave"], ["class", "tds-notification-notice-content bg-white dark:bg-d-neutral-3-300 p-4 shadow-1-xl rounded-md flex flex-col w-96", 4, "ngIf"], ["class", "bg-white dark:bg-d-neutral-3-300 px-4 py-3 shadow-1-xl rounded-md flex flex-col w-96", 4, "ngIf"], ["tabindex", "0", 1, "tds-notification-notice-close", "absolute", "hover:bg-neutral-3-50", "dark:hover:bg-d-neutral-3-300", "rounded-md", "top-4", "right-4", "cursor-pointer", "p-1", "border", "border-transparent", 3, "click"], [1, "tds-notification-notice-close-x", "flex"], [4, "ngIf", "ngIfElse"], ["iconTpl", ""], [1, "tds-notification-notice-content", "bg-white", "dark:bg-d-neutral-3-300", "p-4", "shadow-1-xl", "rounded-md", "flex", "flex-col", "w-96"], [1, "tds-notification-notice-content", "flex", 3, "ngClass"], [3, "ngClass"], [3, "ngSwitch"], ["class", "text-2xl leading-6 text-success-400 tdsi-success-fill  tds-notification-notice-icon tds-notification-notice-icon-success", 4, "ngSwitchCase"], ["class", "text-2xl leading-6 text-info-400 tdsi-information-fill tds-notification-notice-icon tds-notification-notice-icon-info", 4, "ngSwitchCase"], ["class", "text-2xl leading-6 text-warning-400  tdsi-warning-fill tds-notification-notice-icon tds-notification-notice-icon-warning", 4, "ngSwitchCase"], ["class", "text-2xl leading-6 text-error-400  tdsi-error-fill tds-notification-notice-icon tds-notification-notice-icon-error", 4, "ngSwitchCase"], [1, "flex", "flex-col"], [1, "tds-notification-notice-message", "text-title-1", "text-neutral-1-900", "dark:text-d-neutral-1-900", "font-semibold", "break-all", "pr-8", 3, "innerHTML"], [1, "tds-notification-notice-description", "pt-2", "text-body-2", "font-regular", "text-neutral-1-900", "dark:text-d-neutral-1-900", "break-all", 3, "innerHTML"], [1, "text-2xl", "leading-6", "text-success-400", "tdsi-success-fill", "tds-notification-notice-icon", "tds-notification-notice-icon-success"], [1, "text-2xl", "leading-6", "text-info-400", "tdsi-information-fill", "tds-notification-notice-icon", "tds-notification-notice-icon-info"], [1, "text-2xl", "leading-6", "text-warning-400", "tdsi-warning-fill", "tds-notification-notice-icon", "tds-notification-notice-icon-warning"], [1, "text-2xl", "leading-6", "text-error-400", "tdsi-error-fill", "tds-notification-notice-icon", "tds-notification-notice-icon-error"], [1, "bg-white", "dark:bg-d-neutral-3-300", "px-4", "py-3", "shadow-1-xl", "rounded-md", "flex", "flex-col", "w-96"], [3, "ngIf", "ngTemplateOutlet", "ngTemplateOutletContext"], [4, "tdsStringTemplateOutlet"], ["tds-button-close", "", 1, "tds-notification-close-icon", 3, "ngClass"], ["tds-button-close", "", 1, "tds-notification-close-icon"]], template: function TDSNotificationComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵlistener("@notificationMotion.done", function TDSNotificationComponent_Template_div_animation_notificationMotion_done_0_listener($event) { return ctx.animationStateChanged.next($event); })("click", function TDSNotificationComponent_Template_div_click_0_listener($event) { return ctx.onClick($event); })("mouseenter", function TDSNotificationComponent_Template_div_mouseenter_0_listener() { return ctx.onEnter(); })("mouseleave", function TDSNotificationComponent_Template_div_mouseleave_0_listener() { return ctx.onLeave(); });
        i0.ɵɵtemplate(1, TDSNotificationComponent_div_1_Template, 11, 14, "div", 1);
        i0.ɵɵtemplate(2, TDSNotificationComponent_div_2_Template, 2, 6, "div", 2);
        i0.ɵɵelementStart(3, "a", 3);
        i0.ɵɵlistener("click", function TDSNotificationComponent_Template_a_click_3_listener() { return ctx.close(); });
        i0.ɵɵelementStart(4, "span", 4);
        i0.ɵɵtemplate(5, TDSNotificationComponent_ng_container_5_Template, 2, 1, "ng-container", 5);
        i0.ɵɵtemplate(6, TDSNotificationComponent_ng_template_6_Template, 1, 0, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵelementEnd()()();
    } if (rf & 2) {
        const _r3 = i0.ɵɵreference(7);
        i0.ɵɵproperty("ngStyle", ctx.instance.options.style)("ngClass", ctx.instance.options.clss || "")("@notificationMotion", ctx.state);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.instance.template);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.instance.template);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.instance.options == null ? null : ctx.instance.options.closeIcon)("ngIfElse", _r3);
    } }, directives: [i1.NgStyle, i1.NgClass, i1.NgIf, i1.NgSwitch, i1.NgSwitchCase, i1.NgTemplateOutlet, i2.TDSStringTemplateOutletDirective, i3.TDSButtonCloseComponent], encapsulation: 2, data: { animation: [notificationMotion] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSNotificationComponent, [{
        type: Component,
        args: [{
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-notification',
                exportAs: 'tdsNotification',
                preserveWhitespaces: false,
                animations: [notificationMotion],
                host: {
                    class: "mb-2 block"
                },
                template: `
  <div class="tds-notification-notice tds-notification-notice-closable relative mr-4"
  [ngStyle]="instance.options.style!" 
  [ngClass]="instance.options.clss || ''" 
  [@notificationMotion]="state"
  (@notificationMotion.done)="animationStateChanged.next($event)" 
  (click)="onClick($event)" 
  (mouseenter)="onEnter()"
  (mouseleave)="onLeave()">
  <div *ngIf="!instance.template"
      class="tds-notification-notice-content bg-white dark:bg-d-neutral-3-300 p-4 shadow-1-xl rounded-md flex flex-col w-96">
      <div class="tds-notification-notice-content flex  "
          [ngClass]="{ 'tds-notification-notice-with-icon': instance.type !== 'blank' }">
          <div
              [ngClass]="{'tds-notification-notice-with-icon mr-4':instance.type !== 'blank','hidden':instance.type == 'blank'}">
              <ng-container [ngSwitch]="instance.type">
                  <i *ngSwitchCase="'success'"
                      class="text-2xl leading-6 text-success-400 tdsi-success-fill  tds-notification-notice-icon tds-notification-notice-icon-success"></i>
                  <i *ngSwitchCase="'info'"
                      class="text-2xl leading-6 text-info-400 tdsi-information-fill tds-notification-notice-icon tds-notification-notice-icon-info"></i>
                  <i *ngSwitchCase="'warning'"
                      class="text-2xl leading-6 text-warning-400  tdsi-warning-fill tds-notification-notice-icon tds-notification-notice-icon-warning"></i>
                  <i *ngSwitchCase="'error'"
                      class="text-2xl leading-6 text-error-400  tdsi-error-fill tds-notification-notice-icon tds-notification-notice-icon-error"></i>
              </ng-container>
          </div>
          <div class="flex flex-col">
              <div class="tds-notification-notice-message text-title-1 text-neutral-1-900 dark:text-d-neutral-1-900 font-semibold break-all pr-8"
                  [innerHTML]="instance.title"></div>
              <div class="tds-notification-notice-description pt-2 text-body-2 font-regular text-neutral-1-900 dark:text-d-neutral-1-900 break-all"
                  [innerHTML]="instance.content"></div>
          </div>
      </div>
  </div>
  <div class="bg-white dark:bg-d-neutral-3-300 px-4 py-3 shadow-1-xl rounded-md flex flex-col w-96" *ngIf="instance.template">
  <ng-template [ngIf]="instance.template" [ngTemplateOutlet]="instance.template!"
  [ngTemplateOutletContext]="{ $implicit: this, data: instance.options.data! }"></ng-template>
  </div>
 
  <a tabindex="0" class="tds-notification-notice-close absolute hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300  rounded-md top-4 right-4 cursor-pointer p-1 border border-transparent" (click)="close()">
      <span class="tds-notification-notice-close-x flex">
          <ng-container *ngIf="instance.options?.closeIcon; else iconTpl">
              <ng-container *tdsStringTemplateOutlet="instance.options?.closeIcon; let closeIcon">
                  <button [ngClass]="instance.options.closeIcon!" tds-button-close class="tds-notification-close-icon"></button>
              </ng-container>
          </ng-container>
          <ng-template #iconTpl>
            <button tds-button-close class="tds-notification-close-icon"></button>
          </ng-template>
      </span>
  </a>
</div>
  `
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { instance: [{
            type: Input
        }], placement: [{
            type: Input
        }], index: [{
            type: Input
        }], destroyed: [{
            type: Output
        }] }); })();
