// tds-ui/esm2020/notification/notification.service.mjs
import { Injectable } from '@angular/core';
import { TDSMNService } from 'tds-ui/message';
import { TDSNotificationContainerComponent } from './notification-container.component';
import { TDSNotificationServiceModule } from './notification.service.module';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/services";
import * as i2 from "@angular/cdk/overlay";
let notificationId = 0;
export class TDSNotificationService extends TDSMNService {
    constructor(tdsSingletonService, overlay, injector) {
        super(tdsSingletonService, overlay, injector);
        this.componentPrefix = 'notification-';
    }
    success(title, content, options) {
        return this.createInstance({ type: 'success', title, content }, options);
    }
    error(title, content, options) {
        return this.createInstance({ type: 'error', title, content }, options);
    }
    info(title, content, options) {
        return this.createInstance({ type: 'info', title, content }, options);
    }
    warning(title, content, options) {
        return this.createInstance({ type: 'warning', title, content }, options);
    }
    blank(title, content, options) {
        return this.createInstance({ type: 'blank', title, content }, options);
    }
    create(type, title, content, options) {
        return this.createInstance({ type, title, content }, options);
    }
    template(template, options) {
        return this.createInstance({ template }, options);
    }
    generateMessageId() {
        return `${this.componentPrefix}-${notificationId++}`;
    }
    createInstance(message, options) {
        this.container = this.withContainer(TDSNotificationContainerComponent);
        return this.container.create({
            ...message,
            ...{
                createdAt: new Date(),
                messageId: this.generateMessageId(),
                options
            }
        });
    }
}
TDSNotificationService.ɵfac = function TDSNotificationService_Factory(t) { return new (t || TDSNotificationService)(i0.ɵɵinject(i1.TDSSingletonService), i0.ɵɵinject(i2.Overlay), i0.ɵɵinject(i0.Injector)); };
TDSNotificationService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TDSNotificationService, factory: TDSNotificationService.ɵfac, providedIn: TDSNotificationServiceModule });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSNotificationService, [{
        type: Injectable,
        args: [{
                providedIn: TDSNotificationServiceModule
            }]
    }], function () { return [{ type: i1.TDSSingletonService }, { type: i2.Overlay }, { type: i0.Injector }]; }, null); })();
