// tds-ui/esm2020/notification/notification.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSNotificationContainerComponent } from './notification-container.component';
import { TDSNotificationComponent } from './notification.component';
import { TDSNotificationServiceModule } from './notification.service.module';
import * as i0 from "@angular/core";
export class TDSNotificationModule {
}
TDSNotificationModule.ɵfac = function TDSNotificationModule_Factory(t) { return new (t || TDSNotificationModule)(); };
TDSNotificationModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSNotificationModule });
TDSNotificationModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSNotificationServiceModule, TDSButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSNotificationModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSNotificationServiceModule, TDSButtonModule],
                declarations: [TDSNotificationComponent, TDSNotificationContainerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSNotificationModule, { declarations: [TDSNotificationComponent, TDSNotificationContainerComponent], imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSNotificationServiceModule, TDSButtonModule] }); })();
