// tds-ui/esm2020/tree/tree.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSHighlightModule } from 'tds-ui/core/highlight';
import { TDSNoAnimationModule } from 'tds-ui/core/no-animation';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSTreeDropIndicatorComponent } from './tree-drop-indicator.component';
import { TDSTreeIndentComponent } from './tree-indent.component';
import { TDSTreeNodeBuiltinCheckboxComponent } from './tree-node-checkbox.component';
import { TDSTreeNodeSwitcherComponent } from './tree-node-switcher.component';
import { TDSTreeNodeTitleComponent } from './tree-node-title.component';
import { TDSTreeNodeBuiltinComponent } from './tree-node.component';
import { TDSTreeComponent } from './tree.component';
import * as i0 from "@angular/core";
export class TDSTreeModule {
}
TDSTreeModule.ɵfac = function TDSTreeModule_Factory(t) { return new (t || TDSTreeModule)(); };
TDSTreeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSTreeModule });
TDSTreeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            TDSOutletModule,
            TDSNoAnimationModule,
            TDSHighlightModule,
            ScrollingModule,
            TDSSpinnerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTreeModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    TDSOutletModule,
                    TDSNoAnimationModule,
                    TDSHighlightModule,
                    ScrollingModule,
                    TDSSpinnerModule
                ],
                declarations: [
                    TDSTreeComponent,
                    TDSTreeNodeBuiltinComponent,
                    TDSTreeIndentComponent,
                    TDSTreeNodeSwitcherComponent,
                    TDSTreeNodeBuiltinCheckboxComponent,
                    TDSTreeNodeTitleComponent,
                    TDSTreeDropIndicatorComponent
                ],
                exports: [TDSTreeComponent, TDSTreeNodeBuiltinComponent, TDSTreeIndentComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSTreeModule, { declarations: [TDSTreeComponent,
        TDSTreeNodeBuiltinComponent,
        TDSTreeIndentComponent,
        TDSTreeNodeSwitcherComponent,
        TDSTreeNodeBuiltinCheckboxComponent,
        TDSTreeNodeTitleComponent,
        TDSTreeDropIndicatorComponent], imports: [BidiModule,
        CommonModule,
        TDSOutletModule,
        TDSNoAnimationModule,
        TDSHighlightModule,
        ScrollingModule,
        TDSSpinnerModule], exports: [TDSTreeComponent, TDSTreeNodeBuiltinComponent, TDSTreeIndentComponent] }); })();
