// tds-ui/esm2020/tree/tree-node-switcher.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/core/outlet";
import * as i3 from "tds-ui/progress-spinner";
const _c0 = function (a0, a1) { return { "tdsi-arrow-down-fill": a0, "tdsi-arrow-right-fill": a1 }; };
function TDSTreeNodeSwitcherComponent_ng_container_0_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "span", 4);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(1, _c0, !ctx_r5.selectMode && ctx_r5.isSwitcherOpen, !(!ctx_r5.selectMode && ctx_r5.isSwitcherOpen)));
} }
const _c1 = function (a0, a1) { return { $implicit: a0, origin: a1 }; };
function TDSTreeNodeSwitcherComponent_ng_container_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTreeNodeSwitcherComponent_ng_container_0_ng_container_1_ng_container_1_Template, 2, 4, "ng-container", 3);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r4.expandedIcon)("tdsStringTemplateOutletContext", i0.ɵɵpureFunction2(2, _c1, ctx_r4.context, ctx_r4.context.origin));
} }
function TDSTreeNodeSwitcherComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTreeNodeSwitcherComponent_ng_container_0_ng_container_1_Template, 2, 5, "ng-container", 2);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    const _r2 = i0.ɵɵreference(3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r0.isLoading)("ngIfElse", _r2);
} }
function TDSTreeNodeSwitcherComponent_ng_container_1_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTreeNodeSwitcherComponent_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTreeNodeSwitcherComponent_ng_container_1_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 3);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r6.expandedIcon)("tdsStringTemplateOutletContext", i0.ɵɵpureFunction2(2, _c1, ctx_r6.context, ctx_r6.context.origin));
} }
function TDSTreeNodeSwitcherComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTreeNodeSwitcherComponent_ng_container_1_ng_container_1_Template, 2, 5, "ng-container", 2);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    const _r2 = i0.ɵɵreference(3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r1.isLoading)("ngIfElse", _r2);
} }
function TDSTreeNodeSwitcherComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-spinner", 5);
} }
export class TDSTreeNodeSwitcherComponent {
    constructor() {
        this.selectMode = false;
    }
    get isShowLineIcon() {
        return !this.isLeaf && !!this.showLine;
    }
    get isShowSwitchIcon() {
        return !this.isLeaf && !this.showLine;
    }
    get isSwitcherOpen() {
        return !!this.isExpanded && !this.isLeaf;
    }
    get isSwitcherClose() {
        return !this.isExpanded && !this.isLeaf;
    }
}
TDSTreeNodeSwitcherComponent.ɵfac = function TDSTreeNodeSwitcherComponent_Factory(t) { return new (t || TDSTreeNodeSwitcherComponent)(); };
TDSTreeNodeSwitcherComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTreeNodeSwitcherComponent, selectors: [["tds-tree-node-switcher"]], hostAttrs: [1, "flex", "items-center"], hostVars: 16, hostBindings: function TDSTreeNodeSwitcherComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-select-tree-switcher", ctx.selectMode)("tds-select-tree-switcher-noop", ctx.selectMode && ctx.isLeaf)("tds-select-tree-switcher_open", ctx.selectMode && ctx.isSwitcherOpen)("tds-select-tree-switcher_close", ctx.selectMode && ctx.isSwitcherClose)("tds-tree-switcher", !ctx.selectMode)("tds-tree-switcher-noop", !ctx.selectMode && ctx.isLeaf)("tds-tree-switcher_open", !ctx.selectMode && ctx.isSwitcherOpen)("tds-tree-switcher_close", !ctx.selectMode && ctx.isSwitcherClose);
    } }, inputs: { showExpand: "showExpand", showLine: "showLine", expandedIcon: "expandedIcon", selectMode: "selectMode", context: "context", isLeaf: "isLeaf", isLoading: "isLoading", isExpanded: "isExpanded" }, decls: 4, vars: 2, consts: [[4, "ngIf"], ["loadingTemplate", ""], [4, "ngIf", "ngIfElse"], [4, "tdsStringTemplateOutlet", "tdsStringTemplateOutletContext"], [1, "text-neutral-1-900", "text-xl", "leading-none", "mr-2", "cursor-pointer", 3, "ngClass"], [1, "mr-2", "w-4", "h-4"]], template: function TDSTreeNodeSwitcherComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSTreeNodeSwitcherComponent_ng_container_0_Template, 2, 2, "ng-container", 0);
        i0.ɵɵtemplate(1, TDSTreeNodeSwitcherComponent_ng_container_1_Template, 2, 2, "ng-container", 0);
        i0.ɵɵtemplate(2, TDSTreeNodeSwitcherComponent_ng_template_2_Template, 1, 0, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.isShowSwitchIcon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.showLine);
    } }, directives: [i1.NgIf, i2.TDSStringTemplateOutletDirective, i1.NgClass, i3.TDSProgressSpinnerComponent], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTreeNodeSwitcherComponent, [{
        type: Component,
        args: [{ selector: 'tds-tree-node-switcher', changeDetection: ChangeDetectionStrategy.OnPush, preserveWhitespaces: false, host: {
                    '[class.tds-select-tree-switcher]': 'selectMode',
                    '[class.tds-select-tree-switcher-noop]': 'selectMode && isLeaf',
                    '[class.tds-select-tree-switcher_open]': 'selectMode && isSwitcherOpen',
                    '[class.tds-select-tree-switcher_close]': 'selectMode && isSwitcherClose',
                    '[class.tds-tree-switcher]': '!selectMode',
                    '[class.tds-tree-switcher-noop]': '!selectMode && isLeaf',
                    '[class.tds-tree-switcher_open]': '!selectMode && isSwitcherOpen',
                    '[class.tds-tree-switcher_close]': '!selectMode && isSwitcherClose',
                    'class': 'flex items-center'
                }, template: "<ng-container *ngIf=\"isShowSwitchIcon\">\r\n    <ng-container *ngIf=\"!isLoading; else loadingTemplate\">\r\n      <ng-container *tdsStringTemplateOutlet=\"expandedIcon; context: { $implicit: context, origin: context.origin }\">\r\n        <!-- <i\r\n          nz-icon\r\n          nzType=\"caret-down\"\r\n          [class.tds-select-tree-switcher-icon]=\"selectMode\"\r\n          [class.tds-tree-switcher-icon]=\"!selectMode\"\r\n        ></i> -->\r\n        <span class=\"text-neutral-1-900 text-xl leading-none mr-2 cursor-pointer\" [ngClass]=\"{'tdsi-arrow-down-fill':!selectMode && isSwitcherOpen,'tdsi-arrow-right-fill':!(!selectMode && isSwitcherOpen)}\"></span>\r\n      </ng-container>\r\n    </ng-container>\r\n  </ng-container>\r\n  <ng-container *ngIf=\"showLine\">\r\n    <ng-container *ngIf=\"!isLoading; else loadingTemplate\">\r\n      <ng-container *tdsStringTemplateOutlet=\"expandedIcon; context: { $implicit: context, origin: context.origin }\">\r\n        <!-- <i\r\n          *ngIf=\"isShowLineIcon\"\r\n          nz-icon\r\n          [nzType]=\"isSwitcherOpen ? 'minus-square' : 'plus-square'\"\r\n          class=\"tds-tree-switcher-line-icon\"\r\n        ></i>\r\n        <i *ngIf=\"!isShowLineIcon\" nz-icon nzType=\"file\" class=\"tds-tree-switcher-line-icon\"></i> -->\r\n      </ng-container>\r\n    </ng-container>\r\n  </ng-container>\r\n  <ng-template #loadingTemplate>\r\n    <tds-spinner class=\"mr-2 w-4 h-4\">                          \r\n    </tds-spinner>\r\n    \r\n  </ng-template>" }]
    }], null, { showExpand: [{
            type: Input
        }], showLine: [{
            type: Input
        }], expandedIcon: [{
            type: Input
        }], selectMode: [{
            type: Input
        }], context: [{
            type: Input
        }], isLeaf: [{
            type: Input
        }], isLoading: [{
            type: Input
        }], isExpanded: [{
            type: Input
        }] }); })();
