// tds-ui/esm2020/tree/tree-indent.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function TDSTreeIndentComponent_span_0_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 3);
    i0.ɵɵtext(1, " \u00A0 ");
    i0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { "inline-block w-7": a0 }; };
function TDSTreeIndentComponent_span_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 1);
    i0.ɵɵtemplate(1, TDSTreeIndentComponent_span_0_span_1_Template, 2, 0, "span", 2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r2 = ctx.index;
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("tds-tree-indent-unit", !ctx_r0.selectMode)("tds-select-tree-indent-unit", ctx_r0.selectMode)("tds-select-tree-indent-unit-start", ctx_r0.selectMode && ctx_r0.isStart[i_r2])("tds-tree-indent-unit-start", !ctx_r0.selectMode && ctx_r0.isStart[i_r2])("tds-select-tree-indent-unit-end", ctx_r0.selectMode && ctx_r0.isEnd[i_r2])("tds-tree-indent-unit-end", !ctx_r0.selectMode && ctx_r0.isEnd[i_r2]);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(14, _c0, !ctx_r0.selectMode));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !(!ctx_r0.selectMode && ctx_r0.isEnd[i_r2]));
} }
export class TDSTreeIndentComponent {
    constructor() {
        this.treeLevel = 0;
        this.isStart = [];
        this.isEnd = [];
        this.selectMode = false;
        this.listOfUnit = [];
    }
    ngOnChanges(changes) {
        const { treeLevel } = changes;
        if (treeLevel) {
            this.listOfUnit = [...new Array(treeLevel.currentValue || 0)];
        }
    }
}
TDSTreeIndentComponent.ɵfac = function TDSTreeIndentComponent_Factory(t) { return new (t || TDSTreeIndentComponent)(); };
TDSTreeIndentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTreeIndentComponent, selectors: [["tds-tree-indent"]], hostAttrs: [1, "items-stretch", "whitespace-nowrap", "select-none", "flex"], hostVars: 5, hostBindings: function TDSTreeIndentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("aria-hidden", true);
        i0.ɵɵclassProp("tds-tree-indent", !ctx.selectMode)("tds-select-tree-indent", ctx.selectMode);
    } }, inputs: { treeLevel: "treeLevel", isStart: "isStart", isEnd: "isEnd", selectMode: "selectMode" }, exportAs: ["tdsTreeIndent"], features: [i0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [["class", "relative ", 3, "tds-tree-indent-unit", "tds-select-tree-indent-unit", "tds-select-tree-indent-unit-start", "tds-tree-indent-unit-start", "tds-select-tree-indent-unit-end", "tds-tree-indent-unit-end", "ngClass", 4, "ngFor", "ngForOf"], [1, "relative", 3, "ngClass"], ["class", "absolute inset-0  -inset-y-1.5 border-l border-neutral-2-200", "style", "left:10px", 4, "ngIf"], [1, "absolute", "inset-0", "-inset-y-1.5", "border-l", "border-neutral-2-200", 2, "left", "10px"]], template: function TDSTreeIndentComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSTreeIndentComponent_span_0_Template, 2, 16, "span", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("ngForOf", ctx.listOfUnit);
    } }, directives: [i1.NgForOf, i1.NgClass, i1.NgIf], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTreeIndentComponent, [{
        type: Component,
        args: [{
                selector: 'tds-tree-indent',
                exportAs: 'tdsTreeIndent',
                template: `
    <span      
      [class.tds-tree-indent-unit]="!selectMode"
      [class.tds-select-tree-indent-unit]="selectMode"
      [class.tds-select-tree-indent-unit-start]="selectMode && isStart[i]"
      [class.tds-tree-indent-unit-start]="!selectMode && isStart[i]"
      [class.tds-select-tree-indent-unit-end]="selectMode && isEnd[i]"
      [class.tds-tree-indent-unit-end]="!selectMode && isEnd[i]"
      [ngClass] ="{'inline-block w-7':!selectMode}"
      class="relative "
      *ngFor="let _ of listOfUnit; let i = index"
    >    
      <span class="absolute inset-0  -inset-y-1.5 border-l border-neutral-2-200" style="left:10px" *ngIf="!(!selectMode && isEnd[i])">
        &nbsp;        
      </span>
    </span>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                host: {
                    '[attr.aria-hidden]': 'true',
                    '[class.tds-tree-indent]': '!selectMode',
                    '[class.tds-select-tree-indent]': 'selectMode',
                    'class': "items-stretch whitespace-nowrap select-none flex"
                }
            }]
    }], null, { treeLevel: [{
            type: Input
        }], isStart: [{
            type: Input
        }], isEnd: [{
            type: Input
        }], selectMode: [{
            type: Input
        }] }); })();
