// tds-ui/esm2020/tree/tree-drop-indicator.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSTreeDropIndicatorComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.level = 1;
        this.direction = 'ltr';
        this.style = {};
    }
    ngOnChanges(_changes) {
        this.renderIndicator(this.dropPosition, this.direction);
    }
    renderIndicator(dropPosition, direction = 'ltr') {
        const offset = 4;
        const startPosition = direction === 'ltr' ? 'left' : 'right';
        const endPosition = direction === 'ltr' ? 'right' : 'left';
        const style = {
            [startPosition]: `${offset}px`,
            [endPosition]: '0px'
        };
        switch (dropPosition) {
            case -1:
                style.top = `${-3}px`;
                break;
            case 1:
                style.bottom = `${-3}px`;
                break;
            case 0:
                // dropPosition === 0
                style.bottom = `${-3}px`;
                style[startPosition] = `${offset + 24}px`;
                break;
            default:
                style.display = 'none';
                break;
        }
        this.style = style;
        this.cdr.markForCheck();
    }
}
TDSTreeDropIndicatorComponent.ɵfac = function TDSTreeDropIndicatorComponent_Factory(t) { return new (t || TDSTreeDropIndicatorComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSTreeDropIndicatorComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTreeDropIndicatorComponent, selectors: [["tds-tree-drop-indicator"]], hostVars: 4, hostBindings: function TDSTreeDropIndicatorComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵstyleMap(ctx.style);
        i0.ɵɵclassProp("tds-tree-drop-indicator", true);
    } }, inputs: { dropPosition: "dropPosition", level: "level", direction: "direction" }, exportAs: ["TDSTreeDropIndicator"], features: [i0.ɵɵNgOnChangesFeature], decls: 0, vars: 0, template: function TDSTreeDropIndicatorComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTreeDropIndicatorComponent, [{
        type: Component,
        args: [{
                selector: 'tds-tree-drop-indicator',
                exportAs: 'TDSTreeDropIndicator',
                template: ``,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                host: {
                    '[class.tds-tree-drop-indicator]': 'true',
                    '[style]': 'style'
                }
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { dropPosition: [{
            type: Input
        }], level: [{
            type: Input
        }], direction: [{
            type: Input
        }] }); })();
