// tds-ui/esm2020/empty/partial/simple.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSEmptySimpleComponent {
}
TDSEmptySimpleComponent.ɵfac = function TDSEmptySimpleComponent_Factory(t) { return new (t || TDSEmptySimpleComponent)(); };
TDSEmptySimpleComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSEmptySimpleComponent, selectors: [["tds-empty-simple"]], exportAs: ["tdsEmptySimple"], decls: 6, vars: 0, consts: [["width", "64", "height", "41", "viewBox", "0 0 64 41", "xmlns", "http://www.w3.org/2000/svg", 1, "tds-empty-img-simple"], ["transform", "translate(0 1)", "fill", "none", "fill-rule", "evenodd"], ["cx", "32", "cy", "33", "rx", "32", "ry", "7", 1, "tds-empty-img-simple-ellipse", "fill-current", "text-gray-200"], ["fill-rule", "nonzero", 1, "tds-empty-img-simple-g", "stroke-current", "text-gray-400"], ["d", "M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"], ["d", "M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z", 1, "tds-empty-img-simple-path", 2, "fill", "#fafafa"]], template: function TDSEmptySimpleComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵnamespaceSVG();
        i0.ɵɵelementStart(0, "svg", 0)(1, "g", 1);
        i0.ɵɵelement(2, "ellipse", 2);
        i0.ɵɵelementStart(3, "g", 3);
        i0.ɵɵelement(4, "path", 4)(5, "path", 5);
        i0.ɵɵelementEnd()()();
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptySimpleComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-empty-simple',
                exportAs: 'tdsEmptySimple',
                template: `
    <svg class="tds-empty-img-simple" width="64" height="41" viewBox="0 0 64 41" xmlns="http://www.w3.org/2000/svg">
      <g transform="translate(0 1)" fill="none" fill-rule="evenodd">
        <ellipse class="tds-empty-img-simple-ellipse fill-current text-gray-200" cx="32" cy="33" rx="32" ry="7" />
        <g class="tds-empty-img-simple-g stroke-current text-gray-400" fill-rule="nonzero">
          <path
            d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
          />
          <path
            d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
            class="tds-empty-img-simple-path" style="fill: #fafafa;"
          />
        </g>
      </g>
    </svg>
  `
            }]
    }], null, null); })();
