// tds-ui/esm2020/affix/affix.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSAffixComponent } from './affix.component';
import * as i0 from "@angular/core";
export class TDSAffixModule {
}
TDSAffixModule.ɵfac = function TDSAffixModule_Factory(t) { return new (t || TDSAffixModule)(); };
TDSAffixModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSAffixModule });
TDSAffixModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PlatformModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAffixModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSAffixComponent],
                exports: [TDSAffixComponent],
                imports: [BidiModule, CommonModule, PlatformModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSAffixModule, { declarations: [TDSAffixComponent], imports: [BidiModule, CommonModule, PlatformModule], exports: [TDSAffixComponent] }); })();
