// tds-ui/esm2020/anchor/anchor-link.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, Input, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./anchor.component";
import * as i2 from "@angular/cdk/platform";
import * as i3 from "@angular/common";
const _c0 = ["tdsTemplate"];
const _c1 = ["linkTitle"];
function TDSAnchorLinkComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r1.titleStr);
} }
const _c2 = function (a0) { return { "text-body-2 font-semibold text-neutral-1-900 dark:text-d-neutral-1-900": a0 }; };
const _c3 = ["*"];
export class TDSAnchorLinkComponent {
    constructor(elementRef, anchorComp, platform, renderer) {
        this.elementRef = elementRef;
        this.anchorComp = anchorComp;
        this.platform = platform;
        this.renderer = renderer;
        this.tdsHref = '#';
        this.titleStr = '';
        this.isGroup = false;
        this.renderer.addClass(elementRef.nativeElement, 'tds-anchor-link');
    }
    set tdsTitle(value) {
        if (value instanceof TemplateRef) {
            this.titleStr = null;
            this.titleTpl = value;
        }
        else {
            this.titleStr = value;
        }
    }
    ngOnInit() {
        this.anchorComp.registerLink(this);
    }
    getLinkTitleElement() {
        return this.linkTitle.nativeElement;
    }
    setActive() {
        this.renderer.addClass(this.elementRef.nativeElement, 'tds-anchor-link-active');
        this.renderer.addClass(this.elementRef.nativeElement, 'text-primary-1');
        this.renderer.removeClass(this.elementRef.nativeElement, 'text-neutral-1-500');
        this.renderer.addClass(this.elementRef.nativeElement, 'border-primary-1');
        this.renderer.addClass(this.elementRef.nativeElement, 'border-l-2');
        this.renderer.addClass(this.elementRef.nativeElement, 'ml-[-2px]');
        // this.renderer.removeClass(this.elementRef.nativeElement,'border-neutral-2-100');
        // this.renderer.removeClass(this.elementRef.nativeElement,'border-l');
    }
    unsetActive() {
        this.renderer.removeClass(this.elementRef.nativeElement, 'tds-anchor-link-active');
        this.renderer.addClass(this.elementRef.nativeElement, 'text-neutral-1-500');
        this.renderer.removeClass(this.elementRef.nativeElement, 'text-primary-1');
        // this.renderer.addClass(this.elementRef.nativeElement,'border-neutral-2-100');
        // this.renderer.addClass(this.elementRef.nativeElement,'border-l');
        this.renderer.removeClass(this.elementRef.nativeElement, 'border-primary-1');
        this.renderer.removeClass(this.elementRef.nativeElement, 'border-l-2');
        this.renderer.removeClass(this.elementRef.nativeElement, 'ml-[-2px]');
    }
    goToClick(e) {
        e.preventDefault();
        e.stopPropagation();
        if (this.platform.isBrowser) {
            this.anchorComp.handleScrollTo(this);
        }
    }
    ngOnDestroy() {
        this.anchorComp.unregisterLink(this);
    }
}
TDSAnchorLinkComponent.ɵfac = function TDSAnchorLinkComponent_Factory(t) { return new (t || TDSAnchorLinkComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.TDSAnchorComponent), i0.ɵɵdirectiveInject(i2.Platform), i0.ɵɵdirectiveInject(i0.Renderer2)); };
TDSAnchorLinkComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSAnchorLinkComponent, selectors: [["tds-link"]], contentQueries: function TDSAnchorLinkComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, _c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.tdsTemplate = _t.first);
    } }, viewQuery: function TDSAnchorLinkComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.linkTitle = _t.first);
    } }, hostAttrs: [1, "flex", "flex-col", "pl-4", "py-1.5", "pr-2", "text-caption-1", "font-regular", "box-border"], inputs: { tdsHref: "tdsHref", tdsTarget: "tdsTarget", tdsTitle: "tdsTitle", isGroup: "isGroup" }, exportAs: ["tdsLink"], ngContentSelectors: _c3, decls: 4, vars: 8, consts: [[1, "tds-anchor-link-title", "hover:text-primary-1", "dark:hover:text-primary-1", 3, "href", "title", "target", "ngClass", "click"], ["linkTitle", ""], [4, "ngIf", "ngIfElse"]], template: function TDSAnchorLinkComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "a", 0, 1);
        i0.ɵɵlistener("click", function TDSAnchorLinkComponent_Template_a_click_0_listener($event) { return ctx.goToClick($event); });
        i0.ɵɵtemplate(2, TDSAnchorLinkComponent_span_2_Template, 2, 1, "span", 2);
        i0.ɵɵelementEnd();
        i0.ɵɵprojection(3);
    } if (rf & 2) {
        i0.ɵɵproperty("href", ctx.tdsHref, i0.ɵɵsanitizeUrl)("title", ctx.titleStr)("target", ctx.tdsTarget)("ngClass", i0.ɵɵpureFunction1(6, _c2, ctx.isGroup));
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.titleStr)("ngIfElse", ctx.titleTpl || ctx.tdsTemplate);
    } }, directives: [i3.NgClass, i3.NgIf], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAnchorLinkComponent, [{
        type: Component,
        args: [{
                selector: 'tds-link',
                exportAs: 'tdsLink',
                preserveWhitespaces: false,
                template: `
    <a #linkTitle 
    (click)="goToClick($event)"
     [href]="tdsHref"
      class="tds-anchor-link-title hover:text-primary-1 dark:hover:text-primary-1"
       [title]="titleStr"
       [target]="tdsTarget"
        [ngClass]="{'text-body-2 font-semibold text-neutral-1-900 dark:text-d-neutral-1-900':isGroup}" >
      <span *ngIf="titleStr; else titleTpl || tdsTemplate">{{ titleStr }}</span>
    </a>
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    class: "flex flex-col pl-4 py-1.5 pr-2  text-caption-1 font-regular box-border",
                }
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.TDSAnchorComponent }, { type: i2.Platform }, { type: i0.Renderer2 }]; }, { tdsHref: [{
            type: Input
        }], tdsTarget: [{
            type: Input
        }], tdsTitle: [{
            type: Input
        }], isGroup: [{
            type: Input
        }], tdsTemplate: [{
            type: ContentChild,
            args: ['tdsTemplate', { static: false }]
        }], linkTitle: [{
            type: ViewChild,
            args: ['linkTitle']
        }] }); })();
