// tds-ui/esm2020/anchor/anchor.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSAffixModule } from 'tds-ui/affix';
import { TDSAnchorLinkComponent } from './anchor-link.component';
import { TDSAnchorComponent } from './anchor.component';
import * as i0 from "@angular/core";
export class TDSAnchorModule {
}
TDSAnchorModule.ɵfac = function TDSAnchorModule_Factory(t) { return new (t || TDSAnchorModule)(); };
TDSAnchorModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSAnchorModule });
TDSAnchorModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSAffixModule, PlatformModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAnchorModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSAnchorComponent, TDSAnchorLinkComponent],
                exports: [TDSAnchorComponent, TDSAnchorLinkComponent],
                imports: [BidiModule, CommonModule, TDSAffixModule, PlatformModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSAnchorModule, { declarations: [TDSAnchorComponent, TDSAnchorLinkComponent], imports: [BidiModule, CommonModule, TDSAffixModule, PlatformModule], exports: [TDSAnchorComponent, TDSAnchorLinkComponent] }); })();
