// tds-ui/esm2020/tooltip/tooltip.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSOverlayModule } from 'tds-ui/core/overlay';
import { TDSToolTipComponent, TDSTooltipDirective } from './tooltip';
import * as i0 from "@angular/core";
export class TDSToolTipModule {
}
TDSToolTipModule.ɵfac = function TDSToolTipModule_Factory(t) { return new (t || TDSToolTipModule)(); };
TDSToolTipModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSToolTipModule });
TDSToolTipModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSOverlayModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSToolTipModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSToolTipComponent, TDSTooltipDirective],
                exports: [TDSToolTipComponent, TDSTooltipDirective],
                imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSOverlayModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSToolTipModule, { declarations: [TDSToolTipComponent, TDSTooltipDirective], imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSOverlayModule], exports: [TDSToolTipComponent, TDSTooltipDirective] }); })();
