// tds-ui/esm2020/shared/utility/helper-object.mjs
import { TDSHelperArray } from "./helper-array";
// @dynamic
export class TDSHelperObject {
    /**
      * Kiểm tra đối tượng có giá trị hay ko
      * @param value
      *
      */
    static hasValue(value) {
        return !(value === undefined || value === null);
    }
    /**
   * sao chép đối tượng
   * @param obj
   *
   */
    static cloneObject(obj) {
        if (this.hasValue(obj)) {
            return JSON.parse(JSON.stringify(obj));
        }
        else {
            return obj;
        }
    }
    static copyObject(taget, source) {
        taget = Object.assign(taget, source);
    }
    static isObject(k, c) {
        return !!k && "object" === typeof k && (!c || !TDSHelperArray.isArray(k));
    }
    /** Coerces a data-bound value (typically a string) to a boolean. */
    static coerceBooleanProperty(value) {
        return value != null && `${value}` !== 'false';
    }
    /**
     * kiểm tra đối tượng là function
     * @param fun
     * return true/false
     */
    static isFunction(fun) {
        return typeof (fun) === 'function';
    }
}
