// tds-ui/esm2020/tds-input/disabled-control.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
export class TDSDisabledControlDirective {
    constructor(ngControl) {
        this.ngControl = ngControl;
    }
    set disabledControl(state) {
        const action = state ? 'disable' : 'enable';
        this.ngControl.control[action]();
    }
}
TDSDisabledControlDirective.ɵfac = function TDSDisabledControlDirective_Factory(t) { return new (t || TDSDisabledControlDirective)(i0.ɵɵdirectiveInject(i1.NgControl)); };
TDSDisabledControlDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSDisabledControlDirective, selectors: [["", "formControlName", ""], ["", "formControl", "", "disabledControl", ""]], inputs: { disabledControl: "disabledControl" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDisabledControlDirective, [{
        type: Directive,
        args: [{
                selector: '([formControlName], [formControl])[disabledControl]',
            }]
    }], function () { return [{ type: i1.NgControl }]; }, { disabledControl: [{
            type: Input
        }] }); })();
