// tds-ui/esm2020/tds-input/input-module.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from "@angular/core";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { TDSFormFieldModule } from "tds-ui/form-field";
import { ErrorStateMatcher } from "tds-ui/shared/common";
import { TDSAutosizeDirective } from "./autosize.directive";
import { TDSDisabledControlDirective } from "./disabled-control";
import { TDSInput } from "./input";
import { TDSInputDebounceDirective } from "./input-debounce.directive";
import * as i0 from "@angular/core";
export class TDSInputModule {
}
TDSInputModule.ɵfac = function TDSInputModule_Factory(t) { return new (t || TDSInputModule)(); };
TDSInputModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSInputModule });
TDSInputModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [ErrorStateMatcher], imports: [[
            ReactiveFormsModule,
            FormsModule,
            TDSFormFieldModule,
        ], TDSFormFieldModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSInputModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSInput,
                    TDSDisabledControlDirective,
                    TDSAutosizeDirective,
                    TDSInputDebounceDirective
                ],
                imports: [
                    ReactiveFormsModule,
                    FormsModule,
                    TDSFormFieldModule,
                ],
                exports: [
                    TDSFormFieldModule,
                    TDSInput,
                    TDSDisabledControlDirective,
                    TDSAutosizeDirective,
                    TDSInputDebounceDirective
                ],
                providers: [ErrorStateMatcher],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSInputModule, { declarations: [TDSInput,
        TDSDisabledControlDirective,
        TDSAutosizeDirective,
        TDSInputDebounceDirective], imports: [ReactiveFormsModule,
        FormsModule,
        TDSFormFieldModule], exports: [TDSFormFieldModule,
        TDSInput,
        TDSDisabledControlDirective,
        TDSAutosizeDirective,
        TDSInputDebounceDirective] }); })();
