// tds-ui/esm2020/tds-input/input-debounce.directive.mjs
import { Directive, EventEmitter, Input, Output } from "@angular/core";
import { fromEvent, Subject } from "rxjs";
import { debounceTime, distinctUntilChanged, takeUntil } from "rxjs/operators";
import * as i0 from "@angular/core";
export class TDSInputDebounceDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.tdsInputDebounce = 300;
        this.inputKeyup = new EventEmitter();
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        const eventStream = fromEvent(this.elementRef.nativeElement, 'keyup').pipe(debounceTime(this.tdsInputDebounce), distinctUntilChanged());
        eventStream.pipe(takeUntil(this.destroy$)).subscribe(event => this.inputKeyup.emit({
            value: this.elementRef.nativeElement.value,
            keyupEvent: event
        }));
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSInputDebounceDirective.ɵfac = function TDSInputDebounceDirective_Factory(t) { return new (t || TDSInputDebounceDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSInputDebounceDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSInputDebounceDirective, selectors: [["", "tdsInputDebounce", ""]], inputs: { tdsInputDebounce: "tdsInputDebounce" }, outputs: { inputKeyup: "inputKeyup" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSInputDebounceDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsInputDebounce]'
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { tdsInputDebounce: [{
            type: Input
        }], inputKeyup: [{
            type: Output
        }] }); })();
