// tds-ui/esm2020/modal/modal-title.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./modal-types";
import * as i2 from "tds-ui/core/outlet";
const _c0 = ["tds-modal-title", ""];
function TDSModalTitleComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "div", 1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("innerHTML", ctx_r0.config.title, i0.ɵɵsanitizeHtml);
} }
export class TDSModalTitleComponent {
    constructor(config) {
        this.config = config;
    }
}
TDSModalTitleComponent.ɵfac = function TDSModalTitleComponent_Factory(t) { return new (t || TDSModalTitleComponent)(i0.ɵɵdirectiveInject(i1.ModalOptions)); };
TDSModalTitleComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSModalTitleComponent, selectors: [["div", "tds-modal-title", ""]], hostAttrs: [1, "tds-modal-header"], exportAs: ["tdsModalTitleBuiltin"], attrs: _c0, decls: 1, vars: 1, consts: [[4, "tdsStringTemplateOutlet"], [1, "text-title-1", "text-neutral-1-900", "dark:text-d-neutral-1-900", "font-semibold", "break-words", 3, "innerHTML"]], template: function TDSModalTitleComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSModalTitleComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.config.title);
    } }, directives: [i2.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModalTitleComponent, [{
        type: Component,
        args: [{
                selector: 'div[tds-modal-title]',
                exportAs: 'tdsModalTitleBuiltin',
                template: `
    <ng-container *tdsStringTemplateOutlet="config.title">
        <div [innerHTML]="config.title" class="text-title-1 text-neutral-1-900 dark:text-d-neutral-1-900 font-semibold break-words"></div>
      </ng-container>
  `,
                host: {
                    class: 'tds-modal-header'
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: i1.ModalOptions }]; }, null); })();
