// tds-ui/esm2020/modal/modal-close.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./modal-types";
import * as i2 from "tds-ui/core/outlet";
import * as i3 from "tds-ui/button";
const _c0 = ["tds-modal-close", ""];
function TdsModalCloseComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "button", 2);
    i0.ɵɵelementContainerEnd();
} }
export class TdsModalCloseComponent {
    constructor(config) {
        this.config = config;
    }
}
TdsModalCloseComponent.ɵfac = function TdsModalCloseComponent_Factory(t) { return new (t || TdsModalCloseComponent)(i0.ɵɵdirectiveInject(i1.ModalOptions)); };
TdsModalCloseComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TdsModalCloseComponent, selectors: [["div", "tds-modal-close", ""]], hostAttrs: ["aria-label", "Close", 1, "tds-modal-close"], exportAs: ["tdsModalCloseBuiltin"], attrs: _c0, decls: 2, vars: 1, consts: [[1, "tds-modal-close-x", "flex"], [4, "tdsStringTemplateOutlet"], ["tds-button-close", ""]], template: function TdsModalCloseComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 0);
        i0.ɵɵtemplate(1, TdsModalCloseComponent_ng_container_1_Template, 2, 0, "ng-container", 1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.config.closeIcon);
    } }, directives: [i2.TDSStringTemplateOutletDirective, i3.TDSButtonCloseComponent], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TdsModalCloseComponent, [{
        type: Component,
        args: [{
                selector: 'div[tds-modal-close]',
                exportAs: 'tdsModalCloseBuiltin',
                template: `
    <span class="tds-modal-close-x flex">
      <ng-container *tdsStringTemplateOutlet="config.closeIcon; let closeIcon">
       <button tds-button-close></button>
      </ng-container>
    </span>
  `,
                host: {
                    class: 'tds-modal-close',
                    'aria-label': 'Close'
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: i1.ModalOptions }]; }, null); })();
