// tds-ui/esm2020/modal/modal.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSPipesModule } from 'tds-ui/core/pipes';
import { TdsModalCloseComponent } from './modal-close.component';
import { TDSModalConfirmContainerComponent } from './modal-confirm-container.component';
import { TDSModalContainerComponent } from './modal-container.component';
import { TDSModalContentDirective } from './modal-content.directive';
import { TDSModalFooterComponent } from './modal-footer.component';
import { TDSModalFooterDirective } from './modal-footer.directive';
import { TDSModalTitleComponent } from './modal-title.component';
import { TDSModalTitleDirective } from './modal-title.directive';
import { TDSModalComponent } from './modal.component';
import { TDSModalService } from './modal.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/cdk/portal";
import * as i3 from "tds-ui/core/pipes";
export class TDSModalModule {
}
TDSModalModule.ɵfac = function TDSModalModule_Factory(t) { return new (t || TDSModalModule)(); };
TDSModalModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSModalModule });
TDSModalModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [TDSModalService], imports: [[
            CommonModule,
            BidiModule,
            OverlayModule,
            TDSOutletModule,
            PortalModule,
            TDSPipesModule,
            TDSButtonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModalModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    BidiModule,
                    OverlayModule,
                    TDSOutletModule,
                    PortalModule,
                    TDSPipesModule,
                    TDSButtonModule
                ],
                exports: [TDSModalComponent, TDSModalFooterDirective, TDSModalContentDirective, TDSModalTitleDirective],
                providers: [TDSModalService],
                declarations: [
                    TDSModalComponent,
                    TDSModalFooterDirective,
                    TDSModalContentDirective,
                    TdsModalCloseComponent,
                    TDSModalFooterComponent,
                    TDSModalTitleComponent,
                    TDSModalTitleDirective,
                    TDSModalContainerComponent,
                    TDSModalConfirmContainerComponent,
                    TDSModalComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSModalModule, { declarations: [TDSModalComponent,
        TDSModalFooterDirective,
        TDSModalContentDirective,
        TdsModalCloseComponent,
        TDSModalFooterComponent,
        TDSModalTitleComponent,
        TDSModalTitleDirective,
        TDSModalContainerComponent,
        TDSModalConfirmContainerComponent,
        TDSModalComponent], imports: [CommonModule,
        BidiModule,
        OverlayModule,
        TDSOutletModule,
        PortalModule,
        TDSPipesModule,
        TDSButtonModule], exports: [TDSModalComponent, TDSModalFooterDirective, TDSModalContentDirective, TDSModalTitleDirective] }); })();
i0.ɵɵsetComponentScope(TDSModalContainerComponent, [i1.NgClass, i1.NgStyle, i1.NgIf, TdsModalCloseComponent,
    TDSModalTitleComponent, i2.CdkPortalOutlet, TDSModalFooterComponent], [i3.TDSToCssUnitPipe]);
