// tds-ui/esm2020/modal/modal-footer.directive.mjs
import { Directive, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./modal-ref";
export class TDSModalFooterDirective {
    constructor(modalRef, templateRef) {
        this.modalRef = modalRef;
        this.templateRef = templateRef;
        if (this.modalRef) {
            this.modalRef.updateConfig({
                footer: this.templateRef
            });
        }
    }
}
TDSModalFooterDirective.ɵfac = function TDSModalFooterDirective_Factory(t) { return new (t || TDSModalFooterDirective)(i0.ɵɵdirectiveInject(i1.TDSModalRef, 8), i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSModalFooterDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSModalFooterDirective, selectors: [["", "tdsModalFooter", ""]], exportAs: ["tdsModalFooter"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModalFooterDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsModalFooter]',
                exportAs: 'tdsModalFooter'
            }]
    }], function () { return [{ type: i1.TDSModalRef, decorators: [{
                type: Optional
            }] }, { type: i0.TemplateRef }]; }, null); })();
