// tds-ui/esm2020/modal/modal-container.component.mjs
import { CdkPortalOutlet } from '@angular/cdk/portal';
import { DOCUMENT } from '@angular/common';
import { ChangeDetectionStrategy, Component, Inject, Optional, ViewChild } from '@angular/core';
import { ANIMATION_MODULE_TYPE } from '@angular/platform-browser/animations';
import { TDSHelperObject } from 'tds-ui/shared/utility';
import { TDSModalAnimations } from './modal-animations';
import { BaseModalContainerComponent } from './modal-container';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/a11y";
import * as i2 from "@angular/cdk/overlay";
import * as i3 from "tds-ui/core/config";
import * as i4 from "./modal-types";
const _c0 = ["modalElement"];
function TDSModalContainerComponent_div_4_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 11);
    i0.ɵɵlistener("click", function TDSModalContainerComponent_div_4_div_1_Template_div_click_0_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(2); return ctx_r6.onCloseClick(); });
    i0.ɵɵelementEnd();
} }
function TDSModalContainerComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 8);
    i0.ɵɵtemplate(1, TDSModalContainerComponent_div_4_div_1_Template, 1, 0, "div", 9);
    i0.ɵɵelement(2, "div", 10);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.config.closable);
} }
function TDSModalContainerComponent_ng_template_6_Template(rf, ctx) { }
function TDSModalContainerComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 12);
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵproperty("innerHTML", ctx_r3.config.content, i0.ɵɵsanitizeHtml);
} }
function TDSModalContainerComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r9 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 13);
    i0.ɵɵlistener("cancelTriggered", function TDSModalContainerComponent_div_8_Template_div_cancelTriggered_0_listener() { i0.ɵɵrestoreView(_r9); const ctx_r8 = i0.ɵɵnextContext(); return ctx_r8.onCloseClick(); })("okTriggered", function TDSModalContainerComponent_div_8_Template_div_okTriggered_0_listener() { i0.ɵɵrestoreView(_r9); const ctx_r10 = i0.ɵɵnextContext(); return ctx_r10.onOkClick(); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵproperty("modalRef", ctx_r4.modalRef);
} }
const _c1 = function (a0) { return { "top-24": a0 }; };
const _c2 = function (a0) { return { "max-h-[calc(100vh-148px)] overflow-auto tds-custom-scroll": a0 }; };
export class TDSModalContainerComponent extends BaseModalContainerComponent {
    constructor(elementRef, focusTrapFactory, cdr, render, overlayRef, tdsConfigService, config, document, animationType) {
        super(elementRef, focusTrapFactory, cdr, render, overlayRef, tdsConfigService, config, document, animationType);
        this.config = config;
    }
    get width() {
        let result = 520;
        if (TDSHelperObject.hasValue(this.config)) {
            if (TDSHelperObject.hasValue(this.config.size)) {
                switch (this.config.size) {
                    case "sm":
                        result = 300;
                        break;
                    case "md":
                        result = 500;
                        break;
                    case "lg":
                        result = 800;
                        break;
                    case "xl":
                        result = 1140;
                        break;
                    default:
                        result = 500;
                        break;
                }
                return result;
            }
            else {
                return this.config?.width;
            }
        }
        return result;
    }
}
TDSModalContainerComponent.ɵfac = function TDSModalContainerComponent_Factory(t) { return new (t || TDSModalContainerComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.FocusTrapFactory), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i2.OverlayRef), i0.ɵɵdirectiveInject(i3.TDSConfigService), i0.ɵɵdirectiveInject(i4.ModalOptions), i0.ɵɵdirectiveInject(DOCUMENT, 8), i0.ɵɵdirectiveInject(ANIMATION_MODULE_TYPE, 8)); };
TDSModalContainerComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSModalContainerComponent, selectors: [["tds-modal-container"]], viewQuery: function TDSModalContainerComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(CdkPortalOutlet, 7);
        i0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.portalOutlet = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.modalElementRef = _t.first);
    } }, hostAttrs: ["tabindex", "-1", "role", "dialog"], hostVars: 12, hostBindings: function TDSModalContainerComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵsyntheticHostListener("@modalContainer.start", function TDSModalContainerComponent_animation_modalContainer_start_HostBindingHandler($event) { return ctx.onAnimationStart($event); })("@modalContainer.done", function TDSModalContainerComponent_animation_modalContainer_done_HostBindingHandler($event) { return ctx.onAnimationDone($event); });
        i0.ɵɵlistener("click", function TDSModalContainerComponent_click_HostBindingHandler($event) { return ctx.onContainerClick($event); })("mouseup", function TDSModalContainerComponent_mouseup_HostBindingHandler() { return ctx.onMouseup(); });
    } if (rf & 2) {
        i0.ɵɵsyntheticHostProperty("@.disabled", ctx.config.noAnimation)("@modalContainer", ctx.state);
        i0.ɵɵclassMap(ctx.config.wrapClassName ? "tds-modal-wrap fixed top-0 left-0 right-0 bottom-0 overflow-auto outline-none " + ctx.config.wrapClassName : "tds-modal-wrap fixed top-0 left-0 right-0 bottom-0 overflow-auto outline-none");
        i0.ɵɵstyleProp("z-index", ctx.config.zindex);
        i0.ɵɵclassProp("tds-modal-wrap-rtl", ctx.dir === "rtl")("tds-modal-centered", ctx.config.centered)("flex", ctx.config.centered);
    } }, exportAs: ["tdsModalContainer"], features: [i0.ɵɵInheritDefinitionFeature], decls: 9, vars: 18, consts: [["role", "document", 3, "ngClass", "ngStyle", "mousedown"], ["modalElement", ""], [1, "tds-modal-content", "pointer-events-auto"], ["class", "w-full p-3 pl-4 flex  flex-row-reverse items-center justify-between border-b  border-neutral-2-200 dark:border-d-neutral-2-200", 4, "ngIf"], [1, "tds-modal-body", "p-4", 3, "ngStyle", "ngClass"], ["cdkPortalOutlet", ""], [3, "innerHTML", 4, "ngIf"], ["tds-modal-footer", "", 3, "modalRef", "cancelTriggered", "okTriggered", 4, "ngIf"], [1, "w-full", "p-3", "pl-4", "flex", "flex-row-reverse", "items-center", "justify-between", "border-b", "border-neutral-2-200", "dark:border-d-neutral-2-200"], ["class", "justify-self-end pl-2", "tds-modal-close", "", 3, "click", 4, "ngIf"], ["tds-modal-title", "", 1, "flex-auto", "tds-modal-title"], ["tds-modal-close", "", 1, "justify-self-end", "pl-2", 3, "click"], [3, "innerHTML"], ["tds-modal-footer", "", 3, "modalRef", "cancelTriggered", "okTriggered"]], template: function TDSModalContainerComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0, 1);
        i0.ɵɵlistener("mousedown", function TDSModalContainerComponent_Template_div_mousedown_0_listener() { return ctx.onMousedown(); });
        i0.ɵɵpipe(2, "tdsToCssUnit");
        i0.ɵɵelementStart(3, "div", 2);
        i0.ɵɵtemplate(4, TDSModalContainerComponent_div_4_Template, 3, 1, "div", 3);
        i0.ɵɵelementStart(5, "div", 4);
        i0.ɵɵtemplate(6, TDSModalContainerComponent_ng_template_6_Template, 0, 0, "ng-template", 5);
        i0.ɵɵtemplate(7, TDSModalContainerComponent_div_7_Template, 1, 1, "div", 6);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(8, TDSModalContainerComponent_div_8_Template, 1, 1, "div", 7);
        i0.ɵɵelementEnd()();
    } if (rf & 2) {
        i0.ɵɵclassMapInterpolate1("tds-modal bg-white dark:bg-d-neutral-3-300 text-neutral-1-900 dark:text-d-neutral-1-900 rounded-md border-box pointer-events-none relative my-auto mx-auto ", ctx.config.className, "");
        i0.ɵɵstyleProp("width", i0.ɵɵpipeBind1(2, 12, ctx.width));
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(14, _c1, !ctx.config.centered))("ngStyle", ctx.config.style);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngIf", ctx.config.title);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngStyle", ctx.config.bodyStyle)("ngClass", i0.ɵɵpureFunction1(16, _c2, ctx.config.centered));
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.isStringContent);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.config.footer !== null);
    } }, encapsulation: 2, data: { animation: [TDSModalAnimations.modalContainer] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModalContainerComponent, [{
        type: Component,
        args: [{
                selector: 'tds-modal-container',
                exportAs: 'tdsModalContainer',
                template: `
    <div
      #modalElement
      role="document"
      class="tds-modal bg-white dark:bg-d-neutral-3-300 text-neutral-1-900 dark:text-d-neutral-1-900 rounded-md border-box pointer-events-none relative my-auto mx-auto {{config.className!}}"
      (mousedown)="onMousedown()"
      [ngClass]="{'top-24':!config.centered}"
      [ngStyle]="config.style!"
      [style.width]="width! | tdsToCssUnit"
    >  
      <div class="tds-modal-content pointer-events-auto" >
        <div class="w-full p-3 pl-4 flex  flex-row-reverse items-center justify-between border-b  border-neutral-2-200 dark:border-d-neutral-2-200"  *ngIf="config.title">       
        <div class="justify-self-end pl-2" *ngIf="config.closable" tds-modal-close (click)="onCloseClick()"></div>
        <div class="flex-auto tds-modal-title"  tds-modal-title></div>
        
        </div>
        
        <div class="tds-modal-body p-4" [ngStyle]="config.bodyStyle!" [ngClass]="{'max-h-[calc(100vh-148px)] overflow-auto tds-custom-scroll':config.centered}">
          <ng-template cdkPortalOutlet></ng-template>
          <div *ngIf="isStringContent" [innerHTML]="config.content"></div>
         
        </div>
        
        <div
          *ngIf="config.footer !== null"
          tds-modal-footer
          [modalRef]="modalRef"
          (cancelTriggered)="onCloseClick()"
          (okTriggered)="onOkClick()"
        ></div>
      </div>
    </div>
  `,
                animations: [TDSModalAnimations.modalContainer],
                // Using OnPush for modal caused footer can not to detect changes. we can fix it when 8.x.
                changeDetection: ChangeDetectionStrategy.Default,
                host: {
                    tabindex: '-1',
                    role: 'dialog',
                    '[class]': 'config.wrapClassName ? "tds-modal-wrap fixed top-0 left-0 right-0 bottom-0 overflow-auto outline-none " + config.wrapClassName : "tds-modal-wrap fixed top-0 left-0 right-0 bottom-0 overflow-auto outline-none"',
                    '[class.tds-modal-wrap-rtl]': `dir === 'rtl'`,
                    '[class.tds-modal-centered]': 'config.centered',
                    '[class.flex]': 'config.centered',
                    '[style.zIndex]': 'config.zindex',
                    '[@.disabled]': 'config.noAnimation',
                    '[@modalContainer]': 'state',
                    '(@modalContainer.start)': 'onAnimationStart($event)',
                    '(@modalContainer.done)': 'onAnimationDone($event)',
                    '(click)': 'onContainerClick($event)',
                    '(mouseup)': 'onMouseup()'
                }
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.FocusTrapFactory }, { type: i0.ChangeDetectorRef }, { type: i0.Renderer2 }, { type: i2.OverlayRef }, { type: i3.TDSConfigService }, { type: i4.ModalOptions }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [DOCUMENT]
            }] }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [ANIMATION_MODULE_TYPE]
            }] }]; }, { portalOutlet: [{
            type: ViewChild,
            args: [CdkPortalOutlet, { static: true }]
        }], modalElementRef: [{
            type: ViewChild,
            args: ['modalElement', { static: true }]
        }] }); })();
