// tds-ui/esm2020/modal/modal-title.directive.mjs
import { Directive, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./modal-ref";
export class TDSModalTitleDirective {
    constructor(modalRef, templateRef) {
        this.modalRef = modalRef;
        this.templateRef = templateRef;
        if (this.modalRef) {
            this.modalRef.updateConfig({
                title: this.templateRef
            });
        }
    }
}
TDSModalTitleDirective.ɵfac = function TDSModalTitleDirective_Factory(t) { return new (t || TDSModalTitleDirective)(i0.ɵɵdirectiveInject(i1.TDSModalRef, 8), i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSModalTitleDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSModalTitleDirective, selectors: [["", "tdsModalTitle", ""]], exportAs: ["tdsModalTitle"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModalTitleDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsModalTitle]',
                exportAs: 'tdsModalTitle'
            }]
    }], function () { return [{ type: i1.TDSModalRef, decorators: [{
                type: Optional
            }] }, { type: i0.TemplateRef }]; }, null); })();
