// tds-ui/esm2020/modal/modal-footer.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { Subject } from 'rxjs';
import { isPromise } from 'tds-ui/core/util';
import * as i0 from "@angular/core";
import * as i1 from "./modal-types";
import * as i2 from "@angular/common";
import * as i3 from "tds-ui/core/outlet";
import * as i4 from "tds-ui/button";
const _c0 = ["tds-modal-footer", ""];
function TDSModalFooterComponent_ng_container_0_ng_container_1_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 5);
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(3);
    i0.ɵɵproperty("innerHTML", ctx_r4.config.footer, i0.ɵɵsanitizeHtml);
} }
function TDSModalFooterComponent_ng_container_0_ng_container_1_ng_container_2_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r9 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 8);
    i0.ɵɵlistener("click", function TDSModalFooterComponent_ng_container_0_ng_container_1_ng_container_2_button_2_Template_button_click_0_listener() { const restoredCtx = i0.ɵɵrestoreView(_r9); const button_r7 = restoredCtx.$implicit; const ctx_r8 = i0.ɵɵnextContext(4); return ctx_r8.onButtonClick(button_r7); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const button_r7 = ctx.$implicit;
    const ctx_r6 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("hidden", !ctx_r6.getButtonCallableProp(button_r7, "show"))("loading", ctx_r6.getButtonCallableProp(button_r7, "loading"))("disabled", ctx_r6.getButtonCallableProp(button_r7, "disabled"))("size", button_r7.size);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", button_r7.label, " ");
} }
function TDSModalFooterComponent_ng_container_0_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "div", 6);
    i0.ɵɵtemplate(2, TDSModalFooterComponent_ng_container_0_ng_container_1_ng_container_2_button_2_Template, 2, 5, "button", 7);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r5.buttons);
} }
function TDSModalFooterComponent_ng_container_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSModalFooterComponent_ng_container_0_ng_container_1_div_1_Template, 1, 1, "div", 3);
    i0.ɵɵtemplate(2, TDSModalFooterComponent_ng_container_0_ng_container_1_ng_container_2_Template, 3, 1, "ng-container", 4);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r3.buttonsFooter);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r3.buttonsFooter);
} }
const _c1 = function (a0, a1) { return { $implicit: a0, modalRef: a1 }; };
function TDSModalFooterComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSModalFooterComponent_ng_container_0_ng_container_1_Template, 3, 2, "ng-container", 2);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r0.config.footer)("tdsStringTemplateOutletContext", i0.ɵɵpureFunction2(2, _c1, ctx_r0.config.componentParams, ctx_r0.modalRef));
} }
function TDSModalFooterComponent_ng_template_1_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r13 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 11);
    i0.ɵɵlistener("click", function TDSModalFooterComponent_ng_template_1_button_1_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r13); const ctx_r12 = i0.ɵɵnextContext(2); return ctx_r12.onCancel(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("disabled", ctx_r10.config.cancelDisabled);
    i0.ɵɵattribute("cdkFocusInitial", ctx_r10.config.autofocus === "cancel" || null);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r10.config.cancelText || ctx_r10.locale.cancelText, " ");
} }
function TDSModalFooterComponent_ng_template_1_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r15 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 12);
    i0.ɵɵlistener("click", function TDSModalFooterComponent_ng_template_1_button_2_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r15); const ctx_r14 = i0.ɵɵnextContext(2); return ctx_r14.onOk(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("disabled", ctx_r11.config.okDisabled);
    i0.ɵɵattribute("cdkFocusInitial", ctx_r11.config.autofocus === "ok" || null);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r11.config.okText || ctx_r11.locale.okText, " ");
} }
function TDSModalFooterComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 6);
    i0.ɵɵtemplate(1, TDSModalFooterComponent_ng_template_1_button_1_Template, 2, 3, "button", 9);
    i0.ɵɵtemplate(2, TDSModalFooterComponent_ng_template_1_button_2_Template, 2, 3, "button", 10);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.config.cancelText !== null);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.config.okText !== null);
} }
export class TDSModalFooterComponent {
    constructor(config) {
        this.config = config;
        this.buttonsFooter = false;
        this.buttons = [];
        this.cancelTriggered = new EventEmitter();
        this.okTriggered = new EventEmitter();
        this.destroy$ = new Subject();
        this.locale = {
            okText: "Lưu",
            cancelText: "Hủy",
        };
        if (Array.isArray(config.footer)) {
            this.buttonsFooter = true;
            this.buttons = config.footer.map(mergeDefaultOption);
        }
    }
    onCancel() {
        this.cancelTriggered.emit();
    }
    onOk() {
        this.okTriggered.emit();
    }
    /**
     * Returns the value of the specified key.
     * If it is a function, run and return the return value of the function.
     */
    getButtonCallableProp(options, prop) {
        const value = options[prop];
        const componentInstance = this.modalRef.getContentComponent();
        return typeof value === 'function' ? value.apply(options, componentInstance && [componentInstance]) : value;
    }
    /**
     * Run function based on the type and set its `loading` prop if needed.
     */
    onButtonClick(options) {
        const loading = this.getButtonCallableProp(options, 'loading');
        if (!loading) {
            const result = this.getButtonCallableProp(options, 'onClick');
            if (options.autoLoading && isPromise(result)) {
                options.loading = true;
                result.then(() => (options.loading = false)).catch(() => (options.loading = false));
            }
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSModalFooterComponent.ɵfac = function TDSModalFooterComponent_Factory(t) { return new (t || TDSModalFooterComponent)(i0.ɵɵdirectiveInject(i1.ModalOptions)); };
TDSModalFooterComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSModalFooterComponent, selectors: [["div", "tds-modal-footer", ""]], hostAttrs: [1, "tds-modal-footer"], inputs: { modalRef: "modalRef" }, outputs: { cancelTriggered: "cancelTriggered", okTriggered: "okTriggered" }, exportAs: ["tdsModalFooterBuiltin"], attrs: _c0, decls: 3, vars: 2, consts: [[4, "ngIf", "ngIfElse"], ["defaultFooterButtons", ""], [4, "tdsStringTemplateOutlet", "tdsStringTemplateOutletContext"], [3, "innerHTML", 4, "ngIf"], [4, "ngIf"], [3, "innerHTML"], [1, "w-full", "flex", "justify-end", "p-4"], ["class", "ml-2 min-w-[100px]", "tds-button", "", 3, "hidden", "loading", "disabled", "size", "click", 4, "ngFor", "ngForOf"], ["tds-button", "", 1, "ml-2", "min-w-[100px]", 3, "hidden", "loading", "disabled", "size", "click"], ["class", "ml-2", "tds-button", "", "class", "min-w-[100px]", "color", "secondary", 3, "disabled", "click", 4, "ngIf"], ["tds-button", "", "class", "ml-2 min-w-[100px]", 3, "disabled", "click", 4, "ngIf"], ["tds-button", "", "color", "secondary", 1, "min-w-[100px]", 3, "disabled", "click"], ["tds-button", "", 1, "ml-2", "min-w-[100px]", 3, "disabled", "click"]], template: function TDSModalFooterComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSModalFooterComponent_ng_container_0_Template, 2, 5, "ng-container", 0);
        i0.ɵɵtemplate(1, TDSModalFooterComponent_ng_template_1_Template, 3, 2, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = i0.ɵɵreference(2);
        i0.ɵɵproperty("ngIf", ctx.config.footer)("ngIfElse", _r1);
    } }, directives: [i2.NgIf, i3.TDSStringTemplateOutletDirective, i2.NgForOf, i4.TDSButtonComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModalFooterComponent, [{
        type: Component,
        args: [{
                selector: 'div[tds-modal-footer]',
                exportAs: 'tdsModalFooterBuiltin',
                template: `
    <ng-container *ngIf="config.footer; else defaultFooterButtons">
      <ng-container *tdsStringTemplateOutlet="config.footer; context: { $implicit: config.componentParams, modalRef: modalRef }">
        <div *ngIf="!buttonsFooter" [innerHTML]="config.footer"></div>
        <ng-container *ngIf="buttonsFooter">
        <div class="w-full flex justify-end p-4">
        <button
        class="ml-2 min-w-[100px]"
            *ngFor="let button of buttons"
            tds-button
            (click)="onButtonClick(button)"
            [hidden]="!getButtonCallableProp(button, 'show')"
            [loading]="getButtonCallableProp(button, 'loading')"
            [disabled]="getButtonCallableProp(button, 'disabled')"          
            [size]="button.size!"
          >
            {{ button.label }}
          </button>
        </div>
          
        </ng-container>
      </ng-container>
    </ng-container>
    <ng-template #defaultFooterButtons>
     <div class="w-full flex justify-end p-4">
     <button
     class="ml-2"
     *ngIf="config.cancelText !== null"
     [attr.cdkFocusInitial]="config.autofocus === 'cancel' || null"
     tds-button
     class="min-w-[100px]"
     color="secondary"
     (click)="onCancel()"      
     [disabled]="config.cancelDisabled!"
   >
     {{ config.cancelText || locale.cancelText }}
   </button>
   <button
     *ngIf="config.okText !== null"
     [attr.cdkFocusInitial]="config.autofocus === 'ok' || null"
     tds-button   
     (click)="onOk()"        
     [disabled]="config.okDisabled!"
     class="ml-2 min-w-[100px]"
   >
     {{ config.okText || locale.okText }} 
       </button>
     </div>
    </ng-template>
  `,
                host: {
                    class: 'tds-modal-footer'
                },
                changeDetection: ChangeDetectionStrategy.Default
            }]
    }], function () { return [{ type: i1.ModalOptions }]; }, { cancelTriggered: [{
            type: Output
        }], okTriggered: [{
            type: Output
        }], modalRef: [{
            type: Input
        }] }); })();
function mergeDefaultOption(options) {
    return {
        type: null,
        size: 'md',
        autoLoading: true,
        show: true,
        loading: false,
        disabled: false,
        color: 'secondary',
        ...options
    };
}
