// tds-ui/esm2020/input/input.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSAutosizeDirective } from './autosize.directive';
import { TDSInputGroupSlotComponent } from './input-group-slot.component';
import { TDSInputGroupComponent, TDSInputGroupWhitSuffixOrPrefixDirective } from './input-group.component';
import { TDSInputDirective } from './input.directive';
import { TDSTextareaCountComponent } from './textarea-count.component';
import * as i0 from "@angular/core";
export class TDSFormInputModule {
}
TDSFormInputModule.ɵfac = function TDSFormInputModule_Factory(t) { return new (t || TDSFormInputModule)(); };
TDSFormInputModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSFormInputModule });
TDSFormInputModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PlatformModule, TDSOutletModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFormInputModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSTextareaCountComponent,
                    TDSInputDirective,
                    TDSInputGroupComponent,
                    TDSAutosizeDirective,
                    TDSInputGroupSlotComponent,
                    TDSInputGroupWhitSuffixOrPrefixDirective
                ],
                exports: [
                    TDSTextareaCountComponent,
                    TDSInputDirective,
                    TDSInputGroupComponent,
                    TDSAutosizeDirective,
                    TDSInputGroupSlotComponent,
                    TDSInputGroupWhitSuffixOrPrefixDirective
                ],
                imports: [BidiModule, CommonModule, PlatformModule, TDSOutletModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSFormInputModule, { declarations: [TDSTextareaCountComponent,
        TDSInputDirective,
        TDSInputGroupComponent,
        TDSAutosizeDirective,
        TDSInputGroupSlotComponent,
        TDSInputGroupWhitSuffixOrPrefixDirective], imports: [BidiModule, CommonModule, PlatformModule, TDSOutletModule], exports: [TDSTextareaCountComponent,
        TDSInputDirective,
        TDSInputGroupComponent,
        TDSAutosizeDirective,
        TDSInputGroupSlotComponent,
        TDSInputGroupWhitSuffixOrPrefixDirective] }); })();
