// tds-ui/esm2020/input/input-group-slot.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/core/outlet";
const _c0 = ["tds-input-group-slot", ""];
function TDSInputGroupSlotComponent_i_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i");
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r0.icon);
} }
function TDSInputGroupSlotComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r1.template);
} }
export class TDSInputGroupSlotComponent {
    constructor() {
        this.icon = null;
        this.type = null;
        this.template = null;
    }
}
TDSInputGroupSlotComponent.ɵfac = function TDSInputGroupSlotComponent_Factory(t) { return new (t || TDSInputGroupSlotComponent)(); };
TDSInputGroupSlotComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSInputGroupSlotComponent, selectors: [["", "tds-input-group-slot", ""]], hostVars: 6, hostBindings: function TDSInputGroupSlotComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-input-group-addon", ctx.type === "addon")("tds-input-prefix", ctx.type === "prefix")("tds-input-suffix", ctx.type === "suffix");
    } }, inputs: { icon: "icon", type: "type", template: "template" }, attrs: _c0, decls: 2, vars: 2, consts: [[3, "class", 4, "ngIf"], [4, "tdsStringTemplateOutlet"]], template: function TDSInputGroupSlotComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSInputGroupSlotComponent_i_0_Template, 1, 2, "i", 0);
        i0.ɵɵtemplate(1, TDSInputGroupSlotComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.icon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.template);
    } }, directives: [i1.NgIf, i2.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSInputGroupSlotComponent, [{
        type: Component,
        args: [{
                selector: '[tds-input-group-slot]',
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <i  [class]="icon" *ngIf="icon"></i>
    <ng-container *tdsStringTemplateOutlet="template">{{ template }}</ng-container>
  `,
                host: {
                    '[class.tds-input-group-addon]': `type === 'addon'`,
                    '[class.tds-input-prefix]': `type === 'prefix'`,
                    '[class.tds-input-suffix]': `type === 'suffix'`
                }
            }]
    }], null, { icon: [{
            type: Input
        }], type: [{
            type: Input
        }], template: [{
            type: Input
        }] }); })();
