// tds-ui/esm2020/input/input.directive.mjs
import { __decorate } from "tslib";
import { Directive, Input, Optional, Self } from '@angular/core';
import { Subject } from 'rxjs';
import { filter, takeUntil } from 'rxjs/operators';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
import * as i2 from "@angular/cdk/bidi";
export class TDSInputDirective {
    constructor(ngControl, renderer, elementRef, directionality) {
        this.ngControl = ngControl;
        this.directionality = directionality;
        this.tdsBorderless = false;
        this.tdsSize = 'md';
        this._disabled = false;
        this.disabled$ = new Subject();
        this.dir = 'ltr';
        this.destroy$ = new Subject();
        renderer.addClass(elementRef.nativeElement, 'tds-input');
    }
    get disabled() {
        if (this.ngControl && this.ngControl.disabled !== null) {
            return this.ngControl.disabled;
        }
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = value != null && `${value}` !== 'false';
    }
    ngOnInit() {
        if (this.ngControl) {
            this.ngControl.statusChanges
                ?.pipe(filter(() => this.ngControl.disabled !== null), takeUntil(this.destroy$))
                .subscribe(() => {
                this.disabled$.next(this.ngControl.disabled);
            });
        }
        this.dir = this.directionality.value;
        this.directionality.change?.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
        });
    }
    ngOnChanges(changes) {
        const { disabled } = changes;
        if (disabled) {
            this.disabled$.next(this.disabled);
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSInputDirective.ɵfac = function TDSInputDirective_Factory(t) { return new (t || TDSInputDirective)(i0.ɵɵdirectiveInject(i1.NgControl, 10), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i2.Directionality, 8)); };
TDSInputDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSInputDirective, selectors: [["input", "tds-input", ""], ["textarea", "tds-input", ""]], hostVars: 15, hostBindings: function TDSInputDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("disabled", ctx.disabled || null);
        i0.ɵɵclassProp("tds-input-disabled", ctx.disabled)("tds-input-borderless", ctx.tdsBorderless)("tds-input-rtl", ctx.dir === "rtl")("tds-input-sm", ctx.tdsSize === "sm")("tds-input-md", ctx.tdsSize === "md")("tds-input-lg", ctx.tdsSize === "lg")("tds-input-xl", ctx.tdsSize === "xl");
    } }, inputs: { tdsBorderless: "tdsBorderless", tdsSize: "tdsSize", disabled: "disabled" }, exportAs: ["tdsinput"], features: [i0.ɵɵNgOnChangesFeature] });
__decorate([
    InputBoolean()
], TDSInputDirective.prototype, "tdsBorderless", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSInputDirective, [{
        type: Directive,
        args: [{
                selector: 'input[tds-input],textarea[tds-input]',
                exportAs: 'tdsinput',
                host: {
                    '[class.tds-input-disabled]': 'disabled',
                    '[class.tds-input-borderless]': 'tdsBorderless',
                    '[attr.disabled]': 'disabled || null',
                    '[class.tds-input-rtl]': `dir=== 'rtl'`,
                    '[class.tds-input-sm]': `tdsSize=== 'sm'`,
                    '[class.tds-input-md]': `tdsSize=== 'md'`,
                    '[class.tds-input-lg]': `tdsSize=== 'lg'`,
                    '[class.tds-input-xl]': `tdsSize=== 'xl'`
                }
            }]
    }], function () { return [{ type: i1.NgControl, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }, { type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i2.Directionality, decorators: [{
                type: Optional
            }] }]; }, { tdsBorderless: [{
            type: Input
        }], tdsSize: [{
            type: Input
        }], disabled: [{
            type: Input
        }] }); })();
