// tds-ui/esm2020/input/textarea-count.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, Input, isDevMode } from '@angular/core';
import { EMPTY, merge, Subject } from 'rxjs';
import { map, startWith, takeUntil } from 'rxjs/operators';
import { isNotNil } from 'tds-ui/core/util';
import { TDSInputDirective } from './input.directive';
import * as i0 from "@angular/core";
const _c0 = [[["textarea", "tds-input", ""]]];
const _c1 = ["textarea[tds-input]"];
export class TDSTextareaCountComponent {
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.tdsMaxCharacterCount = 0;
        this.tdsComputeCharacterCount = v => v.length;
        this.tdsFormatter = (c, m) => `${c}${m > 0 ? `/${m}` : ``}`;
        this.configChange$ = new Subject();
        this.destroy$ = new Subject();
    }
    ngAfterContentInit() {
        if (!this.TDSInputDirective && isDevMode()) {
            throw new Error('[tds-textarea-count]: Could not find matching textarea[tds-input] child.');
        }
        if (this.TDSInputDirective.ngControl) {
            const valueChanges = this.TDSInputDirective.ngControl.valueChanges || EMPTY;
            merge(valueChanges, this.configChange$)
                .pipe(takeUntil(this.destroy$), map(() => this.TDSInputDirective.ngControl.value), startWith(this.TDSInputDirective.ngControl.value))
                .subscribe(value => {
                this.setDataCount(value);
            });
        }
    }
    setDataCount(value) {
        const inputValue = isNotNil(value) ? String(value) : '';
        const currentCount = this.tdsComputeCharacterCount(inputValue);
        const dataCount = this.tdsFormatter(currentCount, this.tdsMaxCharacterCount);
        this.renderer.setAttribute(this.elementRef.nativeElement, 'data-count', dataCount);
    }
    ngOnDestroy() {
        this.configChange$.complete();
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSTextareaCountComponent.ɵfac = function TDSTextareaCountComponent_Factory(t) { return new (t || TDSTextareaCountComponent)(i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSTextareaCountComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTextareaCountComponent, selectors: [["tds-textarea-count"]], contentQueries: function TDSTextareaCountComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSInputDirective, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.TDSInputDirective = _t.first);
    } }, hostAttrs: [1, "tds-input-textarea-show-count"], inputs: { tdsMaxCharacterCount: "tdsMaxCharacterCount", tdsComputeCharacterCount: "tdsComputeCharacterCount", tdsFormatter: "tdsFormatter" }, ngContentSelectors: _c1, decls: 1, vars: 0, template: function TDSTextareaCountComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c0);
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTextareaCountComponent, [{
        type: Component,
        args: [{
                selector: 'tds-textarea-count',
                template: ` <ng-content select="textarea[tds-input]"></ng-content> `,
                host: {
                    class: 'tds-input-textarea-show-count'
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }]; }, { TDSInputDirective: [{
            type: ContentChild,
            args: [TDSInputDirective, { static: true }]
        }], tdsMaxCharacterCount: [{
            type: Input
        }], tdsComputeCharacterCount: [{
            type: Input
        }], tdsFormatter: [{
            type: Input
        }] }); })();
