// tds-ui/esm2020/card/card.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSCardComponent {
    ngOnInit() {
    }
}
TDSCardComponent.ɵfac = function TDSCardComponent_Factory(t) { return new (t || TDSCardComponent)(); };
TDSCardComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardComponent, selectors: [["tds-card"]], hostAttrs: [1, "block", "overflow-hidden", "rounded-md", "text-neutral-1-900", "dark:text-d-neutral-1-900"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSCardComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardComponent, [{
        type: Component,
        args: [{
                selector: 'tds-card',
                template: `
      <ng-content></ng-content>
  `,
                host: {
                    class: 'block overflow-hidden rounded-md   text-neutral-1-900 dark:text-d-neutral-1-900',
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
