// tds-ui/esm2020/card/card-actions.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSCardActionsComponent {
}
TDSCardActionsComponent.ɵfac = function TDSCardActionsComponent_Factory(t) { return new (t || TDSCardActionsComponent)(); };
TDSCardActionsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardActionsComponent, selectors: [["tds-card-actions"]], hostAttrs: [1, "px-4", "min-h-0", "block", "w-full"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSCardActionsComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardActionsComponent, [{
        type: Component,
        args: [{
                selector: 'tds-card-actions',
                template: `
      <ng-content></ng-content>
  `,
                host: {
                    class: 'px-4 min-h-0 block w-full',
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
