// tds-ui/esm2020/card/card.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSCardComponent } from './card.component';
import { TDSCardHeaderComponent } from './card-header.component';
import { TDSCardContentComponent } from './card-content.component';
import { TDSCardActionsComponent } from './card-actions.component';
import { TDSCardImageDirective } from './card-image.component';
import * as i0 from "@angular/core";
export class TDSCardModule {
}
TDSCardModule.ɵfac = function TDSCardModule_Factory(t) { return new (t || TDSCardModule)(); };
TDSCardModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCardModule });
TDSCardModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSCardHeaderComponent,
                    TDSCardContentComponent,
                    TDSCardActionsComponent,
                    TDSCardComponent,
                    TDSCardImageDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    TDSCardHeaderComponent,
                    TDSCardContentComponent,
                    TDSCardActionsComponent,
                    TDSCardComponent,
                    TDSCardImageDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCardModule, { declarations: [TDSCardHeaderComponent,
        TDSCardContentComponent,
        TDSCardActionsComponent,
        TDSCardComponent,
        TDSCardImageDirective], imports: [CommonModule], exports: [TDSCardHeaderComponent,
        TDSCardContentComponent,
        TDSCardActionsComponent,
        TDSCardComponent,
        TDSCardImageDirective] }); })();
