// tds-ui/esm2020/card/card-header.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSCardHeaderComponent {
}
TDSCardHeaderComponent.ɵfac = function TDSCardHeaderComponent_Factory(t) { return new (t || TDSCardHeaderComponent)(); };
TDSCardHeaderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardHeaderComponent, selectors: [["tds-card-header"]], hostAttrs: [1, "p-4", "block", "w-full", "text-header-1", "font-semibold"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSCardHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'tds-card-header',
                template: `
      <ng-content></ng-content>
  `,
                host: {
                    class: 'p-4 block w-full text-header-1  font-semibold',
                },
                changeDetection: ChangeDetectionStrategy.OnPush,
            }]
    }], null, null); })();
