// tds-ui/esm2020/card/card-content.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSCardContentComponent {
}
TDSCardContentComponent.ɵfac = function TDSCardContentComponent_Factory(t) { return new (t || TDSCardContentComponent)(); };
TDSCardContentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardContentComponent, selectors: [["tds-card-content"]], hostAttrs: [1, "p-4", "h-auto", "block", "w-full", "text-body-2", "font-normal"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSCardContentComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardContentComponent, [{
        type: Component,
        args: [{
                selector: 'tds-card-content',
                template: `
      <ng-content></ng-content>
  `,
                host: {
                    class: 'p-4 h-auto block w-full text-body-2 font-normal',
                },
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, null); })();
