// tds-ui/esm2020/core/option/option.module.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSOption } from './option';
import { TDSOptgroup } from './optgroup';
import { TDSPseudoCheckboxModule } from 'tds-ui/core/selection';
import * as i0 from "@angular/core";
export class TDSOptionModule {
}
TDSOptionModule.ɵfac = function TDSOptionModule_Factory(t) { return new (t || TDSOptionModule)(); };
TDSOptionModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSOptionModule });
TDSOptionModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, TDSPseudoCheckboxModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSOptionModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, TDSPseudoCheckboxModule],
                exports: [TDSOption, TDSOptgroup],
                declarations: [TDSOption, TDSOptgroup]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSOptionModule, { declarations: [TDSOption, TDSOptgroup], imports: [CommonModule, TDSPseudoCheckboxModule], exports: [TDSOption, TDSOptgroup] }); })();
