// tds-ui/esm2020/collapse/collapse.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSNoAnimationModule } from 'tds-ui/core/no-animation';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSCollapsePanelComponent } from './collapse-panel.component';
import { TDSCollapseComponent } from './collapse.component';
import * as i0 from "@angular/core";
export class TDSCollapseModule {
}
TDSCollapseModule.ɵfac = function TDSCollapseModule_Factory(t) { return new (t || TDSCollapseModule)(); };
TDSCollapseModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCollapseModule });
TDSCollapseModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule, TDSNoAnimationModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCollapseModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSCollapsePanelComponent, TDSCollapseComponent],
                exports: [TDSCollapsePanelComponent, TDSCollapseComponent],
                imports: [BidiModule, CommonModule, TDSOutletModule, TDSNoAnimationModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCollapseModule, { declarations: [TDSCollapsePanelComponent, TDSCollapseComponent], imports: [BidiModule, CommonModule, TDSOutletModule, TDSNoAnimationModule], exports: [TDSCollapsePanelComponent, TDSCollapseComponent] }); })();
