// tds-ui/esm2020/divider/divider.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSDividerComponent } from './divider.component';
import * as i0 from "@angular/core";
export class TDSDividerModule {
}
TDSDividerModule.ɵfac = function TDSDividerModule_Factory(t) { return new (t || TDSDividerModule)(); };
TDSDividerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSDividerModule });
TDSDividerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDividerModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, TDSOutletModule],
                declarations: [TDSDividerComponent],
                exports: [TDSDividerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSDividerModule, { declarations: [TDSDividerComponent], imports: [BidiModule, CommonModule, TDSOutletModule], exports: [TDSDividerComponent] }); })();
