// tds-ui/esm2020/zoom-box/tds-zoom-box.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSZoomBoxComponent } from './tds-zoom-box';
import * as i0 from "@angular/core";
export class TDSZoomBoxModule {
}
TDSZoomBoxModule.ɵfac = function TDSZoomBoxModule_Factory(t) { return new (t || TDSZoomBoxModule)(); };
TDSZoomBoxModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSZoomBoxModule });
TDSZoomBoxModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            BidiModule,
            OverlayModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSZoomBoxModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSZoomBoxComponent
                ],
                imports: [
                    CommonModule,
                    BidiModule,
                    OverlayModule
                ],
                exports: [
                    TDSZoomBoxComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSZoomBoxModule, { declarations: [TDSZoomBoxComponent], imports: [CommonModule,
        BidiModule,
        OverlayModule], exports: [TDSZoomBoxComponent] }); })();
