// tds-ui/esm2020/core/util/number.mjs
import { TDSHelperString } from "tds-ui/shared/utility";
export function getPercent(min, max, value) {
    return ((value - min) / (max - min)) * 100;
}
export function getPrecision(num) {
    const numStr = num.toString();
    const dotIndex = numStr.indexOf('.');
    return dotIndex >= 0 ? numStr.length - dotIndex - 1 : 0;
}
export function ensureNumberInRange(num, min, max) {
    if (isNaN(num) || num < min) {
        return min;
    }
    else if (num > max) {
        return max;
    }
    else {
        return num;
    }
}
export function isNumberFinite(value) {
    return typeof value === 'number' && isFinite(value);
}
export function toDecimal(value, decimal) {
    return Math.round(value * Math.pow(10, decimal)) / Math.pow(10, decimal);
}
export function sum(input, initial = 0) {
    return input.reduce((previous, current) => previous + current, initial);
}
export function formatNumberWithCommas(value, commas = ',') {
    if (value != null) {
        return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, commas);
    }
    return value;
}
;
export function parserNumberWithCommas(value, commas = ',') {
    if (value != null) {
        return TDSHelperString.replaceAll(value, commas, '');
    }
    return value;
}
;
