// tds-ui/esm2020/step/steps.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSStepComponent } from './step.component';
import { TDSStepsComponent } from './steps.component';
import * as i0 from "@angular/core";
export class TDSStepsModule {
}
TDSStepsModule.ɵfac = function TDSStepsModule_Factory(t) { return new (t || TDSStepsModule)(); };
TDSStepsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSStepsModule });
TDSStepsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSStepsModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, TDSOutletModule],
                exports: [TDSStepsComponent, TDSStepComponent],
                declarations: [TDSStepsComponent, TDSStepComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSStepsModule, { declarations: [TDSStepsComponent, TDSStepComponent], imports: [BidiModule, CommonModule, TDSOutletModule], exports: [TDSStepsComponent, TDSStepComponent] }); })();
