// tds-ui/esm2020/rate/rate.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSRateItemComponent } from './rate-item.component';
import { TDSRateComponent } from './rate.component';
import * as i0 from "@angular/core";
export class TDSRateModule {
}
TDSRateModule.ɵfac = function TDSRateModule_Factory(t) { return new (t || TDSRateModule)(); };
TDSRateModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSRateModule });
TDSRateModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSToolTipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRateModule, [{
        type: NgModule,
        args: [{
                exports: [TDSRateComponent],
                declarations: [TDSRateComponent, TDSRateItemComponent],
                imports: [BidiModule, CommonModule, TDSToolTipModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSRateModule, { declarations: [TDSRateComponent, TDSRateItemComponent], imports: [BidiModule, CommonModule, TDSToolTipModule], exports: [TDSRateComponent] }); })();
