// tds-ui/esm2020/rate/rate-item.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["tds-rate-item", ""];
function TDSRateItemComponent_ng_template_1_Template(rf, ctx) { }
function TDSRateItemComponent_ng_template_3_Template(rf, ctx) { }
function TDSRateItemComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 4);
} }
const _c1 = function (a0) { return { $implicit: a0 }; };
const _c2 = function (a0, a1) { return { "text-base-amber-400 dark:text-d-base-amber-400 opacity-100": a0, "opacity-0 text-current": a1 }; };
export class TDSRateItemComponent {
    constructor() {
        this.isHalf = false;
        this.index = 0;
        this.allowHalf = false;
        this.itemHover = new EventEmitter();
        this.itemClick = new EventEmitter();
    }
    hoverRate(isHalf) {
        this.itemHover.next(isHalf && this.allowHalf);
    }
    clickRate(isHalf) {
        this.itemClick.next(isHalf && this.allowHalf);
    }
}
TDSRateItemComponent.ɵfac = function TDSRateItemComponent_Factory(t) { return new (t || TDSRateItemComponent)(); };
TDSRateItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSRateItemComponent, selectors: [["", "tds-rate-item", ""]], inputs: { isHalf: "isHalf", character: "character", index: "index", allowHalf: "allowHalf" }, outputs: { itemHover: "itemHover", itemClick: "itemClick" }, exportAs: ["tdsRateItem"], attrs: _c0, decls: 6, vars: 12, consts: [[1, "tds-rate-star-second", "select-none", "text-current", 3, "mouseover", "click"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "tds-rate-star-first", "select-none", "absolute", "left-0", "top-0", "w-1/2", "h-full", "overflow-hidden", 3, "ngClass", "mouseover", "click"], ["defaultCharacter", ""], [1, "tdsi-star-1-fill"]], template: function TDSRateItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵlistener("mouseover", function TDSRateItemComponent_Template_div_mouseover_0_listener($event) { ctx.hoverRate(false); return $event.stopPropagation(); })("click", function TDSRateItemComponent_Template_div_click_0_listener() { return ctx.clickRate(false); });
        i0.ɵɵtemplate(1, TDSRateItemComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(2, "div", 2);
        i0.ɵɵlistener("mouseover", function TDSRateItemComponent_Template_div_mouseover_2_listener($event) { ctx.hoverRate(true); return $event.stopPropagation(); })("click", function TDSRateItemComponent_Template_div_click_2_listener() { return ctx.clickRate(true); });
        i0.ɵɵtemplate(3, TDSRateItemComponent_ng_template_3_Template, 0, 0, "ng-template", 1);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(4, TDSRateItemComponent_ng_template_4_Template, 1, 0, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r2 = i0.ɵɵreference(5);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.character || _r2)("ngTemplateOutletContext", i0.ɵɵpureFunction1(5, _c1, ctx.index));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(7, _c2, ctx.isHalf, !ctx.isHalf));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.character || _r2)("ngTemplateOutletContext", i0.ɵɵpureFunction1(10, _c1, ctx.index));
    } }, directives: [i1.NgTemplateOutlet, i1.NgClass], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSRateItemComponent.prototype, "isHalf", void 0);
__decorate([
    InputBoolean()
], TDSRateItemComponent.prototype, "allowHalf", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRateItemComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: '[tds-rate-item]',
                exportAs: 'tdsRateItem',
                template: `
    <div
      class="tds-rate-star-second select-none text-current"
      (mouseover)="hoverRate(false); $event.stopPropagation()"
      (click)="clickRate(false)"
    >
      <ng-template
        [ngTemplateOutlet]="character || defaultCharacter"
        [ngTemplateOutletContext]="{ $implicit: index }"
      ></ng-template>
    </div>
    <div [ngClass]="{'text-base-amber-400 dark:text-d-base-amber-400 opacity-100':isHalf,'opacity-0 text-current':!isHalf}" class="tds-rate-star-first select-none  absolute left-0 top-0 w-1/2 h-full overflow-hidden " (mouseover)="hoverRate(true); $event.stopPropagation()" (click)="clickRate(true)">
      <ng-template
        [ngTemplateOutlet]="character || defaultCharacter"
        [ngTemplateOutletContext]="{ $implicit: index }"
      ></ng-template>
    </div>

    <ng-template #defaultCharacter>
      <i class="tdsi-star-1-fill"></i>
    </ng-template>
  `
            }]
    }], null, { isHalf: [{
            type: Input
        }], character: [{
            type: Input
        }], index: [{
            type: Input
        }], allowHalf: [{
            type: Input
        }], itemHover: [{
            type: Output
        }], itemClick: [{
            type: Output
        }] }); })();
