// tds-ui/esm2020/core/no-animation/tds-no-animation.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSNoAnimationDirective } from './tds-no-animation.directive';
import * as i0 from "@angular/core";
export class TDSNoAnimationModule {
}
TDSNoAnimationModule.ɵfac = function TDSNoAnimationModule_Factory(t) { return new (t || TDSNoAnimationModule)(); };
TDSNoAnimationModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSNoAnimationModule });
TDSNoAnimationModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSNoAnimationModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSNoAnimationDirective],
                exports: [TDSNoAnimationDirective],
                imports: [CommonModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSNoAnimationModule, { declarations: [TDSNoAnimationDirective], imports: [CommonModule], exports: [TDSNoAnimationDirective] }); })();
