// tds-ui/esm2020/core/resize-observers/resize-observer.module.mjs
import { NgModule } from '@angular/core';
import { TDSResizeObserverDirective } from './resize-observer.directive';
import { TDSResizeObserverFactory } from './resize-observer.service';
import * as i0 from "@angular/core";
export class TDSResizeObserverModule {
}
TDSResizeObserverModule.ɵfac = function TDSResizeObserverModule_Factory(t) { return new (t || TDSResizeObserverModule)(); };
TDSResizeObserverModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSResizeObserverModule });
TDSResizeObserverModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [TDSResizeObserverFactory] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSResizeObserverModule, [{
        type: NgModule,
        args: [{
                providers: [TDSResizeObserverFactory],
                declarations: [TDSResizeObserverDirective],
                exports: [TDSResizeObserverDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSResizeObserverModule, { declarations: [TDSResizeObserverDirective], exports: [TDSResizeObserverDirective] }); })();
