// tds-ui/esm2020/core/resize-observers/resize-observer.directive.mjs
import { __decorate } from "tslib";
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "./resize-observer.service";
export class TDSResizeObserverDirective {
    constructor(resizeObserver, elementRef) {
        this.resizeObserver = resizeObserver;
        this.elementRef = elementRef;
        this.resizeObserve = new EventEmitter();
        this.resizeObserverDisabled = false;
        this.currentSubscription = null;
    }
    subscribe() {
        this.unsubscribe();
        this.currentSubscription = this.resizeObserver.observe(this.elementRef).subscribe(this.resizeObserve);
    }
    unsubscribe() {
        this.currentSubscription?.unsubscribe();
    }
    ngAfterContentInit() {
        if (!this.currentSubscription && !this.resizeObserverDisabled) {
            this.subscribe();
        }
    }
    ngOnDestroy() {
        this.unsubscribe();
    }
    ngOnChanges(changes) {
        const { resizeObserve } = changes;
        if (resizeObserve) {
            if (this.resizeObserverDisabled) {
                this.unsubscribe();
            }
            else {
                this.subscribe();
            }
        }
    }
}
TDSResizeObserverDirective.ɵfac = function TDSResizeObserverDirective_Factory(t) { return new (t || TDSResizeObserverDirective)(i0.ɵɵdirectiveInject(i1.TDSResizeObserver), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSResizeObserverDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSResizeObserverDirective, selectors: [["", "resizeObserver", ""]], inputs: { resizeObserverDisabled: "resizeObserverDisabled" }, outputs: { resizeObserve: "resizeObserve" }, features: [i0.ɵɵNgOnChangesFeature] });
__decorate([
    InputBoolean()
], TDSResizeObserverDirective.prototype, "resizeObserverDisabled", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSResizeObserverDirective, [{
        type: Directive,
        args: [{
                selector: '[resizeObserver]'
            }]
    }], function () { return [{ type: i1.TDSResizeObserver }, { type: i0.ElementRef }]; }, { resizeObserve: [{
            type: Output
        }], resizeObserverDisabled: [{
            type: Input
        }] }); })();
