// tds-ui/esm2020/tabs/tab-add-button.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/outlet";
function TDSTabAddButtonComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
export class TDSTabAddButtonComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.addIcon = 'plus';
        this.element = this.elementRef.nativeElement;
    }
    getElementWidth() {
        return this.element?.offsetWidth || 0;
    }
    getElementHeight() {
        return this.element?.offsetHeight || 0;
    }
}
TDSTabAddButtonComponent.ɵfac = function TDSTabAddButtonComponent_Factory(t) { return new (t || TDSTabAddButtonComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSTabAddButtonComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTabAddButtonComponent, selectors: [["tds-tab-add-button"], ["button", "tds-tab-add-button", ""]], hostAttrs: ["aria-label", "Add tab", "type", "button", 1, "tds-tabs-nav-add"], inputs: { addIcon: "addIcon" }, decls: 1, vars: 1, consts: [[4, "tdsStringTemplateOutlet"]], template: function TDSTabAddButtonComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSTabAddButtonComponent_ng_container_0_Template, 1, 0, "ng-container", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.addIcon);
    } }, directives: [i1.TDSStringTemplateOutletDirective], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabAddButtonComponent, [{
        type: Component,
        args: [{
                selector: 'tds-tab-add-button, button[tds-tab-add-button]',
                template: `
    <ng-container *tdsStringTemplateOutlet="addIcon; let icon">
     
    </ng-container>
  `,
                host: {
                    class: 'tds-tabs-nav-add',
                    'aria-label': 'Add tab',
                    type: 'button'
                }
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { addIcon: [{
            type: Input
        }] }); })();
