// tds-ui/esm2020/tabs/tab-link.directive.mjs
import { Directive, Host, Optional, Self } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * Fix https://github.com/angular/angular/issues/8563
 */
export class TDSTabLinkTemplateDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
}
TDSTabLinkTemplateDirective.ɵfac = function TDSTabLinkTemplateDirective_Factory(t) { return new (t || TDSTabLinkTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef, 1)); };
TDSTabLinkTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTabLinkTemplateDirective, selectors: [["ng-template", "tdsTabLink", ""]], exportAs: ["tdsTabLinkTemplate"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabLinkTemplateDirective, [{
        type: Directive,
        args: [{
                selector: 'ng-template[tdsTabLink]',
                exportAs: 'tdsTabLinkTemplate'
            }]
    }], function () { return [{ type: i0.TemplateRef, decorators: [{
                type: Host
            }] }]; }, null); })();
/**
 * This component is for catching `routerLink` directive.
 */
export class TDSTabLinkDirective {
    constructor(elementRef, routerLink, routerLinkWithHref) {
        this.elementRef = elementRef;
        this.routerLink = routerLink;
        this.routerLinkWithHref = routerLinkWithHref;
    }
}
TDSTabLinkDirective.ɵfac = function TDSTabLinkDirective_Factory(t) { return new (t || TDSTabLinkDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.RouterLink, 10), i0.ɵɵdirectiveInject(i1.RouterLinkWithHref, 10)); };
TDSTabLinkDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTabLinkDirective, selectors: [["a", "tds-tab-link", ""]], exportAs: ["tdsTabLink"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabLinkDirective, [{
        type: Directive,
        args: [{
                selector: 'a[tds-tab-link]',
                exportAs: 'tdsTabLink'
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.RouterLink, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }, { type: i1.RouterLinkWithHref, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }]; }, null); })();
