// tds-ui/esm2020/tabs/tab-nav-item.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSTabNavItemDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.disabled = false;
        this.active = false;
        this.el = elementRef.nativeElement;
        this.parentElement = this.el.parentElement;
    }
    focus() {
        this.el.focus();
    }
    get width() {
        return this.parentElement.offsetWidth;
    }
    get height() {
        return this.parentElement.offsetHeight;
    }
    get left() {
        return this.parentElement.offsetLeft;
    }
    get top() {
        return this.parentElement.offsetTop;
    }
}
TDSTabNavItemDirective.ɵfac = function TDSTabNavItemDirective_Factory(t) { return new (t || TDSTabNavItemDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSTabNavItemDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTabNavItemDirective, selectors: [["", "tdsTabNavItem", ""]], inputs: { disabled: "disabled", tab: "tab", active: "active" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabNavItemDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsTabNavItem]'
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { disabled: [{
            type: Input
        }], tab: [{
            type: Input
        }], active: [{
            type: Input
        }] }); })();
