// tds-ui/esm2020/tabs/tab-close-button.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/outlet";
function TDSTabCloseButtonComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
export class TDSTabCloseButtonComponent {
    constructor() {
        this.closeIcon = 'close';
    }
}
TDSTabCloseButtonComponent.ɵfac = function TDSTabCloseButtonComponent_Factory(t) { return new (t || TDSTabCloseButtonComponent)(); };
TDSTabCloseButtonComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTabCloseButtonComponent, selectors: [["tds-tab-close-button"], ["button", "tds-tab-close-button", ""]], hostAttrs: ["aria-label", "Close tab", "type", "button", 1, "tds-tabs-tab-remove"], inputs: { closeIcon: "closeIcon" }, decls: 1, vars: 1, consts: [[4, "tdsStringTemplateOutlet"]], template: function TDSTabCloseButtonComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSTabCloseButtonComponent_ng_container_0_Template, 1, 0, "ng-container", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.closeIcon);
    } }, directives: [i1.TDSStringTemplateOutletDirective], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabCloseButtonComponent, [{
        type: Component,
        args: [{
                selector: 'tds-tab-close-button, button[tds-tab-close-button]',
                template: `
    <ng-container *tdsStringTemplateOutlet="closeIcon; let icon">
      <!-- <i nz-icon [type]="icon" nzTheme="outline"></i> -->
    </ng-container>
  `,
                host: {
                    class: 'tds-tabs-tab-remove',
                    'aria-label': 'Close tab',
                    type: 'button'
                }
            }]
    }], function () { return []; }, { closeIcon: [{
            type: Input
        }] }); })();
