// tds-ui/esm2020/tabs/tab.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Inject, InjectionToken, Input, Output, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { TDSTabLinkDirective, TDSTabLinkTemplateDirective } from './tab-link.directive';
import { TDSTabDirective } from './tab.directive';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
const _c0 = ["contentTemplate"];
function TDSTabComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0);
} }
function TDSTabComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0, 1);
} }
const _c1 = [[["", "tds-tab-link", ""]], "*"];
const _c2 = ["[tds-tab-link]", "*"];
/**
 * Used to provide a tab set to a tab without causing a circular dependency.
 */
export const TDS_TAB_SET = new InjectionToken('TDS_TAB_SET');
export class TDSTabComponent {
    constructor(closestTabSet) {
        this.closestTabSet = closestTabSet;
        this.title = '';
        this.closable = false;
        this.closeIcon = 'close';
        this.disabled = false;
        this.forceRender = false;
        this.clsContent = "p-4 w-full h-full";
        this.tabSelect = new EventEmitter();
        this.tabDeselect = new EventEmitter();
        this.tabClick = new EventEmitter();
        this.tabContextmenu = new EventEmitter();
        this.template = null;
        this.isActive = false;
        this.position = null;
        this.origin = null;
        this.stateChanges = new Subject();
    }
    get content() {
        return this.template || this.contentTemplate;
    }
    get label() {
        return this.title || this.tabLinkTemplateDirective?.templateRef;
    }
    ngOnChanges(changes) {
        const { title, disabled, forceRender } = changes;
        if (title || disabled || forceRender) {
            this.stateChanges.next();
        }
    }
    ngOnDestroy() {
        this.stateChanges.complete();
    }
}
TDSTabComponent.ɵfac = function TDSTabComponent_Factory(t) { return new (t || TDSTabComponent)(i0.ɵɵdirectiveInject(TDS_TAB_SET)); };
TDSTabComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTabComponent, selectors: [["tds-tab"]], contentQueries: function TDSTabComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSTabLinkTemplateDirective, 5);
        i0.ɵɵcontentQuery(dirIndex, TDSTabDirective, 5, TemplateRef);
        i0.ɵɵcontentQuery(dirIndex, TDSTabLinkDirective, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.tabLinkTemplateDirective = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.template = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.linkDirective = _t.first);
    } }, viewQuery: function TDSTabComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.contentTemplate = _t.first);
    } }, inputs: { title: "title", closable: "closable", closeIcon: "closeIcon", disabled: "disabled", forceRender: "forceRender", clsContent: "clsContent" }, outputs: { tabSelect: "tabSelect", tabDeselect: "tabDeselect", tabClick: "tabClick", tabContextmenu: "tabContextmenu" }, exportAs: ["tdsTab"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c2, decls: 4, vars: 0, consts: [["tabLinkTemplate", ""], ["contentTemplate", ""]], template: function TDSTabComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c1);
        i0.ɵɵtemplate(0, TDSTabComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(2, TDSTabComponent_ng_template_2_Template, 1, 0, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSTabComponent.prototype, "closable", void 0);
__decorate([
    InputBoolean()
], TDSTabComponent.prototype, "disabled", void 0);
__decorate([
    InputBoolean()
], TDSTabComponent.prototype, "forceRender", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabComponent, [{
        type: Component,
        args: [{
                selector: 'tds-tab',
                exportAs: 'tdsTab',
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <ng-template #tabLinkTemplate>
      <ng-content select="[tds-tab-link]"></ng-content>
    </ng-template>
    <ng-template #contentTemplate>
      <ng-content></ng-content>
    </ng-template>
  `
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [TDS_TAB_SET]
            }] }]; }, { title: [{
            type: Input
        }], closable: [{
            type: Input
        }], closeIcon: [{
            type: Input
        }], disabled: [{
            type: Input
        }], forceRender: [{
            type: Input
        }], clsContent: [{
            type: Input
        }], tabSelect: [{
            type: Output
        }], tabDeselect: [{
            type: Output
        }], tabClick: [{
            type: Output
        }], tabContextmenu: [{
            type: Output
        }], tabLinkTemplateDirective: [{
            type: ContentChild,
            args: [TDSTabLinkTemplateDirective, { static: false }]
        }], template: [{
            type: ContentChild,
            args: [TDSTabDirective, { static: false, read: TemplateRef }]
        }], linkDirective: [{
            type: ContentChild,
            args: [TDSTabLinkDirective, { static: false }]
        }], contentTemplate: [{
            type: ViewChild,
            args: ['contentTemplate', { static: true }]
        }] }); })();
