// tds-ui/esm2020/tabs/tabs.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { BidiModule } from '@angular/cdk/bidi';
import { ObserversModule } from '@angular/cdk/observers';
import { PlatformModule } from '@angular/cdk/platform';
import { CdkScrollableModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSDropDownModule } from 'tds-ui/dropdown';
import { TDSTabAddButtonComponent } from './tab-add-button.component';
import { TDSTabBodyComponent } from './tab-body.component';
import { TDSTabCloseButtonComponent } from './tab-close-button.component';
import { TDSTabLinkDirective, TDSTabLinkTemplateDirective } from './tab-link.directive';
import { TDSTabNavBarComponent } from './tab-nav-bar.component';
import { TDSTabNavItemDirective } from './tab-nav-item.directive';
import { TDSTabNavOperationComponent } from './tab-nav-operation.component';
import { TDSTabScrollListDirective } from './tab-scroll-list.directive';
import { TDSTabComponent } from './tab.component';
import { TDSTabDirective } from './tab.directive';
import { TDSTabsInkBarDirective } from './tabs-ink-bar.directive';
import { TDSTabSetComponent } from './tabset.component';
import * as i0 from "@angular/core";
const DIRECTIVES = [
    TDSTabSetComponent,
    TDSTabComponent,
    TDSTabNavBarComponent,
    TDSTabNavItemDirective,
    TDSTabsInkBarDirective,
    TDSTabScrollListDirective,
    TDSTabNavOperationComponent,
    TDSTabAddButtonComponent,
    TDSTabCloseButtonComponent,
    TDSTabDirective,
    TDSTabBodyComponent,
    TDSTabLinkDirective,
    TDSTabLinkTemplateDirective
];
export class TDSTabsModule {
}
TDSTabsModule.ɵfac = function TDSTabsModule_Factory(t) { return new (t || TDSTabsModule)(); };
TDSTabsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSTabsModule });
TDSTabsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            ObserversModule,
            TDSOutletModule,
            PlatformModule,
            A11yModule,
            CdkScrollableModule,
            TDSDropDownModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTabsModule, [{
        type: NgModule,
        args: [{
                declarations: [DIRECTIVES],
                exports: [DIRECTIVES],
                imports: [
                    BidiModule,
                    CommonModule,
                    ObserversModule,
                    TDSOutletModule,
                    PlatformModule,
                    A11yModule,
                    CdkScrollableModule,
                    TDSDropDownModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSTabsModule, { declarations: [TDSTabSetComponent,
        TDSTabComponent,
        TDSTabNavBarComponent,
        TDSTabNavItemDirective,
        TDSTabsInkBarDirective,
        TDSTabScrollListDirective,
        TDSTabNavOperationComponent,
        TDSTabAddButtonComponent,
        TDSTabCloseButtonComponent,
        TDSTabDirective,
        TDSTabBodyComponent,
        TDSTabLinkDirective,
        TDSTabLinkTemplateDirective], imports: [BidiModule,
        CommonModule,
        ObserversModule,
        TDSOutletModule,
        PlatformModule,
        A11yModule,
        CdkScrollableModule,
        TDSDropDownModule], exports: [TDSTabSetComponent,
        TDSTabComponent,
        TDSTabNavBarComponent,
        TDSTabNavItemDirective,
        TDSTabsInkBarDirective,
        TDSTabScrollListDirective,
        TDSTabNavOperationComponent,
        TDSTabAddButtonComponent,
        TDSTabCloseButtonComponent,
        TDSTabDirective,
        TDSTabBodyComponent,
        TDSTabLinkDirective,
        TDSTabLinkTemplateDirective] }); })();
