// tds-ui/esm2020/progress/progress.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSProgressComponent } from './progress.component';
import * as i0 from "@angular/core";
export class TDSProgressModule {
}
TDSProgressModule.ɵfac = function TDSProgressModule_Factory(t) { return new (t || TDSProgressModule)(); };
TDSProgressModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSProgressModule });
TDSProgressModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSProgressModule, [{
        type: NgModule,
        args: [{
                exports: [TDSProgressComponent],
                declarations: [TDSProgressComponent],
                imports: [BidiModule, CommonModule, TDSOutletModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSProgressModule, { declarations: [TDSProgressComponent], imports: [BidiModule, CommonModule, TDSOutletModule], exports: [TDSProgressComponent] }); })();
