// tds-ui/esm2020/core/pipes/tds-bytes.pipe.mjs
import { Pipe } from '@angular/core';
import { isNumberFinite, toDecimal } from 'tds-ui/core/util';
import * as i0 from "@angular/core";
export class TDSBytesPipe {
    transform(input, decimal = 0, from = 'B', to) {
        if (!(isNumberFinite(input) && isNumberFinite(decimal) && decimal % 1 === 0 && decimal >= 0)) {
            return input;
        }
        let bytes = input;
        let unit = from;
        while (unit !== 'B') {
            bytes *= 1024;
            unit = TDSBytesPipe.formats[unit].prev;
        }
        if (to) {
            const format = TDSBytesPipe.formats[to];
            const result = toDecimal(TDSBytesPipe.calculateResult(format, bytes), decimal);
            return TDSBytesPipe.formatResult(result, to);
        }
        for (const key in TDSBytesPipe.formats) {
            if (TDSBytesPipe.formats.hasOwnProperty(key)) {
                const format = TDSBytesPipe.formats[key];
                if (bytes < format.max) {
                    const result = toDecimal(TDSBytesPipe.calculateResult(format, bytes), decimal);
                    return TDSBytesPipe.formatResult(result, key);
                }
            }
        }
    }
    static formatResult(result, unit) {
        return `${result} ${unit}`;
    }
    static calculateResult(format, bytes) {
        const prev = format.prev ? TDSBytesPipe.formats[format.prev] : undefined;
        return prev ? bytes / prev.max : bytes;
    }
}
TDSBytesPipe.formats = {
    B: { max: 1024 },
    kB: { max: Math.pow(1024, 2), prev: 'B' },
    KB: { max: Math.pow(1024, 2), prev: 'B' },
    MB: { max: Math.pow(1024, 3), prev: 'kB' },
    GB: { max: Math.pow(1024, 4), prev: 'MB' },
    TB: { max: Number.MAX_SAFE_INTEGER, prev: 'GB' }
};
TDSBytesPipe.ɵfac = function TDSBytesPipe_Factory(t) { return new (t || TDSBytesPipe)(); };
TDSBytesPipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "tdsBytes", type: TDSBytesPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSBytesPipe, [{
        type: Pipe,
        args: [{
                name: 'tdsBytes'
            }]
    }], null, null); })();
