// tds-ui/esm2020/core/pipes/tds-sanitizer.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class TDSSanitizerPipe {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    transform(value, type = 'html') {
        switch (type) {
            case 'html':
                return this.sanitizer.bypassSecurityTrustHtml(value);
            case 'style':
                return this.sanitizer.bypassSecurityTrustStyle(value);
            case 'url':
                return this.sanitizer.bypassSecurityTrustUrl(value);
            case 'resourceUrl':
                return this.sanitizer.bypassSecurityTrustResourceUrl(value);
            default:
                throw new Error(`Invalid safe type specified`);
        }
    }
}
TDSSanitizerPipe.ɵfac = function TDSSanitizerPipe_Factory(t) { return new (t || TDSSanitizerPipe)(i0.ɵɵdirectiveInject(i1.DomSanitizer, 16)); };
TDSSanitizerPipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "tdsSanitizer", type: TDSSanitizerPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSanitizerPipe, [{
        type: Pipe,
        args: [{
                name: 'tdsSanitizer'
            }]
    }], function () { return [{ type: i1.DomSanitizer }]; }, null); })();
