// tds-ui/esm2020/core/pipes/tds-css-unit.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSToCssUnitPipe {
    transform(value, defaultUnit = 'px') {
        const absoluteLengthUnit = ['cm', 'mm', 'Q', 'in', 'pc', 'pt', 'px'];
        const relativeLengthUnit = ['em', 'ex', 'ch', 'rem', '1h', 'vw', 'vh', 'vmin', 'vmax'];
        const percentagesUnit = ['%'];
        const listOfUnit = [...absoluteLengthUnit, ...relativeLengthUnit, ...percentagesUnit];
        let unit = 'px';
        if (listOfUnit.some(u => u === defaultUnit)) {
            unit = defaultUnit;
        }
        return typeof value === 'number' ? `${value}${unit}` : `${value}`;
    }
}
TDSToCssUnitPipe.ɵfac = function TDSToCssUnitPipe_Factory(t) { return new (t || TDSToCssUnitPipe)(); };
TDSToCssUnitPipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "tdsToCssUnit", type: TDSToCssUnitPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSToCssUnitPipe, [{
        type: Pipe,
        args: [{
                name: 'tdsToCssUnit'
            }]
    }], null, null); })();
