// tds-ui/esm2020/core/pipes/tds-aggregate.pipe.mjs
import { Pipe } from '@angular/core';
import { sum } from 'tds-ui/core/util';
import * as i0 from "@angular/core";
export class TDSAggregatePipe {
    transform(value, method) {
        if (!Array.isArray(value)) {
            return value;
        }
        if (value.length === 0) {
            return undefined;
        }
        switch (method) {
            case 'sum':
                return sum(value);
            case 'avg':
                return sum(value) / value.length;
            case 'max':
                return Math.max(...value);
            case 'min':
                return Math.min(...value);
            default:
                throw Error(`Invalid Pipe Arguments: Aggregate pipe doesn't support this type`);
        }
    }
}
TDSAggregatePipe.ɵfac = function TDSAggregatePipe_Factory(t) { return new (t || TDSAggregatePipe)(); };
TDSAggregatePipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "tdsAggregate", type: TDSAggregatePipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAggregatePipe, [{
        type: Pipe,
        args: [{
                name: 'tdsAggregate'
            }]
    }], null, null); })();
