// tds-ui/esm2020/core/pipes/tds-pipes.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSAggregatePipe } from './tds-aggregate.pipe';
import { TDSBytesPipe } from './tds-bytes.pipe';
import { TDSToCssUnitPipe } from './tds-css-unit.pipe';
import { TDSEllipsisPipe } from './tds-ellipsis.pipe';
import { TDSSafeNullPipe } from './tds-safe-null.pipe';
import { TDSSanitizerPipe } from './tds-sanitizer.pipe';
import { TDSTrimPipe } from './tds-trim.pipe';
import * as i0 from "@angular/core";
const pipes = [TDSToCssUnitPipe,
    TDSSafeNullPipe,
    TDSSanitizerPipe,
    TDSTrimPipe,
    TDSBytesPipe,
    TDSAggregatePipe,
    TDSEllipsisPipe];
export class TDSPipesModule {
}
TDSPipesModule.ɵfac = function TDSPipesModule_Factory(t) { return new (t || TDSPipesModule)(); };
TDSPipesModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSPipesModule });
TDSPipesModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPipesModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [pipes],
                declarations: [pipes]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSPipesModule, { declarations: [TDSToCssUnitPipe,
        TDSSafeNullPipe,
        TDSSanitizerPipe,
        TDSTrimPipe,
        TDSBytesPipe,
        TDSAggregatePipe,
        TDSEllipsisPipe], imports: [CommonModule], exports: [TDSToCssUnitPipe,
        TDSSafeNullPipe,
        TDSSanitizerPipe,
        TDSTrimPipe,
        TDSBytesPipe,
        TDSAggregatePipe,
        TDSEllipsisPipe] }); })();
