// tds-ui/esm2020/table/src/table/table-fixed-row.component.mjs
import { ChangeDetectionStrategy, Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../table-style.service";
import * as i2 from "@angular/common";
const _c0 = ["tdElement"];
const _c1 = ["tds-table-fixed-row", ""];
function TDSTableFixedRowComponent_div_2_ng_template_2_Template(rf, ctx) { }
function TDSTableFixedRowComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 4);
    i0.ɵɵpipe(1, "async");
    i0.ɵɵtemplate(2, TDSTableFixedRowComponent_div_2_ng_template_2_Template, 0, 0, "ng-template", 5);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    const _r2 = i0.ɵɵreference(5);
    i0.ɵɵstyleProp("width", i0.ɵɵpipeBind1(1, 3, ctx_r1.hostWidth$), "px");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function TDSTableFixedRowComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0);
} }
const _c2 = ["*"];
export class TDSTableFixedRowComponent {
    constructor(tdsTableStyleService, renderer, cdr) {
        this.tdsTableStyleService = tdsTableStyleService;
        this.renderer = renderer;
        this.cdr = cdr;
        this.hostWidth$ = new BehaviorSubject(null);
        this.enableAutoMeasure$ = new BehaviorSubject(false);
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        if (this.tdsTableStyleService) {
            const { enableAutoMeasure$, hostWidth$ } = this.tdsTableStyleService;
            enableAutoMeasure$.pipe(takeUntil(this.destroy$)).subscribe(this.enableAutoMeasure$);
            hostWidth$.subscribe({
                next: (res) => {
                    this.hostWidth$.next(res);
                    this.cdr.markForCheck();
                }
            });
        }
    }
    ngAfterViewInit() {
        this.tdsTableStyleService.columnCount$.pipe(takeUntil(this.destroy$)).subscribe(count => {
            this.renderer.setAttribute(this.tdElement.nativeElement, 'colspan', `${count}`);
        });
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSTableFixedRowComponent.ɵfac = function TDSTableFixedRowComponent_Factory(t) { return new (t || TDSTableFixedRowComponent)(i0.ɵɵdirectiveInject(i1.TDSTableStyleService), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSTableFixedRowComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTableFixedRowComponent, selectors: [["tr", "tds-table-fixed-row", ""], ["tr", "expand", ""]], viewQuery: function TDSTableFixedRowComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.tdElement = _t.first);
    } }, attrs: _c1, ngContentSelectors: _c2, decls: 6, vars: 4, consts: [[1, "tds-disable-td", "tds-table-cell", "border-b-0"], ["tdElement", ""], ["class", "tds-table-expanded-row-fixed p-0", "style", "position: sticky; left: 0px; overflow: hidden;", 3, "width", 4, "ngIf", "ngIfElse"], ["contentTemplate", ""], [1, "tds-table-expanded-row-fixed", "p-0", 2, "position", "sticky", "left", "0px", "overflow", "hidden"], [3, "ngTemplateOutlet"]], template: function TDSTableFixedRowComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "td", 0, 1);
        i0.ɵɵtemplate(2, TDSTableFixedRowComponent_div_2_Template, 3, 5, "div", 2);
        i0.ɵɵpipe(3, "async");
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(4, TDSTableFixedRowComponent_ng_template_4_Template, 1, 0, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r2 = i0.ɵɵreference(5);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", i0.ɵɵpipeBind1(3, 2, ctx.enableAutoMeasure$))("ngIfElse", _r2);
    } }, directives: [i2.NgIf, i2.NgTemplateOutlet], pipes: [i2.AsyncPipe], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableFixedRowComponent, [{
        type: Component,
        args: [{
                selector: 'tr[tds-table-fixed-row], tr[expand]',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
    <td class="tds-disable-td tds-table-cell border-b-0" #tdElement>
      <div
        class="tds-table-expanded-row-fixed p-0"
        *ngIf="enableAutoMeasure$ | async; else contentTemplate"
        style="position: sticky; left: 0px; overflow: hidden;"
        [style.width.px]="hostWidth$ | async"
      >
        <ng-template [ngTemplateOutlet]="contentTemplate"></ng-template>
      </div>
    </td>
    <ng-template #contentTemplate><ng-content></ng-content></ng-template>
  `
            }]
    }], function () { return [{ type: i1.TDSTableStyleService }, { type: i0.Renderer2 }, { type: i0.ChangeDetectorRef }]; }, { tdElement: [{
            type: ViewChild,
            args: ['tdElement']
        }] }); })();
