// tds-ui/esm2020/table/src/table/table-content.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["tds-table-content", ""];
function TDSTableContentComponent_col_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "col");
} if (rf & 2) {
    const width_r3 = ctx.$implicit;
    i0.ɵɵstyleProp("width", width_r3)("min-width", width_r3);
} }
function TDSTableContentComponent_thead_1_ng_template_1_Template(rf, ctx) { }
function TDSTableContentComponent_thead_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "thead", 3);
    i0.ɵɵtemplate(1, TDSTableContentComponent_thead_1_ng_template_1_Template, 0, 0, "ng-template", 2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r1.theadTemplate);
} }
function TDSTableContentComponent_ng_template_2_Template(rf, ctx) { }
const _c1 = ["*"];
export class TDSTableContentComponent {
    constructor() {
        this.tableLayout = 'auto';
        this.theadTemplate = null;
        this.contentTemplate = null;
        this.listOfColWidth = [];
        this.scrollX = null;
    }
}
TDSTableContentComponent.ɵfac = function TDSTableContentComponent_Factory(t) { return new (t || TDSTableContentComponent)(); };
TDSTableContentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTableContentComponent, selectors: [["table", "tds-table-content", ""]], hostAttrs: [1, "w-full"], hostVars: 8, hostBindings: function TDSTableContentComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵstyleProp("table-layout", ctx.tableLayout)("width", ctx.scrollX)("min-width", ctx.scrollX ? "100%" : null);
        i0.ɵɵclassProp("tds-table-fixed", ctx.scrollX);
    } }, inputs: { tableLayout: "tableLayout", theadTemplate: "theadTemplate", contentTemplate: "contentTemplate", listOfColWidth: "listOfColWidth", scrollX: "scrollX" }, attrs: _c0, ngContentSelectors: _c1, decls: 4, vars: 3, consts: [[3, "width", "minWidth", 4, "ngFor", "ngForOf"], ["class", "tds-table-thead text-neutral-1-900 text-body-2 font-semibold", 4, "ngIf"], [3, "ngTemplateOutlet"], [1, "tds-table-thead", "text-neutral-1-900", "text-body-2", "font-semibold"]], template: function TDSTableContentComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtemplate(0, TDSTableContentComponent_col_0_Template, 1, 4, "col", 0);
        i0.ɵɵtemplate(1, TDSTableContentComponent_thead_1_Template, 2, 1, "thead", 1);
        i0.ɵɵtemplate(2, TDSTableContentComponent_ng_template_2_Template, 0, 0, "ng-template", 2);
        i0.ɵɵprojection(3);
    } if (rf & 2) {
        i0.ɵɵproperty("ngForOf", ctx.listOfColWidth);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.theadTemplate);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.contentTemplate);
    } }, directives: [i1.NgForOf, i1.NgIf, i1.NgTemplateOutlet], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableContentComponent, [{
        type: Component,
        args: [{
                selector: 'table[tds-table-content]',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
    <col  [style.width]="width" [style.minWidth]="width" *ngFor="let width of listOfColWidth" />
    <thead class="tds-table-thead text-neutral-1-900 text-body-2 font-semibold" *ngIf="theadTemplate">
      <ng-template [ngTemplateOutlet]="theadTemplate"></ng-template>
    </thead>
    <ng-template [ngTemplateOutlet]="contentTemplate"></ng-template>
    <ng-content></ng-content>
  `,
                host: {
                    '[style.table-layout]': 'tableLayout',
                    '[class.tds-table-fixed]': 'scrollX',
                    '[style.width]': 'scrollX',
                    '[style.min-width]': `scrollX ? '100%': null`,
                    'class': "w-full"
                }
            }]
    }], null, { tableLayout: [{
            type: Input
        }], theadTemplate: [{
            type: Input
        }], contentTemplate: [{
            type: Input
        }], listOfColWidth: [{
            type: Input
        }], scrollX: [{
            type: Input
        }] }); })();
