// tds-ui/esm2020/table/src/table/table-virtual-scroll.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSTableVirtualScrollDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ngTemplateContextGuard(_dir, _ctx) {
        return true;
    }
}
TDSTableVirtualScrollDirective.ɵfac = function TDSTableVirtualScrollDirective_Factory(t) { return new (t || TDSTableVirtualScrollDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSTableVirtualScrollDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTableVirtualScrollDirective, selectors: [["", "tds-virtual-scroll", ""]], exportAs: ["tdsVirtualScroll"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableVirtualScrollDirective, [{
        type: Directive,
        args: [{
                selector: '[tds-virtual-scroll]',
                exportAs: 'tdsVirtualScroll'
            }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
