// tds-ui/esm2020/table/src/table/tbody.component.mjs
/* eslint-disable @angular-eslint/component-selector */
import { ChangeDetectionStrategy, Component, Optional, ViewEncapsulation } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../table-style.service";
import * as i2 from "@angular/common";
import * as i3 from "./tr-measure.component";
import * as i4 from "./table-fixed-row.component";
import * as i5 from "tds-ui/empty";
function TDSTbodyComponent_ng_container_0_tr_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "tr", 3);
    i0.ɵɵlistener("listOfAutoWidth", function TDSTbodyComponent_ng_container_0_tr_1_Template_tr_listOfAutoWidth_0_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(2); return ctx_r4.onListOfAutoWidthChange($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const listOfMeasureColumn_r2 = i0.ɵɵnextContext().ngIf;
    i0.ɵɵproperty("listOfMeasureColumn", listOfMeasureColumn_r2);
} }
function TDSTbodyComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTbodyComponent_ng_container_0_tr_1_Template, 1, 1, "tr", 2);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const listOfMeasureColumn_r2 = ctx.ngIf;
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.isInsideTable && listOfMeasureColumn_r2.length);
} }
function TDSTbodyComponent_tr_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "tr", 4);
    i0.ɵɵelement(1, "tds-embed-empty", 5);
    i0.ɵɵpipe(2, "async");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("specificContent", i0.ɵɵpipeBind1(2, 1, ctx_r1.noResult$));
} }
const _c0 = ["*"];
export class TDSTbodyComponent {
    constructor(tdsTableStyleService) {
        this.tdsTableStyleService = tdsTableStyleService;
        this.isInsideTable = false;
        this.showEmpty$ = new BehaviorSubject(false);
        this.noResult$ = new BehaviorSubject(undefined);
        this.listOfMeasureColumn$ = new BehaviorSubject([]);
        this.destroy$ = new Subject();
        this.isInsideTable = !!this.tdsTableStyleService;
        if (this.tdsTableStyleService) {
            const { showEmpty$, noResult$, listOfMeasureColumn$ } = this.tdsTableStyleService;
            noResult$.pipe(takeUntil(this.destroy$)).subscribe(this.noResult$);
            listOfMeasureColumn$.pipe(takeUntil(this.destroy$)).subscribe(this.listOfMeasureColumn$);
            showEmpty$.pipe(takeUntil(this.destroy$)).subscribe(this.showEmpty$);
        }
    }
    onListOfAutoWidthChange(listOfAutoWidth) {
        this.tdsTableStyleService.setListOfAutoWidth(listOfAutoWidth);
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSTbodyComponent.ɵfac = function TDSTbodyComponent_Factory(t) { return new (t || TDSTbodyComponent)(i0.ɵɵdirectiveInject(i1.TDSTableStyleService, 8)); };
TDSTbodyComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTbodyComponent, selectors: [["tbody"]], hostVars: 2, hostBindings: function TDSTbodyComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-table-tbody", ctx.isInsideTable);
    } }, ngContentSelectors: _c0, decls: 5, vars: 6, consts: [[4, "ngIf"], ["class", "tds-table-placeholder", "tds-table-fixed-row", "", 4, "ngIf"], ["tds-table-measure-row", "", 3, "listOfMeasureColumn", "listOfAutoWidth", 4, "ngIf"], ["tds-table-measure-row", "", 3, "listOfMeasureColumn", "listOfAutoWidth"], ["tds-table-fixed-row", "", 1, "tds-table-placeholder"], ["componentName", "table", 3, "specificContent"]], template: function TDSTbodyComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtemplate(0, TDSTbodyComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
        i0.ɵɵpipe(1, "async");
        i0.ɵɵprojection(2);
        i0.ɵɵtemplate(3, TDSTbodyComponent_tr_3_Template, 3, 3, "tr", 1);
        i0.ɵɵpipe(4, "async");
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", i0.ɵɵpipeBind1(1, 2, ctx.listOfMeasureColumn$));
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", i0.ɵɵpipeBind1(4, 4, ctx.showEmpty$));
    } }, directives: [i2.NgIf, i3.TDSTrMeasureComponent, i4.TDSTableFixedRowComponent, i5.TDSEmbedEmptyComponent], pipes: [i2.AsyncPipe], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTbodyComponent, [{
        type: Component,
        args: [{
                selector: 'tbody',
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
    <ng-container *ngIf="listOfMeasureColumn$ | async as listOfMeasureColumn">
      <tr
        tds-table-measure-row
        *ngIf="isInsideTable && listOfMeasureColumn.length"
        [listOfMeasureColumn]="listOfMeasureColumn"
        (listOfAutoWidth)="onListOfAutoWidthChange($event)"
      ></tr>
    </ng-container>
    <ng-content></ng-content>   
    <tr class="tds-table-placeholder" tds-table-fixed-row *ngIf="showEmpty$ | async">
     <tds-embed-empty componentName="table" [specificContent]="(noResult$ | async)!"></tds-embed-empty>
    </tr>
  `,
                host: {
                    '[class.tds-table-tbody]': 'isInsideTable'
                }
            }]
    }], function () { return [{ type: i1.TDSTableStyleService, decorators: [{
                type: Optional
            }] }]; }, null); })();
