// tds-ui/esm2020/table/src/table/title-footer.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/outlet";
function TDSTableTitleFooterComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r0.title);
} }
function TDSTableTitleFooterComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r1.footer);
} }
export class TDSTableTitleFooterComponent {
    constructor() {
        this.title = null;
        this.footer = null;
    }
}
TDSTableTitleFooterComponent.ɵfac = function TDSTableTitleFooterComponent_Factory(t) { return new (t || TDSTableTitleFooterComponent)(); };
TDSTableTitleFooterComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTableTitleFooterComponent, selectors: [["tds-table-title-footer"]], hostAttrs: [1, "block", "w-full", "p-3"], hostVars: 4, hostBindings: function TDSTableTitleFooterComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-table-title", ctx.title !== null)("tds-table-footer", ctx.footer !== null);
    } }, inputs: { title: "title", footer: "footer" }, decls: 2, vars: 2, consts: [[4, "tdsStringTemplateOutlet"]], template: function TDSTableTitleFooterComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSTableTitleFooterComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
        i0.ɵɵtemplate(1, TDSTableTitleFooterComponent_ng_container_1_Template, 2, 1, "ng-container", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.title);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.footer);
    } }, directives: [i1.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableTitleFooterComponent, [{
        type: Component,
        args: [{
                selector: 'tds-table-title-footer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
    <ng-container *tdsStringTemplateOutlet="title">{{ title }}</ng-container>
    <ng-container *tdsStringTemplateOutlet="footer">{{ footer }}</ng-container>
  `,
                host: {
                    '[class.tds-table-title]': `title !== null`,
                    '[class.tds-table-footer]': `footer !== null`,
                    class: "block w-full p-3 "
                }
            }]
    }], null, { title: [{
            type: Input
        }], footer: [{
            type: Input
        }] }); })();
