// tds-ui/esm2020/table/src/table/tr-expand.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSTrExpandDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.expand = true;
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-table-expanded-row');
    }
}
TDSTrExpandDirective.ɵfac = function TDSTrExpandDirective_Factory(t) { return new (t || TDSTrExpandDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSTrExpandDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTrExpandDirective, selectors: [["tr", "expand", ""]], hostVars: 1, hostBindings: function TDSTrExpandDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵhostProperty("hidden", !ctx.expand);
    } }, inputs: { expand: "expand" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTrExpandDirective, [{
        type: Directive,
        args: [{
                selector: 'tr[expand]',
                host: {
                    '[hidden]': `!expand`
                }
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { expand: [{
            type: Input
        }] }); })();
