// tds-ui/esm2020/form-field/error.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Attribute, ChangeDetectionStrategy, Component, InjectionToken, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["*"];
let nextUniqueId = 0;
/**
 * Injection token that can be used to reference instances of `TDSError`. It serves as
 * alternative token to the actual `TDSError` class which could cause unnecessary
 * retention of the class and its directive metadata.
 */
export const TDS_ERROR = new InjectionToken('TDSError');
/** Single error message to be shown underneath the form field. */
export class TDSError {
    constructor(ariaLive, elementRef) {
        this.id = `tds-error-${nextUniqueId++}`;
        this.icon = 'tdsi-warning-2-fill text-base leading-4';
        if (!ariaLive) {
            elementRef.nativeElement.setAttribute('aria-live', 'polite');
        }
    }
}
TDSError.ɵfac = function TDSError_Factory(t) { return new (t || TDSError)(i0.ɵɵinjectAttribute('aria-live'), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSError.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSError, selectors: [["tds-error"]], hostAttrs: ["aria-atomic", "true", 1, "tds-error", "text-xs", "leading-4", "font-normal", "text-error-500", "dark:text-d-error-500", "mt-1", "inline-flex", "items-center"], hostVars: 1, hostBindings: function TDSError_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("id", ctx.id);
    } }, inputs: { id: "id", icon: "icon" }, features: [i0.ɵɵProvidersFeature([{ provide: TDS_ERROR, useExisting: TDSError }])], ngContentSelectors: _c0, decls: 2, vars: 1, consts: [[1, "mr-2", 3, "ngClass"]], template: function TDSError_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelement(0, "i", 0);
        i0.ɵɵprojection(1);
    } if (rf & 2) {
        i0.ɵɵproperty("ngClass", ctx.icon);
    } }, directives: [i1.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSError, [{
        type: Component,
        args: [{
                selector: 'tds-error',
                template: `
    <i class='mr-2' [ngClass]="icon"> 
    </i>
    <ng-content></ng-content>
  `,
                host: {
                    'class': 'tds-error text-xs leading-4 font-normal text-error-500  dark:text-d-error-500 mt-1 inline-flex items-center',
                    '[attr.id]': 'id',
                    'aria-atomic': 'true',
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [{ provide: TDS_ERROR, useExisting: TDSError }],
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Attribute,
                args: ['aria-live']
            }] }, { type: i0.ElementRef }]; }, { id: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
