// tds-ui/esm2020/form-field/form-field.module.mjs
import { ObserversModule } from "@angular/cdk/observers";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { TDSAddOnLeft, TDSAddOnRight } from "./add-on";
import { TDSError } from "./error";
import { TDSFormField } from "./form-field";
import { TDSFormFieldContainer } from "./form-field-container";
import { TDSLabel } from "./label";
import { TDSPrefix } from "./prefix";
import { TDSSuccess } from "./success";
import { TDSSuffix } from "./suffix";
import { TDSWarning } from "./warning";
import * as i0 from "@angular/core";
export class TDSFormFieldModule {
}
TDSFormFieldModule.ɵfac = function TDSFormFieldModule_Factory(t) { return new (t || TDSFormFieldModule)(); };
TDSFormFieldModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSFormFieldModule });
TDSFormFieldModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            ObserversModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFormFieldModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    ObserversModule
                ],
                exports: [
                    TDSFormField,
                    TDSLabel,
                    TDSSuccess,
                    TDSError,
                    TDSWarning,
                    TDSFormFieldContainer,
                    TDSPrefix,
                    TDSSuffix,
                    TDSAddOnRight,
                    TDSAddOnLeft
                ],
                declarations: [
                    TDSFormField,
                    TDSLabel,
                    TDSSuccess,
                    TDSError,
                    TDSWarning,
                    TDSFormFieldContainer,
                    TDSPrefix,
                    TDSSuffix,
                    TDSAddOnRight,
                    TDSAddOnLeft
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSFormFieldModule, { declarations: [TDSFormField,
        TDSLabel,
        TDSSuccess,
        TDSError,
        TDSWarning,
        TDSFormFieldContainer,
        TDSPrefix,
        TDSSuffix,
        TDSAddOnRight,
        TDSAddOnLeft], imports: [CommonModule,
        ObserversModule], exports: [TDSFormField,
        TDSLabel,
        TDSSuccess,
        TDSError,
        TDSWarning,
        TDSFormFieldContainer,
        TDSPrefix,
        TDSSuffix,
        TDSAddOnRight,
        TDSAddOnLeft] }); })();
