// tds-ui/esm2020/form-field/form-field-container.mjs
import { Directive, Input } from "@angular/core";
import * as i0 from "@angular/core";
export class TDSFormFieldContainer {
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        this._color = "default";
        this._focus = false;
        this._disabled = false;
        this.defaultColor = "default";
        this._tdsClass = "";
    }
    get tdsClass() {
        return this._tdsClass;
    }
    set tdsClass(value) {
        this._tdsClass = value;
    }
}
TDSFormFieldContainer.ɵfac = function TDSFormFieldContainer_Factory(t) { return new (t || TDSFormFieldContainer)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSFormFieldContainer.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSFormFieldContainer, selectors: [["", "tds-form-field-container", ""]], hostVars: 2, hostBindings: function TDSFormFieldContainer_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassMap(ctx.tdsClass);
    } }, inputs: { tdsClass: "tdsClass" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFormFieldContainer, [{
        type: Directive,
        args: [{
                selector: `[tds-form-field-container]`,
                host: {
                    "[class]": 'tdsClass'
                }
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { tdsClass: [{
            type: Input
        }] }); })();
