// tds-ui/esm2020/form-field/label.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function TDSLabel_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 2);
    i0.ɵɵtext(1, "*");
    i0.ɵɵelementEnd();
} }
const _c0 = ["*"];
// /** The floating label for a `tds-form-field`. */
// @Directive({
//   selector: 'tds-label',
//   host:{
//     class:"text-body-2 font-semibold text-neutral-1-900  block dark:text-d-neutral-1-900 "
//   }
// })
// export class TDSLabel {}
export class TDSLabel {
    constructor(cdr) {
        this.cdr = cdr;
        this.tdsRequired = false;
    }
    ngOnChanges(changes) {
        const { tdsFor, tdsRequired } = changes;
        if (tdsFor || tdsRequired) {
            this.cdr.markForCheck();
        }
    }
    updateRequired(required) {
        this.tdsRequired = required;
        this.cdr.detectChanges();
    }
}
TDSLabel.ɵfac = function TDSLabel_Factory(t) { return new (t || TDSLabel)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSLabel.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSLabel, selectors: [["tds-label"]], inputs: { tdsFor: "tdsFor", tdsRequired: "tdsRequired" }, exportAs: ["tdsLabel"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 3, vars: 2, consts: [[1, "text-body-2", "font-semibold", "text-neutral-1-900", "block", "dark:text-d-neutral-1-900"], ["class", "ml-1 text-error-500 dark:text-d-error-500 text-body-2 font-semibold", 4, "ngIf"], [1, "ml-1", "text-error-500", "dark:text-d-error-500", "text-body-2", "font-semibold"]], template: function TDSLabel_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "label", 0);
        i0.ɵɵprojection(1);
        i0.ɵɵtemplate(2, TDSLabel_span_2_Template, 2, 0, "span", 1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵattribute("for", ctx.tdsFor);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.tdsRequired);
    } }, directives: [i1.NgIf], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSLabel.prototype, "tdsRequired", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSLabel, [{
        type: Component,
        args: [{
                selector: 'tds-label',
                exportAs: 'tdsLabel',
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <label [attr.for]="tdsFor" class="text-body-2 font-semibold text-neutral-1-900  block dark:text-d-neutral-1-900">
      <ng-content></ng-content>
      <span class="ml-1 text-error-500 dark:text-d-error-500 text-body-2 font-semibold"
                *ngIf="tdsRequired">*</span>
    </label>
  `
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { tdsFor: [{
            type: Input
        }], tdsRequired: [{
            type: Input
        }] }); })();
