// tds-ui/esm2020/form-field/prefix.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, InjectionToken } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Injection token that can be used to reference instances of `TDSPrefix`. It serves as
 * alternative token to the actual `TDSPrefix` class which could cause unnecessary
 * retention of the class and its directive metadata.
 */
export const TDS_PREFIX = new InjectionToken('TDSPrefix');
/** Prefix to be placed in front of the form field. */
export class TDSPrefix {
}
TDSPrefix.ɵfac = function TDSPrefix_Factory(t) { return new (t || TDSPrefix)(); };
TDSPrefix.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSPrefix, selectors: [["", "tdsPrefix", ""]], hostAttrs: [1, "tds-prefix", "text-neutral-1-300", "dark:text-d-neutral-1-900"], features: [i0.ɵɵProvidersFeature([{ provide: TDS_PREFIX, useExisting: TDSPrefix }])] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPrefix, [{
        type: Directive,
        args: [{
                selector: '[tdsPrefix]',
                providers: [{ provide: TDS_PREFIX, useExisting: TDSPrefix }],
                host: {
                    class: 'tds-prefix  text-neutral-1-300 dark:text-d-neutral-1-900'
                }
            }]
    }], null, null); })();
