// tds-ui/esm2020/form-field/suffix.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, InjectionToken } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Injection token that can be used to reference instances of `TDSSuffix`. It serves as
 * alternative token to the actual `TDSSuffix` class which could cause unnecessary
 * retention of the class and its directive metadata.
 */
export const TDS_SUFFIX = new InjectionToken('TDSSuffix');
/** Suffix to be placed at the end of the form field. */
export class TDSSuffix {
}
TDSSuffix.ɵfac = function TDSSuffix_Factory(t) { return new (t || TDSSuffix)(); };
TDSSuffix.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSSuffix, selectors: [["", "tdsSuffix", ""]], hostAttrs: [1, "tds-suffix"], exportAs: ["tdsSuffix"], features: [i0.ɵɵProvidersFeature([{ provide: TDS_SUFFIX, useExisting: TDSSuffix }])] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSuffix, [{
        type: Directive,
        args: [{
                selector: '[tdsSuffix]',
                exportAs: 'tdsSuffix',
                providers: [{ provide: TDS_SUFFIX, useExisting: TDSSuffix }],
                host: {
                    class: 'tds-suffix'
                }
            }]
    }], null, null); })();
