// tds-ui/esm2020/form-field/warning.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Attribute, ChangeDetectionStrategy, Component, InjectionToken, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["*"];
let nextUniqueId = 0;
/**
 * Injection token that can be used to reference instances of `TDSError`. It serves as
 * alternative token to the actual `TDSError` class which could cause unnecessary
 * retention of the class and its directive metadata.
 */
export const TDS_WARNING = new InjectionToken('TDSWarning');
/** Single success message to be shown underneath the form field. */
export class TDSWarning {
    constructor(ariaLive, elementRef) {
        this.id = `tds-warning-${nextUniqueId++}`;
        this.icon = 'tdsi-warning-fill text-base leading-4';
        // If no aria-live value is set add 'polite' as a default. This is preferred over setting
        // role='alert' so that screen readers do not interrupt the current task to read this aloud.
        if (!ariaLive) {
            elementRef.nativeElement.setAttribute('aria-live', 'polite');
        }
    }
}
TDSWarning.ɵfac = function TDSWarning_Factory(t) { return new (t || TDSWarning)(i0.ɵɵinjectAttribute('aria-live'), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSWarning.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSWarning, selectors: [["tds-warning"]], hostAttrs: ["aria-atomic", "true", 1, "tds-warning", "text-warning-500", "text-xs", "leading-4", "font-normal", "mt-1", "inline-flex", "items-center"], hostVars: 1, hostBindings: function TDSWarning_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("id", ctx.id);
    } }, inputs: { id: "id", icon: "icon" }, features: [i0.ɵɵProvidersFeature([{ provide: TDS_WARNING, useExisting: TDSWarning }])], ngContentSelectors: _c0, decls: 2, vars: 1, consts: [[1, "mr-2", 3, "ngClass"]], template: function TDSWarning_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelement(0, "i", 0);
        i0.ɵɵprojection(1);
    } if (rf & 2) {
        i0.ɵɵproperty("ngClass", ctx.icon);
    } }, directives: [i1.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSWarning, [{
        type: Component,
        args: [{
                selector: 'tds-warning',
                template: `
    <i class='mr-2' [ngClass]="icon"> 
    </i>
    <ng-content></ng-content>
  `,
                host: {
                    'class': 'tds-warning text-warning-500 text-xs leading-4 font-normal mt-1 inline-flex items-center',
                    '[attr.id]': 'id',
                    'aria-atomic': 'true',
                },
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [{ provide: TDS_WARNING, useExisting: TDSWarning }],
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Attribute,
                args: ['aria-live']
            }] }, { type: i0.ElementRef }]; }, { id: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
