// tds-ui/esm2020/switch/switch.module.mjs
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { TDSSwitch } from "./switch";
import * as i0 from "@angular/core";
export class TDSSwitchModule {
}
TDSSwitchModule.ɵfac = function TDSSwitchModule_Factory(t) { return new (t || TDSSwitchModule)(); };
TDSSwitchModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSSwitchModule });
TDSSwitchModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            ReactiveFormsModule,
            FormsModule,
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSwitchModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSSwitch],
                imports: [
                    ReactiveFormsModule,
                    FormsModule,
                    CommonModule
                ],
                exports: [
                    TDSSwitch
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSSwitchModule, { declarations: [TDSSwitch], imports: [ReactiveFormsModule,
        FormsModule,
        CommonModule], exports: [TDSSwitch] }); })();
