// tds-ui/esm2020/drawer/drawer.service.mjs
import { ComponentPortal } from '@angular/cdk/portal';
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TDSDrawerComponent } from './drawer.component';
import { TDSDrawerServiceModule } from './drawer.service.module';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
export class DrawerBuilderForService {
    constructor(overlay, options) {
        this.overlay = overlay;
        this.options = options;
        this.unsubscribe$ = new Subject();
        /** pick {@link TDSDrawerOptions.onCancel} and omit this option */
        const { onCancel, ...componentOption } = this.options;
        this.overlayRef = this.overlay.create();
        this.drawerRef = this.overlayRef.attach(new ComponentPortal(TDSDrawerComponent)).instance;
        this.updateOptions(componentOption);
        // Prevent repeatedly open drawer when tap focus element.
        this.drawerRef.savePreviouslyFocusedElement();
        this.drawerRef.onViewInit.pipe(takeUntil(this.unsubscribe$)).subscribe(() => {
            this.drawerRef.open();
        });
        this.drawerRef.onClose.subscribe(() => {
            if (onCancel) {
                onCancel().then(canClose => {
                    if (canClose !== false) {
                        this.drawerRef.close();
                    }
                });
            }
            else {
                this.drawerRef.close();
            }
        });
        this.drawerRef.afterClose.pipe(takeUntil(this.unsubscribe$)).subscribe(() => {
            this.overlayRef.dispose();
            this.drawerRef = null;
            this.unsubscribe$.next();
            this.unsubscribe$.complete();
        });
    }
    getInstance() {
        return this.drawerRef;
    }
    updateOptions(options) {
        Object.assign(this.drawerRef, options);
    }
}
export class TDSDrawerService {
    constructor(overlay) {
        this.overlay = overlay;
    }
    create(options) {
        return new DrawerBuilderForService(this.overlay, options).getInstance();
    }
}
TDSDrawerService.ɵfac = function TDSDrawerService_Factory(t) { return new (t || TDSDrawerService)(i0.ɵɵinject(i1.Overlay)); };
TDSDrawerService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TDSDrawerService, factory: TDSDrawerService.ɵfac, providedIn: TDSDrawerServiceModule });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDrawerService, [{
        type: Injectable,
        args: [{ providedIn: TDSDrawerServiceModule }]
    }], function () { return [{ type: i1.Overlay }]; }, null); })();
