// tds-ui/esm2020/drawer/drawer.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSDrawerContentDirective } from './drawer-content.directive';
import { TDSDrawerComponent } from './drawer.component';
import { TDSDrawerServiceModule } from './drawer.service.module';
import * as i0 from "@angular/core";
export class TDSDrawerModule {
}
TDSDrawerModule.ɵfac = function TDSDrawerModule_Factory(t) { return new (t || TDSDrawerModule)(); };
TDSDrawerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSDrawerModule });
TDSDrawerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            OverlayModule,
            PortalModule,
            TDSOutletModule,
            TDSDrawerServiceModule,
            TDSButtonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDrawerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    OverlayModule,
                    PortalModule,
                    TDSOutletModule,
                    TDSDrawerServiceModule,
                    TDSButtonModule
                ],
                exports: [TDSDrawerComponent, TDSDrawerContentDirective],
                declarations: [TDSDrawerComponent, TDSDrawerContentDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSDrawerModule, { declarations: [TDSDrawerComponent, TDSDrawerContentDirective], imports: [BidiModule,
        CommonModule,
        OverlayModule,
        PortalModule,
        TDSOutletModule,
        TDSDrawerServiceModule,
        TDSButtonModule], exports: [TDSDrawerComponent, TDSDrawerContentDirective] }); })();
