// tds-ui/esm2020/message/message-container.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TDSMNContainerComponent } from './base';
import { toCssPixel } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/config";
import * as i2 from "@angular/common";
import * as i3 from "./message.component";
function TDSMessageContainerComponent_tds_message_2_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "tds-message", 3);
    i0.ɵɵlistener("destroyed", function TDSMessageContainerComponent_tds_message_2_Template_tds_message_destroyed_0_listener($event) { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.remove($event.id, $event.userAction); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const instance_r1 = ctx.$implicit;
    i0.ɵɵproperty("instance", instance_r1);
} }
const TDS_CONFIG_COMPONENT_NAME = 'message';
const TDS_MESSAGE_DEFAULT_CONFIG = {
    animate: true,
    duration: 3000,
    maxStack: 7,
    pauseOnHover: true,
    top: 24,
    direction: 'ltr'
};
export class TDSMessageContainerComponent extends TDSMNContainerComponent {
    constructor(cdr, tdsConfigService) {
        super(cdr, tdsConfigService);
        this.destroy$ = new Subject();
        this.dir = 'ltr';
        this.instances = [];
        const config = TDS_MESSAGE_DEFAULT_CONFIG;
        this.dir = config?.direction || 'ltr';
    }
    subscribeConfigChange() {
        this.tdsConfigService
            .getConfigChangeEventForComponent(TDS_CONFIG_COMPONENT_NAME)
            .pipe(takeUntil(this.destroy$))
            .subscribe(() => {
            this.updateConfig();
            const config = this.tdsConfigService.getConfigForComponent(TDS_CONFIG_COMPONENT_NAME);
            if (config) {
                const { direction } = config;
                this.dir = direction || this.dir;
            }
        });
    }
    updateConfig() {
        this.config = {
            ...TDS_MESSAGE_DEFAULT_CONFIG,
            ...this.config,
            ...this.tdsConfigService.getConfigForComponent(TDS_CONFIG_COMPONENT_NAME)
        };
        this.top = toCssPixel(this.config.top);
        this.cdr.markForCheck();
    }
}
TDSMessageContainerComponent.ɵfac = function TDSMessageContainerComponent_Factory(t) { return new (t || TDSMessageContainerComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSConfigService)); };
TDSMessageContainerComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSMessageContainerComponent, selectors: [["tds-message-container"]], exportAs: ["tdsMessageContainer"], features: [i0.ɵɵInheritDefinitionFeature], decls: 3, vars: 5, consts: [[1, "tds-message", "pointer-events-none", "w-full", "left-0", "fixed"], [1, "w-full", "flex", "items-center", "flex-col"], [3, "instance", "destroyed", 4, "ngFor", "ngForOf"], [3, "instance", "destroyed"]], template: function TDSMessageContainerComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0)(1, "div", 1);
        i0.ɵɵtemplate(2, TDSMessageContainerComponent_tds_message_2_Template, 1, 1, "tds-message", 2);
        i0.ɵɵelementEnd()();
    } if (rf & 2) {
        i0.ɵɵstyleProp("top", ctx.top);
        i0.ɵɵclassProp("tds-message-rtl", ctx.dir === "rtl");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngForOf", ctx.instances);
    } }, directives: [i2.NgForOf, i3.TDSMessageComponent], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMessageContainerComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-message-container',
                exportAs: 'tdsMessageContainer',
                preserveWhitespaces: false,
                template: `
    <div class="tds-message pointer-events-none w-full left-0 fixed "  [class.tds-message-rtl]="dir === 'rtl'" [style.top]="top">
      <div class="w-full flex items-center flex-col">
          <tds-message *ngFor="let instance of instances" [instance]="instance" (destroyed)="remove($event.id, $event.userAction)"></tds-message>
      </div>
    </div>
  `
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.TDSConfigService }]; }, null); })();
