// tds-ui/esm2020/message/message.service.mjs
import { Injectable } from '@angular/core';
import { TDSMNService } from './base';
import { TDSMessageContainerComponent } from './message-container.component';
import { TDSMessageServiceModule } from './message.service.module';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/services";
import * as i2 from "@angular/cdk/overlay";
export class TDSMessageService extends TDSMNService {
    constructor(tdsSingletonService, overlay, injector) {
        super(tdsSingletonService, overlay, injector);
        this.componentPrefix = 'message-';
    }
    success(content, options) {
        return this.createInstance({ type: 'success', content }, options);
    }
    error(content, options) {
        return this.createInstance({ type: 'error', content }, options);
    }
    info(content, options) {
        return this.createInstance({ type: 'info', content }, options);
    }
    warning(content, options) {
        return this.createInstance({ type: 'warning', content }, options);
    }
    loading(content, options) {
        return this.createInstance({ type: 'loading', content }, options);
    }
    create(type, content, options) {
        return this.createInstance({ type, content }, options);
    }
    createInstance(message, options) {
        this.container = this.withContainer(TDSMessageContainerComponent);
        return this.container.create({
            ...message,
            ...{
                createdAt: new Date(),
                messageId: this.getInstanceId(),
                options
            }
        });
    }
}
TDSMessageService.ɵfac = function TDSMessageService_Factory(t) { return new (t || TDSMessageService)(i0.ɵɵinject(i1.TDSSingletonService), i0.ɵɵinject(i2.Overlay), i0.ɵɵinject(i0.Injector)); };
TDSMessageService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TDSMessageService, factory: TDSMessageService.ɵfac, providedIn: TDSMessageServiceModule });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMessageService, [{
        type: Injectable,
        args: [{
                providedIn: TDSMessageServiceModule
            }]
    }], function () { return [{ type: i1.TDSSingletonService }, { type: i2.Overlay }, { type: i0.Injector }]; }, null); })();
