// tds-ui/esm2020/message/message.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSMessageContainerComponent } from './message-container.component';
import { TDSMessageComponent } from './message.component';
import { TDSMessageServiceModule } from './message.service.module';
import * as i0 from "@angular/core";
export class TDSMessageModule {
}
TDSMessageModule.ɵfac = function TDSMessageModule_Factory(t) { return new (t || TDSMessageModule)(); };
TDSMessageModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSMessageModule });
TDSMessageModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSMessageServiceModule, TDSSpinnerModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMessageModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSMessageServiceModule, TDSSpinnerModule],
                declarations: [TDSMessageContainerComponent, TDSMessageComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSMessageModule, { declarations: [TDSMessageContainerComponent, TDSMessageComponent], imports: [BidiModule, CommonModule, OverlayModule, TDSOutletModule, TDSMessageServiceModule, TDSSpinnerModule] }); })();
