// tds-ui/esm2020/message/message.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { moveUpMotion } from 'tds-ui/core/animation';
import { TDSMNComponent } from './base';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/progress-spinner";
import * as i3 from "tds-ui/core/outlet";
function TDSMessageComponent_i_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 10);
} }
function TDSMessageComponent_i_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 11);
} }
function TDSMessageComponent_i_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 12);
} }
function TDSMessageComponent_i_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 13);
} }
function TDSMessageComponent_tds_spinner_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-spinner", 14);
} if (rf & 2) {
    i0.ɵɵproperty("color", "primary");
} }
function TDSMessageComponent_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "span", 15);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("innerHTML", ctx_r5.instance.content, i0.ɵɵsanitizeHtml);
} }
export class TDSMessageComponent extends TDSMNComponent {
    constructor(cdr) {
        super(cdr);
        this.destroyed = new EventEmitter();
    }
}
TDSMessageComponent.ɵfac = function TDSMessageComponent_Factory(t) { return new (t || TDSMessageComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSMessageComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSMessageComponent, selectors: [["tds-message"]], inputs: { instance: "instance" }, outputs: { destroyed: "destroyed" }, exportAs: ["tdsMessage"], features: [i0.ɵɵInheritDefinitionFeature], decls: 10, vars: 9, consts: [[1, "tds-message-notice", "pb-1", 3, "mouseenter", "mouseleave"], [1, "tds-message-notice-content", "bg-white", "dark:bg-d-neutral-3-300", "px-4", "py-3", "shadow-1-xl", "rounded-md"], [1, "tds-message-custom-content", "flex", "items-center", "text-center", 3, "ngClass"], [3, "ngSwitch"], ["class", "text-success-400 tdsi-success-fill text-2xl leading-6 mr-2.5", 4, "ngSwitchCase"], ["class", "text-info-400 tdsi-information-fill text-2xl leading-6 mr-2.5", 4, "ngSwitchCase"], ["class", "text-warning-400  tdsi-warning-fill text-2xl leading-6 mr-2.5", 4, "ngSwitchCase"], ["class", "text-error-400  tdsi-error-fill text-2xl leading-6 mr-2.5", 4, "ngSwitchCase"], ["class", "w-5 h-5 mr-2.5", 3, "color", 4, "ngSwitchCase"], [4, "tdsStringTemplateOutlet"], [1, "text-success-400", "tdsi-success-fill", "text-2xl", "leading-6", "mr-2.5"], [1, "text-info-400", "tdsi-information-fill", "text-2xl", "leading-6", "mr-2.5"], [1, "text-warning-400", "tdsi-warning-fill", "text-2xl", "leading-6", "mr-2.5"], [1, "text-error-400", "tdsi-error-fill", "text-2xl", "leading-6", "mr-2.5"], [1, "w-5", "h-5", "mr-2.5", 3, "color"], [1, "text-body-2", "text-neutral-1-900", "dark:text-d-neutral-1-900", "font-regular", 3, "innerHTML"]], template: function TDSMessageComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵlistener("@moveUpMotion.done", function TDSMessageComponent_Template_div_animation_moveUpMotion_done_0_listener($event) { return ctx.animationStateChanged.next($event); })("mouseenter", function TDSMessageComponent_Template_div_mouseenter_0_listener() { return ctx.onEnter(); })("mouseleave", function TDSMessageComponent_Template_div_mouseleave_0_listener() { return ctx.onLeave(); });
        i0.ɵɵelementStart(1, "div", 1)(2, "div", 2);
        i0.ɵɵelementContainerStart(3, 3);
        i0.ɵɵtemplate(4, TDSMessageComponent_i_4_Template, 1, 0, "i", 4);
        i0.ɵɵtemplate(5, TDSMessageComponent_i_5_Template, 1, 0, "i", 5);
        i0.ɵɵtemplate(6, TDSMessageComponent_i_6_Template, 1, 0, "i", 6);
        i0.ɵɵtemplate(7, TDSMessageComponent_i_7_Template, 1, 0, "i", 7);
        i0.ɵɵtemplate(8, TDSMessageComponent_tds_spinner_8_Template, 1, 1, "tds-spinner", 8);
        i0.ɵɵelementContainerEnd();
        i0.ɵɵtemplate(9, TDSMessageComponent_ng_container_9_Template, 2, 1, "ng-container", 9);
        i0.ɵɵelementEnd()()();
    } if (rf & 2) {
        i0.ɵɵproperty("@moveUpMotion", ctx.instance.state);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngClass", "tds-message-" + ctx.instance.type);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitch", ctx.instance.type);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "success");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "info");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "warning");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "error");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "loading");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.instance.content);
    } }, directives: [i1.NgClass, i1.NgSwitch, i1.NgSwitchCase, i2.TDSProgressSpinnerComponent, i3.TDSStringTemplateOutletDirective], encapsulation: 2, data: { animation: [moveUpMotion] }, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMessageComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-message',
                exportAs: 'tdsMessage',
                preserveWhitespaces: false,
                animations: [moveUpMotion],
                template: `
    <div
      class="tds-message-notice pb-1"
      [@moveUpMotion]="instance.state"
      (@moveUpMotion.done)="animationStateChanged.next($event)"
      (mouseenter)="onEnter()"
      (mouseleave)="onLeave()"
    >
      <div class="tds-message-notice-content  bg-white dark:bg-d-neutral-3-300 px-4 py-3 shadow-1-xl rounded-md ">
        <div class="tds-message-custom-content flex items-center text-center" [ngClass]="'tds-message-' + instance.type">
          <ng-container [ngSwitch]="instance.type">
            <i *ngSwitchCase="'success'" class="text-success-400 tdsi-success-fill text-2xl leading-6 mr-2.5"></i>
            <i *ngSwitchCase="'info'" class="text-info-400 tdsi-information-fill text-2xl leading-6 mr-2.5"></i>
            <i *ngSwitchCase="'warning'" class="text-warning-400  tdsi-warning-fill text-2xl leading-6 mr-2.5"></i>
            <i *ngSwitchCase="'error'" class="text-error-400  tdsi-error-fill text-2xl leading-6 mr-2.5"></i> 
            <tds-spinner *ngSwitchCase="'loading'"  [color]="'primary'" class="w-5 h-5 mr-2.5"></tds-spinner>
               
          </ng-container>
          <ng-container *tdsStringTemplateOutlet="instance.content">
            <span class="text-body-2 text-neutral-1-900 dark:text-d-neutral-1-900 font-regular" [innerHTML]="instance.content"></span>
          </ng-container>
        </div>
      </div>
    </div>
  `,
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { instance: [{
            type: Input
        }], destroyed: [{
            type: Output
        }] }); })();
