// tds-ui/esm2020/upload/upload.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSProgressModule } from 'tds-ui/progress';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSUploadBtnComponent } from './upload-btn.component';
import { TDSUploadListComponent } from './upload-list.component';
import { TDSUploadComponent } from './upload.component';
import * as i0 from "@angular/core";
export class TDSUploadModule {
}
TDSUploadModule.ɵfac = function TDSUploadModule_Factory(t) { return new (t || TDSUploadModule)(); };
TDSUploadModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSUploadModule });
TDSUploadModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            FormsModule,
            PlatformModule,
            TDSToolTipModule,
            TDSProgressModule,
            TDSI18nModule,
            TDSButtonModule,
            TDSSpinnerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSUploadModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    FormsModule,
                    PlatformModule,
                    TDSToolTipModule,
                    TDSProgressModule,
                    TDSI18nModule,
                    TDSButtonModule,
                    TDSSpinnerModule
                ],
                declarations: [TDSUploadComponent, TDSUploadBtnComponent, TDSUploadListComponent],
                exports: [TDSUploadComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSUploadModule, { declarations: [TDSUploadComponent, TDSUploadBtnComponent, TDSUploadListComponent], imports: [BidiModule,
        CommonModule,
        FormsModule,
        PlatformModule,
        TDSToolTipModule,
        TDSProgressModule,
        TDSI18nModule,
        TDSButtonModule,
        TDSSpinnerModule], exports: [TDSUploadComponent] }); })();
