// tds-ui/esm2020/button-menu/menu-module.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CdkScrollableModule } from '@angular/cdk/scrolling';
import { TDSButtonMenu } from './menu';
import { TDSButtonMenuContent } from './menu-content';
import { TDSButtonMenuItem } from './menu-item';
import { TDSBUTTON_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER, TDSButtonMenuTrigger } from './menu-trigger';
import * as i0 from "@angular/core";
/**
 * Used by both the current `TDSButtonMenuModule` and the MDC `TDSButtonMenuModule`
 * to declare the menu-related directives.
 */
export class _TDSButtonMenuDirectivesModule {
}
_TDSButtonMenuDirectivesModule.ɵfac = function _TDSButtonMenuDirectivesModule_Factory(t) { return new (t || _TDSButtonMenuDirectivesModule)(); };
_TDSButtonMenuDirectivesModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: _TDSButtonMenuDirectivesModule });
_TDSButtonMenuDirectivesModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [TDSBUTTON_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(_TDSButtonMenuDirectivesModule, [{
        type: NgModule,
        args: [{
                exports: [TDSButtonMenuTrigger, TDSButtonMenuContent],
                declarations: [
                    TDSButtonMenuTrigger,
                    TDSButtonMenuContent,
                ],
                providers: [TDSBUTTON_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(_TDSButtonMenuDirectivesModule, { declarations: [TDSButtonMenuTrigger,
        TDSButtonMenuContent], exports: [TDSButtonMenuTrigger, TDSButtonMenuContent] }); })();
export class TDSButtonMenuModule {
}
TDSButtonMenuModule.ɵfac = function TDSButtonMenuModule_Factory(t) { return new (t || TDSButtonMenuModule)(); };
TDSButtonMenuModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSButtonMenuModule });
TDSButtonMenuModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [TDSBUTTON_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER], imports: [[
            CommonModule,
            OverlayModule,
            _TDSButtonMenuDirectivesModule,
        ], CdkScrollableModule, _TDSButtonMenuDirectivesModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonMenuModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    OverlayModule,
                    _TDSButtonMenuDirectivesModule,
                ],
                exports: [CdkScrollableModule, TDSButtonMenu, TDSButtonMenuItem, _TDSButtonMenuDirectivesModule],
                declarations: [TDSButtonMenu, TDSButtonMenuItem],
                providers: [TDSBUTTON_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSButtonMenuModule, { declarations: [TDSButtonMenu, TDSButtonMenuItem], imports: [CommonModule,
        OverlayModule, _TDSButtonMenuDirectivesModule], exports: [CdkScrollableModule, TDSButtonMenu, TDSButtonMenuItem, _TDSButtonMenuDirectivesModule] }); })();
