// tds-ui/esm2020/badges/badge.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { ObserversModule } from '@angular/cdk/observers';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSNoAnimationModule } from 'tds-ui/core/no-animation';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSBadgeSupComponent } from './badge-sup.component';
import { TDSBadgeComponent } from './badge.component';
import { TDSRibbonComponent } from './ribbon.component';
import * as i0 from "@angular/core";
export class TDSBadgeModule {
}
TDSBadgeModule.ɵfac = function TDSBadgeModule_Factory(t) { return new (t || TDSBadgeModule)(); };
TDSBadgeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSBadgeModule });
TDSBadgeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, ObserversModule, TDSOutletModule, TDSNoAnimationModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSBadgeModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSBadgeComponent, TDSBadgeSupComponent, TDSRibbonComponent],
                exports: [TDSBadgeComponent, TDSRibbonComponent],
                imports: [BidiModule, CommonModule, ObserversModule, TDSOutletModule, TDSNoAnimationModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSBadgeModule, { declarations: [TDSBadgeComponent, TDSBadgeSupComponent, TDSRibbonComponent], imports: [BidiModule, CommonModule, ObserversModule, TDSOutletModule, TDSNoAnimationModule], exports: [TDSBadgeComponent, TDSRibbonComponent] }); })();
