// tds-ui/esm2020/badges/ribbon.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { badgePresetColors } from './preset-colors';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/outlet";
function TDSRibbonComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r0.text);
} }
const _c0 = ["*"];
export class TDSRibbonComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.placement = 'end';
        this.text = null;
        this.presetColor = null;
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-ribbon-wrapper');
    }
    ngOnChanges(changes) {
        const { color } = changes;
        if (color) {
            this.presetColor = this.color && badgePresetColors.indexOf(this.color) !== -1 ? this.color : null;
        }
    }
}
TDSRibbonComponent.ɵfac = function TDSRibbonComponent_Factory(t) { return new (t || TDSRibbonComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSRibbonComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSRibbonComponent, selectors: [["tds-ribbon"]], inputs: { color: "color", placement: "placement", text: "text" }, exportAs: ["tdsRibbon"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 4, vars: 11, consts: [[1, "tds-ribbon"], [4, "tdsStringTemplateOutlet"], [1, "tds-ribbon-corner"]], template: function TDSRibbonComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
        i0.ɵɵelementStart(1, "div", 0);
        i0.ɵɵtemplate(2, TDSRibbonComponent_ng_container_2_Template, 2, 1, "ng-container", 1);
        i0.ɵɵelement(3, "div", 2);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵclassMap(ctx.presetColor && "tds-ribbon-color-" + ctx.presetColor);
        i0.ɵɵstyleProp("background-color", !ctx.presetColor && ctx.color);
        i0.ɵɵclassProp("tds-ribbon-placement-end", ctx.placement === "end")("tds-ribbon-placement-start", ctx.placement === "start");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.text);
        i0.ɵɵadvance(1);
        i0.ɵɵstyleProp("color", !ctx.presetColor && ctx.color);
    } }, directives: [i1.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRibbonComponent, [{
        type: Component,
        args: [{
                selector: 'tds-ribbon',
                exportAs: 'tdsRibbon',
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <ng-content></ng-content>
    <div
      class="tds-ribbon"
      [class]="presetColor && 'tds-ribbon-color-' + presetColor"
      [class.tds-ribbon-placement-end]="placement === 'end'"
      [class.tds-ribbon-placement-start]="placement === 'start'"
      [style.background-color]="!presetColor && color"
    >
      <ng-container *tdsStringTemplateOutlet="text">{{ text }}</ng-container>
      <div class="tds-ribbon-corner" [style.color]="!presetColor && color"></div>
    </div>
  `
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { color: [{
            type: Input
        }], placement: [{
            type: Input
        }], text: [{
            type: Input
        }] }); })();
