// tds-ui/fesm2020/tds-ui-header.mjs
import * as i0 from '@angular/core';
import { EventEmitter, Component, Input, Output, ContentChildren, NgModule } from '@angular/core';
import { TDSFormField, TDSFormFieldModule } from 'tds-ui/form-field';
import { TDSSelect } from 'tds-ui/select';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i2 from 'tds-ui/avatar';
import { TDSAvatarModule } from 'tds-ui/avatar';
import { TDSInputModule } from 'tds-ui/tds-input';
import { TDSButtonModule } from 'tds-ui/button';

function TDSHeaderComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 12)(1, "img", 13);
    i0.ɵɵlistener("click", function TDSHeaderComponent_div_1_Template_img_click_1_listener($event) { i0.ɵɵrestoreView(_r6); const ctx_r5 = i0.ɵɵnextContext(); return ctx_r5.onClickLogo($event); });
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("src", ctx_r0.urlLogo, i0.ɵɵsanitizeUrl);
} }
function TDSHeaderComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r8 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 14)(1, "img", 15);
    i0.ɵɵlistener("click", function TDSHeaderComponent_div_2_Template_img_click_1_listener($event) { i0.ɵɵrestoreView(_r8); const ctx_r7 = i0.ɵɵnextContext(); return ctx_r7.onClickLogo($event); });
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("src", ctx_r1.urlLogoText, i0.ɵɵsanitizeUrl);
} }
function TDSHeaderComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 16);
    i0.ɵɵprojection(1, 2);
    i0.ɵɵelementEnd();
} }
function TDSHeaderComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 17)(1, "div", 18);
    i0.ɵɵprojection(2, 3);
    i0.ɵɵelementEnd()();
} }
function TDSHeaderComponent_div_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 19);
    i0.ɵɵelement(1, "tds-avatar", 20);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsSrc", ctx_r4.urlAvatar);
} }
const _c0 = [[["", 8, "tds-header-toolbar-left"]], [["", 8, "tds-header-toolbar-right"]], [["tds-form-field"]], [["tds-select"]]];
const _c1 = [".tds-header-toolbar-left", ".tds-header-toolbar-right", "tds-form-field", "tds-select"];
class TDSHeaderComponent {
    constructor() {
        this.urlAvatar = null;
        this.urlLogo = null;
        this.urlLogoText = null;
        this.onToggle = new EventEmitter();
        this.clickLogo = new EventEmitter();
    }
    ngOnInit() {
    }
    onClickToggle(e) {
        this.onToggle.emit();
    }
    onClickLogo(e) {
        this.clickLogo.emit(e);
    }
}
TDSHeaderComponent.ɵfac = function TDSHeaderComponent_Factory(t) { return new (t || TDSHeaderComponent)(); };
TDSHeaderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSHeaderComponent, selectors: [["tds-header"]], contentQueries: function TDSHeaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSSelect, 4);
        i0.ɵɵcontentQuery(dirIndex, TDSFormField, 4);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx._selectChildren = _t);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx._formFieldChildren = _t);
    } }, hostAttrs: [1, "flex", "w-full", "dark:bg-d-neutral-3-200", "items-center", "justify-between", "py-1"], inputs: { urlAvatar: "urlAvatar", urlLogo: "urlLogo", urlLogoText: "urlLogoText" }, outputs: { onToggle: "onToggle", clickLogo: "clickLogo" }, ngContentSelectors: _c1, decls: 15, vars: 5, consts: [[1, "tds-header-left", "flex", "items-center", "pr-4"], ["class", "logo pl-2 block lg:hidden", 4, "ngIf"], ["class", "logo-text pl-10 hidden lg:block", 4, "ngIf"], [1, "tds-header-toggle", "px-2", "block", "md:hidden"], ["tds-flat-button-icon", "", "size", "md", 3, "click"], [1, "flex", "items-center"], [1, "tdsi-menu-2-fill"], ["class", "tds-header-search pl-4 lg:w-96", 4, "ngIf"], [1, "inline-flex"], [1, "tds-header-right", "flex", "pr-4"], ["class", "pl-4", 4, "ngIf"], ["class", "inline-flex tds-header-avatar pl-4", 4, "ngIf"], [1, "logo", "pl-2", "block", "lg:hidden"], ["logo", "", "width", "48", "height", "48", "viewBox", "0 0 48 48", "fill", "none", 1, "cursor-pointer", 3, "src", "click"], [1, "logo-text", "pl-10", "hidden", "lg:block"], ["width", "156", "height", "48", "viewBox", "0 0 156 48", "fill", "none", 1, "cursor-pointer", 3, "src", "click"], [1, "tds-header-search", "pl-4", "lg:w-96"], [1, "pl-4"], [1, "w-40"], [1, "inline-flex", "tds-header-avatar", "pl-4"], [3, "tdsSrc"]], template: function TDSHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c0);
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, TDSHeaderComponent_div_1_Template, 2, 1, "div", 1);
        i0.ɵɵtemplate(2, TDSHeaderComponent_div_2_Template, 2, 1, "div", 2);
        i0.ɵɵelementStart(3, "div", 3)(4, "button", 4);
        i0.ɵɵlistener("click", function TDSHeaderComponent_Template_button_click_4_listener($event) { return ctx.onClickToggle($event); });
        i0.ɵɵelementStart(5, "span", 5);
        i0.ɵɵelement(6, "i", 6);
        i0.ɵɵelementEnd()()();
        i0.ɵɵtemplate(7, TDSHeaderComponent_div_7_Template, 2, 0, "div", 7);
        i0.ɵɵelementStart(8, "div", 8);
        i0.ɵɵprojection(9);
        i0.ɵɵelementEnd()();
        i0.ɵɵelementStart(10, "div", 9)(11, "div", 8);
        i0.ɵɵprojection(12, 1);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(13, TDSHeaderComponent_div_13_Template, 3, 0, "div", 10);
        i0.ɵɵtemplate(14, TDSHeaderComponent_div_14_Template, 2, 1, "div", 11);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.urlLogo);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.urlLogoText);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", ctx._formFieldChildren.length > 0);
        i0.ɵɵadvance(6);
        i0.ɵɵproperty("ngIf", ctx._selectChildren.length);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.urlAvatar);
    } }, directives: [i1.NgIf, i2.TDSAvatarComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSHeaderComponent, [{
        type: Component,
        args: [{ selector: 'tds-header', host: {
                    class: "flex w-full dark:bg-d-neutral-3-200  items-center justify-between py-1"
                }, template: "<div class=\"tds-header-left flex items-center pr-4\">\r\n    <div class=\"logo pl-2 block lg:hidden\" *ngIf=\"urlLogo\">\r\n        <img class=\"cursor-pointer\" (click)=\"onClickLogo($event)\" [src]=\"urlLogo\" logo width=\"48\" height=\"48\" viewBox=\"0 0 48 48\" fill=\"none\">\r\n    </div>\r\n    <div class=\"logo-text pl-10 hidden lg:block\" *ngIf=\"urlLogoText\">\r\n        <img class=\"cursor-pointer\" (click)=\"onClickLogo($event)\" [src]=\"urlLogoText\" width=\"156\" height=\"48\" viewBox=\"0 0 156 48\" fill=\"none\">\r\n    </div>\r\n    <div class=\"tds-header-toggle px-2 block md:hidden\">\r\n        <button tds-flat-button-icon size=\"md\" (click)=\"onClickToggle($event)\">\r\n            <span class=\"flex items-center\">\r\n                <i class=\"tdsi-menu-2-fill\"></i>\r\n            </span>\r\n        </button>\r\n    </div>\r\n    <div class=\"tds-header-search pl-4 lg:w-96\" *ngIf=\"_formFieldChildren.length > 0\">\r\n        <ng-content select='tds-form-field'></ng-content>\r\n    </div>\r\n    <div class=\"inline-flex\">\r\n        <ng-content select='.tds-header-toolbar-left'></ng-content>\r\n    </div>\r\n</div>\r\n\r\n<div class=\"tds-header-right flex pr-4\">\r\n    <div class=\"inline-flex\">\r\n        <ng-content select='.tds-header-toolbar-right'></ng-content>\r\n    </div>\r\n\r\n    <div class=\"pl-4\" *ngIf=\"_selectChildren.length\">\r\n        <div class=\"w-40\">\r\n            <ng-content select='tds-select'></ng-content>\r\n        </div>\r\n    </div>\r\n    <div class=\"inline-flex tds-header-avatar pl-4\" *ngIf=\"urlAvatar\">\r\n        <tds-avatar [tdsSrc]=\"urlAvatar\">\r\n        </tds-avatar>\r\n    </div>\r\n</div>" }]
    }], function () { return []; }, { urlAvatar: [{
            type: Input
        }], urlLogo: [{
            type: Input
        }], urlLogoText: [{
            type: Input
        }], onToggle: [{
            type: Output
        }], clickLogo: [{
            type: Output
        }], _selectChildren: [{
            type: ContentChildren,
            args: [TDSSelect, { descendants: false }]
        }], _formFieldChildren: [{
            type: ContentChildren,
            args: [TDSFormField, { descendants: false }]
        }] }); })();

class TDSHeaderModule {
}
TDSHeaderModule.ɵfac = function TDSHeaderModule_Factory(t) { return new (t || TDSHeaderModule)(); };
TDSHeaderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSHeaderModule });
TDSHeaderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            TDSButtonModule,
            TDSFormFieldModule,
            TDSAvatarModule,
            TDSInputModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSHeaderModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSHeaderComponent
                ],
                exports: [
                    TDSHeaderComponent
                ],
                imports: [
                    CommonModule,
                    TDSButtonModule,
                    TDSFormFieldModule,
                    TDSAvatarModule,
                    TDSInputModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSHeaderModule, { declarations: [TDSHeaderComponent], imports: [CommonModule,
        TDSButtonModule,
        TDSFormFieldModule,
        TDSAvatarModule,
        TDSInputModule], exports: [TDSHeaderComponent] }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { TDSHeaderComponent, TDSHeaderModule };
