// tds-ui/fesm2020/tds-ui.mjs
import * as i0 from '@angular/core';
import { NgModule } from '@angular/core';
import { TDSAffixModule } from 'tds-ui/affix';
import { TDSAlertModule } from 'tds-ui/alert';
import { TDSAnchorModule } from 'tds-ui/anchor';
import { TDSAutocompleteModule } from 'tds-ui/auto-complete';
import { TDSAvatarModule } from 'tds-ui/avatar';
import { TDSBadgeModule } from 'tds-ui/badges';
import { TDSBreadCrumbModule } from 'tds-ui/breadcrumb';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSButtonMenuModule } from 'tds-ui/button-menu';
import { TDSButtonSpitModule } from 'tds-ui/buttton-split';
import { TDSCalendarModule } from 'tds-ui/calendar';
import { TDSCardModule } from 'tds-ui/card';
import { TDSCarouselModule } from 'tds-ui/carousel';
import { TDSCascaderModule } from 'tds-ui/cascader';
import { TDSCollapseModule } from 'tds-ui/collapse';
import { TDSDatePickerModule } from 'tds-ui/date-picker';
import { TDSDividerModule } from 'tds-ui/divider';
import { TDSDrawerModule } from 'tds-ui/drawer';
import { TDSDropDownModule } from 'tds-ui/dropdown';
import { TDSEmptyModule } from 'tds-ui/empty';
import { TDSFilterStatusModule } from 'tds-ui/filter-status';
import { TDSFlexModule } from 'tds-ui/flex';
import { TDSFormFieldModule } from 'tds-ui/form-field';
import { TDSHeaderModule } from 'tds-ui/header';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSImageModule } from 'tds-ui/image';
import { TDSFormInputModule } from 'tds-ui/input';
import { TDSInputNumberModule } from 'tds-ui/input-number';
import { TDSMenuModule } from 'tds-ui/menu';
import { TDSMessageModule } from 'tds-ui/message';
import { TDSModalModule } from 'tds-ui/modal';
import { TDSNotificationModule } from 'tds-ui/notification';
import { TDSPageHeaderModule } from 'tds-ui/page-header';
import { TDSPaginationModule } from 'tds-ui/pagination';
import { TDSPopoverModule } from 'tds-ui/popover';
import { TDSProgressModule } from 'tds-ui/progress';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSRadioModule } from 'tds-ui/radio';
import { TDSRateModule } from 'tds-ui/rate';
import { TDSScrollIntoViewModule } from 'tds-ui/scroll-into-view';
import { TDSSelectModule } from 'tds-ui/select';
import { TDSSkeletonModule } from 'tds-ui/skeleton';
import { TDSSliderModule } from 'tds-ui/slider';
import { TDSStepsModule } from 'tds-ui/step';
import { TDSSwitchModule } from 'tds-ui/switch';
import { TDSTableModule } from 'tds-ui/table';
import { TDSTabsModule } from 'tds-ui/tabs';
import { TDSTagModule } from 'tds-ui/tag';
import { TDSCheckBoxModule } from 'tds-ui/tds-checkbox';
import { TDSInputModule } from 'tds-ui/tds-input';
import { TDSTimePickerModule } from 'tds-ui/time-picker';
import { TDSTimelineModule } from 'tds-ui/timeline';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSTreeModule } from 'tds-ui/tree';
import { TDSTypographyModule } from 'tds-ui/typography';
import { TDSUploadModule } from 'tds-ui/upload';
import { TDSZoomBoxModule } from 'tds-ui/zoom-box';

const TMT_Module = [
    TDSButtonModule,
    TDSCardModule,
    // TInputModule,
    TDSSpinnerModule,
    TDSScrollIntoViewModule,
    TDSTabsModule,
    TDSBadgeModule,
    TDSFlexModule,
    TDSAvatarModule,
    TDSFlexModule,
    TDSButtonMenuModule,
    TDSFormFieldModule,
    TDSInputModule,
    TDSCheckBoxModule,
    TDSSelectModule,
    TDSMenuModule,
    TDSUploadModule,
    TDSTagModule,
    TDSToolTipModule,
    TDSRadioModule,
    TDSPopoverModule,
    TDSAlertModule,
    TDSMessageModule,
    TDSNotificationModule,
    TDSModalModule,
    TDSDrawerModule,
    TDSProgressModule,
    TDSTimePickerModule,
    TDSI18nModule,
    TDSDatePickerModule,
    TDSSwitchModule,
    TDSDropDownModule,
    TDSBreadCrumbModule,
    TDSButtonSpitModule,
    TDSAffixModule,
    TDSHeaderModule,
    TDSPageHeaderModule,
    TDSStepsModule,
    TDSCollapseModule,
    TDSTreeModule,
    TDSPaginationModule,
    TDSImageModule,
    TDSAutocompleteModule,
    TDSTypographyModule,
    TDSInputNumberModule,
    TDSRateModule,
    TDSTableModule,
    TDSEmptyModule,
    TDSFilterStatusModule,
    TDSAnchorModule,
    TDSDividerModule,
    TDSSkeletonModule,
    TDSCarouselModule,
    TDSTimelineModule,
    TDSCascaderModule,
    TDSZoomBoxModule,
    TDSSliderModule,
    TDSFormInputModule,
    TDSCalendarModule
];
class TDSModule {
}
TDSModule.ɵfac = function TDSModule_Factory(t) { return new (t || TDSModule)(); };
TDSModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSModule });
TDSModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            TMT_Module
        ], TDSButtonModule,
        TDSCardModule,
        // TInputModule,
        TDSSpinnerModule,
        TDSScrollIntoViewModule,
        TDSTabsModule,
        TDSBadgeModule,
        TDSFlexModule,
        TDSAvatarModule,
        TDSFlexModule,
        TDSButtonMenuModule,
        TDSFormFieldModule,
        TDSInputModule,
        TDSCheckBoxModule,
        TDSSelectModule,
        TDSMenuModule,
        TDSUploadModule,
        TDSTagModule,
        TDSToolTipModule,
        TDSRadioModule,
        TDSPopoverModule,
        TDSAlertModule,
        TDSMessageModule,
        TDSNotificationModule,
        TDSModalModule,
        TDSDrawerModule,
        TDSProgressModule,
        TDSTimePickerModule,
        TDSI18nModule,
        TDSDatePickerModule,
        TDSSwitchModule,
        TDSDropDownModule,
        TDSBreadCrumbModule,
        TDSButtonSpitModule,
        TDSAffixModule,
        TDSHeaderModule,
        TDSPageHeaderModule,
        TDSStepsModule,
        TDSCollapseModule,
        TDSTreeModule,
        TDSPaginationModule,
        TDSImageModule,
        TDSAutocompleteModule,
        TDSTypographyModule,
        TDSInputNumberModule,
        TDSRateModule,
        TDSTableModule,
        TDSEmptyModule,
        TDSFilterStatusModule,
        TDSAnchorModule,
        TDSDividerModule,
        TDSSkeletonModule,
        TDSCarouselModule,
        TDSTimelineModule,
        TDSCascaderModule,
        TDSZoomBoxModule,
        TDSSliderModule,
        TDSFormInputModule,
        TDSCalendarModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSModule, [{
        type: NgModule,
        args: [{
                imports: [
                    TMT_Module
                ],
                exports: [
                    TMT_Module,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSModule, { imports: [TDSButtonModule,
        TDSCardModule,
        // TInputModule,
        TDSSpinnerModule,
        TDSScrollIntoViewModule,
        TDSTabsModule,
        TDSBadgeModule,
        TDSFlexModule,
        TDSAvatarModule,
        TDSFlexModule,
        TDSButtonMenuModule,
        TDSFormFieldModule,
        TDSInputModule,
        TDSCheckBoxModule,
        TDSSelectModule,
        TDSMenuModule,
        TDSUploadModule,
        TDSTagModule,
        TDSToolTipModule,
        TDSRadioModule,
        TDSPopoverModule,
        TDSAlertModule,
        TDSMessageModule,
        TDSNotificationModule,
        TDSModalModule,
        TDSDrawerModule,
        TDSProgressModule,
        TDSTimePickerModule,
        TDSI18nModule,
        TDSDatePickerModule,
        TDSSwitchModule,
        TDSDropDownModule,
        TDSBreadCrumbModule,
        TDSButtonSpitModule,
        TDSAffixModule,
        TDSHeaderModule,
        TDSPageHeaderModule,
        TDSStepsModule,
        TDSCollapseModule,
        TDSTreeModule,
        TDSPaginationModule,
        TDSImageModule,
        TDSAutocompleteModule,
        TDSTypographyModule,
        TDSInputNumberModule,
        TDSRateModule,
        TDSTableModule,
        TDSEmptyModule,
        TDSFilterStatusModule,
        TDSAnchorModule,
        TDSDividerModule,
        TDSSkeletonModule,
        TDSCarouselModule,
        TDSTimelineModule,
        TDSCascaderModule,
        TDSZoomBoxModule,
        TDSSliderModule,
        TDSFormInputModule,
        TDSCalendarModule], exports: [TDSButtonModule,
        TDSCardModule,
        // TInputModule,
        TDSSpinnerModule,
        TDSScrollIntoViewModule,
        TDSTabsModule,
        TDSBadgeModule,
        TDSFlexModule,
        TDSAvatarModule,
        TDSFlexModule,
        TDSButtonMenuModule,
        TDSFormFieldModule,
        TDSInputModule,
        TDSCheckBoxModule,
        TDSSelectModule,
        TDSMenuModule,
        TDSUploadModule,
        TDSTagModule,
        TDSToolTipModule,
        TDSRadioModule,
        TDSPopoverModule,
        TDSAlertModule,
        TDSMessageModule,
        TDSNotificationModule,
        TDSModalModule,
        TDSDrawerModule,
        TDSProgressModule,
        TDSTimePickerModule,
        TDSI18nModule,
        TDSDatePickerModule,
        TDSSwitchModule,
        TDSDropDownModule,
        TDSBreadCrumbModule,
        TDSButtonSpitModule,
        TDSAffixModule,
        TDSHeaderModule,
        TDSPageHeaderModule,
        TDSStepsModule,
        TDSCollapseModule,
        TDSTreeModule,
        TDSPaginationModule,
        TDSImageModule,
        TDSAutocompleteModule,
        TDSTypographyModule,
        TDSInputNumberModule,
        TDSRateModule,
        TDSTableModule,
        TDSEmptyModule,
        TDSFilterStatusModule,
        TDSAnchorModule,
        TDSDividerModule,
        TDSSkeletonModule,
        TDSCarouselModule,
        TDSTimelineModule,
        TDSCascaderModule,
        TDSZoomBoxModule,
        TDSSliderModule,
        TDSFormInputModule,
        TDSCalendarModule] }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { TDSModule };
