// tds-ui/esm2020/calendar/calendar.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSButtonModule } from 'tds-ui/button';
import { LibPackerModule, TDSDatePickerModule } from 'tds-ui/date-picker';
import { TDSDropDownModule } from 'tds-ui/dropdown';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSRadioModule } from 'tds-ui/radio';
import { TDSSelectModule } from 'tds-ui/select';
import { TDSDateCellDirective, TDSDateFullCellDirective, TDSMonthCellDirective, TDSMonthFullCellDirective } from './calendar-cells';
import { TDSCalendarHeaderComponent } from './calendar-header.component';
import { TDSCalendarComponent } from './calendar.component';
import { TDSCalendarDatepickerComponent } from './tds-calendar-datepicker.component';
import { TDSCalendarModeDayComponent } from './tds-calendar-mode-day.component';
import { TDSCalendarModeWeekComponent } from './tds-calendar-mode-week.component';
import { TDSCalendarModeMonthComponent } from './tds-calendar-mode-month.component';
import { TDSCalendarModeYearComponent } from './tds-calendar-mode-year.component';
import { TDSCalendarEventComponent } from './tds-calendar-event.component';
import { TDSCalendarCurrentTimeComponent } from './tds-calendar-current-time.component';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import * as i0 from "@angular/core";
export class TDSCalendarModule {
}
TDSCalendarModule.ɵfac = function TDSCalendarModule_Factory(t) { return new (t || TDSCalendarModule)(); };
TDSCalendarModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCalendarModule });
TDSCalendarModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule,
            CommonModule,
            FormsModule,
            TDSI18nModule,
            TDSRadioModule,
            TDSSelectModule,
            LibPackerModule,
            TDSDropDownModule,
            TDSDatePickerModule,
            TDSRadioModule,
            TDSButtonModule,
            TDSOutletModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCalendarModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSCalendarHeaderComponent,
                    TDSCalendarComponent,
                    TDSDateCellDirective,
                    TDSDateFullCellDirective,
                    TDSMonthCellDirective,
                    TDSMonthFullCellDirective,
                    TDSCalendarDatepickerComponent,
                    TDSCalendarModeDayComponent,
                    TDSCalendarModeWeekComponent,
                    TDSCalendarModeMonthComponent,
                    TDSCalendarModeYearComponent,
                    TDSCalendarEventComponent,
                    TDSCalendarCurrentTimeComponent
                ],
                exports: [
                    TDSCalendarComponent,
                    TDSDateCellDirective,
                    TDSDateFullCellDirective,
                    TDSMonthCellDirective,
                    TDSMonthFullCellDirective,
                    TDSCalendarDatepickerComponent,
                    TDSCalendarModeDayComponent,
                    TDSCalendarModeWeekComponent,
                    TDSCalendarModeMonthComponent,
                    TDSCalendarModeYearComponent,
                    TDSCalendarEventComponent,
                    TDSCalendarHeaderComponent,
                    TDSCalendarCurrentTimeComponent
                ],
                imports: [BidiModule,
                    CommonModule,
                    FormsModule,
                    TDSI18nModule,
                    TDSRadioModule,
                    TDSSelectModule,
                    LibPackerModule,
                    TDSDropDownModule,
                    TDSDatePickerModule,
                    TDSRadioModule,
                    TDSButtonModule,
                    TDSOutletModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCalendarModule, { declarations: [TDSCalendarHeaderComponent,
        TDSCalendarComponent,
        TDSDateCellDirective,
        TDSDateFullCellDirective,
        TDSMonthCellDirective,
        TDSMonthFullCellDirective,
        TDSCalendarDatepickerComponent,
        TDSCalendarModeDayComponent,
        TDSCalendarModeWeekComponent,
        TDSCalendarModeMonthComponent,
        TDSCalendarModeYearComponent,
        TDSCalendarEventComponent,
        TDSCalendarCurrentTimeComponent], imports: [BidiModule,
        CommonModule,
        FormsModule,
        TDSI18nModule,
        TDSRadioModule,
        TDSSelectModule,
        LibPackerModule,
        TDSDropDownModule,
        TDSDatePickerModule,
        TDSRadioModule,
        TDSButtonModule,
        TDSOutletModule], exports: [TDSCalendarComponent,
        TDSDateCellDirective,
        TDSDateFullCellDirective,
        TDSMonthCellDirective,
        TDSMonthFullCellDirective,
        TDSCalendarDatepickerComponent,
        TDSCalendarModeDayComponent,
        TDSCalendarModeWeekComponent,
        TDSCalendarModeMonthComponent,
        TDSCalendarModeYearComponent,
        TDSCalendarEventComponent,
        TDSCalendarHeaderComponent,
        TDSCalendarCurrentTimeComponent] }); })();
