// tds-ui/esm2020/calendar/calendar-cells.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSDateCellDirective {
}
TDSDateCellDirective.ɵfac = function TDSDateCellDirective_Factory(t) { return new (t || TDSDateCellDirective)(); };
TDSDateCellDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSDateCellDirective, selectors: [["", "tdsDateCell", ""]], exportAs: ["tdsDateCell"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDateCellDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsDateCell]',
                exportAs: 'tdsDateCell'
            }]
    }], null, null); })();
export class TDSMonthCellDirective {
}
TDSMonthCellDirective.ɵfac = function TDSMonthCellDirective_Factory(t) { return new (t || TDSMonthCellDirective)(); };
TDSMonthCellDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSMonthCellDirective, selectors: [["", "tdsMonthCell", ""]], exportAs: ["tdsMonthCell"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMonthCellDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsMonthCell]',
                exportAs: 'tdsMonthCell'
            }]
    }], null, null); })();
export class TDSDateFullCellDirective {
}
TDSDateFullCellDirective.ɵfac = function TDSDateFullCellDirective_Factory(t) { return new (t || TDSDateFullCellDirective)(); };
TDSDateFullCellDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSDateFullCellDirective, selectors: [["", "tdsDateFullCell", ""]], exportAs: ["tdsDateFullCell"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDateFullCellDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsDateFullCell]',
                exportAs: 'tdsDateFullCell'
            }]
    }], null, null); })();
export class TDSMonthFullCellDirective {
}
TDSMonthFullCellDirective.ɵfac = function TDSMonthFullCellDirective_Factory(t) { return new (t || TDSMonthFullCellDirective)(); };
TDSMonthFullCellDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSMonthFullCellDirective, selectors: [["", "tdsMonthFullCell", ""]], exportAs: ["tdsMonthFullCell"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMonthFullCellDirective, [{
        type: Directive,
        args: [{
                selector: '[tdsMonthFullCell]',
                exportAs: 'tdsMonthFullCell'
            }]
    }], null, null); })();
