// tds-ui/esm2020/calendar/tds-calendar-current-time.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { differenceInMinutes, isSameDay, setHours, setMinutes } from 'date-fns';
import { BehaviorSubject, interval } from 'rxjs';
import { map, startWith, switchMapTo, takeUntil } from 'rxjs/operators';
import { TDSDestroyService } from 'tds-ui/core/services';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/services";
import * as i2 from "@angular/common";
function TDSCalendarCurrentTimeComponent_ng_template_0_div_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 3)(1, "div", 4)(2, "div", 5);
    i0.ɵɵtext(3);
    i0.ɵɵpipe(4, "date");
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext();
    const topPx_r9 = ctx_r12.topPx;
    const now_r10 = ctx_r12.now;
    const ctx_r11 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("top", topPx_r9, "px")("left", ctx_r11.widthColTime, "px");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(i0.ɵɵpipeBind2(4, 5, now_r10, "HH:mm"));
} }
function TDSCalendarCurrentTimeComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSCalendarCurrentTimeComponent_ng_template_0_div_0_Template, 5, 8, "div", 2);
} if (rf & 2) {
    const isVisible_r8 = ctx.isVisible;
    i0.ɵɵproperty("ngIf", isVisible_r8);
} }
function TDSCalendarCurrentTimeComponent_ng_template_2_Template(rf, ctx) { }
const _c0 = function (a0, a1, a2, a3, a4, a5, a6, a7) { return { columnDate: a0, dayStartHour: a1, dayStartMinute: a2, dayEndHour: a3, dayEndMinute: a4, isVisible: a5, topPx: a6, now: a7 }; };
export class TDSCalendarCurrentTimeComponent {
    constructor(zone, cdr, destroy$) {
        this.zone = zone;
        this.cdr = cdr;
        this.destroy$ = destroy$;
        this.columnDate$ = new BehaviorSubject(undefined);
        //1 phút hiển thị 1 lần
        this.marker$ = interval(60 * 1000).pipe(
        // switchMap(() => ),
        startWith(0), switchMapTo(this.columnDate$), map((columnDate) => {
            const now = new Date();
            const startOfDay = setMinutes(setHours(columnDate, this.dayStartHour), this.dayStartMinute);
            const endOfDay = setMinutes(setHours(columnDate, this.dayEndHour), this.dayEndMinute);
            const hourHeightModifier = (this.hourSegments * this.hourSegmentHeight) /
                (this.hourDuration || 60);
            return {
                isVisible: isSameDay(columnDate, now) &&
                    now >= startOfDay &&
                    now <= endOfDay,
                top: differenceInMinutes(now, startOfDay) *
                    hourHeightModifier,
                now: now
            };
        }), takeUntil(this.destroy$));
    }
    ngOnChanges(changes) {
        if (changes.columnDate) {
            this.columnDate$.next(changes.columnDate.currentValue);
        }
    }
}
TDSCalendarCurrentTimeComponent.ɵfac = function TDSCalendarCurrentTimeComponent_Factory(t) { return new (t || TDSCalendarCurrentTimeComponent)(i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSDestroyService)); };
TDSCalendarCurrentTimeComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCalendarCurrentTimeComponent, selectors: [["tds-calendar-current-time"]], inputs: { columnDate: "columnDate", dayStartHour: "dayStartHour", dayStartMinute: "dayStartMinute", dayEndHour: "dayEndHour", dayEndMinute: "dayEndMinute", hourSegments: "hourSegments", hourDuration: "hourDuration", hourSegmentHeight: "hourSegmentHeight", customTemplate: "customTemplate", widthColTime: "widthColTime" }, features: [i0.ɵɵProvidersFeature([
            TDSDestroyService
        ]), i0.ɵɵNgOnChangesFeature], decls: 6, vars: 17, consts: [["defaultTemplate", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["class", "cal-current-time-marker bg-primary-1 absolute inset-0 h-px ", 3, "top", "left", 4, "ngIf"], [1, "cal-current-time-marker", "bg-primary-1", "absolute", "inset-0", "h-px"], [1, "w-full", "flex", "items-center", "justify-center", "relative"], [1, "absolute", "-left-[60px]", "text-primary-1", "z-10", "bg-white"]], template: function TDSCalendarCurrentTimeComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSCalendarCurrentTimeComponent_ng_template_0_Template, 1, 1, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(2, TDSCalendarCurrentTimeComponent_ng_template_2_Template, 0, 0, "ng-template", 1);
        i0.ɵɵpipe(3, "async");
        i0.ɵɵpipe(4, "async");
        i0.ɵɵpipe(5, "async");
    } if (rf & 2) {
        const _r0 = i0.ɵɵreference(1);
        let tmp_1_0;
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.customTemplate || _r0)("ngTemplateOutletContext", i0.ɵɵpureFunction8(8, _c0, ctx.columnDate, ctx.dayStartHour, ctx.dayStartMinute, ctx.dayEndHour, ctx.dayEndMinute, (tmp_1_0 = i0.ɵɵpipeBind1(3, 2, ctx.marker$)) == null ? null : tmp_1_0.isVisible, (tmp_1_0 = i0.ɵɵpipeBind1(4, 4, ctx.marker$)) == null ? null : tmp_1_0.top, (tmp_1_0 = i0.ɵɵpipeBind1(5, 6, ctx.marker$)) == null ? null : tmp_1_0.now));
    } }, directives: [i2.NgIf, i2.NgTemplateOutlet], pipes: [i2.DatePipe, i2.AsyncPipe], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCalendarCurrentTimeComponent, [{
        type: Component,
        args: [{ selector: 'tds-calendar-current-time', changeDetection: ChangeDetectionStrategy.OnPush, providers: [
                    TDSDestroyService
                ], template: "<ng-template #defaultTemplate let-columnDate=\"columnDate\" let-dayStartHour=\"dayStartHour\"\n    let-dayStartMinute=\"dayStartMinute\" let-dayEndHour=\"dayEndHour\" let-dayEndMinute=\"dayEndMinute\"\n    let-isVisible=\"isVisible\" let-topPx=\"topPx\" let-now=\"now\">\n\n    <div class=\"cal-current-time-marker bg-primary-1 absolute inset-0 h-px \" *ngIf=\"isVisible\"\n        [style.top.px]=\"topPx\" [style.left.px]=\"widthColTime\">\n        <div class=\"w-full flex items-center justify-center relative \">\n            <div class=\"absolute -left-[60px] text-primary-1 z-10  bg-white\">{{ now | date:\"HH:mm\"}}</div>            \n        </div>\n    </div>\n</ng-template>\n<ng-template [ngTemplateOutlet]=\"customTemplate || defaultTemplate\" [ngTemplateOutletContext]=\"{\n  columnDate: columnDate,\n  dayStartHour: dayStartHour,\n  dayStartMinute: dayStartMinute,\n  dayEndHour: dayEndHour,\n  dayEndMinute: dayEndMinute,\n  isVisible: (marker$ | async)?.isVisible,\n  topPx: (marker$ | async)?.top,\n  now: (marker$ | async)?.now\n}\">\n</ng-template>" }]
    }], function () { return [{ type: i0.NgZone }, { type: i0.ChangeDetectorRef }, { type: i1.TDSDestroyService }]; }, { columnDate: [{
            type: Input
        }], dayStartHour: [{
            type: Input
        }], dayStartMinute: [{
            type: Input
        }], dayEndHour: [{
            type: Input
        }], dayEndMinute: [{
            type: Input
        }], hourSegments: [{
            type: Input
        }], hourDuration: [{
            type: Input
        }], hourSegmentHeight: [{
            type: Input
        }], customTemplate: [{
            type: Input
        }], widthColTime: [{
            type: Input
        }] }); })();
