// tds-ui/esm2020/calendar/tds-calendar-mode-week.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSCalendarModeWeekComponent {
    constructor() { }
    ngOnInit() {
    }
}
TDSCalendarModeWeekComponent.ɵfac = function TDSCalendarModeWeekComponent_Factory(t) { return new (t || TDSCalendarModeWeekComponent)(); };
TDSCalendarModeWeekComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCalendarModeWeekComponent, selectors: [["tds-calendar-mode-week"]], decls: 2, vars: 0, template: function TDSCalendarModeWeekComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "p");
        i0.ɵɵtext(1, "tds-calendar-mode-week works!");
        i0.ɵɵelementEnd();
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCalendarModeWeekComponent, [{
        type: Component,
        args: [{ selector: 'tds-calendar-mode-week', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<p>tds-calendar-mode-week works!</p>\n" }]
    }], function () { return []; }, null); })();
