// tds-ui/esm2020/calendar/tds-calendar-mode-day.component.mjs
import { ChangeDetectionStrategy, Component, ContentChildren, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { CandyDate } from 'tds-ui/core/time';
import { getDayViewHourGrid, trackByHourSegment, getDayView } from './calendar.utils';
import { TDSCalendarEventComponent } from './tds-calendar-event.component';
import { startWith, switchMap, takeUntil } from 'rxjs/operators';
import { BehaviorSubject, merge } from 'rxjs';
import { TDSDestroyService } from 'tds-ui/core/services';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/services";
import * as i2 from "@angular/common";
import * as i3 from "./tds-calendar-current-time.component";
import * as i4 from "tds-ui/core/outlet";
function TDSCalendarModeDayComponent_ng_template_6_ng_template_0_tr_0_Template(rf, ctx) { if (rf & 1) {
    const _r11 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "tr", 9)(1, "td", 10);
    i0.ɵɵtext(2);
    i0.ɵɵpipe(3, "date");
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "td", 11);
    i0.ɵɵlistener("click", function TDSCalendarModeDayComponent_ng_template_6_ng_template_0_tr_0_Template_td_click_4_listener($event) { i0.ɵɵrestoreView(_r11); const segment_r6 = i0.ɵɵnextContext().$implicit; const ctx_r9 = i0.ɵɵnextContext(2); return ctx_r9.onClickSegment($event, segment_r6); });
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const segment_r6 = i0.ɵɵnextContext().$implicit;
    const week_r2 = i0.ɵɵnextContext().$implicit;
    const ctx_r7 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵattribute("rowspan", week_r2.segments.length);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", i0.ɵɵpipeBind2(3, 4, segment_r6.displayDate, "HH:mm"), " ");
    i0.ɵɵadvance(2);
    i0.ɵɵstyleProp("height", ctx_r7.hourSegmentHeight, "px");
} }
function TDSCalendarModeDayComponent_ng_template_6_ng_template_0_tr_1_Template(rf, ctx) { if (rf & 1) {
    const _r16 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "tr", 12)(1, "td", 11);
    i0.ɵɵlistener("click", function TDSCalendarModeDayComponent_ng_template_6_ng_template_0_tr_1_Template_td_click_1_listener($event) { i0.ɵɵrestoreView(_r16); const segment_r6 = i0.ɵɵnextContext().$implicit; const ctx_r14 = i0.ɵɵnextContext(2); return ctx_r14.onClickSegment($event, segment_r6); });
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵstyleProp("height", ctx_r8.hourSegmentHeight, "px");
} }
function TDSCalendarModeDayComponent_ng_template_6_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSCalendarModeDayComponent_ng_template_6_ng_template_0_tr_0_Template, 5, 7, "tr", 7);
    i0.ɵɵtemplate(1, TDSCalendarModeDayComponent_ng_template_6_ng_template_0_tr_1_Template, 2, 2, "tr", 8);
} if (rf & 2) {
    const segment_r6 = ctx.$implicit;
    i0.ɵɵproperty("ngIf", segment_r6.isStart);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !segment_r6.isStart);
} }
function TDSCalendarModeDayComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSCalendarModeDayComponent_ng_template_6_ng_template_0_Template, 2, 2, "ng-template", 6);
} if (rf & 2) {
    const week_r2 = ctx.$implicit;
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngForOf", week_r2.segments)("ngForTrackBy", ctx_r0.trackByHourSegment);
} }
function TDSCalendarModeDayComponent_div_8_ng_template_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const dataEvent_r18 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", dataEvent_r18.event.template, " ");
} }
function TDSCalendarModeDayComponent_div_8_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 16);
    i0.ɵɵtemplate(1, TDSCalendarModeDayComponent_div_8_ng_template_3_ng_container_1_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const dataEvent_r18 = ctx.$implicit;
    i0.ɵɵstyleProp("height", dataEvent_r18.height, "px")("margin-top", dataEvent_r18.top, "px")("margin-left", dataEvent_r18.left, "px")("width", dataEvent_r18.width, "px");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", dataEvent_r18.event.template);
} }
function TDSCalendarModeDayComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 13)(1, "div", 14);
    i0.ɵɵelement(2, "div", 15);
    i0.ɵɵtemplate(3, TDSCalendarModeDayComponent_div_8_ng_template_3_Template, 2, 9, "ng-template", 3);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("left", ctx_r1.widthColTime, "px");
    i0.ɵɵadvance(2);
    i0.ɵɵstyleProp("width", ctx_r1.maxWidth, "px");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", ctx_r1.dayView.events);
} }
const _c0 = ["*"];
export class TDSCalendarModeDayComponent {
    constructor(destroy$, cdr, elementRef) {
        this.destroy$ = destroy$;
        this.cdr = cdr;
        this.elementRef = elementRef;
        this.viewDate = new Date();
        this.start = new CandyDate();
        /**
         * Giờ kết thúc ngày trong thời gian 24 giờ. Phải là 0-23
         */
        this.dayEndHour = 17;
        /**
        * Phút cuối ngày. Phải là 0-59
        */
        this.dayEndMinute = 30;
        /**
        * Giờ bắt đầu ngày trong thời gian 24 giờ. Phải là 0-23
        */
        this.dayStartHour = 8;
        /**
        * Phút bắt đầu ngày. Phải là 0-59
        */
        this.dayStartMinute = 0;
        /**
         * The number of segments in an hour. Must divide equally into 60.
         */
        this.hourSegments = 4;
        /**
         * The height in pixels of each hour segment
         */
        this.hourSegmentHeight = 30;
        /**
         * The minimum height in pixels of each event
         */
        this.minimumEventHeight = 100;
        /**
        * The width in pixels of each hour segment
        */
        this.hourSegmentWidth = 200;
        /**
         * The duration of each segment group in minutes
         */
        this.hourDuration = 60;
        this.widthColTime = 80;
        /**
         * @hidden
         */
        this.trackByHourSegment = trackByHourSegment;
        this.weekViewHour = [];
        this.activeDate = new CandyDate();
        this.maxWidth = 0;
        this.widthTable = null;
        this.listOfTemplateEventItem$ = new BehaviorSubject([]);
        this.tdsClickSegment = new EventEmitter();
        this.listOfEvent = [];
        this.paddingLeftContent = 16;
    }
    ngOnInit() {
        this.listOfTemplateEventItem$.pipe(takeUntil(this.destroy$))
            .subscribe((value) => {
            let dayStart = {
                hour: this.dayStartHour,
                minute: this.dayStartMinute,
            };
            let dayEnd = {
                hour: this.dayEndHour,
                minute: this.dayEndMinute,
            };
            this.listOfEvent = [...value];
            this.dayView = getDayView({
                events: [...value], viewDate: this.viewDate, hourSegments: this.hourSegments, dayStart: dayStart,
                dayEnd: dayEnd, eventWidth: this.hourSegmentWidth, hourDuration: this.hourDuration, minimumEventHeight: this.minimumEventHeight, segmentHeight: this.hourSegmentHeight
            });
            let width = this.paddingLeftContent;
            let widthComp = this.elementRef.nativeElement.getBoundingClientRect().width;
            this.dayView.events.forEach(e => {
                const widtCol = e.left + e.width;
                if (width < widtCol) {
                    width = widtCol;
                }
            });
            this.widthTable = width > widthComp ? width + this.widthColTime + this.paddingLeftContent : null;
            this.maxWidth = width == 0 ? 0 : width;
            this.cdr.markForCheck();
        });
    }
    ngOnChanges(changes) {
        if (changes.viewDate) {
            this.buildTime();
            this.cdr.markForCheck();
        }
    }
    ngAfterContentInit() {
        this.listEventComp.changes.pipe(startWith(true), switchMap(() => merge(...[
            this.listEventComp.changes,
            ...this.listEventComp.map(option => option.changes),
        ]).pipe(startWith(true))), takeUntil(this.destroy$)).subscribe(() => {
            const listOfEventInterface = this.listEventComp.toArray().map(item => {
                const { tdsEndDate, tdsStartDate, tdsData, tdsTemplate } = item;
                return {
                    start: tdsStartDate,
                    end: tdsEndDate,
                    data: tdsData,
                    template: tdsTemplate,
                    allDay: false
                };
            });
            this.listOfTemplateEventItem$.next(listOfEventInterface);
        });
    }
    onClickSegment(e, date) {
        e.stopPropagation();
        e.preventDefault();
        this.tdsClickSegment.emit(date);
    }
    buildTime() {
        this.activeDate = new CandyDate(this.viewDate);
        let dayStart = {
            hour: this.dayStartHour,
            minute: this.dayStartMinute,
        };
        let dayEnd = {
            hour: this.dayEndHour,
            minute: this.dayEndMinute,
        };
        this.weekViewHour = getDayViewHourGrid({
            dayStart: dayStart,
            dayEnd: dayEnd,
            hourDuration: this.hourDuration,
            viewDate: this.viewDate,
            hourSegments: this.hourSegments
        });
        this.listOfTemplateEventItem$.next(this.listOfEvent);
    }
}
TDSCalendarModeDayComponent.ɵfac = function TDSCalendarModeDayComponent_Factory(t) { return new (t || TDSCalendarModeDayComponent)(i0.ɵɵdirectiveInject(i1.TDSDestroyService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSCalendarModeDayComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCalendarModeDayComponent, selectors: [["tds-calendar-mode-day"]], contentQueries: function TDSCalendarModeDayComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSCalendarEventComponent, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.listEventComp = _t);
    } }, hostAttrs: [1, "w-full", "block", "border", "border-neutral-2-200"], inputs: { viewDate: "viewDate", locale: "locale", start: "start", dayEndHour: "dayEndHour", dayEndMinute: "dayEndMinute", dayStartHour: "dayStartHour", dayStartMinute: "dayStartMinute", hourSegments: "hourSegments", hourSegmentHeight: "hourSegmentHeight", minimumEventHeight: "minimumEventHeight", hourSegmentWidth: "hourSegmentWidth", hourDuration: "hourDuration", widthColTime: "widthColTime", currentTimeMarkerTemplate: "currentTimeMarkerTemplate" }, outputs: { tdsClickSegment: "tdsClickSegment" }, features: [i0.ɵɵProvidersFeature([TDSDestroyService]), i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 10, vars: 15, consts: [[1, "w-full", "h-full", "relative", "overflow-auto", "tds-custom-scroll"], [1, "table-fixed", "min-w-full"], [3, "width"], ["ngFor", "", 3, "ngForOf"], [3, "columnDate", "dayStartHour", "dayStartMinute", "dayEndHour", "dayEndMinute", "hourSegments", "hourDuration", "hourSegmentHeight", "customTemplate", "widthColTime"], ["class", "tds-calendar-event-container pointer-events-none absolute inset-0 pl-4", 3, "left", 4, "ngIf"], ["ngFor", "", 3, "ngForOf", "ngForTrackBy"], ["class", "tds-calendar-segment-start", 4, "ngIf"], ["class", "tds-calendar-segment", 4, "ngIf"], [1, "tds-calendar-segment-start"], [1, "tds-calendar-col-lock", "align-top", "text-center", "sticky", "left-0", "z-10", "bg-white"], [1, "cursor-pointer", 3, "click"], [1, "tds-calendar-segment"], [1, "tds-calendar-event-container", "pointer-events-none", "absolute", "inset-0", "pl-4"], [1, "w-full", "h-full", "relative", "pointer-events-none"], [1, "h-px", "pointer-events-none"], [1, "absolute", "pointer-events-auto"], [4, "tdsStringTemplateOutlet"]], template: function TDSCalendarModeDayComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "div", 0)(1, "table", 1)(2, "colgroup");
        i0.ɵɵelement(3, "col", 2)(4, "col");
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(5, "tbody");
        i0.ɵɵtemplate(6, TDSCalendarModeDayComponent_ng_template_6_Template, 1, 2, "ng-template", 3);
        i0.ɵɵelementEnd()();
        i0.ɵɵelement(7, "tds-calendar-current-time", 4);
        i0.ɵɵtemplate(8, TDSCalendarModeDayComponent_div_8_Template, 4, 5, "div", 5);
        i0.ɵɵelementEnd();
        i0.ɵɵprojection(9);
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵstyleProp("width", ctx.widthTable, "px");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("width", ctx.widthColTime);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngForOf", ctx.weekViewHour);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("columnDate", ctx.viewDate)("dayStartHour", ctx.dayStartHour)("dayStartMinute", ctx.dayStartMinute)("dayEndHour", ctx.dayEndHour)("dayEndMinute", ctx.dayEndMinute)("hourSegments", ctx.hourSegments)("hourDuration", ctx.hourDuration)("hourSegmentHeight", ctx.hourSegmentHeight)("customTemplate", ctx.currentTimeMarkerTemplate)("widthColTime", ctx.widthColTime);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.dayView);
    } }, directives: [i2.NgForOf, i2.NgIf, i3.TDSCalendarCurrentTimeComponent, i4.TDSStringTemplateOutletDirective], pipes: [i2.DatePipe], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCalendarModeDayComponent, [{
        type: Component,
        args: [{ selector: 'tds-calendar-mode-day', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, providers: [TDSDestroyService], host: {
                    class: "w-full block border  border-neutral-2-200"
                }, template: "<div class=\"w-full h-full  relative overflow-auto tds-custom-scroll\">\n    <table class=\"table-fixed min-w-full\" [style.width.px]=\"widthTable\">\n        <colgroup>\n            <col [width]=\"widthColTime\">\n            <col>\n        </colgroup>\n        <tbody>\n            <ng-template ngFor [ngForOf]=\"weekViewHour\" let-week let-idx=\"index\" let-total=\"count\">               \n                <ng-template ngFor [ngForOf]=\"week.segments\" let-segment [ngForTrackBy]=\"trackByHourSegment\">\n                    <tr  *ngIf=\"segment.isStart\" class=\"tds-calendar-segment-start\">\n                        <td [attr.rowspan]=\"week.segments.length\" class=\"tds-calendar-col-lock align-top text-center sticky left-0 z-10 bg-white \">\n                            {{segment.displayDate | date:\"HH:mm\"}}\n                        </td>\n                        <td class=\"cursor-pointer\" [style.height.px]=\"hourSegmentHeight\" (click)=\"onClickSegment($event,segment)\">\n                            \n                        </td>\n                    </tr>\n                    <tr *ngIf=\"!segment.isStart\" class=\"tds-calendar-segment\">\n                        <td class=\"cursor-pointer\" [style.height.px]=\"hourSegmentHeight\" (click)=\"onClickSegment($event,segment)\">                           \n                        </td>\n                    </tr>                    \n                </ng-template>\n            </ng-template>\n        </tbody>\n    </table>\n    <tds-calendar-current-time\n    [columnDate]=\"viewDate\"\n    [dayStartHour]=\"dayStartHour\"\n    [dayStartMinute]=\"dayStartMinute\"\n    [dayEndHour]=\"dayEndHour\"\n    [dayEndMinute]=\"dayEndMinute\"\n    [hourSegments]=\"hourSegments\"\n    [hourDuration]=\"hourDuration\"\n    [hourSegmentHeight]=\"hourSegmentHeight\"\n    [customTemplate]=\"currentTimeMarkerTemplate\"\n    [widthColTime]=\"widthColTime\"\n  ></tds-calendar-current-time>\n    <!-- <ng-template ngFor [ngForOf]=\"weekViewHour\" let-week let-idx=\"index\" let-total=\"count\">\n        <ng-template ngFor [ngForOf]=\"week.segments\" let-segment [ngForTrackBy]=\"trackByHourSegment\">\n            <div class=\"w-full flex\" [ngClass]=\"{'border-t border-neutral-2-200':segment.isStart}\"\n                [style.height.px]=\"hourSegmentHeight\">\n                <div class=\"w-20 text-center\">\n                    <span *ngIf=\"segment.isStart\">\n                        {{segment.displayDate | date:\"HH:mm\"}}\n                    </span>\n                </div>\n                <div class=\"flex-auto\">\n\n                </div>\n            </div>\n        </ng-template>\n    </ng-template> -->\n    <div class=\"tds-calendar-event-container pointer-events-none absolute inset-0 pl-4\" [style.left.px]=\"widthColTime\" *ngIf=\"dayView\">\n        <div class=\"w-full h-full relative pointer-events-none\">\n            <div class=\"h-px pointer-events-none\" [style.width.px]=\"maxWidth\"></div>\n            <ng-template ngFor [ngForOf]=\"dayView.events\" let-dataEvent>\n                <div class=\"absolute pointer-events-auto\" [style.height.px]=\"dataEvent.height\" [style.marginTop.px]=\"dataEvent.top\"\n                    [style.marginLeft.px]=\"dataEvent.left\" [style.width.px]=\"dataEvent.width\">\n                    <ng-container *tdsStringTemplateOutlet=\"dataEvent.event.template\">{{ dataEvent.event.template }}\n                    </ng-container>\n                </div>\n            </ng-template>   \n        </div>\n    </div>     \n  \n</div>\n<ng-content></ng-content>" }]
    }], function () { return [{ type: i1.TDSDestroyService }, { type: i0.ChangeDetectorRef }, { type: i0.ElementRef }]; }, { viewDate: [{
            type: Input
        }], locale: [{
            type: Input
        }], start: [{
            type: Input
        }], dayEndHour: [{
            type: Input
        }], dayEndMinute: [{
            type: Input
        }], dayStartHour: [{
            type: Input
        }], dayStartMinute: [{
            type: Input
        }], hourSegments: [{
            type: Input
        }], hourSegmentHeight: [{
            type: Input
        }], minimumEventHeight: [{
            type: Input
        }], hourSegmentWidth: [{
            type: Input
        }], hourDuration: [{
            type: Input
        }], widthColTime: [{
            type: Input
        }], currentTimeMarkerTemplate: [{
            type: Input
        }], listEventComp: [{
            type: ContentChildren,
            args: [TDSCalendarEventComponent, { descendants: true }]
        }], tdsClickSegment: [{
            type: Output
        }] }); })();
