// tds-ui/esm2020/calendar/tds-calendar-event.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSCalendarEventComponent {
    constructor() {
        this.tdsStartDate = new Date();
        this.tdsEndDate = new Date();
        this.changes = new Subject();
    }
    ngOnInit() {
    }
    ngOnChanges() {
        this.changes.next();
    }
}
TDSCalendarEventComponent.ɵfac = function TDSCalendarEventComponent_Factory(t) { return new (t || TDSCalendarEventComponent)(); };
TDSCalendarEventComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCalendarEventComponent, selectors: [["tds-calendar-event"]], inputs: { tdsStartDate: "tdsStartDate", tdsEndDate: "tdsEndDate", tdsData: "tdsData", tdsTemplate: "tdsTemplate" }, features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSCalendarEventComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCalendarEventComponent, [{
        type: Component,
        args: [{
                selector: 'tds-calendar-event',
                template: `
    <ng-content></ng-content>
  `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
            }]
    }], function () { return []; }, { tdsStartDate: [{
            type: Input
        }], tdsEndDate: [{
            type: Input
        }], tdsData: [{
            type: Input
        }], tdsTemplate: [{
            type: Input
        }] }); })();
