// tds-ui/esm2020/calendar/tds-calendar-mode-month.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSCalendarModeMonthComponent {
    constructor() { }
    ngOnInit() {
    }
}
TDSCalendarModeMonthComponent.ɵfac = function TDSCalendarModeMonthComponent_Factory(t) { return new (t || TDSCalendarModeMonthComponent)(); };
TDSCalendarModeMonthComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCalendarModeMonthComponent, selectors: [["tds-calendar-mode-month"]], decls: 2, vars: 0, template: function TDSCalendarModeMonthComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "p");
        i0.ɵɵtext(1, "tds-calendar-mode-month works!");
        i0.ɵɵelementEnd();
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCalendarModeMonthComponent, [{
        type: Component,
        args: [{ selector: 'tds-calendar-mode-month', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<p>tds-calendar-mode-month works!</p>\n" }]
    }], function () { return []; }, null); })();
