// tds-ui/esm2020/flex/flex.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSFlexComponent } from './flex';
import * as i0 from "@angular/core";
export class TDSFlexModule {
}
TDSFlexModule.ɵfac = function TDSFlexModule_Factory(t) { return new (t || TDSFlexModule)(); };
TDSFlexModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSFlexModule });
TDSFlexModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFlexModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSFlexComponent],
                exports: [TDSFlexComponent],
                imports: [
                    CommonModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSFlexModule, { declarations: [TDSFlexComponent], imports: [CommonModule], exports: [TDSFlexComponent] }); })();
