// tds-ui/esm2020/auto-complete/autocomplete.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSNoAnimationModule } from 'tds-ui/core/no-animation';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSAutocompleteOptgroupComponent } from './autocomplete-optgroup.component';
import { TDSAutocompleteOptionComponent } from './autocomplete-option.component';
import { TDSAutocompleteTriggerDirective } from './autocomplete-trigger.directive';
import { TDSAutocompleteComponent } from './autocomplete.component';
import * as i0 from "@angular/core";
export class TDSAutocompleteModule {
}
TDSAutocompleteModule.ɵfac = function TDSAutocompleteModule_Factory(t) { return new (t || TDSAutocompleteModule)(); };
TDSAutocompleteModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSAutocompleteModule });
TDSAutocompleteModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, OverlayModule, FormsModule, TDSOutletModule, TDSNoAnimationModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAutocompleteModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSAutocompleteComponent,
                    TDSAutocompleteOptionComponent,
                    TDSAutocompleteTriggerDirective,
                    TDSAutocompleteOptgroupComponent
                ],
                exports: [
                    TDSAutocompleteComponent,
                    TDSAutocompleteOptionComponent,
                    TDSAutocompleteTriggerDirective,
                    TDSAutocompleteOptgroupComponent
                ],
                imports: [BidiModule, CommonModule, OverlayModule, FormsModule, TDSOutletModule, TDSNoAnimationModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSAutocompleteModule, { declarations: [TDSAutocompleteComponent,
        TDSAutocompleteOptionComponent,
        TDSAutocompleteTriggerDirective,
        TDSAutocompleteOptgroupComponent], imports: [BidiModule, CommonModule, OverlayModule, FormsModule, TDSOutletModule, TDSNoAnimationModule], exports: [TDSAutocompleteComponent,
        TDSAutocompleteOptionComponent,
        TDSAutocompleteTriggerDirective,
        TDSAutocompleteOptgroupComponent] }); })();
