// tds-ui/esm2020/auto-complete/autocomplete-optgroup.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/outlet";
function TDSAutocompleteOptgroupComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r0.label);
} }
const _c0 = [[["tds-auto-option"]]];
const _c1 = ["tds-auto-option"];
export class TDSAutocompleteOptgroupComponent {
    constructor() { }
}
TDSAutocompleteOptgroupComponent.ɵfac = function TDSAutocompleteOptgroupComponent_Factory(t) { return new (t || TDSAutocompleteOptgroupComponent)(); };
TDSAutocompleteOptgroupComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSAutocompleteOptgroupComponent, selectors: [["tds-auto-optgroup"]], hostAttrs: [1, "flex", "w-full", "flex-col", "px-3", "text-neutral-1-900", "text-body-2", "font-regular"], inputs: { label: "label" }, exportAs: ["tdsAutoOptgroup"], ngContentSelectors: _c1, decls: 3, vars: 1, consts: [[1, "tds-select-item", "tds-select-item-group"], [4, "tdsStringTemplateOutlet"]], template: function TDSAutocompleteOptgroupComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c0);
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, TDSAutocompleteOptgroupComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
        i0.ɵɵelementEnd();
        i0.ɵɵprojection(2);
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.label);
    } }, directives: [i1.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAutocompleteOptgroupComponent, [{
        type: Component,
        args: [{
                selector: 'tds-auto-optgroup',
                exportAs: 'tdsAutoOptgroup',
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
    <div class="tds-select-item tds-select-item-group">
      <ng-container *tdsStringTemplateOutlet="label">{{ label }}</ng-container>
    </div>
    <ng-content select="tds-auto-option"></ng-content>
  `,
                host: {
                    class: "flex w-full flex-col px-3 text-neutral-1-900 text-body-2 font-regular"
                }
            }]
    }], function () { return []; }, { label: [{
            type: Input
        }] }); })();
