// tds-ui/fesm2015/tds-ui-pagination.mjs
import { __decorate } from 'tslib';
import * as i0 from '@angular/core';
import { EventEmitter, Component, ViewEncapsulation, ChangeDetectionStrategy, Input, Output, Optional, ViewChild, NgModule } from '@angular/core';
import { Subject, ReplaySubject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i3$1 from 'tds-ui/core/config';
import { WithConfig } from 'tds-ui/core/config';
import * as i2$2 from 'tds-ui/core/services';
import { gridResponsiveMap, TDSBreakpointEnum } from 'tds-ui/core/services';
import { toNumber, InputBoolean, InputNumber } from 'tds-ui/shared/utility';
import * as i1$1 from 'tds-ui/i18n';
import { TDSI18nModule } from 'tds-ui/i18n';
import * as i1 from '@angular/cdk/bidi';
import { BidiModule } from '@angular/cdk/bidi';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i4 from 'tds-ui/form-field';
import { TDSFormFieldModule } from 'tds-ui/form-field';
import * as i5 from 'tds-ui/tds-input';
import { TDSInputModule } from 'tds-ui/tds-input';
import * as i2$1 from 'tds-ui/select';
import { TDSSelectModule } from 'tds-ui/select';
import * as i3 from '@angular/forms';
import { FormsModule } from '@angular/forms';
import * as i4$1 from 'tds-ui/button';
import { TDSButtonModule } from 'tds-ui/button';

/* eslint-disable */
const _c0$3 = ["tds-pagination-item", ""];
function TDSPaginationItemComponent_ng_template_0_a_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "a", 6);
        i0.ɵɵtext(1);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const page_r4 = i0.ɵɵnextContext().page;
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(page_r4);
    }
}
const _c1$2 = function (a0, a1) { return { "cursor-pointer": a0, "cursor-not-allowed": a1 }; };
const _c2$1 = function (a0, a1) { return { "text-neutral-1-100 dark:text-d-neutral-1-100": a0, "text-neutral-1-500 dark:text-d-neutral-1-500": a1 }; };
function TDSPaginationItemComponent_ng_template_0_button_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "button", 7);
        i0.ɵɵelement(1, "i", 8);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r6 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("disabled", ctx_r6.disabled)("ngClass", i0.ɵɵpureFunction2(3, _c1$2, !ctx_r6.disabled, ctx_r6.disabled));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(6, _c2$1, ctx_r6.disabled, !ctx_r6.disabled));
    }
}
function TDSPaginationItemComponent_ng_template_0_button_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "button", 7);
        i0.ɵɵelement(1, "i", 9);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r7 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("disabled", ctx_r7.disabled)("ngClass", i0.ɵɵpureFunction2(3, _c1$2, ctx_r7.disabled, ctx_r7.disabled));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(6, _c2$1, ctx_r7.disabled, !ctx_r7.disabled));
    }
}
function TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 2);
        i0.ɵɵelement(1, "span", 15);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r11 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngSwitch", ctx_r11.direction);
    }
}
function TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 2);
        i0.ɵɵelement(1, "span", 16);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r12 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngSwitch", ctx_r12.direction);
    }
}
function TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 12);
        i0.ɵɵelementContainerStart(1, 2);
        i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_2_Template, 2, 1, "ng-container", 13);
        i0.ɵɵtemplate(3, TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_ng_container_3_Template, 2, 1, "ng-container", 13);
        i0.ɵɵelementContainerEnd();
        i0.ɵɵelement(4, "span", 14);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const type_r3 = i0.ɵɵnextContext(2).$implicit;
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitch", type_r3);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "prev_5");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "next_5");
    }
}
function TDSPaginationItemComponent_ng_template_0_ng_container_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "a", 10);
        i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_0_ng_container_4_div_2_Template, 5, 3, "div", 11);
        i0.ɵɵelementEnd();
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const type_r3 = i0.ɵɵnextContext().$implicit;
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitch", type_r3);
    }
}
function TDSPaginationItemComponent_ng_template_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 2);
        i0.ɵɵtemplate(1, TDSPaginationItemComponent_ng_template_0_a_1_Template, 2, 1, "a", 3);
        i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_0_button_2_Template, 2, 9, "button", 4);
        i0.ɵɵtemplate(3, TDSPaginationItemComponent_ng_template_0_button_3_Template, 2, 9, "button", 4);
        i0.ɵɵtemplate(4, TDSPaginationItemComponent_ng_template_0_ng_container_4_Template, 3, 1, "ng-container", 5);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const type_r3 = ctx.$implicit;
        i0.ɵɵproperty("ngSwitch", type_r3);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "page");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "prev");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "next");
    }
}
function TDSPaginationItemComponent_ng_template_2_Template(rf, ctx) { }
const _c3$1 = function (a0, a1) { return { $implicit: a0, page: a1 }; };
class TDSPaginationItemComponent {
    constructor() {
        this.active = false;
        this.index = null;
        this.disabled = false;
        this.direction = 'ltr';
        this.type = null;
        this.itemRender = null;
        this.diffIndex = new EventEmitter();
        this.gotoIndex = new EventEmitter();
        this.title = null;
    }
    clickItem() {
        if (!this.disabled) {
            if (this.type === 'page') {
                this.gotoIndex.emit(this.index);
            }
            else {
                this.diffIndex.emit({
                    next: 1,
                    prev: -1,
                    prev_5: -5,
                    next_5: 5
                }[this.type]);
            }
        }
    }
    ngOnChanges(changes) {
        var _a, _b, _c, _d;
        const { locale, index, type } = changes;
        if (locale || index || type) {
            this.title = {
                page: `${this.index}`,
                next: (_a = this.locale) === null || _a === void 0 ? void 0 : _a.next_page,
                prev: (_b = this.locale) === null || _b === void 0 ? void 0 : _b.prev_page,
                prev_5: (_c = this.locale) === null || _c === void 0 ? void 0 : _c.prev_5,
                next_5: (_d = this.locale) === null || _d === void 0 ? void 0 : _d.next_5
            }[this.type];
        }
    }
}
TDSPaginationItemComponent.ɵfac = function TDSPaginationItemComponent_Factory(t) { return new (t || TDSPaginationItemComponent)(); };
TDSPaginationItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationItemComponent, selectors: [["li", "tds-pagination-item", ""]], hostVars: 19, hostBindings: function TDSPaginationItemComponent_HostBindings(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵlistener("click", function TDSPaginationItemComponent_click_HostBindingHandler() { return ctx.clickItem(); });
        }
        if (rf & 2) {
            i0.ɵɵattribute("title", ctx.title);
            i0.ɵɵclassProp("tds-pagination-prev", ctx.type === "prev")("tds-pagination-next", ctx.type === "next")("tds-pagination-item", ctx.type === "page")("tds-pagination-jump-prev", ctx.type === "prev_5")("tds-pagination-jump-prev-custom-icon", ctx.type === "prev_5")("tds-pagination-jump-next", ctx.type === "next_5")("tds-pagination-jump-next-custom-icon", ctx.type === "next_5")("tds-pagination-disabled", ctx.disabled)("tds-pagination-item-active", ctx.active);
        }
    }, inputs: { active: "active", locale: "locale", index: "index", disabled: "disabled", direction: "direction", type: "type", itemRender: "itemRender" }, outputs: { diffIndex: "diffIndex", gotoIndex: "gotoIndex" }, features: [i0.ɵɵNgOnChangesFeature], attrs: _c0$3, decls: 3, vars: 5, consts: [["renderItemTemplate", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [3, "ngSwitch"], ["class", "px-0.5", 4, "ngSwitchCase"], ["type", "button", "class", "tds-pagination-item-link flex", 3, "disabled", "ngClass", 4, "ngSwitchCase"], [4, "ngSwitchDefault"], [1, "px-0.5"], ["type", "button", 1, "tds-pagination-item-link", "flex", 3, "disabled", "ngClass"], [1, "tdsi-arrow-left-fill", "leading-4", "text-base", 3, "ngClass"], [1, "tdsi-arrow-right-fill", "leading-4", "text-base", 3, "ngClass"], [1, "tds-pagination-item-link", "flex", 3, "ngSwitch"], ["class", "tds-pagination-item-container flex", 4, "ngSwitchDefault"], [1, "tds-pagination-item-container", "flex"], [3, "ngSwitch", 4, "ngSwitchCase"], [1, "tds-pagination-item-ellipsis", "tdsi-three-dots-horizon-fill", "group-hover:hidden", "leading-4", "text-base"], [1, "tds-pagination-item-link-icon", "tdsi-chevron-previous-line", "group-hover:inline", "hidden", "leading-4", "text-base"], [1, "tds-pagination-item-link-icon", "tdsi-chevron-next-line", "group-hover:inline", "hidden", "leading-4", "text-base"]], template: function TDSPaginationItemComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵtemplate(0, TDSPaginationItemComponent_ng_template_0_Template, 5, 4, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(2, TDSPaginationItemComponent_ng_template_2_Template, 0, 0, "ng-template", 1);
        }
        if (rf & 2) {
            const _r0 = i0.ɵɵreference(1);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("ngTemplateOutlet", ctx.itemRender || _r0)("ngTemplateOutletContext", i0.ɵɵpureFunction2(2, _c3$1, ctx.type, ctx.index));
        }
    }, directives: [i2.NgSwitch, i2.NgSwitchCase, i2.NgClass, i2.NgSwitchDefault, i2.NgTemplateOutlet], encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationItemComponent, [{
            type: Component,
            args: [{ selector: 'li[tds-pagination-item]', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.tds-pagination-prev]': `type === 'prev'`,
                        '[class.tds-pagination-next]': `type === 'next'`,
                        '[class.tds-pagination-item]': `type === 'page'`,
                        '[class.tds-pagination-jump-prev]': `type === 'prev_5'`,
                        '[class.tds-pagination-jump-prev-custom-icon]': `type === 'prev_5'`,
                        '[class.tds-pagination-jump-next]': `type === 'next_5'`,
                        '[class.tds-pagination-jump-next-custom-icon]': `type === 'next_5'`,
                        '[class.tds-pagination-disabled]': 'disabled',
                        '[class.tds-pagination-item-active]': 'active',
                        '[attr.title]': 'title',
                        '(click)': 'clickItem()'
                    }, template: "<ng-template #renderItemTemplate let-type let-page=\"page\">\r\n  <ng-container [ngSwitch]=\"type\">\r\n    <a class=\"px-0.5\" *ngSwitchCase=\"'page'\">{{ page }}</a>\r\n    <button type=\"button\" [disabled]=\"disabled\" class=\"tds-pagination-item-link flex\"\r\n      [ngClass]=\"{'cursor-pointer':!disabled,'cursor-not-allowed':disabled}\" *ngSwitchCase=\"'prev'\">\r\n      <i class=\"tdsi-arrow-left-fill leading-4 text-base\" [ngClass]=\"{'text-neutral-1-100 dark:text-d-neutral-1-100':disabled,'text-neutral-1-500 dark:text-d-neutral-1-500':!disabled}\"></i>\r\n\r\n    </button>\r\n    <button type=\"button\" [disabled]=\"disabled\" [ngClass]=\"{'cursor-pointer':disabled,'cursor-not-allowed':disabled}\"\r\n      class=\"tds-pagination-item-link flex\" *ngSwitchCase=\"'next'\">\r\n      <i class=\"tdsi-arrow-right-fill leading-4 text-base\" [ngClass]=\"{'text-neutral-1-100 dark:text-d-neutral-1-100':disabled,'text-neutral-1-500 dark:text-d-neutral-1-500':!disabled}\"></i>\r\n\r\n    </button>\r\n    <ng-container *ngSwitchDefault>\r\n      <a class=\"tds-pagination-item-link flex\" [ngSwitch]=\"type\">\r\n        <div class=\"tds-pagination-item-container flex\" *ngSwitchDefault>\r\n          <ng-container [ngSwitch]=\"type\">\r\n            <ng-container *ngSwitchCase=\"'prev_5'\" [ngSwitch]=\"direction\">\r\n\r\n              <span class=\"tds-pagination-item-link-icon tdsi-chevron-previous-line group-hover:inline hidden leading-4 text-base\"></span>\r\n            </ng-container>\r\n            <ng-container *ngSwitchCase=\"'next_5'\" [ngSwitch]=\"direction\">\r\n\r\n              <span class=\"tds-pagination-item-link-icon tdsi-chevron-next-line group-hover:inline hidden leading-4 text-base\"></span>\r\n            </ng-container>\r\n          </ng-container>\r\n          <span class=\"tds-pagination-item-ellipsis tdsi-three-dots-horizon-fill group-hover:hidden leading-4 text-base\"></span>\r\n        </div>\r\n      </a>\r\n    </ng-container>\r\n  </ng-container>\r\n</ng-template>\r\n<ng-template [ngTemplateOutlet]=\"itemRender || renderItemTemplate\"\r\n  [ngTemplateOutletContext]=\"{ $implicit: type, page: index }\"></ng-template>" }]
        }], null, { active: [{
                type: Input
            }], locale: [{
                type: Input
            }], index: [{
                type: Input
            }], disabled: [{
                type: Input
            }], direction: [{
                type: Input
            }], type: [{
                type: Input
            }], itemRender: [{
                type: Input
            }], diffIndex: [{
                type: Output
            }], gotoIndex: [{
                type: Output
            }] });
})();

const _c0$2 = ["containerTemplate"];
const _c1$1 = function (a0, a1) { return { "cursor-pointer": a0, "cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500": a1 }; };
function TDSPaginationSimpleComponent_ng_template_0_Template(rf, ctx) {
    if (rf & 1) {
        const _r3 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "div", 1)(1, "ul", 2)(2, "li", 3);
        i0.ɵɵlistener("click", function TDSPaginationSimpleComponent_ng_template_0_Template_li_click_2_listener() { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.prePage(); });
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(3, "li", 4)(4, "div", 5)(5, "tds-form-field", 6)(6, "input", 7);
        i0.ɵɵlistener("keydown.enter", function TDSPaginationSimpleComponent_ng_template_0_Template_input_keydown_enter_6_listener($event) { i0.ɵɵrestoreView(_r3); const ctx_r4 = i0.ɵɵnextContext(); return ctx_r4.jumpToPageViaInput($event); });
        i0.ɵɵelementEnd()();
        i0.ɵɵelementStart(7, "div", 8)(8, "span", 9);
        i0.ɵɵtext(9, "/");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(10);
        i0.ɵɵelementEnd()()();
        i0.ɵɵelementStart(11, "li", 10);
        i0.ɵɵlistener("click", function TDSPaginationSimpleComponent_ng_template_0_Template_li_click_11_listener() { i0.ɵɵrestoreView(_r3); const ctx_r5 = i0.ɵɵnextContext(); return ctx_r5.nextPage(); });
        i0.ɵɵelementEnd()()();
    }
    if (rf & 2) {
        const ctx_r1 = i0.ɵɵnextContext();
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("disabled", ctx_r1.isFirstIndex)("direction", ctx_r1.dir)("itemRender", ctx_r1.itemRender)("ngClass", i0.ɵɵpureFunction2(15, _c1$1, !ctx_r1.isFirstIndex, ctx_r1.isFirstIndex));
        i0.ɵɵattribute("title", ctx_r1.locale.prev_page);
        i0.ɵɵadvance(1);
        i0.ɵɵattribute("title", ctx_r1.pageIndex + "/" + ctx_r1.lastIndex);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("disabled", ctx_r1.disabled)("value", ctx_r1.pageIndex)("size", ctx_r1.size);
        i0.ɵɵadvance(4);
        i0.ɵɵtextInterpolate1(" ", ctx_r1.lastIndex, " ");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("disabled", ctx_r1.isLastIndex)("direction", ctx_r1.dir)("ngClass", i0.ɵɵpureFunction2(18, _c1$1, !ctx_r1.isLastIndex, ctx_r1.isLastIndex))("itemRender", ctx_r1.itemRender);
        i0.ɵɵattribute("title", ctx_r1.locale == null ? null : ctx_r1.locale.next_page);
    }
}
class TDSPaginationSimpleComponent {
    constructor(cdr, renderer, elementRef, directionality) {
        this.cdr = cdr;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.directionality = directionality;
        this.itemRender = null;
        this.disabled = false;
        this.total = 0;
        this.pageIndex = 1;
        this.pageSize = 10;
        this.size = 'md';
        this.pageIndexChange = new EventEmitter();
        this.lastIndex = 0;
        this.isFirstIndex = false;
        this.isLastIndex = false;
        this.dir = 'ltr';
        this.destroy$ = new Subject();
        renderer.removeChild(renderer.parentNode(elementRef.nativeElement), elementRef.nativeElement);
    }
    ngOnInit() {
        var _a;
        (_a = this.directionality.change) === null || _a === void 0 ? void 0 : _a.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.updateRtlStyle();
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
        this.updateRtlStyle();
    }
    updateRtlStyle() {
        if (this.dir === 'rtl') {
            this.renderer.addClass(this.elementRef.nativeElement, 'tds-pagination-rtl');
        }
        else {
            this.renderer.removeClass(this.elementRef.nativeElement, 'tds-pagination-rtl');
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    jumpToPageViaInput($event) {
        const target = $event.target;
        const index = toNumber(target.value, this.pageIndex);
        this.onPageIndexChange(index);
        target.value = `${this.pageIndex}`;
    }
    prePage() {
        this.onPageIndexChange(this.pageIndex - 1);
    }
    nextPage() {
        this.onPageIndexChange(this.pageIndex + 1);
    }
    onPageIndexChange(index) {
        this.pageIndexChange.next(index);
    }
    updateBindingValue() {
        this.lastIndex = Math.ceil(this.total / this.pageSize);
        this.isFirstIndex = this.pageIndex === 1;
        this.isLastIndex = this.pageIndex === this.lastIndex;
    }
    ngOnChanges(changes) {
        const { pageIndex, total, pageSize } = changes;
        if (pageIndex || total || pageSize) {
            this.updateBindingValue();
        }
    }
}
TDSPaginationSimpleComponent.ɵfac = function TDSPaginationSimpleComponent_Factory(t) { return new (t || TDSPaginationSimpleComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.Directionality, 8)); };
TDSPaginationSimpleComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationSimpleComponent, selectors: [["tds-pagination-simple"]], viewQuery: function TDSPaginationSimpleComponent_Query(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵviewQuery(_c0$2, 7);
        }
        if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.template = _t.first);
        }
    }, inputs: { itemRender: "itemRender", disabled: "disabled", locale: "locale", total: "total", pageIndex: "pageIndex", pageSize: "pageSize", size: "size" }, outputs: { pageIndexChange: "pageIndexChange" }, features: [i0.ɵɵNgOnChangesFeature], decls: 2, vars: 0, consts: [["containerTemplate", ""], [1, "flex", "items-center", "w-full", "select-none"], [1, "flex", "items-center"], ["tds-pagination-item", "", "type", "prev", 1, "min-w-7", "min-h-7", "text-sm", "text-center", "align-middle", "mx-0.5", "leading-7", "box-content", "select-auto", "group", "border", "border-transparent", "hover:bg-neutral-3-50", "dark:hover:bg-d-neutral-3-300", "rounded-md", "flex", "items-center", "justify-center", 3, "disabled", "direction", "itemRender", "ngClass", "click"], [1, "tds-pagination-simple-pager", "mx-3"], [1, "flex", "items-center", "justify-center"], [1, "w-12"], ["autocomplete", "off", "tdsInput", "", "autocomplete", "off", 1, "text-center", "text-body-2", "font-regular", 3, "disabled", "value", "size", "keydown.enter"], [1, "ml-2", "text-caption-1", "font-regular", "text-neutral-1-900", "dark:text-d-neutral-1-900"], [1, "tds-pagination-slash"], ["tds-pagination-item", "", "type", "next", 1, "min-w-7", "min-h-7", "text-sm", "text-center", "align-middle", "mx-0.5", "leading-7", "box-content", "select-auto", "group", "border", "border-transparent", "hover:bg-neutral-3-50", "dark:hover:bg-d-neutral-3-300", "rounded-md", "flex", "items-center", "justify-center", 3, "disabled", "direction", "ngClass", "itemRender", "click"]], template: function TDSPaginationSimpleComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵtemplate(0, TDSPaginationSimpleComponent_ng_template_0_Template, 12, 21, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        }
    }, directives: [TDSPaginationItemComponent, i2.NgClass, i4.TDSFormField, i5.TDSInput], encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationSimpleComponent, [{
            type: Component,
            args: [{ selector: 'tds-pagination-simple', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-template #containerTemplate>\r\n    <div class=\"flex items-center w-full select-none\">\r\n        <ul class=\"flex items-center\">\r\n            <li tds-pagination-item [attr.title]=\"locale.prev_page\" [disabled]=\"isFirstIndex\" [direction]=\"dir\"\r\n                (click)=\"prePage()\" type=\"prev\" [itemRender]=\"itemRender\"\r\n                class=\"min-w-7 min-h-7 text-sm text-center align-middle mx-0.5 leading-7 box-content select-auto group border border-transparent hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300  rounded-md flex items-center justify-center\"\r\n                [ngClass]=\"{'cursor-pointer':!isFirstIndex,'cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500':isFirstIndex}\"></li>\r\n            <li [attr.title]=\"pageIndex + '/' + lastIndex\" class=\"tds-pagination-simple-pager mx-3\">\r\n                <div class=\"flex items-center justify-center\">\r\n                    <tds-form-field class=\"w-12\">\r\n                        <input autocomplete=\"off\" class=\"text-center text-body-2 font-regular\" tdsInput autocomplete=\"off\" [disabled]=\"disabled\"\r\n                            [value]=\"pageIndex\" (keydown.enter)=\"jumpToPageViaInput($event)\" [size]=\"size\" />\r\n                    </tds-form-field>\r\n                    <div class=\"ml-2 text-caption-1 font-regular text-neutral-1-900 dark:text-d-neutral-1-900\">\r\n                        <span class=\"tds-pagination-slash\">/</span>\r\n                        {{ lastIndex }}\r\n                    </div>\r\n\r\n                </div>\r\n            </li>\r\n            <li tds-pagination-item [attr.title]=\"locale?.next_page\" [disabled]=\"isLastIndex\" [direction]=\"dir\"\r\n                class=\"min-w-7 min-h-7 text-sm text-center align-middle mx-0.5 leading-7 box-content select-auto group border border-transparent hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300 rounded-md flex items-center justify-center\"\r\n                [ngClass]=\"{'cursor-pointer':!isLastIndex,'cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500':isLastIndex}\"\r\n                (click)=\"nextPage()\" type=\"next\" [itemRender]=\"itemRender\"></li>\r\n        </ul>\r\n    </div>\r\n\r\n</ng-template>" }]
        }], function () {
        return [{ type: i0.ChangeDetectorRef }, { type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i1.Directionality, decorators: [{
                        type: Optional
                    }] }];
    }, { template: [{
                type: ViewChild,
                args: ['containerTemplate', { static: true }]
            }], itemRender: [{
                type: Input
            }], disabled: [{
                type: Input
            }], locale: [{
                type: Input
            }], total: [{
                type: Input
            }], pageIndex: [{
                type: Input
            }], pageSize: [{
                type: Input
            }], size: [{
                type: Input
            }], pageIndexChange: [{
                type: Output
            }] });
})();

const _c0$1 = ["tds-pagination-options", ""];
function TDSPaginationOptionsComponent_div_0_Template(rf, ctx) {
    if (rf & 1) {
        const _r3 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "div", 2)(1, "tds-select", 3);
        i0.ɵɵlistener("ngModelChange", function TDSPaginationOptionsComponent_div_0_Template_tds_select_ngModelChange_1_listener($event) { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.onPageSizeChange($event); });
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(2, "div", 4);
        i0.ɵɵtext(3);
        i0.ɵɵelementEnd()();
    }
    if (rf & 2) {
        const ctx_r0 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("disabled", ctx_r0.disabled)("size", ctx_r0.size)("ngModel", ctx_r0.pageSize)("data", ctx_r0.listOfPageSizeOption);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx_r0.locale.items_per_page, " ");
    }
}
function TDSPaginationOptionsComponent_div_1_Template(rf, ctx) {
    if (rf & 1) {
        const _r5 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "div", 5);
        i0.ɵɵtext(1);
        i0.ɵɵelementStart(2, "tds-form-field", 6)(3, "input", 7);
        i0.ɵɵlistener("keydown.enter", function TDSPaginationOptionsComponent_div_1_Template_input_keydown_enter_3_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return ctx_r4.jumpToPageViaInput($event); });
        i0.ɵɵelementEnd()();
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r1 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx_r1.locale.jump_to, " ");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("disabled", ctx_r1.disabled)("size", ctx_r1.size);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx_r1.locale.page, " ");
    }
}
class TDSPaginationOptionsComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.size = 'md';
        this.disabled = false;
        this.showSizeChanger = false;
        this.showQuickJumper = false;
        this.total = 0;
        this.pageIndex = 1;
        this.pageSize = 10;
        this.pageSizeOptions = [];
        this.pageIndexChange = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.listOfPageSizeOption = [];
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-pagination-options');
    }
    onPageSizeChange(size) {
        if (this.pageSize !== size) {
            this.pageSizeChange.next(size);
        }
    }
    jumpToPageViaInput($event) {
        const target = $event.target;
        const index = Math.floor(toNumber(target.value, this.pageIndex));
        this.pageIndexChange.next(index);
        target.value = '';
    }
    trackByOption(_, option) {
        return option.value;
    }
    ngOnChanges(changes) {
        const { pageSize, pageSizeOptions, locale } = changes;
        if (pageSize || pageSizeOptions || locale) {
            this.listOfPageSizeOption = [...new Set([...this.pageSizeOptions, this.pageSize])].map(item => ({
                value: item,
                label: `${item}`
            }));
        }
    }
}
TDSPaginationOptionsComponent.ɵfac = function TDSPaginationOptionsComponent_Factory(t) { return new (t || TDSPaginationOptionsComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSPaginationOptionsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationOptionsComponent, selectors: [["div", "tds-pagination-options", ""]], inputs: { size: "size", disabled: "disabled", showSizeChanger: "showSizeChanger", showQuickJumper: "showQuickJumper", locale: "locale", total: "total", pageIndex: "pageIndex", pageSize: "pageSize", pageSizeOptions: "pageSizeOptions" }, outputs: { pageIndexChange: "pageIndexChange", pageSizeChange: "pageSizeChange" }, features: [i0.ɵɵNgOnChangesFeature], attrs: _c0$1, decls: 2, vars: 2, consts: [["class", "inline-flex items-center", 4, "ngIf"], ["class", "tds-pagination-options-quick-jumper flex items-center ", 4, "ngIf"], [1, "inline-flex", "items-center"], ["textField", "label", "valueField", "value", 1, "tds-pagination-options-size-changer", 3, "disabled", "size", "ngModel", "data", "ngModelChange"], [1, "ml-2", "min-w-max"], [1, "tds-pagination-options-quick-jumper", "flex", "items-center"], [1, "w-20", "mx-3"], ["autocomplete", "off", "tdsInput", "", "autocomplete", "off", 3, "disabled", "size", "keydown.enter"]], template: function TDSPaginationOptionsComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵtemplate(0, TDSPaginationOptionsComponent_div_0_Template, 4, 5, "div", 0);
            i0.ɵɵtemplate(1, TDSPaginationOptionsComponent_div_1_Template, 5, 4, "div", 1);
        }
        if (rf & 2) {
            i0.ɵɵproperty("ngIf", ctx.showSizeChanger);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.showQuickJumper);
        }
    }, directives: [i2.NgIf, i2$1.TDSSelect, i3.NgControlStatus, i3.NgModel, i4.TDSFormField, i5.TDSInput], encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationOptionsComponent, [{
            type: Component,
            args: [{
                    selector: 'div[tds-pagination-options]',
                    preserveWhitespaces: false,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
    <div class='inline-flex items-center' *ngIf="showSizeChanger">
    <tds-select
    class="tds-pagination-options-size-changer"
    
    [disabled]="disabled"
    [size]="size"
    [ngModel]="pageSize"
    (ngModelChange)="onPageSizeChange($event)"
    [data]="listOfPageSizeOption"
    textField="label"
    valueField="value"
  >     
  </tds-select>
      <div class='ml-2 min-w-max'>
        {{locale.items_per_page}}
      </div>
    </div>
   
    <div class="tds-pagination-options-quick-jumper flex items-center " *ngIf="showQuickJumper">
      {{ locale.jump_to }}
      <tds-form-field class="w-20 mx-3">     
        <input autocomplete="off" tdsInput autocomplete="off" 
         [disabled]="disabled" [size]="size" (keydown.enter)="jumpToPageViaInput($event)" />    
      </tds-form-field>
      {{ locale.page }}
    </div>
  `
                }]
        }], function () { return [{ type: i0.ElementRef }]; }, { size: [{
                type: Input
            }], disabled: [{
                type: Input
            }], showSizeChanger: [{
                type: Input
            }], showQuickJumper: [{
                type: Input
            }], locale: [{
                type: Input
            }], total: [{
                type: Input
            }], pageIndex: [{
                type: Input
            }], pageSize: [{
                type: Input
            }], pageSizeOptions: [{
                type: Input
            }], pageIndexChange: [{
                type: Output
            }], pageSizeChange: [{
                type: Output
            }] });
})();

const _c0 = ["containerTemplate"];
const _c1 = function (a0, a1, a2, a3) { return { $implicit: a0, locale: a1, bordered: a2, dir: a3 }; };
function TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementContainer(1, 13);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const page_r10 = i0.ɵɵnextContext().$implicit;
        const ctx_r11 = i0.ɵɵnextContext(2);
        const _r2 = i0.ɵɵreference(3);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r2)("ngTemplateOutletContext", i0.ɵɵpureFunction4(2, _c1, page_r10, ctx_r11.locale, ctx_r11.bordered, ctx_r11.dir));
    }
}
function TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementContainer(1, 13);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const page_r10 = i0.ɵɵnextContext().$implicit;
        const ctx_r12 = i0.ɵɵnextContext(2);
        const _r4 = i0.ɵɵreference(5);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r4)("ngTemplateOutletContext", i0.ɵɵpureFunction4(2, _c1, page_r10, ctx_r12.locale, ctx_r12.bordered, ctx_r12.dir));
    }
}
function TDSPaginationDefaultComponent_ng_template_0_ng_template_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 10);
        i0.ɵɵtemplate(1, TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_1_Template, 2, 7, "ng-container", 11);
        i0.ɵɵtemplate(2, TDSPaginationDefaultComponent_ng_template_0_ng_template_2_ng_container_2_Template, 2, 7, "ng-container", 12);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const page_r10 = ctx.$implicit;
        i0.ɵɵproperty("ngSwitch", page_r10.type);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "page");
    }
}
function TDSPaginationDefaultComponent_ng_template_0_div_3_Template(rf, ctx) {
    if (rf & 1) {
        const _r16 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "div", 14);
        i0.ɵɵlistener("pageIndexChange", function TDSPaginationDefaultComponent_ng_template_0_div_3_Template_div_pageIndexChange_0_listener($event) { i0.ɵɵrestoreView(_r16); const ctx_r15 = i0.ɵɵnextContext(2); return ctx_r15.onPageIndexChange($event); })("pageSizeChange", function TDSPaginationDefaultComponent_ng_template_0_div_3_Template_div_pageSizeChange_0_listener($event) { i0.ɵɵrestoreView(_r16); const ctx_r17 = i0.ɵɵnextContext(2); return ctx_r17.onPageSizeChange($event); });
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r7 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("total", ctx_r7.total)("locale", ctx_r7.locale)("disabled", ctx_r7.disabled)("size", ctx_r7.size)("showSizeChanger", ctx_r7.showSizeChanger)("showQuickJumper", ctx_r7.showQuickJumper)("pageIndex", ctx_r7.pageIndex)("pageSize", ctx_r7.pageSize)("pageSizeOptions", ctx_r7.pageSizeOptions);
    }
}
function TDSPaginationDefaultComponent_ng_template_0_div_5_ng_template_1_Template(rf, ctx) { }
const _c2 = function (a0, a1) { return { $implicit: a0, range: a1 }; };
function TDSPaginationDefaultComponent_ng_template_0_div_5_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 15);
        i0.ɵɵtemplate(1, TDSPaginationDefaultComponent_ng_template_0_div_5_ng_template_1_Template, 0, 0, "ng-template", 13);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r8 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", ctx_r8.showTotal)("ngTemplateOutletContext", i0.ɵɵpureFunction2(2, _c2, ctx_r8.total, ctx_r8.ranges));
    }
}
function TDSPaginationDefaultComponent_ng_template_0_button_6_Template(rf, ctx) {
    if (rf & 1) {
        const _r20 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "button", 16);
        i0.ɵɵlistener("click", function TDSPaginationDefaultComponent_ng_template_0_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r20); const ctx_r19 = i0.ɵɵnextContext(2); return ctx_r19.onRefresh($event); });
        i0.ɵɵelementStart(1, "span", 4);
        i0.ɵɵelement(2, "i", 17);
        i0.ɵɵelementEnd()();
    }
}
function TDSPaginationDefaultComponent_ng_template_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 3)(1, "ul", 4);
        i0.ɵɵtemplate(2, TDSPaginationDefaultComponent_ng_template_0_ng_template_2_Template, 3, 2, "ng-template", 5);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(3, TDSPaginationDefaultComponent_ng_template_0_div_3_Template, 1, 9, "div", 6);
        i0.ɵɵelementStart(4, "div", 7);
        i0.ɵɵtemplate(5, TDSPaginationDefaultComponent_ng_template_0_div_5_Template, 2, 5, "div", 8);
        i0.ɵɵtemplate(6, TDSPaginationDefaultComponent_ng_template_0_button_6_Template, 3, 0, "button", 9);
        i0.ɵɵelementEnd()();
    }
    if (rf & 2) {
        const ctx_r1 = i0.ɵɵnextContext();
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngForOf", ctx_r1.listOfPageItem)("ngForTrackBy", ctx_r1.trackByPageItem);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r1.showQuickJumper || ctx_r1.showSizeChanger);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx_r1.showTotal);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r1.showRefresh);
    }
}
const _c3 = function (a0, a1, a2, a3, a4, a5, a6) { return { "bg-primary-1 text-white": a0, "text-neutral-1-600 dark:text-d-neutral-1-600 hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300": a1, "cursor-pointer": a2, "cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500": a3, "border-transparent": a4, " border-primary-1": a5, "border-neutral-2-200 dark:border-d-neutral-2-200": a6 }; };
function TDSPaginationDefaultComponent_ng_template_2_Template(rf, ctx) {
    if (rf & 1) {
        const _r26 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "li", 18);
        i0.ɵɵlistener("gotoIndex", function TDSPaginationDefaultComponent_ng_template_2_Template_li_gotoIndex_0_listener($event) { i0.ɵɵrestoreView(_r26); const ctx_r25 = i0.ɵɵnextContext(); return ctx_r25.jumpPage($event); })("diffIndex", function TDSPaginationDefaultComponent_ng_template_2_Template_li_diffIndex_0_listener($event) { i0.ɵɵrestoreView(_r26); const ctx_r27 = i0.ɵɵnextContext(); return ctx_r27.jumpDiff($event); });
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const page_r21 = ctx.$implicit;
        const dir_r23 = ctx.dir;
        const bordered_r24 = ctx.bordered;
        const ctx_r3 = i0.ɵɵnextContext();
        i0.ɵɵproperty("locale", ctx_r3.locale)("ngClass", i0.ɵɵpureFunction7(8, _c3, ctx_r3.pageIndex === page_r21.index, ctx_r3.pageIndex !== page_r21.index, !page_r21.disabled, !!page_r21.disabled, !bordered_r24, ctx_r3.pageIndex === page_r21.index && bordered_r24, ctx_r3.pageIndex !== page_r21.index && bordered_r24))("type", page_r21.type)("index", page_r21.index)("disabled", !!page_r21.disabled)("itemRender", ctx_r3.itemRender)("active", ctx_r3.pageIndex === page_r21.index)("direction", dir_r23);
    }
}
const _c4 = function (a0, a1) { return { "cursor-pointer": a0, "cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500": a1 }; };
function TDSPaginationDefaultComponent_ng_template_4_Template(rf, ctx) {
    if (rf & 1) {
        const _r33 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementStart(0, "li", 19);
        i0.ɵɵlistener("gotoIndex", function TDSPaginationDefaultComponent_ng_template_4_Template_li_gotoIndex_0_listener($event) { i0.ɵɵrestoreView(_r33); const ctx_r32 = i0.ɵɵnextContext(); return ctx_r32.jumpPage($event); })("diffIndex", function TDSPaginationDefaultComponent_ng_template_4_Template_li_diffIndex_0_listener($event) { i0.ɵɵrestoreView(_r33); const ctx_r34 = i0.ɵɵnextContext(); return ctx_r34.jumpDiff($event); });
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const page_r28 = ctx.$implicit;
        const dir_r30 = ctx.dir;
        const ctx_r5 = i0.ɵɵnextContext();
        i0.ɵɵproperty("locale", ctx_r5.locale)("ngClass", i0.ɵɵpureFunction2(8, _c4, !page_r28.disabled, !!page_r28.disabled))("type", page_r28.type)("index", page_r28.index)("disabled", !!page_r28.disabled)("itemRender", ctx_r5.itemRender)("active", ctx_r5.pageIndex === page_r28.index)("direction", dir_r30);
    }
}
class TDSPaginationDefaultComponent {
    constructor(cdr, renderer, elementRef, directionality) {
        this.cdr = cdr;
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.directionality = directionality;
        this.size = 'md';
        this.itemRender = null;
        this.showTotal = null;
        this.disabled = false;
        this.showSizeChanger = false;
        this.showQuickJumper = false;
        this.total = 0;
        this.pageIndex = 1;
        this.pageSize = 10;
        this.pageSizeOptions = [10, 20, 30, 40];
        this.showRefresh = false;
        this.bordered = true;
        this.pageIndexChange = new EventEmitter();
        this.pageSizeChange = new EventEmitter();
        this.clickRefresh = new EventEmitter();
        this.ranges = [0, 0];
        this.listOfPageItem = [];
        this.dir = 'ltr';
        this.destroy$ = new Subject();
        renderer.removeChild(renderer.parentNode(elementRef.nativeElement), elementRef.nativeElement);
    }
    ngOnInit() {
        var _a;
        (_a = this.directionality.change) === null || _a === void 0 ? void 0 : _a.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.updateRtlStyle();
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
        this.updateRtlStyle();
    }
    updateRtlStyle() {
        if (this.dir === 'rtl') {
            this.renderer.addClass(this.elementRef.nativeElement, 'tds-pagination-rtl');
        }
        else {
            this.renderer.removeClass(this.elementRef.nativeElement, 'tds-pagination-rtl');
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    jumpPage(index) {
        this.onPageIndexChange(index);
    }
    jumpDiff(diff) {
        this.jumpPage(this.pageIndex + diff);
    }
    trackByPageItem(_, value) {
        return `${value.type}-${value.index}`;
    }
    onPageIndexChange(index) {
        this.pageIndexChange.next(index);
    }
    onPageSizeChange(size) {
        this.pageSizeChange.next(size);
    }
    getLastIndex(total, pageSize) {
        return Math.ceil(total / pageSize);
    }
    buildIndexes() {
        const lastIndex = this.getLastIndex(this.total, this.pageSize);
        this.listOfPageItem = this.getListOfPageItem(this.pageIndex, lastIndex);
    }
    getListOfPageItem(pageIndex, lastIndex) {
        const concatWithPrevNext = (listOfPage) => {
            const prevItem = {
                type: 'prev',
                disabled: pageIndex === 1
            };
            const nextItem = {
                type: 'next',
                disabled: pageIndex === lastIndex
            };
            return [prevItem, ...listOfPage, nextItem];
        };
        const generatePage = (start, end) => {
            const list = [];
            for (let i = start; i <= end; i++) {
                list.push({
                    index: i,
                    type: 'page'
                });
            }
            return list;
        };
        if (lastIndex <= 9) {
            return concatWithPrevNext(generatePage(1, lastIndex));
        }
        else {
            const generateRangeItem = (selected, last) => {
                let listOfRange = [];
                const prevFiveItem = {
                    type: 'prev_5'
                };
                const nextFiveItem = {
                    type: 'next_5'
                };
                const firstPageItem = generatePage(1, 1);
                const lastPageItem = generatePage(lastIndex, lastIndex);
                if (selected < 4) {
                    listOfRange = [...generatePage(2, 5), nextFiveItem];
                }
                else if (selected < last - 3) {
                    listOfRange = [prevFiveItem, ...generatePage(selected - 2, selected + 2), nextFiveItem];
                }
                else {
                    listOfRange = [prevFiveItem, ...generatePage(last - 4, last - 1)];
                }
                return [...firstPageItem, ...listOfRange, ...lastPageItem];
            };
            return concatWithPrevNext(generateRangeItem(pageIndex, lastIndex));
        }
    }
    ngOnChanges(changes) {
        const { pageIndex, pageSize, total } = changes;
        if (pageIndex || pageSize || total) {
            this.ranges = [(this.pageIndex - 1) * this.pageSize + 1, Math.min(this.pageIndex * this.pageSize, this.total)];
            this.buildIndexes();
        }
    }
    onRefresh(e) {
        this.clickRefresh.emit(e);
    }
}
TDSPaginationDefaultComponent.ɵfac = function TDSPaginationDefaultComponent_Factory(t) { return new (t || TDSPaginationDefaultComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.Directionality, 8)); };
TDSPaginationDefaultComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationDefaultComponent, selectors: [["tds-pagination-default"]], viewQuery: function TDSPaginationDefaultComponent_Query(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵviewQuery(_c0, 7);
        }
        if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.template = _t.first);
        }
    }, inputs: { size: "size", itemRender: "itemRender", showTotal: "showTotal", disabled: "disabled", locale: "locale", showSizeChanger: "showSizeChanger", showQuickJumper: "showQuickJumper", total: "total", pageIndex: "pageIndex", pageSize: "pageSize", pageSizeOptions: "pageSizeOptions", showRefresh: "showRefresh", bordered: "bordered" }, outputs: { pageIndexChange: "pageIndexChange", pageSizeChange: "pageSizeChange", clickRefresh: "clickRefresh" }, features: [i0.ɵɵNgOnChangesFeature], decls: 6, vars: 0, consts: [["containerTemplate", ""], ["pageTemplate", ""], ["nextPrevTemplate", ""], [1, "flex", "items-center", "w-full", "select-none"], [1, "flex", "items-center"], ["ngFor", "", 3, "ngForOf", "ngForTrackBy"], ["class", "ml-10", "tds-pagination-options", "", 3, "total", "locale", "disabled", "size", "showSizeChanger", "showQuickJumper", "pageIndex", "pageSize", "pageSizeOptions", "pageIndexChange", "pageSizeChange", 4, "ngIf"], [1, "ml-auto", "flex", "items-center"], ["class", "tds-pagination-total-text", 4, "ngIf"], ["class", "mr-2.5 ml-3 ", "tds-button-icon", "", "color", "secondary", "size", "sm", 3, "click", 4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["tds-pagination-options", "", 1, "ml-10", 3, "total", "locale", "disabled", "size", "showSizeChanger", "showQuickJumper", "pageIndex", "pageSize", "pageSizeOptions", "pageIndexChange", "pageSizeChange"], [1, "tds-pagination-total-text"], ["tds-button-icon", "", "color", "secondary", "size", "sm", 1, "mr-2.5", "ml-3", 3, "click"], [1, "tdsi-loading-fill"], ["tds-pagination-item", "", 1, "min-w-7", "min-h-7", "text-sm", "text-center", "align-middle", "mx-0.5", "leading-7", "box-content", "select-auto", "group", "border", "rounded-md", 3, "locale", "ngClass", "type", "index", "disabled", "itemRender", "active", "direction", "gotoIndex", "diffIndex"], ["tds-pagination-item", "", 1, "min-w-7", "min-h-7", "text-sm", "text-center", "align-middle", "mx-0.5", "leading-7", "box-content", "select-auto", "group", "border", "border-transparent", "hover:bg-neutral-3-50", "dark:hover:bg-d-neutral-3-300", "rounded-md", "flex", "items-center", "justify-center", 3, "locale", "ngClass", "type", "index", "disabled", "itemRender", "active", "direction", "gotoIndex", "diffIndex"]], template: function TDSPaginationDefaultComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵtemplate(0, TDSPaginationDefaultComponent_ng_template_0_Template, 7, 5, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(2, TDSPaginationDefaultComponent_ng_template_2_Template, 1, 16, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(4, TDSPaginationDefaultComponent_ng_template_4_Template, 1, 11, "ng-template", null, 2, i0.ɵɵtemplateRefExtractor);
        }
    }, directives: [i2.NgForOf, i2.NgSwitch, i2.NgSwitchCase, i2.NgTemplateOutlet, i2.NgSwitchDefault, i2.NgIf, TDSPaginationOptionsComponent, i4$1.TDSButtonIconComponent, TDSPaginationItemComponent, i2.NgClass], encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationDefaultComponent, [{
            type: Component,
            args: [{ selector: 'tds-pagination-default', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-template #containerTemplate>\r\n    <div class=\"flex items-center w-full select-none\">\r\n        <ul class=\"flex items-center\">\r\n            <ng-template ngFor let-page [ngForOf]=\"listOfPageItem\" [ngForTrackBy]=\"trackByPageItem\">\r\n                <ng-container [ngSwitch]=\"page.type\">\r\n                    <ng-container *ngSwitchCase=\"'page'\">\r\n                        <ng-container [ngTemplateOutlet]=\"pageTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale,bordered:bordered, dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container>\r\n                    <!-- <ng-container *ngSwitchCase=\"'prev'\">\r\n                        <ng-container [ngTemplateOutlet]=\"nextPrevTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale, dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container>\r\n                    <ng-container *ngSwitchCase=\"'next'\">\r\n                        <ng-container [ngTemplateOutlet]=\"nextPrevTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale, dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container> -->\r\n                    <ng-container *ngSwitchDefault>\r\n                        <ng-container [ngTemplateOutlet]=\"nextPrevTemplate\"\r\n                            [ngTemplateOutletContext]=\"{ $implicit: page, locale:locale, bordered:bordered ,  dir:dir}\">\r\n                        </ng-container>\r\n                    </ng-container>\r\n\r\n                </ng-container>\r\n\r\n\r\n            </ng-template>\r\n\r\n        </ul>\r\n        <div class=\"ml-10\" tds-pagination-options *ngIf=\"showQuickJumper || showSizeChanger\" [total]=\"total\"\r\n            [locale]=\"locale\" [disabled]=\"disabled\" [size]=\"size\" [showSizeChanger]=\"showSizeChanger\"\r\n            [showQuickJumper]=\"showQuickJumper\" [pageIndex]=\"pageIndex\" [pageSize]=\"pageSize\"\r\n            [pageSizeOptions]=\"pageSizeOptions\" (pageIndexChange)=\"onPageIndexChange($event)\"\r\n            (pageSizeChange)=\"onPageSizeChange($event)\"></div>\r\n\r\n        <div class=\"ml-auto flex items-center\">\r\n            <div class=\"tds-pagination-total-text\" *ngIf=\"showTotal\">\r\n                <ng-template [ngTemplateOutlet]=\"showTotal\"\r\n                    [ngTemplateOutletContext]=\"{ $implicit: total, range: ranges }\">\r\n                </ng-template>\r\n            </div>\r\n            <button class=\"mr-2.5 ml-3 \" (click)=\"onRefresh($event)\" tds-button-icon color=\"secondary\" *ngIf=\"showRefresh\"\r\n                size=\"sm\">\r\n                <span class=\"flex items-center\">\r\n                    <i class=\"tdsi-loading-fill\"></i>\r\n                </span>\r\n            </button>\r\n        </div>\r\n    </div>\r\n\r\n\r\n</ng-template>\r\n<ng-template #pageTemplate let-page let-local=\"local\" let-dir=\"dir\" let-bordered=\"bordered\">\r\n    <li tds-pagination-item [locale]=\"locale\"\r\n        class=\"min-w-7 min-h-7 text-sm  text-center align-middle mx-0.5 leading-7 box-content select-auto group border rounded-md\"\r\n        [ngClass]=\"{'bg-primary-1 text-white':pageIndex === page.index,\r\n     'text-neutral-1-600 dark:text-d-neutral-1-600 hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300':pageIndex !== page.index,\r\n        'cursor-pointer':!page.disabled,'cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500':!!page.disabled,\r\n        'border-transparent':!bordered,\r\n        ' border-primary-1':pageIndex === page.index && bordered,\r\n        'border-neutral-2-200 dark:border-d-neutral-2-200':pageIndex !== page.index && bordered\r\n    }\" [type]=\"page.type\" [index]=\"page.index\" [disabled]=\"!!page.disabled\" [itemRender]=\"itemRender\"\r\n        [active]=\"pageIndex === page.index\" (gotoIndex)=\"jumpPage($event)\" (diffIndex)=\"jumpDiff($event)\"\r\n        [direction]=\"dir\"></li>\r\n</ng-template>\r\n<ng-template #nextPrevTemplate let-page let-local=\"local\" let-dir=\"dir\" let-bordered=\"bordered\">\r\n    <li tds-pagination-item [locale]=\"locale\"\r\n        class=\"min-w-7 min-h-7 text-sm text-center align-middle mx-0.5 leading-7 box-content select-auto group border border-transparent hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300 rounded-md flex items-center justify-center\"\r\n        [ngClass]=\"{'cursor-pointer':!page.disabled,'cursor-not-allowed text-neutral-1-500 dark:text-d-neutral-1-500':!!page.disabled}\"\r\n        [type]=\"page.type\" [index]=\"page.index\" [disabled]=\"!!page.disabled\" [itemRender]=\"itemRender\"\r\n        [active]=\"pageIndex === page.index\" (gotoIndex)=\"jumpPage($event)\" (diffIndex)=\"jumpDiff($event)\"\r\n        [direction]=\"dir\"></li>\r\n</ng-template>" }]
        }], function () {
        return [{ type: i0.ChangeDetectorRef }, { type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i1.Directionality, decorators: [{
                        type: Optional
                    }] }];
    }, { template: [{
                type: ViewChild,
                args: ['containerTemplate', { static: true }]
            }], size: [{
                type: Input
            }], itemRender: [{
                type: Input
            }], showTotal: [{
                type: Input
            }], disabled: [{
                type: Input
            }], locale: [{
                type: Input
            }], showSizeChanger: [{
                type: Input
            }], showQuickJumper: [{
                type: Input
            }], total: [{
                type: Input
            }], pageIndex: [{
                type: Input
            }], pageSize: [{
                type: Input
            }], pageSizeOptions: [{
                type: Input
            }], showRefresh: [{
                type: Input
            }], bordered: [{
                type: Input
            }], pageIndexChange: [{
                type: Output
            }], pageSizeChange: [{
                type: Output
            }], clickRefresh: [{
                type: Output
            }] });
})();

function TDSPaginationComponent_ng_container_0_ng_container_1_ng_template_1_Template(rf, ctx) { }
function TDSPaginationComponent_ng_container_0_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSPaginationComponent_ng_container_0_ng_container_1_ng_template_1_Template, 0, 0, "ng-template", 6);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(2);
        const _r1 = i0.ɵɵreference(2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r1.template);
    }
}
function TDSPaginationComponent_ng_container_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSPaginationComponent_ng_container_0_ng_container_1_Template, 2, 1, "ng-container", 5);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r0 = i0.ɵɵnextContext();
        const _r2 = i0.ɵɵreference(4);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r0.simple)("ngIfElse", _r2.template);
    }
}
const TDS_CONFIG_MODULE_NAME = 'pagination';
class TDSPaginationComponent {
    constructor(i18n, cdr, breakpointService, tdsConfigService, directionality, elementRef) {
        this.i18n = i18n;
        this.cdr = cdr;
        this.breakpointService = breakpointService;
        this.tdsConfigService = tdsConfigService;
        this.directionality = directionality;
        this.elementRef = elementRef;
        this._tdsModuleName = TDS_CONFIG_MODULE_NAME;
        this.pageSizeChange = new EventEmitter();
        this.pageIndexChange = new EventEmitter();
        this.showTotal = null;
        this.itemRender = null;
        this.size = 'sm';
        this.pageSizeOptions = [10, 20, 30, 40];
        this.showSizeChanger = false;
        this.showQuickJumper = false;
        this.simple = false;
        this.disabled = false;
        this.responsive = false;
        this.hideOnSinglePage = false;
        this.total = 0;
        this.pageIndex = 1;
        this.pageSize = 10;
        this.showRefresh = false;
        this.bordered = true;
        this.clickRefresh = new EventEmitter();
        this.showPagination = true;
        this._size = "sm";
        this.dir = 'ltr';
        this.destroy$ = new Subject();
        this.total$ = new ReplaySubject(1);
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-pagination');
    }
    validatePageIndex(value, lastIndex) {
        if (value > lastIndex) {
            return lastIndex;
        }
        else if (value < 1) {
            return 1;
        }
        else {
            return value;
        }
    }
    onPageIndexChange(index) {
        const lastIndex = this.getLastIndex(this.total, this.pageSize);
        const validIndex = this.validatePageIndex(index, lastIndex);
        if (validIndex !== this.pageIndex && !this.disabled) {
            this.pageIndex = validIndex;
            this.pageIndexChange.emit(this.pageIndex);
        }
    }
    onPageSizeChange(size) {
        this.pageSize = size;
        this.pageSizeChange.emit(size);
        const lastIndex = this.getLastIndex(this.total, this.pageSize);
        if (this.pageIndex > lastIndex) {
            this.onPageIndexChange(lastIndex);
        }
    }
    onTotalChange(total) {
        const lastIndex = this.getLastIndex(total, this.pageSize);
        if (this.pageIndex > lastIndex) {
            Promise.resolve().then(() => {
                this.onPageIndexChange(lastIndex);
                this.cdr.markForCheck();
            });
        }
    }
    getLastIndex(total, pageSize) {
        return Math.ceil(total / pageSize);
    }
    ngOnInit() {
        var _a;
        this.i18n.localeChange.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.locale = this.i18n.getLocaleData('Pagination');
            this.cdr.markForCheck();
        });
        this.total$.pipe(takeUntil(this.destroy$)).subscribe(total => {
            this.onTotalChange(total);
        });
        this.breakpointService
            .subscribe(gridResponsiveMap)
            .pipe(takeUntil(this.destroy$))
            .subscribe(bp => {
            if (this.responsive) {
                this._size = bp === TDSBreakpointEnum.xs ? 'sm' : 'md';
                this.cdr.markForCheck();
            }
        });
        (_a = this.directionality.change) === null || _a === void 0 ? void 0 : _a.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    ngOnChanges(changes) {
        const { hideOnSinglePage, total, pageSize, size } = changes;
        if (total) {
            this.total$.next(this.total);
        }
        if (hideOnSinglePage || total || pageSize) {
            this.showPagination =
                (this.hideOnSinglePage && this.total > this.pageSize) || (this.total > 0 && !this.hideOnSinglePage);
        }
        if (size) {
            this._size = size.currentValue;
        }
    }
    onClickRefresh(e) {
        this.clickRefresh.emit(e);
    }
}
TDSPaginationComponent.ɵfac = function TDSPaginationComponent_Factory(t) { return new (t || TDSPaginationComponent)(i0.ɵɵdirectiveInject(i1$1.TDSI18nService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i2$2.TDSBreakpointService), i0.ɵɵdirectiveInject(i3$1.TDSConfigService), i0.ɵɵdirectiveInject(i1.Directionality, 8), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSPaginationComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSPaginationComponent, selectors: [["tds-pagination"]], hostVars: 8, hostBindings: function TDSPaginationComponent_HostBindings(rf, ctx) {
        if (rf & 2) {
            i0.ɵɵclassProp("tds-pagination-simple", ctx.simple)("tds-pagination-disabled", ctx.disabled)("mini", !ctx.simple && ctx._size === "small")("tds-pagination-rtl", ctx.dir === "rtl");
        }
    }, inputs: { showTotal: "showTotal", itemRender: "itemRender", size: "size", pageSizeOptions: "pageSizeOptions", showSizeChanger: "showSizeChanger", showQuickJumper: "showQuickJumper", simple: "simple", disabled: "disabled", responsive: "responsive", hideOnSinglePage: "hideOnSinglePage", total: "total", pageIndex: "pageIndex", pageSize: "pageSize", showRefresh: "showRefresh", bordered: "bordered" }, outputs: { pageSizeChange: "pageSizeChange", pageIndexChange: "pageIndexChange", clickRefresh: "clickRefresh" }, exportAs: ["tdsPagination"], features: [i0.ɵɵNgOnChangesFeature], decls: 5, vars: 21, consts: [[4, "ngIf"], [3, "disabled", "itemRender", "locale", "pageSize", "total", "pageIndex", "size", "pageIndexChange"], ["simplePagination", ""], [3, "size", "itemRender", "showTotal", "showRefresh", "disabled", "locale", "showSizeChanger", "showQuickJumper", "total", "pageIndex", "pageSize", "pageSizeOptions", "bordered", "pageIndexChange", "pageSizeChange", "clickRefresh"], ["defaultPagination", ""], [4, "ngIf", "ngIfElse"], [3, "ngTemplateOutlet"]], template: function TDSPaginationComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵtemplate(0, TDSPaginationComponent_ng_container_0_Template, 2, 2, "ng-container", 0);
            i0.ɵɵelementStart(1, "tds-pagination-simple", 1, 2);
            i0.ɵɵlistener("pageIndexChange", function TDSPaginationComponent_Template_tds_pagination_simple_pageIndexChange_1_listener($event) { return ctx.onPageIndexChange($event); });
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(3, "tds-pagination-default", 3, 4);
            i0.ɵɵlistener("pageIndexChange", function TDSPaginationComponent_Template_tds_pagination_default_pageIndexChange_3_listener($event) { return ctx.onPageIndexChange($event); })("pageSizeChange", function TDSPaginationComponent_Template_tds_pagination_default_pageSizeChange_3_listener($event) { return ctx.onPageSizeChange($event); })("clickRefresh", function TDSPaginationComponent_Template_tds_pagination_default_clickRefresh_3_listener($event) { return ctx.onClickRefresh($event); });
            i0.ɵɵelementEnd();
        }
        if (rf & 2) {
            i0.ɵɵproperty("ngIf", ctx.showPagination);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("disabled", ctx.disabled)("itemRender", ctx.itemRender)("locale", ctx.locale)("pageSize", ctx.pageSize)("total", ctx.total)("pageIndex", ctx.pageIndex)("size", ctx._size);
            i0.ɵɵadvance(2);
            i0.ɵɵproperty("size", ctx._size)("itemRender", ctx.itemRender)("showTotal", ctx.showTotal)("showRefresh", ctx.showRefresh)("disabled", ctx.disabled)("locale", ctx.locale)("showSizeChanger", ctx.showSizeChanger)("showQuickJumper", ctx.showQuickJumper)("total", ctx.total)("pageIndex", ctx.pageIndex)("pageSize", ctx.pageSize)("pageSizeOptions", ctx.pageSizeOptions)("bordered", ctx.bordered);
        }
    }, directives: [i2.NgIf, i2.NgTemplateOutlet, TDSPaginationSimpleComponent, TDSPaginationDefaultComponent], encapsulation: 2, changeDetection: 0 });
__decorate([
    WithConfig()
], TDSPaginationComponent.prototype, "size", void 0);
__decorate([
    WithConfig()
], TDSPaginationComponent.prototype, "pageSizeOptions", void 0);
__decorate([
    WithConfig(),
    InputBoolean()
], TDSPaginationComponent.prototype, "showSizeChanger", void 0);
__decorate([
    WithConfig(),
    InputBoolean()
], TDSPaginationComponent.prototype, "showQuickJumper", void 0);
__decorate([
    WithConfig(),
    InputBoolean()
], TDSPaginationComponent.prototype, "simple", void 0);
__decorate([
    InputBoolean()
], TDSPaginationComponent.prototype, "disabled", void 0);
__decorate([
    InputBoolean()
], TDSPaginationComponent.prototype, "responsive", void 0);
__decorate([
    InputBoolean()
], TDSPaginationComponent.prototype, "hideOnSinglePage", void 0);
__decorate([
    InputNumber()
], TDSPaginationComponent.prototype, "total", void 0);
__decorate([
    InputNumber()
], TDSPaginationComponent.prototype, "pageIndex", void 0);
__decorate([
    InputNumber()
], TDSPaginationComponent.prototype, "pageSize", void 0);
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationComponent, [{
            type: Component,
            args: [{
                    selector: 'tds-pagination',
                    exportAs: 'tdsPagination',
                    preserveWhitespaces: false,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
    <ng-container *ngIf="showPagination">
      <ng-container *ngIf="simple; else defaultPagination.template">
        <ng-template [ngTemplateOutlet]="simplePagination.template"></ng-template>
      </ng-container>
    </ng-container>
    <tds-pagination-simple
      #simplePagination
      [disabled]="disabled"
      [itemRender]="itemRender"
      [locale]="locale"
      [pageSize]="pageSize"
      [total]="total"
      [pageIndex]="pageIndex"
      [size]="_size"
      (pageIndexChange)="onPageIndexChange($event)"
    ></tds-pagination-simple>
    <tds-pagination-default
      #defaultPagination
      [size]="_size"
      [itemRender]="itemRender"
      [showTotal]="showTotal"
      [showRefresh]="showRefresh"
      [disabled]="disabled"
      [locale]="locale"
      [showSizeChanger]="showSizeChanger"
      [showQuickJumper]="showQuickJumper"
      [total]="total"
      [pageIndex]="pageIndex"
      [pageSize]="pageSize"
      [pageSizeOptions]="pageSizeOptions"
      [bordered]="bordered"
      (pageIndexChange)="onPageIndexChange($event)"
      (pageSizeChange)="onPageSizeChange($event)"
      (clickRefresh)="onClickRefresh($event)"
    ></tds-pagination-default>
  `,
                    host: {
                        '[class.tds-pagination-simple]': 'simple',
                        '[class.tds-pagination-disabled]': 'disabled',
                        '[class.mini]': `!simple && _size === 'small'`,
                        '[class.tds-pagination-rtl]': `dir === 'rtl'`
                    }
                }]
        }], function () {
        return [{ type: i1$1.TDSI18nService }, { type: i0.ChangeDetectorRef }, { type: i2$2.TDSBreakpointService }, { type: i3$1.TDSConfigService }, { type: i1.Directionality, decorators: [{
                        type: Optional
                    }] }, { type: i0.ElementRef }];
    }, { pageSizeChange: [{
                type: Output
            }], pageIndexChange: [{
                type: Output
            }], showTotal: [{
                type: Input
            }], itemRender: [{
                type: Input
            }], size: [{
                type: Input
            }], pageSizeOptions: [{
                type: Input
            }], showSizeChanger: [{
                type: Input
            }], showQuickJumper: [{
                type: Input
            }], simple: [{
                type: Input
            }], disabled: [{
                type: Input
            }], responsive: [{
                type: Input
            }], hideOnSinglePage: [{
                type: Input
            }], total: [{
                type: Input
            }], pageIndex: [{
                type: Input
            }], pageSize: [{
                type: Input
            }], showRefresh: [{
                type: Input
            }], bordered: [{
                type: Input
            }], clickRefresh: [{
                type: Output
            }] });
})();

class TDSPaginationModule {
}
TDSPaginationModule.ɵfac = function TDSPaginationModule_Factory(t) { return new (t || TDSPaginationModule)(); };
TDSPaginationModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSPaginationModule });
TDSPaginationModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, FormsModule, TDSSelectModule, TDSI18nModule, TDSFormFieldModule, TDSInputModule, TDSButtonModule]] });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPaginationModule, [{
            type: NgModule,
            args: [{
                    declarations: [
                        TDSPaginationComponent,
                        TDSPaginationSimpleComponent,
                        TDSPaginationOptionsComponent,
                        TDSPaginationItemComponent,
                        TDSPaginationDefaultComponent
                    ],
                    exports: [TDSPaginationComponent],
                    imports: [BidiModule, CommonModule, FormsModule, TDSSelectModule, TDSI18nModule, TDSFormFieldModule, TDSInputModule, TDSButtonModule]
                }]
        }], null, null);
})();
(function () {
    (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSPaginationModule, { declarations: [TDSPaginationComponent,
            TDSPaginationSimpleComponent,
            TDSPaginationOptionsComponent,
            TDSPaginationItemComponent,
            TDSPaginationDefaultComponent], imports: [BidiModule, CommonModule, FormsModule, TDSSelectModule, TDSI18nModule, TDSFormFieldModule, TDSInputModule, TDSButtonModule], exports: [TDSPaginationComponent] });
})();

/**
 * Generated bundle index. Do not edit.
 */

export { TDSPaginationComponent, TDSPaginationDefaultComponent, TDSPaginationItemComponent, TDSPaginationModule, TDSPaginationOptionsComponent, TDSPaginationSimpleComponent };
