// tds-ui/fesm2015/tds-ui-step.mjs
import * as i0 from '@angular/core';
import { TemplateRef, Component, ChangeDetectionStrategy, ViewEncapsulation, ViewChild, Input, EventEmitter, Optional, ContentChildren, Output, NgModule } from '@angular/core';
import { Subject, merge } from 'rxjs';
import { takeUntil, startWith } from 'rxjs/operators';
import { __decorate } from 'tslib';
import { TDSHelperString, InputBoolean, toBoolean } from 'tds-ui/shared/utility';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i2 from 'tds-ui/core/outlet';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import * as i1$1 from '@angular/cdk/bidi';
import { BidiModule } from '@angular/cdk/bidi';

const _c0$1 = ["processDotTemplate"];
const _c1 = function (a0, a1) { return { "border-t-success-400 dark:border-t-d-success-400 border-b-success-400 dark:border-b-d-success-400": a0, "border-t-neutral-3-100 dark:border-t-d-neutral-3-300 border-b-neutral-3-100 dark:border-b-d-neutral-3-300": a1 }; };
function TDSStepComponent_ng_container_2_div_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "div", 13);
    }
    if (rf & 2) {
        const ctx_r12 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(1, _c1, ctx_r12.activeBar, !ctx_r12.activeBar));
    }
}
function TDSStepComponent_ng_container_2_ng_template_4_ng_container_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r16 = i0.ɵɵnextContext(3);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r16.tdsTitle);
    }
}
function TDSStepComponent_ng_container_2_ng_template_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSStepComponent_ng_container_2_ng_template_4_ng_container_0_Template, 2, 1, "ng-container", 14);
    }
    if (rf & 2) {
        const ctx_r13 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r13.tdsTitle);
    }
}
const _c2 = function (a0, a1) { return { "border-l-success-400 dark:border-l-d-success-400": a0, "border-l-neutral-3-100 dark:border-l-d-neutral-3-300": a1 }; };
function TDSStepComponent_ng_container_2_div_5_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "div", 15);
    }
    if (rf & 2) {
        const ctx_r14 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(1, _c2, ctx_r14.activeBar, !ctx_r14.activeBar));
    }
}
function TDSStepComponent_ng_container_2_div_6_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "div", 17);
        i0.ɵɵtext(2);
        i0.ɵɵelementEnd();
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r17 = i0.ɵɵnextContext(3);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx_r17.description, " ");
    }
}
function TDSStepComponent_ng_container_2_div_6_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 16);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_2_div_6_ng_container_1_Template, 3, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r15 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r15.description);
    }
}
const _c3 = function (a0, a1, a2, a3) { return { "rounded-r tds-step-rounded-right": a0, "rounded-l tds-step-rounded-left": a1, "bg-primary-1 dark:bg-d-primary-1 text-white": a2, "bg-neutral-3-100 dark:bg-d-neutral-3-300 text-neutral-1-600 dark:text-d-neutral-1-600": a3 }; };
function TDSStepComponent_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        const _r19 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "div", 7);
        i0.ɵɵlistener("click", function TDSStepComponent_ng_container_2_Template_div_click_1_listener() { i0.ɵɵrestoreView(_r19); const ctx_r18 = i0.ɵɵnextContext(); return ctx_r18.onClick(); });
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_2_div_2_Template, 1, 4, "div", 8);
        i0.ɵɵelementStart(3, "div", 9);
        i0.ɵɵtemplate(4, TDSStepComponent_ng_container_2_ng_template_4_Template, 1, 1, "ng-template", 10);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(5, TDSStepComponent_ng_container_2_div_5_Template, 1, 4, "div", 11);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(6, TDSStepComponent_ng_container_2_div_6_Template, 2, 1, "div", 12);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r0 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tabindex", ctx_r0.clickable && !ctx_r0.disabled ? 0 : null)("ngClass", ctx_r0.disabled ? "cursor-not-allowed" : "cursor-pointer");
        i0.ɵɵattribute("role", ctx_r0.clickable && !ctx_r0.disabled ? "button" : null);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r0.first !== true);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction4(8, _c3, ctx_r0.last == true, ctx_r0.first == true, ctx_r0.activeBar, !ctx_r0.activeBar));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r0.size !== "sm");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r0.last !== true);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r0.description);
    }
}
const _c4 = function (a0, a1, a2, a3, a4) { return { "border-neutral-2-200 dark:border-d-neutral-2-200": a0, "border-primary-1": a1, "top-4": a2, "top-3": a3, "top-5": a4 }; };
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_div_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "div", 26);
    }
    if (rf & 2) {
        const ctx_r24 = i0.ɵɵnextContext(4);
        i0.ɵɵclassMapInterpolate1("tds-steps-item-tail w-full border-t  absolute left-0 right-0 ", ctx_r24.borderStyle, "");
        i0.ɵɵstyleProp("margin-left", ctx_r24.size == "md" ? 89 : ctx_r24.size == "sm" ? 85 : 93, "px")("width", ctx_r24.size == "md" ? "calc(100% - 33px)" : ctx_r24.size == "lg" ? "calc(100% - 41px)" : "calc(100% - 25px)");
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction5(8, _c4, ctx_r24.status !== "finish", ctx_r24.status == "finish", ctx_r24.size == "md", ctx_r24.size == "sm", ctx_r24.size == "lg"));
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_1_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r2 = i0.ɵɵreference(5);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r2);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_2_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r4 = i0.ɵɵreference(7);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r4);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_3_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_3_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r6 = i0.ɵɵreference(9);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r6);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_4_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_4_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r8 = i0.ɵɵreference(11);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r8);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_1_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_2_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_3_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(4, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_ng_container_4_Template, 2, 1, "ng-container", 1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r30 = i0.ɵɵnextContext(5);
        i0.ɵɵproperty("ngSwitch", ctx_r30.status);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "wait");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "finish");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "error");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "process");
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_1_ng_container_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_1_ng_container_0_Template, 1, 0, "ng-container", 27);
    }
    if (rf & 2) {
        i0.ɵɵnextContext(5);
        const _r10 = i0.ɵɵreference(13);
        i0.ɵɵproperty("ngTemplateOutlet", _r10);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_0_Template, 5, 5, "ng-template", 10);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_ng_template_1_Template, 1, 1, "ng-template", 10);
    }
    if (rf & 2) {
        const ctx_r25 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngIf", !ctx_r25.icon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r25.icon);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_5_ng_template_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "span", 31);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_5_ng_template_3_Template(rf, ctx) { }
const _c5 = function (a0, a1, a2) { return { $implicit: a0, status: a1, index: a2 }; };
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_5_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 28);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_5_ng_template_1_Template, 1, 0, "ng-template", null, 29, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_5_ng_template_3_Template, 0, 0, "ng-template", 30);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const _r41 = i0.ɵɵreference(2);
        const ctx_r26 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngTemplateOutlet", ctx_r26.customProcessTemplate || _r41)("ngTemplateOutletContext", i0.ɵɵpureFunction3(2, _c5, _r41, ctx_r26.status, ctx_r26.index));
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_container_9_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r27 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r27.tdsTitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_div_10_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r44 = i0.ɵɵnextContext(5);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r44.subtitle);
    }
}
const _c6 = function (a0, a1) { return { "text-error-400": a0, "text-neutral-1-700 dark:text-d-neutral-1-700": a1 }; };
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_div_10_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 32);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_div_10_ng_container_1_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r28 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(2, _c6, ctx_r28.status == "error", ctx_r28.status !== "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r28.subtitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_container_12_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r29 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r29.description);
    }
}
const _c7 = function (a0, a1, a2, a3) { return { "font-semibold": a0, "text-neutral-1-900 dark:text-d-neutral-1-900": a1, "text-neutral-1-600 dark:text-d-neutral-1-600": a2, "text-error-400": a3 }; };
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        const _r46 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "div", 18);
        i0.ɵɵlistener("click", function TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_Template_div_click_1_listener() { i0.ɵɵrestoreView(_r46); const ctx_r45 = i0.ɵɵnextContext(3); return ctx_r45.onClick(); });
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_div_2_Template, 1, 14, "div", 19);
        i0.ɵɵelementStart(3, "div", 20);
        i0.ɵɵtemplate(4, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_4_Template, 2, 2, "ng-template", 10);
        i0.ɵɵtemplate(5, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_template_5_Template, 4, 6, "ng-template", 10);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(6, "div", 21)(7, "div", 22)(8, "div", 23);
        i0.ɵɵtemplate(9, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_container_9_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(10, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_div_10_Template, 2, 5, "div", 24);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(11, "div", 25);
        i0.ɵɵtemplate(12, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_ng_container_12_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd()()();
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r22 = i0.ɵɵnextContext(3);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tabindex", ctx_r22.clickable && !ctx_r22.disabled ? 0 : null)("ngClass", ctx_r22.disabled ? "cursor-not-allowed" : "");
        i0.ɵɵattribute("role", ctx_r22.clickable && !ctx_r22.disabled ? "button" : null);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r22.last !== true);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx_r22.showProcessDot);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r22.showProcessDot);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction4(11, _c7, ctx_r22.size != "sm", ctx_r22.currentIndex === ctx_r22.index && ctx_r22.status !== "error" || ctx_r22.status === "finish", ctx_r22.currentIndex !== ctx_r22.index && (ctx_r22.status !== "finish" && ctx_r22.status !== "error"), ctx_r22.status == "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r22.tdsTitle);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r22.subtitle);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(16, _c6, ctx_r22.status == "error", ctx_r22.status !== "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r22.description);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_1_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r2 = i0.ɵɵreference(5);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r2);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_2_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r4 = i0.ɵɵreference(7);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r4);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_3_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_3_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r6 = i0.ɵɵreference(9);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r6);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_4_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_4_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(6);
        const _r8 = i0.ɵɵreference(11);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r8);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_1_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_2_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_3_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(4, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_ng_container_4_Template, 2, 1, "ng-container", 1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r53 = i0.ɵɵnextContext(5);
        i0.ɵɵproperty("ngSwitch", ctx_r53.status);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "wait");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "finish");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "error");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "process");
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_1_ng_container_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_1_ng_container_0_Template, 1, 0, "ng-container", 27);
    }
    if (rf & 2) {
        i0.ɵɵnextContext(5);
        const _r10 = i0.ɵɵreference(13);
        i0.ɵɵproperty("ngTemplateOutlet", _r10);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_0_Template, 5, 5, "ng-template", 10);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_ng_template_1_Template, 1, 1, "ng-template", 10);
    }
    if (rf & 2) {
        const ctx_r47 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngIf", !ctx_r47.icon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r47.icon);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_4_ng_template_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "span", 31);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_4_ng_template_3_Template(rf, ctx) { }
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 28);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_4_ng_template_1_Template, 1, 0, "ng-template", null, 29, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_4_ng_template_3_Template, 0, 0, "ng-template", 30);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const _r64 = i0.ɵɵreference(2);
        const ctx_r48 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngTemplateOutlet", ctx_r48.customProcessTemplate || _r64)("ngTemplateOutletContext", i0.ɵɵpureFunction3(2, _c5, _r64, ctx_r48.status, ctx_r48.index));
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_container_8_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r49 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r49.tdsTitle);
    }
}
const _c8 = function (a0, a1, a2, a3, a4) { return { "bg-neutral-2-200 dark:bg-d-neutral-2-200": a0, "bg-primary-1": a1, "top-4": a2, "top-3": a3, "top-5": a4 }; };
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_div_10_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "div", 41);
    }
    if (rf & 2) {
        const ctx_r50 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction5(1, _c8, ctx_r50.status !== "finish", ctx_r50.status == "finish", ctx_r50.size == "md", ctx_r50.size == "sm", ctx_r50.size == "lg"));
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_div_11_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r67 = i0.ɵɵnextContext(5);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r67.subtitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_div_11_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 42);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_div_11_ng_container_1_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r51 = i0.ɵɵnextContext(4);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(2, _c6, ctx_r51.status == "error", ctx_r51.status !== "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r51.subtitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_container_13_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r52 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r52.description);
    }
}
const _c9 = function (a0) { return { "flex items-stretch": a0 }; };
const _c10 = function (a0, a1, a2, a3, a4) { return { "font-semibold": a0, "text-neutral-1-900 dark:text-d-neutral-1-900": a1, "text-neutral-1-600 dark:text-d-neutral-1-600": a2, "text-error-400": a3, "items-center flex": a4 }; };
function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        const _r69 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "div", 33);
        i0.ɵɵlistener("click", function TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_Template_div_click_1_listener() { i0.ɵɵrestoreView(_r69); const ctx_r68 = i0.ɵɵnextContext(3); return ctx_r68.onClick(); });
        i0.ɵɵelementStart(2, "div", 34);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_3_Template, 2, 2, "ng-template", 10);
        i0.ɵɵtemplate(4, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_template_4_Template, 4, 6, "ng-template", 10);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(5, "div", 35)(6, "div", 36)(7, "div", 37);
        i0.ɵɵtemplate(8, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_container_8_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(9, "div", 38);
        i0.ɵɵtemplate(10, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_div_10_Template, 1, 7, "div", 39);
        i0.ɵɵelementEnd()();
        i0.ɵɵtemplate(11, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_div_11_Template, 2, 5, "div", 40);
        i0.ɵɵelementStart(12, "div", 25);
        i0.ɵɵtemplate(13, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_ng_container_13_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd()()();
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r23 = i0.ɵɵnextContext(3);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tabindex", ctx_r23.clickable && !ctx_r23.disabled ? 0 : null);
        i0.ɵɵattribute("role", ctx_r23.clickable && !ctx_r23.disabled ? "button" : null);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx_r23.showProcessDot);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r23.showProcessDot);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(11, _c9, ctx_r23.onlyTitle));
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction5(13, _c10, ctx_r23.size != "sm", ctx_r23.currentIndex === ctx_r23.index && ctx_r23.status !== "error" || ctx_r23.status === "finish", ctx_r23.currentIndex !== ctx_r23.index && (ctx_r23.status !== "finish" && ctx_r23.status !== "error"), ctx_r23.status == "error", ctx_r23.onlyTitle));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r23.tdsTitle);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx_r23.last !== true);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r23.subtitle);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(19, _c6, ctx_r23.status == "error", ctx_r23.status !== "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r23.description);
    }
}
function TDSStepComponent_ng_container_3_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0)(1, 0);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_2_ng_container_2_Template, 13, 19, "ng-container", 1);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_2_ng_container_3_Template, 14, 22, "ng-container", 1);
        i0.ɵɵelementContainerEnd()();
    }
    if (rf & 2) {
        const ctx_r20 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitch", ctx_r20.labelPlacement);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "horizontal");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "vertical");
    }
}
const _c11 = function (a0, a1, a2, a3, a4) { return { "border-neutral-2-200 dark:border-d-neutral-2-200": a0, "border-primary-1": a1, "left-4 mt-8": a2, "left-3 mt-6": a3, "left-5 mt-10": a4 }; };
function TDSStepComponent_ng_container_3_ng_container_3_div_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "div", 26);
    }
    if (rf & 2) {
        const ctx_r70 = i0.ɵɵnextContext(3);
        i0.ɵɵclassMapInterpolate1("tds-steps-item-tail  border-l top-0 bottom-0 absolute ", ctx_r70.borderStyle, "");
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction5(4, _c11, ctx_r70.status !== "finish", ctx_r70.status == "finish", ctx_r70.size == "md", ctx_r70.size == "sm", ctx_r70.size == "lg"));
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_1_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(4);
        const _r2 = i0.ɵɵreference(5);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r2);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_2_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(4);
        const _r4 = i0.ɵɵreference(7);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r4);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_3_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_3_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(4);
        const _r6 = i0.ɵɵreference(9);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r6);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_4_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_4_ng_container_1_Template, 1, 0, "ng-container", 27);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        i0.ɵɵnextContext(4);
        const _r8 = i0.ɵɵreference(11);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", _r8);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 0);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_1_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_2_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_3_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(4, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_ng_container_4_Template, 2, 1, "ng-container", 1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r71 = i0.ɵɵnextContext(3);
        i0.ɵɵproperty("ngSwitch", ctx_r71.status);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "wait");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "finish");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "error");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "process");
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_6_ng_container_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainer(0);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_template_6_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSStepComponent_ng_container_3_ng_container_3_ng_template_6_ng_container_0_Template, 1, 0, "ng-container", 27);
    }
    if (rf & 2) {
        i0.ɵɵnextContext(3);
        const _r10 = i0.ɵɵreference(13);
        i0.ɵɵproperty("ngTemplateOutlet", _r10);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_container_10_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r73 = i0.ɵɵnextContext(3);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r73.tdsTitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_div_11_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r85 = i0.ɵɵnextContext(4);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r85.subtitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_div_11_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 32);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_container_3_ng_container_3_div_11_ng_container_1_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r74 = i0.ɵɵnextContext(3);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(2, _c6, ctx_r74.status == "error", ctx_r74.status !== "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r74.subtitle);
    }
}
function TDSStepComponent_ng_container_3_ng_container_3_ng_container_13_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r75 = i0.ɵɵnextContext(3);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx_r75.description);
    }
}
const _c12 = function (a0) { return { "pb-2": a0 }; };
const _c13 = function (a0, a1, a2, a3) { return { "flex items-center": a0, "h-8": a1, "h-10": a2, "h-6": a3 }; };
function TDSStepComponent_ng_container_3_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        const _r87 = i0.ɵɵgetCurrentView();
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "div", 43);
        i0.ɵɵlistener("click", function TDSStepComponent_ng_container_3_ng_container_3_Template_div_click_1_listener() { i0.ɵɵrestoreView(_r87); const ctx_r86 = i0.ɵɵnextContext(2); return ctx_r86.onClick(); });
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_3_div_2_Template, 1, 10, "div", 44);
        i0.ɵɵelementStart(3, "div", 45)(4, "div", 46);
        i0.ɵɵtemplate(5, TDSStepComponent_ng_container_3_ng_container_3_ng_template_5_Template, 5, 5, "ng-template", 10);
        i0.ɵɵtemplate(6, TDSStepComponent_ng_container_3_ng_container_3_ng_template_6_Template, 1, 1, "ng-template", 10);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(7, "div", 47)(8, "div", 48)(9, "div", 23);
        i0.ɵɵtemplate(10, TDSStepComponent_ng_container_3_ng_container_3_ng_container_10_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(11, TDSStepComponent_ng_container_3_ng_container_3_div_11_Template, 2, 5, "div", 24);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(12, "div", 25);
        i0.ɵɵtemplate(13, TDSStepComponent_ng_container_3_ng_container_3_ng_container_13_Template, 2, 1, "ng-container", 14);
        i0.ɵɵelementEnd()()()();
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r21 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵstyleProp("min-height", ctx_r21.last !== true ? 64 : null, "px");
        i0.ɵɵproperty("ngClass", ctx_r21.disabled ? "cursor-not-allowed" : "")("tabindex", ctx_r21.clickable && !ctx_r21.disabled ? 0 : null);
        i0.ɵɵattribute("role", ctx_r21.clickable && !ctx_r21.disabled ? "button" : null);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r21.last !== true);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", !ctx_r21.icon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r21.icon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(15, _c12, ctx_r21.last !== true));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction4(17, _c13, ctx_r21.onlyTitle, ctx_r21.size == "md" && ctx_r21.onlyTitle, ctx_r21.size == "lg" && ctx_r21.onlyTitle, ctx_r21.size == "sm" && ctx_r21.onlyTitle));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction4(22, _c7, ctx_r21.size != "sm", ctx_r21.currentIndex === ctx_r21.index && ctx_r21.status !== "error" || ctx_r21.status === "finish", ctx_r21.currentIndex !== ctx_r21.index && (ctx_r21.status !== "finish" && ctx_r21.status !== "error"), ctx_r21.status == "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r21.tdsTitle);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r21.subtitle);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(27, _c6, ctx_r21.status == "error", ctx_r21.status !== "error"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r21.description);
    }
}
function TDSStepComponent_ng_container_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0)(1, 0);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_container_3_ng_container_2_Template, 4, 3, "ng-container", 1);
        i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_ng_container_3_Template, 14, 30, "ng-container", 1);
        i0.ɵɵelementContainerEnd()();
    }
    if (rf & 2) {
        const ctx_r1 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitch", ctx_r1.direction);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "horizontal");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "vertical");
    }
}
function TDSStepComponent_ng_template_4_span_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 52);
        i0.ɵɵtext(1);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r88 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx_r88.index + 1, " ");
    }
}
const _c14 = function (a0, a1, a2) { return { "h-8 w-8": a0, "h-10 w-10": a1, "h-6 w-6": a2 }; };
function TDSStepComponent_ng_template_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 49)(1, "span", 50);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_template_4_span_2_Template, 2, 1, "span", 51);
        i0.ɵɵelementEnd()();
    }
    if (rf & 2) {
        const ctx_r3 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(2, _c14, ctx_r3.size == "md", ctx_r3.size == "lg", ctx_r3.size == "sm"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r3.showStep);
    }
}
function TDSStepComponent_ng_template_6_span_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 55);
        i0.ɵɵelement(1, "span", 56);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r89 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(1, _c14, ctx_r89.size == "md", ctx_r89.size == "lg", ctx_r89.size == "sm"));
    }
}
function TDSStepComponent_ng_template_6_span_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 57);
        i0.ɵɵelement(1, "span", 58);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r90 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(1, _c14, ctx_r90.size == "md", ctx_r90.size == "lg", ctx_r90.size == "sm"));
    }
}
function TDSStepComponent_ng_template_6_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 49);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_template_6_span_1_Template, 2, 5, "span", 53);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_template_6_span_2_Template, 2, 5, "span", 54);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r5 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx_r5.showStep);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r5.showStep);
    }
}
function TDSStepComponent_ng_template_8_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 49)(1, "span", 59);
        i0.ɵɵelement(2, "span", 60);
        i0.ɵɵelementEnd()();
    }
    if (rf & 2) {
        const ctx_r7 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(1, _c14, ctx_r7.size == "md", ctx_r7.size == "lg", ctx_r7.size == "sm"));
    }
}
function TDSStepComponent_ng_template_10_span_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 55)(1, "span", 62);
        i0.ɵɵtext(2);
        i0.ɵɵelementEnd()();
    }
    if (rf & 2) {
        const ctx_r91 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(2, _c14, ctx_r91.size == "md", ctx_r91.size == "lg", ctx_r91.size == "sm"));
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx_r91.index + 1, " ");
    }
}
function TDSStepComponent_ng_template_10_span_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "span", 57);
    }
    if (rf & 2) {
        const ctx_r92 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(1, _c14, ctx_r92.size == "md", ctx_r92.size == "lg", ctx_r92.size == "sm"));
    }
}
function TDSStepComponent_ng_template_10_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 49);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_template_10_span_1_Template, 3, 6, "span", 53);
        i0.ɵɵtemplate(2, TDSStepComponent_ng_template_10_span_2_Template, 1, 5, "span", 61);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r9 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r9.showStep);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx_r9.showStep);
    }
}
function TDSStepComponent_ng_template_12_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelementStart(1, "div", 63);
        i0.ɵɵelement(2, "span", 26);
        i0.ɵɵelementEnd();
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const icon_r94 = ctx.$implicit;
        const ctx_r93 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(2, _c14, ctx_r93.size == "md", ctx_r93.size == "lg", ctx_r93.size == "sm"));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", icon_r94);
    }
}
function TDSStepComponent_ng_template_12_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 49);
        i0.ɵɵtemplate(1, TDSStepComponent_ng_template_12_ng_container_1_Template, 3, 6, "ng-container", 14);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r11 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r11.icon);
    }
}
class TDSStepComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.disabled = false;
        this.isCustomStatus = false;
        this._status = 'wait';
        this.oldAPIIcon = true;
        this.mode = 'default';
        this.direction = 'horizontal';
        this.first = false;
        this.index = 0;
        this.last = false;
        this.outStatus = 'process';
        this.showProcessDot = false;
        this.clickable = false;
        this.showStep = false;
        this.click$ = new Subject();
        this.classMapTail = {};
        this.borderStyle = 'border-solid';
        this.backgroundColor = "";
        this._size = "md";
        this._labelPlacement = 'horizontal';
        this._currentIndex = 0;
    }
    get status() {
        return this._status;
    }
    set status(status) {
        this._status = status;
        this.isCustomStatus = true;
    }
    get icon() {
        return this._icon;
    }
    set icon(value) {
        if (!(value instanceof TemplateRef)) {
            this.oldAPIIcon = typeof value === 'string' && value.indexOf('anticon') > -1;
        }
        else {
        }
        this._icon = value;
    }
    get size() {
        return this._size;
    }
    set size(size) {
        this._size = size;
    }
    get labelPlacement() {
        return this._labelPlacement;
    }
    set labelPlacement(labelPlacement) {
        this._labelPlacement = labelPlacement;
    }
    get currentIndex() {
        return this._currentIndex;
    }
    set currentIndex(current) {
        this._currentIndex = current;
        if (!this.isCustomStatus) {
            this._status = current > this.index ? 'finish' : current === this.index ? this.outStatus || '' : 'wait';
        }
    }
    get activeBar() {
        return (this.currentIndex === this.index);
    }
    onClick() {
        if (this.clickable && this.currentIndex !== this.index && !this.disabled) {
            this.click$.next(this.index);
        }
    }
    enable() {
        this.disabled = false;
        this.cdr.markForCheck();
    }
    disable() {
        this.disabled = true;
        this.cdr.markForCheck();
    }
    markForCheck() {
        this.cdr.markForCheck();
    }
    ngOnDestroy() {
        this.click$.complete();
    }
    get onlyTitle() {
        return !TDSHelperString.hasValueString(this.subtitle) && !TDSHelperString.hasValueString(this.description);
    }
}
TDSStepComponent.ɵfac = function TDSStepComponent_Factory(t) { return new (t || TDSStepComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSStepComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSStepComponent, selectors: [["tds-step"]], viewQuery: function TDSStepComponent_Query(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵviewQuery(_c0$1, 5);
        }
        if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.processDotTemplate = _t.first);
        }
    }, hostAttrs: [1, "tds-steps-item", "overflow-visible", "relative", "inline-block", "align-top", "last:flex-none"], hostVars: 18, hostBindings: function TDSStepComponent_HostBindings(rf, ctx) {
        if (rf & 2) {
            i0.ɵɵclassProp("flex-1", ctx.last !== true && ctx.mode === "default")("tds-steps-item-wait", ctx.status === "wait")("tds-steps-item-process", ctx.status === "process")("tds-steps-item-finish", ctx.status === "finish")("tds-steps-item-error", ctx.status === "error")("tds-steps-item-active", ctx.currentIndex === ctx.index)("tds-steps-item-disabled", ctx.disabled)("tds-steps-item-custom", !!ctx.icon)("tds-steps-next-error", ctx.outStatus === "error" && ctx.currentIndex === ctx.index + 1);
        }
    }, inputs: { tdsTitle: "tdsTitle", subtitle: "subtitle", description: "description", disabled: "disabled", status: "status", icon: "icon" }, exportAs: ["tdsStep"], decls: 14, vars: 3, consts: [[3, "ngSwitch"], [4, "ngSwitchCase"], ["waitTemplate", ""], ["finishTemplate", ""], ["errorTemplate", ""], ["processTemplate", ""], ["customIConTemplate", ""], [1, "tds-steps-item-container", "flex", "outline-none", "min-w-32", "items-stretch", "relative", 3, "tabindex", "ngClass", "click"], ["class", "w-0 h-0  border-l-transparent tds-step-part-1", 3, "ngClass", 4, "ngIf"], [1, "flex-auto", "text-caption-1", "font-semibold", "flex", "items-center", "justify-center", "truncate", "tds-step-part-2", 3, "ngClass"], [3, "ngIf"], ["class", "w-0 h-0    border-t-transparent border-b-transparent tds-step-part-3", 3, "ngClass", 4, "ngIf"], ["class", "tds-steps-description w-full text-center", 4, "ngIf"], [1, "w-0", "h-0", "border-l-transparent", "tds-step-part-1", 3, "ngClass"], [4, "tdsStringTemplateOutlet"], [1, "w-0", "h-0", "border-t-transparent", "border-b-transparent", "tds-step-part-3", 3, "ngClass"], [1, "tds-steps-description", "w-full", "text-center"], [1, "p-4"], [1, "tds-steps-item-container", "relative", "outline-none", 3, "tabindex", "ngClass", "click"], [3, "class", "ngClass", "marginLeft", "width", 4, "ngIf"], [1, "tds-steps-item-icon", "w-36"], [1, "tds-steps-item-content", "text-center", "w-36", "mt-1"], [1, "tds-steps-item-title"], [1, "text-body-2", 3, "ngClass"], ["class", "tds-steps-item-subtitle  text-caption-2 font-regular", 3, "ngClass", 4, "ngIf"], [1, "tds-steps-item-description", "text-caption-2", "font-regular", 3, "ngClass"], [3, "ngClass"], [4, "ngTemplateOutlet"], [1, "tds-steps-icon"], ["processDotTemplate", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "tds-steps-icon-dot"], [1, "tds-steps-item-subtitle", "text-caption-2", "font-regular", 3, "ngClass"], [1, "tds-steps-item-container", "relative", "outline-none", "flex", "items-stretch", 3, "tabindex", "click"], [1, "tds-steps-item-icon"], [1, "tds-steps-item-content", "text-left", "ml-2", "flex-auto", 3, "ngClass"], [1, "tds-steps-item-title", "w-full", "flex"], [1, "text-body-2", "pr-2", 3, "ngClass"], [1, "flex-auto", "relative"], ["class", "tds-steps-item-tail w-full h-px  absolute left-0 right-0", 3, "ngClass", 4, "ngIf"], ["class", "tds-steps-item-subtitle w-full text-caption-2 font-regular", 3, "ngClass", 4, "ngIf"], [1, "tds-steps-item-tail", "w-full", "h-px", "absolute", "left-0", "right-0", 3, "ngClass"], [1, "tds-steps-item-subtitle", "w-full", "text-caption-2", "font-regular", 3, "ngClass"], [1, "tds-steps-item-container", "relative", "outline-none", "flex", "items-stretch", "h-full", 3, "ngClass", "tabindex", "click"], [3, "class", "ngClass", 4, "ngIf"], [1, "flex"], [1, "tds-steps-item-icon", "flex", "flex-col", "items-stretch"], [1, "tds-steps-item-content", "pl-2", 3, "ngClass"], [1, "tds-steps-item-title", 3, "ngClass"], [1, "tds-steps-icon", "flex", "w-full", "items-center", "justify-center"], [1, "rounded-full", "border", "border-neutral-2-200", "dark:border-d-neutral-2-200", "flex", "items-center", "justify-center", 3, "ngClass"], ["class", "text-neutral-1-300 text-body-2 font-semibold", 4, "ngIf"], [1, "text-neutral-1-300", "text-body-2", "font-semibold"], ["class", "rounded-full bg-primary-1 flex items-center justify-center", 3, "ngClass", 4, "ngIf"], ["class", "rounded-full border  border-primary-1  flex items-center justify-center", 3, "ngClass", 4, "ngIf"], [1, "rounded-full", "bg-primary-1", "flex", "items-center", "justify-center", 3, "ngClass"], [1, "tdsi-tick-fill", "text-white", "text-base", "leading-4"], [1, "rounded-full", "border", "border-primary-1", "flex", "items-center", "justify-center", 3, "ngClass"], [1, "tdsi-tick-fill", "text-primary-1", "text-base", "leading-4"], [1, "border", "border-error-400", "rounded-full", "flex", "items-center", "justify-center", 3, "ngClass"], [1, "tdsi-close-fill", "text-error-400", "text-base", "leading-4"], ["class", "rounded-full  border border-primary-1 flex items-center justify-center", 3, "ngClass", 4, "ngIf"], [1, "text-white", "font-semibold"], [1, "flex", "items-center", "justify-center", 3, "ngClass"]], template: function TDSStepComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵelementContainerStart(0)(1, 0);
            i0.ɵɵtemplate(2, TDSStepComponent_ng_container_2_Template, 7, 13, "ng-container", 1);
            i0.ɵɵtemplate(3, TDSStepComponent_ng_container_3_Template, 4, 3, "ng-container", 1);
            i0.ɵɵelementContainerEnd()();
            i0.ɵɵtemplate(4, TDSStepComponent_ng_template_4_Template, 3, 6, "ng-template", null, 2, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(6, TDSStepComponent_ng_template_6_Template, 3, 2, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(8, TDSStepComponent_ng_template_8_Template, 3, 5, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(10, TDSStepComponent_ng_template_10_Template, 3, 2, "ng-template", null, 5, i0.ɵɵtemplateRefExtractor);
            i0.ɵɵtemplate(12, TDSStepComponent_ng_template_12_Template, 2, 1, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
        }
        if (rf & 2) {
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngSwitch", ctx.mode);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngSwitchCase", "bar");
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngSwitchCase", "default");
        }
    }, directives: [i1.NgSwitch, i1.NgSwitchCase, i1.NgClass, i1.NgIf, i2.TDSStringTemplateOutletDirective, i1.NgTemplateOutlet], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSStepComponent.prototype, "disabled", void 0);
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSStepComponent, [{
            type: Component,
            args: [{ changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, selector: 'tds-step', exportAs: 'tdsStep', preserveWhitespaces: false, host: {
                        class: 'tds-steps-item overflow-visible relative inline-block  align-top last:flex-none',
                        '[class.flex-1]': 'last !== true && mode==="default"',
                        '[class.tds-steps-item-wait]': 'status === "wait"',
                        '[class.tds-steps-item-process]': 'status === "process"',
                        '[class.tds-steps-item-finish]': 'status === "finish"',
                        '[class.tds-steps-item-error]': 'status === "error"',
                        '[class.tds-steps-item-active]': 'currentIndex === index',
                        '[class.tds-steps-item-disabled]': 'disabled',
                        '[class.tds-steps-item-custom]': '!!icon',
                        '[class.tds-steps-next-error]': '(outStatus === "error") && (currentIndex === index + 1)',
                        // '[class.-ml-2.5]': 'mode==="bar"',
                    }, template: "<ng-container>\r\n  <ng-container [ngSwitch]=\"mode\">\r\n    <ng-container *ngSwitchCase=\"'bar'\">\r\n      <div class=\"tds-steps-item-container flex outline-none min-w-32  items-stretch relative\"\r\n        [attr.role]=\"clickable && !disabled ? 'button' : null\" [tabindex]=\"clickable && !disabled ? 0 : null\"\r\n        [ngClass]=\"disabled ? 'cursor-not-allowed' : 'cursor-pointer'\" (click)=\"onClick()\">\r\n        <div *ngIf=\"first !== true\"\r\n          class=\"w-0 h-0  border-l-transparent tds-step-part-1\"\r\n          [ngClass]=\"{'border-t-success-400 dark:border-t-d-success-400 border-b-success-400 dark:border-b-d-success-400':activeBar,\r\n          'border-t-neutral-3-100 dark:border-t-d-neutral-3-300 border-b-neutral-3-100 dark:border-b-d-neutral-3-300':!activeBar\r\n        }\"></div>\r\n        <div\r\n          class=\"flex-auto text-caption-1  font-semibold flex items-center justify-center  truncate tds-step-part-2\"\r\n          [ngClass]=\"{\r\n          'rounded-r tds-step-rounded-right':last == true,\r\n          'rounded-l tds-step-rounded-left':first == true,\r\n          'bg-primary-1 dark:bg-d-primary-1 text-white':activeBar,\r\n          'bg-neutral-3-100 dark:bg-d-neutral-3-300 text-neutral-1-600 dark:text-d-neutral-1-600':!activeBar}\">\r\n          <ng-template [ngIf]=\"size !== 'sm'\">\r\n            <ng-container *tdsStringTemplateOutlet=\"tdsTitle\">{{ tdsTitle }}</ng-container>\r\n          </ng-template>\r\n        </div>\r\n        <div *ngIf=\"last !== true\"\r\n          class=\"w-0 h-0    border-t-transparent border-b-transparent tds-step-part-3\"\r\n          [ngClass]=\"{'border-l-success-400 dark:border-l-d-success-400':activeBar,\r\n          'border-l-neutral-3-100 dark:border-l-d-neutral-3-300':!activeBar\r\n        }\">\r\n        </div>\r\n\r\n      </div>\r\n      <div class=\"tds-steps-description w-full text-center\" *ngIf=\"description\">\r\n        <ng-container *tdsStringTemplateOutlet=\"description\">\r\n          <div class=\"p-4\">\r\n            {{ description }}\r\n          </div>\r\n        </ng-container>\r\n      </div>\r\n\r\n    </ng-container>\r\n    <ng-container *ngSwitchCase=\"'default'\">\r\n      <ng-container [ngSwitch]=\"direction\">\r\n        <!--direction  horizontal -->\r\n        <ng-container *ngSwitchCase=\"'horizontal'\">\r\n          <ng-container [ngSwitch]=\"labelPlacement\">\r\n            <ng-container *ngSwitchCase=\"'horizontal'\">\r\n              <div class=\"tds-steps-item-container relative outline-none \"\r\n                [attr.role]=\"clickable && !disabled ? 'button' : null\" [tabindex]=\"clickable && !disabled ? 0 : null\"\r\n                [ngClass]=\"disabled ? 'cursor-not-allowed' : ''\" (click)=\"onClick()\">\r\n                <div class=\"tds-steps-item-tail w-full border-t  absolute left-0 right-0 {{borderStyle}}\"\r\n                  *ngIf=\"last !== true\" [ngClass]=\"{'border-neutral-2-200 dark:border-d-neutral-2-200':status !== 'finish',\r\n           'border-primary-1':status == 'finish',\r\n           'top-4' : size == 'md',\r\n           'top-3' : size == 'sm',\r\n           'top-5' : size == 'lg'\r\n          }\" [style.marginLeft.px]=\"size == 'md' ? 89 : size == 'sm' ? 85 : 93 \"\r\n                  [style.width]=\"size == 'md' ?  'calc(100% - 33px)' : size=='lg' ? 'calc(100% - 41px)' : 'calc(100% - 25px)'\">\r\n                </div>\r\n                <div class=\"tds-steps-item-icon w-36 \">\r\n                  <ng-template [ngIf]=\"!showProcessDot\">\r\n                    <ng-template [ngIf]=\"!icon\">\r\n                      <ng-container [ngSwitch]=\"status\">\r\n                        <ng-container *ngSwitchCase=\"'wait'\">\r\n                          <ng-container *ngTemplateOutlet=\"waitTemplate\"></ng-container>\r\n                        </ng-container>\r\n                        <ng-container *ngSwitchCase=\"'finish'\">\r\n                          <ng-container *ngTemplateOutlet=\"finishTemplate\"></ng-container>\r\n                        </ng-container>\r\n                        <ng-container *ngSwitchCase=\"'error'\">\r\n                          <ng-container *ngTemplateOutlet=\"errorTemplate\"></ng-container>\r\n                        </ng-container>\r\n                        <ng-container *ngSwitchCase=\"'process'\">\r\n                          <ng-container *ngTemplateOutlet=\"processTemplate\"></ng-container>\r\n                        </ng-container>\r\n                      </ng-container>\r\n                    </ng-template>\r\n                    <ng-template [ngIf]=\"icon\">\r\n                      <ng-container *ngTemplateOutlet=\"customIConTemplate\"></ng-container>\r\n                    </ng-template>\r\n\r\n\r\n                  </ng-template>\r\n                  <ng-template [ngIf]=\"showProcessDot\">\r\n                    <span class=\"tds-steps-icon\">\r\n                      <ng-template #processDotTemplate>\r\n                        <span class=\"tds-steps-icon-dot\"></span>\r\n                      </ng-template>\r\n                      <ng-template [ngTemplateOutlet]=\"customProcessTemplate || processDotTemplate\"\r\n                        [ngTemplateOutletContext]=\"{\r\n                    $implicit: processDotTemplate,\r\n                    status: status,\r\n                    index: index\r\n                  }\"></ng-template>\r\n                    </span>\r\n                  </ng-template>\r\n                </div>\r\n                <div class=\"tds-steps-item-content text-center w-36 mt-1\">\r\n                  <div class=\"tds-steps-item-title \">\r\n                    <div class=\"text-body-2 \" [ngClass]=\"{\r\n                'font-semibold':  size != 'sm',\r\n                'text-neutral-1-900 dark:text-d-neutral-1-900': (currentIndex === index && status !== 'error') || status === 'finish',\r\n                'text-neutral-1-600 dark:text-d-neutral-1-600':currentIndex !== index &&  (status !== 'finish' &&  status !== 'error'),\r\n                'text-error-400': status == 'error'\r\n              }\">\r\n                      <ng-container *tdsStringTemplateOutlet=\"tdsTitle\">{{ tdsTitle }}</ng-container>\r\n                    </div>\r\n                    <div *ngIf=\"subtitle\"\r\n                      [ngClass]=\"{'text-error-400': status == 'error','text-neutral-1-700 dark:text-d-neutral-1-700': status !== 'error'}\"\r\n                      class=\"tds-steps-item-subtitle  text-caption-2 font-regular\">\r\n                      <ng-container *tdsStringTemplateOutlet=\"subtitle\">{{ subtitle }}</ng-container>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"tds-steps-item-description text-caption-2 font-regular\"\r\n                    [ngClass]=\"{'text-error-400': status == 'error','text-neutral-1-700 dark:text-d-neutral-1-700': status !== 'error'}\">\r\n                    <ng-container *tdsStringTemplateOutlet=\"description\">{{ description }}</ng-container>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </ng-container>\r\n            <ng-container *ngSwitchCase=\"'vertical'\">\r\n              <div class=\"tds-steps-item-container relative outline-none flex items-stretch\"\r\n                [attr.role]=\"clickable && !disabled ? 'button' : null\" [tabindex]=\"clickable && !disabled ? 0 : null\"\r\n                (click)=\"onClick()\">\r\n\r\n                <div class=\"tds-steps-item-icon\">\r\n                  <ng-template [ngIf]=\"!showProcessDot\">\r\n                    <ng-template [ngIf]=\"!icon\">\r\n                      <ng-container [ngSwitch]=\"status\">\r\n                        <ng-container *ngSwitchCase=\"'wait'\">\r\n                          <ng-container *ngTemplateOutlet=\"waitTemplate\"></ng-container>\r\n                        </ng-container>\r\n                        <ng-container *ngSwitchCase=\"'finish'\">\r\n                          <ng-container *ngTemplateOutlet=\"finishTemplate\"></ng-container>\r\n                        </ng-container>\r\n                        <ng-container *ngSwitchCase=\"'error'\">\r\n                          <ng-container *ngTemplateOutlet=\"errorTemplate\"></ng-container>\r\n                        </ng-container>\r\n                        <ng-container *ngSwitchCase=\"'process'\">\r\n                          <ng-container *ngTemplateOutlet=\"processTemplate\"></ng-container>\r\n                        </ng-container>\r\n                      </ng-container>\r\n                    </ng-template>\r\n                    <ng-template [ngIf]=\"icon\">\r\n                      <ng-container *ngTemplateOutlet=\"customIConTemplate\"></ng-container>\r\n                    </ng-template>\r\n                  </ng-template>\r\n                  <ng-template [ngIf]=\"showProcessDot\">\r\n                    <span class=\"tds-steps-icon\">\r\n                      <ng-template #processDotTemplate>\r\n                        <span class=\"tds-steps-icon-dot\"></span>\r\n                      </ng-template>\r\n                      <ng-template [ngTemplateOutlet]=\"customProcessTemplate || processDotTemplate\"\r\n                        [ngTemplateOutletContext]=\"{\r\n                $implicit: processDotTemplate,\r\n                status: status,\r\n                index: index\r\n              }\"></ng-template>\r\n                    </span>\r\n                  </ng-template>\r\n                </div>\r\n                <div class=\"tds-steps-item-content text-left ml-2 flex-auto\"\r\n                  [ngClass]=\"{'flex items-stretch':onlyTitle}\">\r\n                  <div class=\"tds-steps-item-title w-full flex \">\r\n                    <div class=\"text-body-2 pr-2\" [ngClass]=\"{'font-semibold': size != 'sm' ,\r\n                                'text-neutral-1-900 dark:text-d-neutral-1-900': (currentIndex === index && status !== 'error') || status === 'finish',\r\n                                'text-neutral-1-600 dark:text-d-neutral-1-600':currentIndex !== index &&  (status !== 'finish' &&  status !== 'error'),\r\n                                'text-error-400': status == 'error','items-center flex':onlyTitle}\">\r\n                      <ng-container *tdsStringTemplateOutlet=\"tdsTitle\">{{ tdsTitle }}</ng-container>\r\n                    </div>\r\n                    <div class=\"flex-auto relative\">\r\n                      <div class=\"tds-steps-item-tail w-full h-px  absolute left-0 right-0\" *ngIf=\"last !== true\"\r\n                        [ngClass]=\"{'bg-neutral-2-200 dark:bg-d-neutral-2-200':status !== 'finish',\r\n                                    'bg-primary-1':status == 'finish',\r\n                                    'top-4' : size == 'md',\r\n                                    'top-3' : size == 'sm',\r\n                                    'top-5' : size == 'lg'}\">\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"tds-steps-item-subtitle w-full text-caption-2 font-regular\" *ngIf=\"subtitle\"\r\n                    [ngClass]=\"{'text-error-400': status == 'error','text-neutral-1-700 dark:text-d-neutral-1-700': status !== 'error'}\">\r\n                    <ng-container *tdsStringTemplateOutlet=\"subtitle\">{{ subtitle }}</ng-container>\r\n                  </div>\r\n                  <div class=\"tds-steps-item-description text-caption-2 font-regular\"\r\n                    [ngClass]=\"{'text-error-400': status == 'error','text-neutral-1-700 dark:text-d-neutral-1-700': status !== 'error'}\">\r\n                    <ng-container *tdsStringTemplateOutlet=\"description\">{{ description }}</ng-container>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </ng-container>\r\n          </ng-container>\r\n        </ng-container>\r\n        <!--direction vertical -->\r\n        <ng-container *ngSwitchCase=\"'vertical'\">\r\n          <div class=\"tds-steps-item-container relative outline-none flex items-stretch h-full\"\r\n            [ngClass]=\"disabled ? 'cursor-not-allowed' : ''\" [style.minHeight.px]=\"last !== true ? 64 : null\"\r\n            [attr.role]=\"clickable && !disabled ? 'button' : null\" [tabindex]=\"clickable && !disabled ? 0 : null\"\r\n            (click)=\"onClick()\">\r\n            <div class=\"tds-steps-item-tail  border-l top-0 bottom-0 absolute {{borderStyle}}\" *ngIf=\"last !== true\"\r\n              [ngClass]=\"{'border-neutral-2-200 dark:border-d-neutral-2-200':status !== 'finish',\r\n            'border-primary-1':status == 'finish',\r\n            'left-4 mt-8' : size == 'md',\r\n            'left-3 mt-6' : size == 'sm',\r\n            'left-5 mt-10' : size == 'lg'\r\n            \r\n           }\">\r\n            </div>\r\n            <div class=\"flex \">\r\n              <div class=\"tds-steps-item-icon flex flex-col items-stretch\">\r\n                <ng-template [ngIf]=\"!icon\">\r\n                  <ng-container [ngSwitch]=\"status\">\r\n                    <ng-container *ngSwitchCase=\"'wait'\">\r\n                      <ng-container *ngTemplateOutlet=\"waitTemplate\"></ng-container>\r\n                    </ng-container>\r\n                    <ng-container *ngSwitchCase=\"'finish'\">\r\n                      <ng-container *ngTemplateOutlet=\"finishTemplate\"></ng-container>\r\n                    </ng-container>\r\n                    <ng-container *ngSwitchCase=\"'error'\">\r\n                      <ng-container *ngTemplateOutlet=\"errorTemplate\"></ng-container>\r\n                    </ng-container>\r\n                    <ng-container *ngSwitchCase=\"'process'\">\r\n                      <ng-container *ngTemplateOutlet=\"processTemplate\"></ng-container>\r\n                    </ng-container>\r\n                  </ng-container>\r\n                </ng-template>\r\n                <ng-template [ngIf]=\"icon\">\r\n                  <ng-container *ngTemplateOutlet=\"customIConTemplate\"></ng-container>\r\n                </ng-template>\r\n              </div>\r\n              <div class=\"tds-steps-item-content pl-2 \" [ngClass]=\"{'pb-2': last !== true}\">\r\n                <div class=\"tds-steps-item-title \"\r\n                  [ngClass]=\"{'flex items-center':onlyTitle,'h-8': size == 'md' && onlyTitle,'h-10': size == 'lg' && onlyTitle,'h-6': size == 'sm' && onlyTitle}\">\r\n                  <div class=\"text-body-2 \" [ngClass]=\"{\r\n            'font-semibold':  size != 'sm',\r\n            'text-neutral-1-900 dark:text-d-neutral-1-900': (currentIndex === index && status !== 'error') || status === 'finish',\r\n            'text-neutral-1-600 dark:text-d-neutral-1-600':currentIndex !== index &&  (status !== 'finish' &&  status !== 'error'),\r\n            'text-error-400': status == 'error'    }\">\r\n                    <ng-container *tdsStringTemplateOutlet=\"tdsTitle\">{{ tdsTitle }}</ng-container>\r\n                  </div>\r\n                  <div *ngIf=\"subtitle\"\r\n                    [ngClass]=\"{'text-error-400': status == 'error','text-neutral-1-700 dark:text-d-neutral-1-700': status !== 'error'}\"\r\n                    class=\"tds-steps-item-subtitle  text-caption-2 font-regular\">\r\n                    <ng-container *tdsStringTemplateOutlet=\"subtitle\">{{ subtitle }}</ng-container>\r\n                  </div>\r\n                </div>\r\n                <div class=\"tds-steps-item-description text-caption-2 font-regular\"\r\n                  [ngClass]=\"{'text-error-400': status == 'error','text-neutral-1-700 dark:text-d-neutral-1-700': status !== 'error'}\">\r\n                  <ng-container *tdsStringTemplateOutlet=\"description\">{{ description }}</ng-container>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </ng-container>\r\n      </ng-container>\r\n    </ng-container>\r\n  </ng-container>\r\n</ng-container>\r\n\r\n<ng-template #waitTemplate>\r\n  <div class=\"tds-steps-icon flex w-full items-center justify-center\">\r\n    <span [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\"\r\n      class=\"rounded-full  border border-neutral-2-200 dark:border-d-neutral-2-200  flex items-center justify-center\">\r\n      <span class=\"text-neutral-1-300 text-body-2 font-semibold\" *ngIf=\"showStep\">\r\n        {{ index + 1 }}\r\n      </span>\r\n    </span>\r\n  </div>\r\n</ng-template>\r\n<ng-template #finishTemplate>\r\n  <div class=\"tds-steps-icon flex w-full items-center justify-center\">\r\n    <span *ngIf=\"!showStep\" [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\"\r\n      class=\"rounded-full bg-primary-1 flex items-center justify-center\">\r\n      <span class=\"tdsi-tick-fill text-white text-base leading-4\"></span>\r\n    </span>\r\n    <span *ngIf=\"showStep\" [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\"\r\n      class=\"rounded-full border  border-primary-1  flex items-center justify-center\">\r\n      <span class=\"tdsi-tick-fill  text-primary-1 text-base leading-4\"></span>\r\n    </span>\r\n  </div>\r\n</ng-template>\r\n<ng-template #errorTemplate>\r\n  <div class=\"tds-steps-icon flex w-full items-center justify-center\">\r\n    <span [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\"\r\n      class=\"border  border-error-400 rounded-full  flex items-center justify-center\">\r\n      <span class=\"tdsi-close-fill text-error-400 text-base leading-4\"></span>\r\n    </span>\r\n  </div>\r\n</ng-template>\r\n<ng-template #processTemplate>\r\n  <div class=\"tds-steps-icon flex w-full items-center justify-center\">\r\n    <span *ngIf=\"showStep\" [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\"\r\n      class=\"rounded-full bg-primary-1 flex items-center justify-center\">\r\n      <span class=\"text-white font-semibold\">\r\n        {{ index + 1 }}\r\n      </span>\r\n    </span>\r\n    <span *ngIf=\"!showStep\" [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\"\r\n      class=\"rounded-full  border border-primary-1 flex items-center justify-center\">\r\n\r\n    </span>\r\n  </div>\r\n</ng-template>\r\n\r\n<ng-template #customIConTemplate>\r\n  <div class=\"tds-steps-icon flex w-full items-center justify-center\">\r\n    <ng-container *tdsStringTemplateOutlet=\"icon; let icon\">\r\n\r\n      <div class=\"flex items-center justify-center\"\r\n        [ngClass]=\"{'h-8 w-8': size == 'md','h-10 w-10': size == 'lg','h-6 w-6': size == 'sm' }\">\r\n        <span [ngClass]=\"icon\"></span>\r\n      </div>\r\n    </ng-container>\r\n  </div>\r\n</ng-template>" }]
        }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { processDotTemplate: [{
                type: ViewChild,
                args: ['processDotTemplate', { static: false }]
            }], tdsTitle: [{
                type: Input
            }], subtitle: [{
                type: Input
            }], description: [{
                type: Input
            }], disabled: [{
                type: Input
            }], status: [{
                type: Input
            }], icon: [{
                type: Input
            }] });
})();

const _c0 = ["*"];
class TDSStepsComponent {
    constructor(cdr, directionality, elementRef) {
        this.cdr = cdr;
        this.directionality = directionality;
        this.elementRef = elementRef;
        this.mode = 'default';
        this.current = 0;
        this.direction = 'horizontal';
        this.labelPlacement = 'horizontal';
        this.type = 'default';
        this.size = 'md';
        this.startIndex = 0;
        this.status = 'process';
        this.showStep = false;
        this.borderStyle = 'border-solid';
        this.tdsClss = 'bg-white dark:bg-d-neutral-3-50';
        this.indexChange = new EventEmitter();
        this.destroy$ = new Subject();
        this.showProcessDot = false;
        this.classMap = {};
        this.dir = 'ltr';
        this.setClassMap();
    }
    set progressDot(value) {
        if (value instanceof TemplateRef) {
            this.showProcessDot = true;
            this.customProcessDotTemplate = value;
        }
        else {
            this.showProcessDot = toBoolean(value);
        }
        this.updateChildrenSteps();
    }
    ngOnChanges(changes) {
        if (changes.startIndex || changes.direction || changes.status || changes.current || changes.size || changes.labelPlacement || changes.showStep) {
            this.updateChildrenSteps();
        }
        if (changes.progressDot || changes.direction) {
            this.setClassMap();
        }
    }
    ngOnInit() {
        var _a;
        (_a = this.directionality.change) === null || _a === void 0 ? void 0 : _a.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.setClassMap();
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
        this.setClassMap();
        this.updateChildrenSteps();
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
        if (this.indexChangeSubscription) {
            this.indexChangeSubscription.unsubscribe();
        }
    }
    ngAfterContentInit() {
        if (this.steps) {
            this.steps.changes.pipe(startWith(null), takeUntil(this.destroy$)).subscribe(() => {
                this.updateChildrenSteps();
            });
        }
    }
    updateChildrenSteps() {
        if (this.steps) {
            const length = this.steps.length;
            this.steps.toArray().forEach((step, index) => {
                Promise.resolve().then(() => {
                    step.outStatus = this.status;
                    step.showProcessDot = this.showProcessDot;
                    if (this.customProcessDotTemplate) {
                        step.customProcessTemplate = this.customProcessDotTemplate;
                    }
                    step.clickable = this.indexChange.observers.length > 0;
                    step.direction = this.direction;
                    step.index = index + this.startIndex;
                    step.currentIndex = this.current;
                    step.last = length === index + 1;
                    step.size = this.size;
                    step.labelPlacement = this.labelPlacement;
                    step.showStep = this.showStep;
                    step.mode = this.mode;
                    step.first = index == 0;
                    step.borderStyle = this.borderStyle;
                    step.backgroundColor = this.tdsClss;
                    step.markForCheck();
                });
            });
            if (this.indexChangeSubscription) {
                this.indexChangeSubscription.unsubscribe();
            }
            this.indexChangeSubscription = merge(...this.steps.map(step => step.click$)).subscribe(index => this.indexChange.emit(index));
        }
    }
    setClassMap() {
        this.classMap = {
            [`tds-steps`]: true,
            [`flex`]: true,
            [`w-full`]: true,
            [`tds-steps-${this.direction}`]: true,
            [`flex-col`]: this.direction === 'vertical',
            [`h-full`]: this.direction === 'vertical',
            [`tds-steps-label-horizontal`]: this.direction === 'horizontal',
            [`tds-steps-label-vertical`]: (this.showProcessDot || this.labelPlacement === 'vertical') && this.direction === 'horizontal',
            [`tds-steps-dot`]: this.showProcessDot,
            ['tds-steps-navigation']: this.type === 'navigation',
            ['tds-steps-rtl']: this.dir === 'rtl',
            [`tds-steps-bar`]: this.mode === 'bar',
            [`tds-steps-sm`]: this.size === 'sm',
            [`tds-steps-md`]: this.size === 'md',
            // [`tds-steps-xl`]: this.size === 'xl',
            [`tds-steps-lg`]: this.size === 'lg',
        };
    }
}
TDSStepsComponent.ɵfac = function TDSStepsComponent_Factory(t) { return new (t || TDSStepsComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1$1.Directionality, 8), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSStepsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSStepsComponent, selectors: [["tds-steps"]], contentQueries: function TDSStepsComponent_ContentQueries(rf, ctx, dirIndex) {
        if (rf & 1) {
            i0.ɵɵcontentQuery(dirIndex, TDSStepComponent, 4);
        }
        if (rf & 2) {
            let _t;
            i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.steps = _t);
        }
    }, hostVars: 2, hostBindings: function TDSStepsComponent_HostBindings(rf, ctx) {
        if (rf & 2) {
            i0.ɵɵclassMap(ctx.classMap);
        }
    }, inputs: { mode: "mode", current: "current", direction: "direction", labelPlacement: "labelPlacement", type: "type", size: "size", startIndex: "startIndex", status: "status", showStep: "showStep", borderStyle: "borderStyle", tdsClss: "tdsClss", progressDot: "progressDot" }, outputs: { indexChange: "indexChange" }, exportAs: ["tdsSteps"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSStepsComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵprojection(0);
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSStepsComponent, [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    preserveWhitespaces: false,
                    selector: 'tds-steps',
                    exportAs: 'tdsSteps',
                    template: `
    <ng-content></ng-content>
  `,
                    host: {
                        '[class]': "classMap",
                    }
                }]
        }], function () {
        return [{ type: i0.ChangeDetectorRef }, { type: i1$1.Directionality, decorators: [{
                        type: Optional
                    }] }, { type: i0.ElementRef }];
    }, { steps: [{
                type: ContentChildren,
                args: [TDSStepComponent]
            }], mode: [{
                type: Input
            }], current: [{
                type: Input
            }], direction: [{
                type: Input
            }], labelPlacement: [{
                type: Input
            }], type: [{
                type: Input
            }], size: [{
                type: Input
            }], startIndex: [{
                type: Input
            }], status: [{
                type: Input
            }], showStep: [{
                type: Input
            }], borderStyle: [{
                type: Input
            }], tdsClss: [{
                type: Input
            }], progressDot: [{
                type: Input
            }], indexChange: [{
                type: Output
            }] });
})();

class TDSStepsModule {
}
TDSStepsModule.ɵfac = function TDSStepsModule_Factory(t) { return new (t || TDSStepsModule)(); };
TDSStepsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSStepsModule });
TDSStepsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule]] });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSStepsModule, [{
            type: NgModule,
            args: [{
                    imports: [BidiModule, CommonModule, TDSOutletModule],
                    exports: [TDSStepsComponent, TDSStepComponent],
                    declarations: [TDSStepsComponent, TDSStepComponent]
                }]
        }], null, null);
})();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSStepsModule, { declarations: [TDSStepsComponent, TDSStepComponent], imports: [BidiModule, CommonModule, TDSOutletModule], exports: [TDSStepsComponent, TDSStepComponent] }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { TDSStepComponent, TDSStepsComponent, TDSStepsModule };
