// tds-ui/fesm2015/tds-ui-empty.mjs
import * as i4 from '@angular/cdk/portal';
import { TemplatePortal, ComponentPortal, PortalModule } from '@angular/cdk/portal';
import * as i0 from '@angular/core';
import { InjectionToken, Component, ChangeDetectionStrategy, ViewEncapsulation, Input, TemplateRef, Type, Injector, NgModule } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil, startWith } from 'rxjs/operators';
import * as i1$1 from 'tds-ui/core/config';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i1 from 'tds-ui/i18n';
import { TDSI18nModule } from 'tds-ui/i18n';
import * as i3 from 'tds-ui/core/outlet';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { BidiModule } from '@angular/cdk/bidi';

const TDS_EMPTY_COMPONENT_NAME = new InjectionToken('tds-empty-component-name');

class TDSEmptyDefaultComponent {
}
TDSEmptyDefaultComponent.ɵfac = function TDSEmptyDefaultComponent_Factory(t) { return new (t || TDSEmptyDefaultComponent)(); };
TDSEmptyDefaultComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSEmptyDefaultComponent, selectors: [["tds-empty-default"]], hostAttrs: [1, "text-center"], exportAs: ["tdsEmptyDefault"], decls: 23, vars: 0, consts: [["width", "106", "height", "105", "viewBox", "0 0 106 105", "fill", "none", "xmlns", "http://www.w3.org/2000/svg", 1, "tds-empty-default"], ["d", "M53.433 92.4755C77.1442 92.4755 96.366 87.1611 96.366 80.6055C96.366 74.0499 77.1442 68.7355 53.433 68.7355C29.7218 68.7355 10.5 74.0499 10.5 80.6055C10.5 87.1611 29.7218 92.4755 53.433 92.4755Z", "fill", "#F8F9FB"], ["d", "M67.677 21.2565H39.693C34.1701 21.2565 29.693 25.7336 29.693 31.2565V70.8575C29.693 76.3803 34.1701 80.8575 39.693 80.8575H67.677C73.1998 80.8575 77.677 76.3803 77.677 70.8575V31.2565C77.677 25.7336 73.1998 21.2565 67.677 21.2565Z", "fill", "#FDFDFD"], ["d", "M72.626 81.4895H34.744C33.2376 81.4877 31.7935 80.8884 30.7283 79.8232C29.6631 78.7581 29.0639 77.3139 29.062 75.8075V26.3075C29.0649 24.8018 29.6646 23.3587 30.7297 22.2944C31.7948 21.23 33.2383 20.6314 34.744 20.6295H72.626C74.1324 20.6303 75.577 21.2285 76.6429 22.293C77.7089 23.3574 78.3091 24.8011 78.312 26.3075V75.8075C78.3102 77.3146 77.7104 78.7594 76.6443 79.8247C75.5783 80.89 74.1331 81.4887 72.626 81.4895ZM34.744 21.8895C33.5721 21.8906 32.4485 22.3566 31.6198 23.1853C30.7911 24.014 30.3251 25.1376 30.324 26.3095V75.8095C30.3251 76.9814 30.7911 78.1051 31.6198 78.9338C32.4485 79.7624 33.5721 80.2285 34.744 80.2295H72.626C73.7979 80.2285 74.9216 79.7624 75.7503 78.9338C76.5789 78.1051 77.045 76.9814 77.046 75.8095V26.3095C77.045 25.1376 76.5789 24.014 75.7503 23.1853C74.9216 22.3566 73.7979 21.8906 72.626 21.8895H34.744Z", "fill", "#72818D"], ["d", "M71.132 65.2006H37.25C36.1454 65.2006 35.25 66.096 35.25 67.2006V67.2416C35.25 68.3461 36.1454 69.2416 37.25 69.2416H71.132C72.2366 69.2416 73.132 68.3461 73.132 67.2416V67.2006C73.132 66.096 72.2366 65.2006 71.132 65.2006Z", "fill", "#ECEFF4"], ["d", "M71.132 54.5935H37.25C36.1454 54.5935 35.25 55.4889 35.25 56.5935V56.6345C35.25 57.7391 36.1454 58.6345 37.25 58.6345H71.132C72.2366 58.6345 73.132 57.7391 73.132 56.6345V56.5935C73.132 55.4889 72.2366 54.5935 71.132 54.5935Z", "fill", "#ECEFF4"], ["d", "M71.132 42.9766H37.25C36.1454 42.9766 35.25 43.872 35.25 44.9766V45.0176C35.25 46.1221 36.1454 47.0176 37.25 47.0176H71.132C72.2366 47.0176 73.132 46.1221 73.132 45.0176V44.9766C73.132 43.872 72.2366 42.9766 71.132 42.9766Z", "fill", "#ECEFF4"], ["d", "M71.132 32.3695H37.25C36.1454 32.3695 35.25 33.2649 35.25 34.3695V34.4105C35.25 35.5151 36.1454 36.4105 37.25 36.4105H71.132C72.2366 36.4105 73.132 35.5151 73.132 34.4105V34.3695C73.132 33.2649 72.2366 32.3695 71.132 32.3695Z", "fill", "#ECEFF4"], ["d", "M65.555 59.1405C75.0393 59.1405 82.728 51.5647 82.728 42.2195C82.728 32.8743 75.0393 25.2985 65.555 25.2985C56.0706 25.2985 48.382 32.8743 48.382 42.2195C48.382 51.5647 56.0706 59.1405 65.555 59.1405Z", "fill", "white"], ["d", "M65.555 55.0985C72.8076 55.0985 78.687 49.2191 78.687 41.9665C78.687 34.7139 72.8076 28.8345 65.555 28.8345C58.3024 28.8345 52.423 34.7139 52.423 41.9665C52.423 49.2191 58.3024 55.0985 65.555 55.0985Z", "fill", "white"], ["d", "M65.451 55.5014C62.8537 55.5014 60.3148 54.7313 58.1552 53.2883C55.9957 51.8453 54.3125 49.7944 53.3186 47.3948C52.3247 44.9953 52.0646 42.3549 52.5713 39.8075C53.078 37.2601 54.3287 34.9202 56.1653 33.0837C58.0018 31.2472 60.3417 29.9965 62.8891 29.4898C65.4364 28.9831 68.0768 29.2431 70.4764 30.237C72.8759 31.231 74.9269 32.9141 76.3698 35.0737C77.8128 37.2332 78.583 39.7722 78.583 42.3694C78.579 45.851 77.1942 49.1889 74.7323 51.6508C72.2704 54.1126 68.9326 55.4975 65.451 55.5014ZM65.451 30.2464C63.0535 30.2464 60.7098 30.9574 58.7164 32.2894C56.7229 33.6213 55.1692 35.5145 54.2517 37.7295C53.3342 39.9445 53.0942 42.3819 53.5619 44.7333C54.0296 47.0848 55.1841 49.2447 56.8794 50.94C58.5747 52.6353 60.7347 53.7898 63.0861 54.2575C65.4375 54.7252 67.8749 54.4852 70.0899 53.5677C72.3049 52.6502 74.1981 51.0965 75.5301 49.1031C76.862 47.1096 77.573 44.7659 77.573 42.3684C77.5693 39.1546 76.2909 36.0735 74.0184 33.801C71.7459 31.5285 68.6648 30.2501 65.451 30.2464Z", "fill", "#72818D"], ["id", "mask0", "maskUnits", "userSpaceOnUse", "x", "54", "y", "31", "width", "23", "height", "23", 2, "mask-type", "alpha"], ["cx", "65.3189", "cy", "42.2374", "r", "11", "fill", "#C4C4C4"], ["mask", "url(#mask0)"], ["d", "M66.9869 34.3895H53.3499C51.6759 34.3895 50.3189 35.7463 50.3189 37.42C50.3189 39.0937 51.6759 40.4505 53.3499 40.4505H66.9869C68.6609 40.4505 70.0179 39.0937 70.0179 37.42C70.0179 35.7463 68.6609 34.3895 66.9869 34.3895Z", "fill", "#F5F7F9"], ["d", "M67.2399 47.0166H53.0969C51.5627 47.0166 50.3189 48.2604 50.3189 49.7946C50.3189 51.3288 51.5627 52.5726 53.0969 52.5726H67.2399C68.7742 52.5726 70.0179 51.3288 70.0179 49.7946C70.0179 48.2604 68.7742 47.0166 67.2399 47.0166Z", "fill", "#F5F7F9"], ["d", "M77.9832 53.4633L75.927 55.5195L79.3536 58.9462L81.4099 56.8899L77.9832 53.4633Z", "fill", "white"], ["d", "M83.9751 57.4C82.8394 56.2644 80.9983 56.2642 79.8629 57.3997C78.7274 58.5351 78.7276 60.3762 79.8632 61.5118L83.6321 65.2807C84.7678 66.4164 86.6089 66.4165 87.7443 65.2811C88.8798 64.1456 88.8796 62.3045 87.7439 61.1689L83.9751 57.4Z", "fill", "#BDC9D4"], ["d", "M88.014 61.4385L83.705 57.1285C83.3656 56.7912 82.9372 56.5574 82.47 56.4544C82.0027 56.3513 81.5158 56.3833 81.066 56.5465L78.266 53.7465C81.2625 50.4545 82.8712 46.132 82.7561 41.6819C82.641 37.2318 80.811 32.9983 77.6483 29.8655C74.4856 26.7328 70.2349 24.9432 65.7839 24.8705C61.3329 24.7977 57.0259 26.4476 53.7626 29.4753C50.4992 32.503 48.5319 36.6745 48.2715 41.1185C48.0111 45.5625 49.4777 49.9352 52.3651 53.3233C55.2525 56.7115 59.3373 58.8531 63.7663 59.3007C68.1953 59.7483 72.626 58.4674 76.133 55.7255L79.006 58.5985C78.8429 59.0484 78.8111 59.5354 78.9143 60.0027C79.0176 60.4699 79.2515 60.8982 79.589 61.2375L83.898 65.5465C84.1325 65.7811 84.4109 65.9673 84.7173 66.0943C85.0238 66.2213 85.3523 66.2866 85.684 66.2866C86.0157 66.2866 86.3442 66.2213 86.6506 66.0943C86.9571 65.9673 87.2355 65.7811 87.47 65.5465L88.01 65.0065C88.2446 64.772 88.4307 64.4936 88.5577 64.1872C88.6847 63.8807 88.7501 63.5522 88.7501 63.2205C88.7501 62.8888 88.6847 62.5603 88.5577 62.2539C88.4307 61.9474 88.2446 61.669 88.01 61.4345L88.014 61.4385ZM49.645 42.2185C49.645 39.0718 50.5781 35.9958 52.3263 33.3794C54.0745 30.763 56.5593 28.7238 59.4665 27.5196C62.3737 26.3154 65.5726 26.0003 68.6589 26.6142C71.7451 27.2281 74.58 28.7434 76.805 30.9684C79.0301 33.1935 80.5454 36.0284 81.1593 39.1146C81.7732 42.2009 81.4581 45.3998 80.2539 48.307C79.0497 51.2142 77.0105 53.699 74.3941 55.4472C71.7777 57.1954 68.7017 58.1285 65.555 58.1285C61.3369 58.1235 57.2931 56.4456 54.3105 53.463C51.3278 50.4804 49.65 46.4366 49.645 42.2185ZM77.982 54.5345L80.337 56.8895L79.352 57.8745L76.997 55.5195L77.982 54.5345ZM87.12 64.1175L86.581 64.6575C86.3405 64.8875 86.0207 65.0158 85.688 65.0158C85.3553 65.0158 85.0354 64.8875 84.795 64.6575L80.482 60.3485C80.3645 60.2314 80.2713 60.0922 80.2077 59.939C80.144 59.7857 80.1113 59.6214 80.1113 59.4555C80.1113 59.2896 80.144 59.1253 80.2077 58.9721C80.2713 58.8188 80.3645 58.6797 80.482 58.5625L81.021 58.0225C81.258 57.7859 81.5791 57.6531 81.914 57.6531C82.2488 57.6531 82.57 57.7859 82.807 58.0225L87.116 62.3325C87.2333 62.4496 87.3264 62.5888 87.3899 62.7419C87.4535 62.895 87.4862 63.0592 87.4862 63.225C87.4862 63.3908 87.4535 63.555 87.3899 63.7081C87.3264 63.8613 87.2333 64.0004 87.116 64.1175H87.12Z", "fill", "#72818D"], ["d", "M63.846 48.5465C63.845 48.3239 63.8892 48.1034 63.976 47.8985C64.0586 47.7004 64.18 47.5209 64.333 47.3705C64.6539 47.058 65.0851 46.8848 65.533 46.8885C65.7589 46.8868 65.9827 46.931 66.191 47.0185C66.3893 47.1011 66.5702 47.2205 66.724 47.3705C66.8778 47.5215 67.001 47.7008 67.087 47.8985C67.176 48.1036 67.2219 48.3249 67.2219 48.5485C67.2219 48.7721 67.176 48.9933 67.087 49.1985C67.0002 49.3942 66.877 49.5717 66.724 49.7215C66.5708 49.8706 66.3896 49.9882 66.191 50.0675C65.9818 50.1513 65.7583 50.1934 65.533 50.1915C65.3044 50.1941 65.0775 50.1519 64.865 50.0675C64.5612 49.9464 64.3012 49.7362 64.1192 49.4645C63.9371 49.1929 63.8415 48.8725 63.845 48.5455L63.846 48.5465ZM66.901 35.0225V40.9455C66.9008 41.5681 66.8674 42.1903 66.801 42.8095C66.7343 43.4235 66.643 44.0618 66.527 44.7245H64.601C64.4837 44.0625 64.392 43.4241 64.326 42.8095C64.2596 42.1903 64.2262 41.5681 64.226 40.9455V35.0225H66.901Z", "fill", "#F3A72E"], ["d", "M81.787 19.8545C81.6583 19.8549 81.5326 19.8159 81.4267 19.7428C81.3208 19.6697 81.2398 19.566 81.1946 19.4455C81.1493 19.3251 81.1419 19.1936 81.1735 19.0689C81.205 18.9441 81.2739 18.832 81.371 18.7475L84.619 15.9145C84.6815 15.86 84.7541 15.8184 84.8328 15.792C84.9114 15.7655 84.9944 15.7548 85.0771 15.7605C85.1599 15.7662 85.2407 15.7881 85.315 15.825C85.3892 15.8619 85.4555 15.913 85.51 15.9755C85.5645 16.038 85.6062 16.1107 85.6326 16.1893C85.659 16.2679 85.6697 16.3509 85.664 16.4337C85.6584 16.5164 85.6365 16.5972 85.5996 16.6715C85.5627 16.7458 85.5115 16.812 85.449 16.8665L82.201 19.6995C82.0864 19.7998 81.9393 19.8549 81.787 19.8545Z", "fill", "#72818D"], ["d", "M83.41 22.1225C83.2444 22.1259 83.0842 22.0641 82.9638 21.9503C82.8434 21.8366 82.7725 21.6801 82.7665 21.5146C82.7605 21.3491 82.8197 21.1879 82.9315 21.0657C83.0433 20.9435 83.1986 20.8702 83.364 20.8615L87.793 20.5345C87.8759 20.5274 87.9593 20.537 88.0384 20.5626C88.1176 20.5882 88.1908 20.6294 88.2537 20.6837C88.3167 20.738 88.3682 20.8044 88.4052 20.8789C88.4421 20.9534 88.4638 21.0345 88.469 21.1175C88.4811 21.2844 88.4266 21.4493 88.3173 21.576C88.208 21.7028 88.0529 21.781 87.886 21.7935L83.457 22.1205L83.41 22.1225Z", "fill", "#72818D"], ["d", "M79.466 18.4385C79.2987 18.4382 79.1384 18.3717 79.0201 18.2534C78.9018 18.1351 78.8353 17.9748 78.835 17.8075V13.6315C78.835 13.464 78.9015 13.3034 79.02 13.185C79.1384 13.0665 79.299 13 79.4665 13C79.634 13 79.7946 13.0665 79.9131 13.185C80.0315 13.3034 80.098 13.464 80.098 13.6315V17.8065C80.098 17.8895 80.0817 17.9717 80.0499 18.0484C80.0182 18.125 79.9716 18.1947 79.9129 18.2534C79.8542 18.3121 79.7845 18.3586 79.7079 18.3904C79.6312 18.4222 79.549 18.4385 79.466 18.4385Z", "fill", "#72818D"]], template: function TDSEmptyDefaultComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵnamespaceSVG();
            i0.ɵɵelementStart(0, "svg", 0);
            i0.ɵɵelement(1, "path", 1)(2, "path", 2)(3, "path", 3)(4, "path", 4)(5, "path", 5)(6, "path", 6)(7, "path", 7)(8, "path", 8)(9, "path", 9)(10, "path", 10);
            i0.ɵɵelementStart(11, "mask", 11);
            i0.ɵɵelement(12, "circle", 12);
            i0.ɵɵelementEnd();
            i0.ɵɵelementStart(13, "g", 13);
            i0.ɵɵelement(14, "path", 14)(15, "path", 15);
            i0.ɵɵelementEnd();
            i0.ɵɵelement(16, "path", 16)(17, "path", 17)(18, "path", 18)(19, "path", 19)(20, "path", 20)(21, "path", 21)(22, "path", 22);
            i0.ɵɵelementEnd();
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptyDefaultComponent, [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    selector: 'tds-empty-default',
                    exportAs: 'tdsEmptyDefault',
                    template: `
  <svg class="tds-empty-default" width="106" height="105" viewBox="0 0 106 105" fill="none" xmlns="http://www.w3.org/2000/svg">
  <path d="M53.433 92.4755C77.1442 92.4755 96.366 87.1611 96.366 80.6055C96.366 74.0499 77.1442 68.7355 53.433 68.7355C29.7218 68.7355 10.5 74.0499 10.5 80.6055C10.5 87.1611 29.7218 92.4755 53.433 92.4755Z" fill="#F8F9FB"/>
  <path d="M67.677 21.2565H39.693C34.1701 21.2565 29.693 25.7336 29.693 31.2565V70.8575C29.693 76.3803 34.1701 80.8575 39.693 80.8575H67.677C73.1998 80.8575 77.677 76.3803 77.677 70.8575V31.2565C77.677 25.7336 73.1998 21.2565 67.677 21.2565Z" fill="#FDFDFD"/>
  <path d="M72.626 81.4895H34.744C33.2376 81.4877 31.7935 80.8884 30.7283 79.8232C29.6631 78.7581 29.0639 77.3139 29.062 75.8075V26.3075C29.0649 24.8018 29.6646 23.3587 30.7297 22.2944C31.7948 21.23 33.2383 20.6314 34.744 20.6295H72.626C74.1324 20.6303 75.577 21.2285 76.6429 22.293C77.7089 23.3574 78.3091 24.8011 78.312 26.3075V75.8075C78.3102 77.3146 77.7104 78.7594 76.6443 79.8247C75.5783 80.89 74.1331 81.4887 72.626 81.4895ZM34.744 21.8895C33.5721 21.8906 32.4485 22.3566 31.6198 23.1853C30.7911 24.014 30.3251 25.1376 30.324 26.3095V75.8095C30.3251 76.9814 30.7911 78.1051 31.6198 78.9338C32.4485 79.7624 33.5721 80.2285 34.744 80.2295H72.626C73.7979 80.2285 74.9216 79.7624 75.7503 78.9338C76.5789 78.1051 77.045 76.9814 77.046 75.8095V26.3095C77.045 25.1376 76.5789 24.014 75.7503 23.1853C74.9216 22.3566 73.7979 21.8906 72.626 21.8895H34.744Z" fill="#72818D"/>
  <path d="M71.132 65.2006H37.25C36.1454 65.2006 35.25 66.096 35.25 67.2006V67.2416C35.25 68.3461 36.1454 69.2416 37.25 69.2416H71.132C72.2366 69.2416 73.132 68.3461 73.132 67.2416V67.2006C73.132 66.096 72.2366 65.2006 71.132 65.2006Z" fill="#ECEFF4"/>
  <path d="M71.132 54.5935H37.25C36.1454 54.5935 35.25 55.4889 35.25 56.5935V56.6345C35.25 57.7391 36.1454 58.6345 37.25 58.6345H71.132C72.2366 58.6345 73.132 57.7391 73.132 56.6345V56.5935C73.132 55.4889 72.2366 54.5935 71.132 54.5935Z" fill="#ECEFF4"/>
  <path d="M71.132 42.9766H37.25C36.1454 42.9766 35.25 43.872 35.25 44.9766V45.0176C35.25 46.1221 36.1454 47.0176 37.25 47.0176H71.132C72.2366 47.0176 73.132 46.1221 73.132 45.0176V44.9766C73.132 43.872 72.2366 42.9766 71.132 42.9766Z" fill="#ECEFF4"/>
  <path d="M71.132 32.3695H37.25C36.1454 32.3695 35.25 33.2649 35.25 34.3695V34.4105C35.25 35.5151 36.1454 36.4105 37.25 36.4105H71.132C72.2366 36.4105 73.132 35.5151 73.132 34.4105V34.3695C73.132 33.2649 72.2366 32.3695 71.132 32.3695Z" fill="#ECEFF4"/>
  <path d="M65.555 59.1405C75.0393 59.1405 82.728 51.5647 82.728 42.2195C82.728 32.8743 75.0393 25.2985 65.555 25.2985C56.0706 25.2985 48.382 32.8743 48.382 42.2195C48.382 51.5647 56.0706 59.1405 65.555 59.1405Z" fill="white"/>
  <path d="M65.555 55.0985C72.8076 55.0985 78.687 49.2191 78.687 41.9665C78.687 34.7139 72.8076 28.8345 65.555 28.8345C58.3024 28.8345 52.423 34.7139 52.423 41.9665C52.423 49.2191 58.3024 55.0985 65.555 55.0985Z" fill="white"/>
  <path d="M65.451 55.5014C62.8537 55.5014 60.3148 54.7313 58.1552 53.2883C55.9957 51.8453 54.3125 49.7944 53.3186 47.3948C52.3247 44.9953 52.0646 42.3549 52.5713 39.8075C53.078 37.2601 54.3287 34.9202 56.1653 33.0837C58.0018 31.2472 60.3417 29.9965 62.8891 29.4898C65.4364 28.9831 68.0768 29.2431 70.4764 30.237C72.8759 31.231 74.9269 32.9141 76.3698 35.0737C77.8128 37.2332 78.583 39.7722 78.583 42.3694C78.579 45.851 77.1942 49.1889 74.7323 51.6508C72.2704 54.1126 68.9326 55.4975 65.451 55.5014ZM65.451 30.2464C63.0535 30.2464 60.7098 30.9574 58.7164 32.2894C56.7229 33.6213 55.1692 35.5145 54.2517 37.7295C53.3342 39.9445 53.0942 42.3819 53.5619 44.7333C54.0296 47.0848 55.1841 49.2447 56.8794 50.94C58.5747 52.6353 60.7347 53.7898 63.0861 54.2575C65.4375 54.7252 67.8749 54.4852 70.0899 53.5677C72.3049 52.6502 74.1981 51.0965 75.5301 49.1031C76.862 47.1096 77.573 44.7659 77.573 42.3684C77.5693 39.1546 76.2909 36.0735 74.0184 33.801C71.7459 31.5285 68.6648 30.2501 65.451 30.2464Z" fill="#72818D"/>
  <mask id="mask0" style="mask-type:alpha" maskUnits="userSpaceOnUse" x="54" y="31" width="23" height="23">
  <circle cx="65.3189" cy="42.2374" r="11" fill="#C4C4C4"/>
  </mask>
  <g mask="url(#mask0)">
  <path d="M66.9869 34.3895H53.3499C51.6759 34.3895 50.3189 35.7463 50.3189 37.42C50.3189 39.0937 51.6759 40.4505 53.3499 40.4505H66.9869C68.6609 40.4505 70.0179 39.0937 70.0179 37.42C70.0179 35.7463 68.6609 34.3895 66.9869 34.3895Z" fill="#F5F7F9"/>
  <path d="M67.2399 47.0166H53.0969C51.5627 47.0166 50.3189 48.2604 50.3189 49.7946C50.3189 51.3288 51.5627 52.5726 53.0969 52.5726H67.2399C68.7742 52.5726 70.0179 51.3288 70.0179 49.7946C70.0179 48.2604 68.7742 47.0166 67.2399 47.0166Z" fill="#F5F7F9"/>
  </g>
  <path d="M77.9832 53.4633L75.927 55.5195L79.3536 58.9462L81.4099 56.8899L77.9832 53.4633Z" fill="white"/>
  <path d="M83.9751 57.4C82.8394 56.2644 80.9983 56.2642 79.8629 57.3997C78.7274 58.5351 78.7276 60.3762 79.8632 61.5118L83.6321 65.2807C84.7678 66.4164 86.6089 66.4165 87.7443 65.2811C88.8798 64.1456 88.8796 62.3045 87.7439 61.1689L83.9751 57.4Z" fill="#BDC9D4"/>
  <path d="M88.014 61.4385L83.705 57.1285C83.3656 56.7912 82.9372 56.5574 82.47 56.4544C82.0027 56.3513 81.5158 56.3833 81.066 56.5465L78.266 53.7465C81.2625 50.4545 82.8712 46.132 82.7561 41.6819C82.641 37.2318 80.811 32.9983 77.6483 29.8655C74.4856 26.7328 70.2349 24.9432 65.7839 24.8705C61.3329 24.7977 57.0259 26.4476 53.7626 29.4753C50.4992 32.503 48.5319 36.6745 48.2715 41.1185C48.0111 45.5625 49.4777 49.9352 52.3651 53.3233C55.2525 56.7115 59.3373 58.8531 63.7663 59.3007C68.1953 59.7483 72.626 58.4674 76.133 55.7255L79.006 58.5985C78.8429 59.0484 78.8111 59.5354 78.9143 60.0027C79.0176 60.4699 79.2515 60.8982 79.589 61.2375L83.898 65.5465C84.1325 65.7811 84.4109 65.9673 84.7173 66.0943C85.0238 66.2213 85.3523 66.2866 85.684 66.2866C86.0157 66.2866 86.3442 66.2213 86.6506 66.0943C86.9571 65.9673 87.2355 65.7811 87.47 65.5465L88.01 65.0065C88.2446 64.772 88.4307 64.4936 88.5577 64.1872C88.6847 63.8807 88.7501 63.5522 88.7501 63.2205C88.7501 62.8888 88.6847 62.5603 88.5577 62.2539C88.4307 61.9474 88.2446 61.669 88.01 61.4345L88.014 61.4385ZM49.645 42.2185C49.645 39.0718 50.5781 35.9958 52.3263 33.3794C54.0745 30.763 56.5593 28.7238 59.4665 27.5196C62.3737 26.3154 65.5726 26.0003 68.6589 26.6142C71.7451 27.2281 74.58 28.7434 76.805 30.9684C79.0301 33.1935 80.5454 36.0284 81.1593 39.1146C81.7732 42.2009 81.4581 45.3998 80.2539 48.307C79.0497 51.2142 77.0105 53.699 74.3941 55.4472C71.7777 57.1954 68.7017 58.1285 65.555 58.1285C61.3369 58.1235 57.2931 56.4456 54.3105 53.463C51.3278 50.4804 49.65 46.4366 49.645 42.2185ZM77.982 54.5345L80.337 56.8895L79.352 57.8745L76.997 55.5195L77.982 54.5345ZM87.12 64.1175L86.581 64.6575C86.3405 64.8875 86.0207 65.0158 85.688 65.0158C85.3553 65.0158 85.0354 64.8875 84.795 64.6575L80.482 60.3485C80.3645 60.2314 80.2713 60.0922 80.2077 59.939C80.144 59.7857 80.1113 59.6214 80.1113 59.4555C80.1113 59.2896 80.144 59.1253 80.2077 58.9721C80.2713 58.8188 80.3645 58.6797 80.482 58.5625L81.021 58.0225C81.258 57.7859 81.5791 57.6531 81.914 57.6531C82.2488 57.6531 82.57 57.7859 82.807 58.0225L87.116 62.3325C87.2333 62.4496 87.3264 62.5888 87.3899 62.7419C87.4535 62.895 87.4862 63.0592 87.4862 63.225C87.4862 63.3908 87.4535 63.555 87.3899 63.7081C87.3264 63.8613 87.2333 64.0004 87.116 64.1175H87.12Z" fill="#72818D"/>
  <path d="M63.846 48.5465C63.845 48.3239 63.8892 48.1034 63.976 47.8985C64.0586 47.7004 64.18 47.5209 64.333 47.3705C64.6539 47.058 65.0851 46.8848 65.533 46.8885C65.7589 46.8868 65.9827 46.931 66.191 47.0185C66.3893 47.1011 66.5702 47.2205 66.724 47.3705C66.8778 47.5215 67.001 47.7008 67.087 47.8985C67.176 48.1036 67.2219 48.3249 67.2219 48.5485C67.2219 48.7721 67.176 48.9933 67.087 49.1985C67.0002 49.3942 66.877 49.5717 66.724 49.7215C66.5708 49.8706 66.3896 49.9882 66.191 50.0675C65.9818 50.1513 65.7583 50.1934 65.533 50.1915C65.3044 50.1941 65.0775 50.1519 64.865 50.0675C64.5612 49.9464 64.3012 49.7362 64.1192 49.4645C63.9371 49.1929 63.8415 48.8725 63.845 48.5455L63.846 48.5465ZM66.901 35.0225V40.9455C66.9008 41.5681 66.8674 42.1903 66.801 42.8095C66.7343 43.4235 66.643 44.0618 66.527 44.7245H64.601C64.4837 44.0625 64.392 43.4241 64.326 42.8095C64.2596 42.1903 64.2262 41.5681 64.226 40.9455V35.0225H66.901Z" fill="#F3A72E"/>
  <path d="M81.787 19.8545C81.6583 19.8549 81.5326 19.8159 81.4267 19.7428C81.3208 19.6697 81.2398 19.566 81.1946 19.4455C81.1493 19.3251 81.1419 19.1936 81.1735 19.0689C81.205 18.9441 81.2739 18.832 81.371 18.7475L84.619 15.9145C84.6815 15.86 84.7541 15.8184 84.8328 15.792C84.9114 15.7655 84.9944 15.7548 85.0771 15.7605C85.1599 15.7662 85.2407 15.7881 85.315 15.825C85.3892 15.8619 85.4555 15.913 85.51 15.9755C85.5645 16.038 85.6062 16.1107 85.6326 16.1893C85.659 16.2679 85.6697 16.3509 85.664 16.4337C85.6584 16.5164 85.6365 16.5972 85.5996 16.6715C85.5627 16.7458 85.5115 16.812 85.449 16.8665L82.201 19.6995C82.0864 19.7998 81.9393 19.8549 81.787 19.8545Z" fill="#72818D"/>
  <path d="M83.41 22.1225C83.2444 22.1259 83.0842 22.0641 82.9638 21.9503C82.8434 21.8366 82.7725 21.6801 82.7665 21.5146C82.7605 21.3491 82.8197 21.1879 82.9315 21.0657C83.0433 20.9435 83.1986 20.8702 83.364 20.8615L87.793 20.5345C87.8759 20.5274 87.9593 20.537 88.0384 20.5626C88.1176 20.5882 88.1908 20.6294 88.2537 20.6837C88.3167 20.738 88.3682 20.8044 88.4052 20.8789C88.4421 20.9534 88.4638 21.0345 88.469 21.1175C88.4811 21.2844 88.4266 21.4493 88.3173 21.576C88.208 21.7028 88.0529 21.781 87.886 21.7935L83.457 22.1205L83.41 22.1225Z" fill="#72818D"/>
  <path d="M79.466 18.4385C79.2987 18.4382 79.1384 18.3717 79.0201 18.2534C78.9018 18.1351 78.8353 17.9748 78.835 17.8075V13.6315C78.835 13.464 78.9015 13.3034 79.02 13.185C79.1384 13.0665 79.299 13 79.4665 13C79.634 13 79.7946 13.0665 79.9131 13.185C80.0315 13.3034 80.098 13.464 80.098 13.6315V17.8065C80.098 17.8895 80.0817 17.9717 80.0499 18.0484C80.0182 18.125 79.9716 18.1947 79.9129 18.2534C79.8542 18.3121 79.7845 18.3586 79.7079 18.3904C79.6312 18.4222 79.549 18.4385 79.466 18.4385Z" fill="#72818D"/>
  </svg>
  `,
                    host: {
                        class: "text-center"
                    }
                }]
        }], null, null);
})();

class TDSEmptySimpleComponent {
}
TDSEmptySimpleComponent.ɵfac = function TDSEmptySimpleComponent_Factory(t) { return new (t || TDSEmptySimpleComponent)(); };
TDSEmptySimpleComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSEmptySimpleComponent, selectors: [["tds-empty-simple"]], exportAs: ["tdsEmptySimple"], decls: 6, vars: 0, consts: [["width", "64", "height", "41", "viewBox", "0 0 64 41", "xmlns", "http://www.w3.org/2000/svg", 1, "tds-empty-img-simple"], ["transform", "translate(0 1)", "fill", "none", "fill-rule", "evenodd"], ["cx", "32", "cy", "33", "rx", "32", "ry", "7", 1, "tds-empty-img-simple-ellipse", "fill-current", "text-gray-200"], ["fill-rule", "nonzero", 1, "tds-empty-img-simple-g", "stroke-current", "text-gray-400"], ["d", "M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"], ["d", "M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z", 1, "tds-empty-img-simple-path", 2, "fill", "#fafafa"]], template: function TDSEmptySimpleComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵnamespaceSVG();
            i0.ɵɵelementStart(0, "svg", 0)(1, "g", 1);
            i0.ɵɵelement(2, "ellipse", 2);
            i0.ɵɵelementStart(3, "g", 3);
            i0.ɵɵelement(4, "path", 4)(5, "path", 5);
            i0.ɵɵelementEnd()()();
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptySimpleComponent, [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    selector: 'tds-empty-simple',
                    exportAs: 'tdsEmptySimple',
                    template: `
    <svg class="tds-empty-img-simple" width="64" height="41" viewBox="0 0 64 41" xmlns="http://www.w3.org/2000/svg">
      <g transform="translate(0 1)" fill="none" fill-rule="evenodd">
        <ellipse class="tds-empty-img-simple-ellipse fill-current text-gray-200" cx="32" cy="33" rx="32" ry="7" />
        <g class="tds-empty-img-simple-g stroke-current text-gray-400" fill-rule="nonzero">
          <path
            d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
          />
          <path
            d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
            class="tds-empty-img-simple-path" style="fill: #fafafa;"
          />
        </g>
      </g>
    </svg>
  `
                }]
        }], null, null);
})();

function TDSEmptyComponent_ng_container_1_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵelement(1, "img", 5);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r5 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("src", ctx_r5.notFoundImage, i0.ɵɵsanitizeUrl)("alt", ctx_r5.isContentString ? ctx_r5.notFoundContent : "empty");
    }
}
function TDSEmptyComponent_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSEmptyComponent_ng_container_1_ng_container_1_Template, 2, 2, "ng-container", 4);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r0 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r0.notFoundImage);
    }
}
function TDSEmptyComponent_tds_empty_default_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "tds-empty-default");
    }
}
function TDSEmptyComponent_tds_empty_simple_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "tds-empty-simple");
    }
}
function TDSEmptyComponent_p_4_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r6 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx_r6.isContentString ? ctx_r6.notFoundContent : ctx_r6.locale["description"], " ");
    }
}
function TDSEmptyComponent_p_4_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "p", 6);
        i0.ɵɵtemplate(1, TDSEmptyComponent_p_4_ng_container_1_Template, 2, 1, "ng-container", 4);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r3 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r3.notFoundContent);
    }
}
function TDSEmptyComponent_div_5_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r7 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx_r7.notFoundFooter, " ");
    }
}
function TDSEmptyComponent_div_5_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 7);
        i0.ɵɵtemplate(1, TDSEmptyComponent_div_5_ng_container_1_Template, 2, 1, "ng-container", 4);
        i0.ɵɵelementEnd();
    }
    if (rf & 2) {
        const ctx_r4 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r4.notFoundFooter);
    }
}
const TDSEmptyDefaultImages = ['default', 'simple'];
class TDSEmptyComponent {
    constructor(i18n, cdr) {
        this.i18n = i18n;
        this.cdr = cdr;
        this.notFoundImage = 'default';
        this.isContentString = false;
        this.isImageBuildIn = true;
        this.destroy$ = new Subject();
    }
    ngOnChanges(changes) {
        const { notFoundContent, notFoundImage } = changes;
        if (notFoundContent) {
            const content = notFoundContent.currentValue;
            this.isContentString = typeof content === 'string';
        }
        if (notFoundImage) {
            const image = notFoundImage.currentValue || 'default';
            this.isImageBuildIn = TDSEmptyDefaultImages.findIndex(i => i === image) > -1;
        }
    }
    ngOnInit() {
        this.i18n.localeChange.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.locale = this.i18n.getLocaleData('Empty');
            this.cdr.markForCheck();
        });
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSEmptyComponent.ɵfac = function TDSEmptyComponent_Factory(t) { return new (t || TDSEmptyComponent)(i0.ɵɵdirectiveInject(i1.TDSI18nService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSEmptyComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSEmptyComponent, selectors: [["tds-empty"]], hostAttrs: [1, "tds-empty", "flex", "flex-col", "items-center"], inputs: { notFoundImage: "notFoundImage", notFoundContent: "notFoundContent", notFoundFooter: "notFoundFooter" }, exportAs: ["tdsEmpty"], features: [i0.ɵɵNgOnChangesFeature], decls: 6, vars: 5, consts: [[1, "tds-empty-image"], [4, "ngIf"], ["class", "tds-empty-description text-neutral-1-900 text-title-1 font-semibold mt-3", 4, "ngIf"], ["class", "tds-empty-footer mt-3", 4, "ngIf"], [4, "tdsStringTemplateOutlet"], [1, "h-full", 3, "src", "alt"], [1, "tds-empty-description", "text-neutral-1-900", "text-title-1", "font-semibold", "mt-3"], [1, "tds-empty-footer", "mt-3"]], template: function TDSEmptyComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵelementStart(0, "div", 0);
            i0.ɵɵtemplate(1, TDSEmptyComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
            i0.ɵɵtemplate(2, TDSEmptyComponent_tds_empty_default_2_Template, 1, 0, "tds-empty-default", 1);
            i0.ɵɵtemplate(3, TDSEmptyComponent_tds_empty_simple_3_Template, 1, 0, "tds-empty-simple", 1);
            i0.ɵɵelementEnd();
            i0.ɵɵtemplate(4, TDSEmptyComponent_p_4_Template, 2, 1, "p", 2);
            i0.ɵɵtemplate(5, TDSEmptyComponent_div_5_Template, 2, 1, "div", 3);
        }
        if (rf & 2) {
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", !ctx.isImageBuildIn);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.isImageBuildIn && ctx.notFoundImage !== "simple");
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.isImageBuildIn && ctx.notFoundImage === "simple");
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.notFoundContent !== null);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.notFoundFooter);
        }
    }, directives: [i2.NgIf, i3.TDSStringTemplateOutletDirective, TDSEmptyDefaultComponent, TDSEmptySimpleComponent], encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptyComponent, [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    selector: 'tds-empty',
                    exportAs: 'tdsEmpty',
                    template: `
    <div class="tds-empty-image">
      <ng-container *ngIf="!isImageBuildIn">
        <ng-container *tdsStringTemplateOutlet="notFoundImage">
          <img class="h-full" [src]="notFoundImage" [alt]="isContentString ? notFoundContent : 'empty'" />
        </ng-container>
      </ng-container>
      <tds-empty-default *ngIf="isImageBuildIn && notFoundImage !== 'simple'"></tds-empty-default>
      <tds-empty-simple *ngIf="isImageBuildIn && notFoundImage === 'simple'"></tds-empty-simple>
    </div>
    <p class="tds-empty-description text-neutral-1-900 text-title-1 font-semibold mt-3" *ngIf="notFoundContent !== null">
      <ng-container *tdsStringTemplateOutlet="notFoundContent">
        {{ isContentString ? notFoundContent : locale['description'] }}
      </ng-container>
    </p>
    <div class="tds-empty-footer mt-3" *ngIf="notFoundFooter">
      <ng-container *tdsStringTemplateOutlet="notFoundFooter">
        {{ notFoundFooter }}
      </ng-container>
    </div>
  `,
                    host: {
                        class: 'tds-empty  flex flex-col items-center'
                    }
                }]
        }], function () { return [{ type: i1.TDSI18nService }, { type: i0.ChangeDetectorRef }]; }, { notFoundImage: [{
                type: Input
            }], notFoundContent: [{
                type: Input
            }], notFoundFooter: [{
                type: Input
            }] });
})();

function TDSEmbedEmptyComponent_ng_container_0_tds_empty_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "tds-empty", 6);
    }
    if (rf & 2) {
        i0.ɵɵproperty("notFoundImage", "simple");
    }
}
function TDSEmbedEmptyComponent_ng_container_0_tds_empty_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "tds-empty", 7);
    }
    if (rf & 2) {
        i0.ɵɵproperty("notFoundImage", "simple");
    }
}
function TDSEmbedEmptyComponent_ng_container_0_tds_empty_3_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelement(0, "tds-empty");
    }
}
function TDSEmbedEmptyComponent_ng_container_0_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 2);
        i0.ɵɵtemplate(1, TDSEmbedEmptyComponent_ng_container_0_tds_empty_1_Template, 1, 1, "tds-empty", 3);
        i0.ɵɵtemplate(2, TDSEmbedEmptyComponent_ng_container_0_tds_empty_2_Template, 1, 1, "tds-empty", 4);
        i0.ɵɵtemplate(3, TDSEmbedEmptyComponent_ng_container_0_tds_empty_3_Template, 1, 0, "tds-empty", 5);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r0 = i0.ɵɵnextContext();
        i0.ɵɵproperty("ngSwitch", ctx_r0.size);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "lg");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", "md");
    }
}
function TDSEmbedEmptyComponent_ng_container_1_1_ng_template_0_Template(rf, ctx) { }
function TDSEmbedEmptyComponent_ng_container_1_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵtemplate(0, TDSEmbedEmptyComponent_ng_container_1_1_ng_template_0_Template, 0, 0, "ng-template", 8);
    }
    if (rf & 2) {
        const ctx_r5 = i0.ɵɵnextContext(2);
        i0.ɵɵproperty("cdkPortalOutlet", ctx_r5.contentPortal);
    }
}
function TDSEmbedEmptyComponent_ng_container_1_ng_container_2_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtext(1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r6 = i0.ɵɵnextContext(2);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx_r6.content, " ");
    }
}
function TDSEmbedEmptyComponent_ng_container_1_Template(rf, ctx) {
    if (rf & 1) {
        i0.ɵɵelementContainerStart(0);
        i0.ɵɵtemplate(1, TDSEmbedEmptyComponent_ng_container_1_1_Template, 1, 1, null, 1);
        i0.ɵɵtemplate(2, TDSEmbedEmptyComponent_ng_container_1_ng_container_2_Template, 2, 1, "ng-container", 1);
        i0.ɵɵelementContainerEnd();
    }
    if (rf & 2) {
        const ctx_r1 = i0.ɵɵnextContext();
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r1.contentType !== "string");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx_r1.contentType === "string");
    }
}
function getEmptySize(componentName) {
    switch (componentName) {
        case 'table':
        case 'list':
            return 'normal';
        case 'select':
        case 'tree-select':
        case 'cascader':
        case 'transfer':
            return 'small';
        default:
            return '';
    }
}
class TDSEmbedEmptyComponent {
    constructor(configService, viewContainerRef, cdr, injector) {
        this.configService = configService;
        this.viewContainerRef = viewContainerRef;
        this.cdr = cdr;
        this.injector = injector;
        this.contentType = 'string';
        this.size = '';
        this.destroy$ = new Subject();
    }
    ngOnChanges(changes) {
        if (changes.componentName) {
            this.size = getEmptySize(changes.componentName.currentValue);
        }
        if (changes.specificContent && !changes.specificContent.isFirstChange()) {
            this.content = changes.specificContent.currentValue;
            this.renderEmpty();
        }
    }
    ngOnInit() {
        this.subscribeDefaultEmptyContentChange();
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    renderEmpty() {
        const content = this.content;
        if (typeof content === 'string') {
            this.contentType = 'string';
        }
        else if (content instanceof TemplateRef) {
            const context = { $implicit: this.componentName };
            this.contentType = 'template';
            this.contentPortal = new TemplatePortal(content, this.viewContainerRef, context);
        }
        else if (content instanceof Type) {
            const injector = Injector.create({
                parent: this.injector,
                providers: [{ provide: TDS_EMPTY_COMPONENT_NAME, useValue: this.componentName }]
            });
            this.contentType = 'component';
            this.contentPortal = new ComponentPortal(content, this.viewContainerRef, injector);
        }
        else {
            this.contentType = 'string';
            this.contentPortal = undefined;
        }
        this.cdr.detectChanges();
    }
    subscribeDefaultEmptyContentChange() {
        this.configService
            .getConfigChangeEventForComponent('empty')
            .pipe(startWith(true), takeUntil(this.destroy$))
            .subscribe(() => {
            this.content = this.specificContent || this.getUserDefaultEmptyContent();
            this.renderEmpty();
        });
    }
    getUserDefaultEmptyContent() {
        return (this.configService.getConfigForComponent('empty') || {}).TDSDefaultEmptyContent;
    }
}
TDSEmbedEmptyComponent.ɵfac = function TDSEmbedEmptyComponent_Factory(t) { return new (t || TDSEmbedEmptyComponent)(i0.ɵɵdirectiveInject(i1$1.TDSConfigService), i0.ɵɵdirectiveInject(i0.ViewContainerRef), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.Injector)); };
TDSEmbedEmptyComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSEmbedEmptyComponent, selectors: [["tds-embed-empty"]], inputs: { componentName: "componentName", specificContent: "specificContent" }, exportAs: ["tdsEmbedEmpty"], features: [i0.ɵɵNgOnChangesFeature], decls: 2, vars: 2, consts: [[3, "ngSwitch", 4, "ngIf"], [4, "ngIf"], [3, "ngSwitch"], ["class", "tds-empty-normal", 3, "notFoundImage", 4, "ngSwitchCase"], ["class", "tds-empty-small", 3, "notFoundImage", 4, "ngSwitchCase"], [4, "ngSwitchDefault"], [1, "tds-empty-normal", 3, "notFoundImage"], [1, "tds-empty-small", 3, "notFoundImage"], [3, "cdkPortalOutlet"]], template: function TDSEmbedEmptyComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵtemplate(0, TDSEmbedEmptyComponent_ng_container_0_Template, 4, 3, "ng-container", 0);
            i0.ɵɵtemplate(1, TDSEmbedEmptyComponent_ng_container_1_Template, 3, 2, "ng-container", 1);
        }
        if (rf & 2) {
            i0.ɵɵproperty("ngIf", !ctx.content && ctx.specificContent !== null);
            i0.ɵɵadvance(1);
            i0.ɵɵproperty("ngIf", ctx.content);
        }
    }, directives: [i2.NgIf, i2.NgSwitch, i2.NgSwitchCase, TDSEmptyComponent, i2.NgSwitchDefault, i4.CdkPortalOutlet], encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmbedEmptyComponent, [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    selector: 'tds-embed-empty',
                    exportAs: 'tdsEmbedEmpty',
                    template: `
    <ng-container *ngIf="!content && specificContent !== null" [ngSwitch]="size">
      <tds-empty *ngSwitchCase="'lg'" class="tds-empty-normal" [notFoundImage]="'simple'"></tds-empty>
      <tds-empty *ngSwitchCase="'md'" class="tds-empty-small" [notFoundImage]="'simple'"></tds-empty>
      <tds-empty *ngSwitchDefault></tds-empty>
    </ng-container>
    <ng-container *ngIf="content">
      <ng-template *ngIf="contentType !== 'string'" [cdkPortalOutlet]="contentPortal"></ng-template>
      <ng-container *ngIf="contentType === 'string'">
        {{ content }}
      </ng-container>
    </ng-container>
  `
                }]
        }], function () { return [{ type: i1$1.TDSConfigService }, { type: i0.ViewContainerRef }, { type: i0.ChangeDetectorRef }, { type: i0.Injector }]; }, { componentName: [{
                type: Input
            }], specificContent: [{
                type: Input
            }] });
})();

class TDSEmptyModule {
}
TDSEmptyModule.ɵfac = function TDSEmptyModule_Factory(t) { return new (t || TDSEmptyModule)(); };
TDSEmptyModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSEmptyModule });
TDSEmptyModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PortalModule, TDSOutletModule, TDSI18nModule]] });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptyModule, [{
            type: NgModule,
            args: [{
                    imports: [BidiModule, CommonModule, PortalModule, TDSOutletModule, TDSI18nModule],
                    declarations: [TDSEmptyComponent, TDSEmbedEmptyComponent, TDSEmptyDefaultComponent, TDSEmptySimpleComponent],
                    exports: [TDSEmptyComponent, TDSEmbedEmptyComponent]
                }]
        }], null, null);
})();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSEmptyModule, { declarations: [TDSEmptyComponent, TDSEmbedEmptyComponent, TDSEmptyDefaultComponent, TDSEmptySimpleComponent], imports: [BidiModule, CommonModule, PortalModule, TDSOutletModule, TDSI18nModule], exports: [TDSEmptyComponent, TDSEmbedEmptyComponent] }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { TDSEmbedEmptyComponent, TDSEmptyComponent, TDSEmptyDefaultComponent, TDSEmptyModule, TDSEmptySimpleComponent, TDS_EMPTY_COMPONENT_NAME };
