// tds-ui/fesm2015/tds-ui-card.mjs
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Directive, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

const _c0$3 = ["*"];
class TDSCardComponent {
    ngOnInit() {
    }
}
TDSCardComponent.ɵfac = function TDSCardComponent_Factory(t) { return new (t || TDSCardComponent)(); };
TDSCardComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardComponent, selectors: [["tds-card"]], hostAttrs: [1, "block", "overflow-hidden", "rounded-md", "text-neutral-1-900", "dark:text-d-neutral-1-900"], ngContentSelectors: _c0$3, decls: 1, vars: 0, template: function TDSCardComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵprojection(0);
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardComponent, [{
            type: Component,
            args: [{
                    selector: 'tds-card',
                    template: `
      <ng-content></ng-content>
  `,
                    host: {
                        class: 'block overflow-hidden rounded-md   text-neutral-1-900 dark:text-d-neutral-1-900',
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], null, null);
})();

const _c0$2 = ["*"];
class TDSCardHeaderComponent {
}
TDSCardHeaderComponent.ɵfac = function TDSCardHeaderComponent_Factory(t) { return new (t || TDSCardHeaderComponent)(); };
TDSCardHeaderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardHeaderComponent, selectors: [["tds-card-header"]], hostAttrs: [1, "p-4", "block", "w-full", "text-header-1", "font-semibold"], ngContentSelectors: _c0$2, decls: 1, vars: 0, template: function TDSCardHeaderComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵprojection(0);
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardHeaderComponent, [{
            type: Component,
            args: [{
                    selector: 'tds-card-header',
                    template: `
      <ng-content></ng-content>
  `,
                    host: {
                        class: 'p-4 block w-full text-header-1  font-semibold',
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], null, null);
})();

const _c0$1 = ["*"];
class TDSCardContentComponent {
}
TDSCardContentComponent.ɵfac = function TDSCardContentComponent_Factory(t) { return new (t || TDSCardContentComponent)(); };
TDSCardContentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardContentComponent, selectors: [["tds-card-content"]], hostAttrs: [1, "p-4", "h-auto", "block", "w-full", "text-body-2", "font-normal"], ngContentSelectors: _c0$1, decls: 1, vars: 0, template: function TDSCardContentComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵprojection(0);
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardContentComponent, [{
            type: Component,
            args: [{
                    selector: 'tds-card-content',
                    template: `
      <ng-content></ng-content>
  `,
                    host: {
                        class: 'p-4 h-auto block w-full text-body-2 font-normal',
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], null, null);
})();

const _c0 = ["*"];
class TDSCardActionsComponent {
}
TDSCardActionsComponent.ɵfac = function TDSCardActionsComponent_Factory(t) { return new (t || TDSCardActionsComponent)(); };
TDSCardActionsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSCardActionsComponent, selectors: [["tds-card-actions"]], hostAttrs: [1, "px-4", "min-h-0", "block", "w-full"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSCardActionsComponent_Template(rf, ctx) {
        if (rf & 1) {
            i0.ɵɵprojectionDef();
            i0.ɵɵprojection(0);
        }
    }, encapsulation: 2, changeDetection: 0 });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardActionsComponent, [{
            type: Component,
            args: [{
                    selector: 'tds-card-actions',
                    template: `
      <ng-content></ng-content>
  `,
                    host: {
                        class: 'px-4 min-h-0 block w-full',
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], null, null);
})();

class TDSCardImageDirective {
    ngOnInit() {
    }
}
TDSCardImageDirective.ɵfac = function TDSCardImageDirective_Factory(t) { return new (t || TDSCardImageDirective)(); };
TDSCardImageDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSCardImageDirective, selectors: [["", "tds-card-image", ""]], hostAttrs: [1, "block", "w-full", "overflow-hidden"] });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardImageDirective, [{
            type: Directive,
            args: [{
                    selector: '[tds-card-image]',
                    host: {
                        class: 'block w-full overflow-hidden',
                    },
                }]
        }], null, null);
})();

class TDSCardModule {
}
TDSCardModule.ɵfac = function TDSCardModule_Factory(t) { return new (t || TDSCardModule)(); };
TDSCardModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCardModule });
TDSCardModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () {
    (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCardModule, [{
            type: NgModule,
            args: [{
                    declarations: [
                        TDSCardHeaderComponent,
                        TDSCardContentComponent,
                        TDSCardActionsComponent,
                        TDSCardComponent,
                        TDSCardImageDirective
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        TDSCardHeaderComponent,
                        TDSCardContentComponent,
                        TDSCardActionsComponent,
                        TDSCardComponent,
                        TDSCardImageDirective
                    ]
                }]
        }], null, null);
})();
(function () {
    (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCardModule, { declarations: [TDSCardHeaderComponent,
            TDSCardContentComponent,
            TDSCardActionsComponent,
            TDSCardComponent,
            TDSCardImageDirective], imports: [CommonModule], exports: [TDSCardHeaderComponent,
            TDSCardContentComponent,
            TDSCardActionsComponent,
            TDSCardComponent,
            TDSCardImageDirective] });
})();

/**
 * Generated bundle index. Do not edit.
 */

export { TDSCardActionsComponent, TDSCardComponent, TDSCardContentComponent, TDSCardHeaderComponent, TDSCardImageDirective, TDSCardModule };
