// tds-ui/esm2020/table/src/styled/align.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSCellAlignDirective {
    constructor() {
        this.align = null;
    }
}
TDSCellAlignDirective.ɵfac = function TDSCellAlignDirective_Factory(t) { return new (t || TDSCellAlignDirective)(); };
TDSCellAlignDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSCellAlignDirective, selectors: [["th", "align", ""], ["td", "align", ""]], hostVars: 2, hostBindings: function TDSCellAlignDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵstyleProp("text-align", ctx.align);
    } }, inputs: { align: "align" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCellAlignDirective, [{
        type: Directive,
        args: [{
                selector: 'th[align],td[align]',
                host: {
                    '[style.text-align]': 'align'
                }
            }]
    }], null, { align: [{
            type: Input
        }] }); })();
