// tds-ui/esm2020/table/src/styled/ellipsis.directive.mjs
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
export class TDSCellEllipsisDirective {
    constructor() {
        this.ellipsis = true;
    }
}
TDSCellEllipsisDirective.ɵfac = function TDSCellEllipsisDirective_Factory(t) { return new (t || TDSCellEllipsisDirective)(); };
TDSCellEllipsisDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSCellEllipsisDirective, selectors: [["th", "ellipsis", ""], ["td", "ellipsis", ""]], hostVars: 2, hostBindings: function TDSCellEllipsisDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-table-cell-ellipsis", ctx.ellipsis);
    } }, inputs: { ellipsis: "ellipsis" } });
__decorate([
    InputBoolean()
], TDSCellEllipsisDirective.prototype, "ellipsis", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCellEllipsisDirective, [{
        type: Directive,
        args: [{
                selector: 'th[ellipsis],td[ellipsis]',
                host: {
                    '[class.tds-table-cell-ellipsis]': 'ellipsis'
                }
            }]
    }], null, { ellipsis: [{
            type: Input
        }] }); })();
