// tds-ui/esm2020/table/src/styled/word-break.directive.mjs
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
export class TDSCellBreakWordDirective {
    constructor() {
        this.breakWord = true;
    }
}
TDSCellBreakWordDirective.ɵfac = function TDSCellBreakWordDirective_Factory(t) { return new (t || TDSCellBreakWordDirective)(); };
TDSCellBreakWordDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSCellBreakWordDirective, selectors: [["th", "breakWord", ""], ["td", "breakWord", ""]], hostVars: 2, hostBindings: function TDSCellBreakWordDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵstyleProp("word-break", ctx.breakWord ? "break-all" : "");
    } }, inputs: { breakWord: "breakWord" } });
__decorate([
    InputBoolean()
], TDSCellBreakWordDirective.prototype, "breakWord", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCellBreakWordDirective, [{
        type: Directive,
        args: [{
                selector: 'th[breakWord],td[breakWord]',
                host: {
                    '[style.word-break]': `breakWord ? 'break-all' : ''`
                }
            }]
    }], null, { breakWord: [{
            type: Input
        }] }); })();
