// tds-ui/esm2020/menu/menu.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSMenuComponent } from './tds-menu/tds-menu.component';
import { TDSMenuItemComponent } from './tds-menu-item/tds-menu-item.component';
import { RouterModule } from '@angular/router';
import { TDSMenuGroupInlineComponent } from './tds-menu-group-inline/tds-menu-group-inline.component';
import { TDSMenuGroupPopupComponent } from './tds-menu-group-popup/tds-menu-group-popup.component';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSTagModule } from 'tds-ui/tag';
import { TDSBadgeModule } from 'tds-ui/badges';
import * as i0 from "@angular/core";
export class TDSMenuModule {
}
TDSMenuModule.ɵfac = function TDSMenuModule_Factory(t) { return new (t || TDSMenuModule)(); };
TDSMenuModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSMenuModule });
TDSMenuModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            PlatformModule,
            OverlayModule,
            RouterModule,
            TDSOutletModule,
            TDSBadgeModule,
            TDSToolTipModule,
            TDSTagModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMenuModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    PlatformModule,
                    OverlayModule,
                    RouterModule,
                    TDSOutletModule,
                    TDSBadgeModule,
                    TDSToolTipModule,
                    TDSTagModule
                ],
                declarations: [
                    TDSMenuComponent,
                    TDSMenuItemComponent,
                    TDSMenuGroupInlineComponent,
                    TDSMenuGroupPopupComponent,
                ],
                exports: [
                    TDSMenuComponent,
                    TDSMenuItemComponent,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSMenuModule, { declarations: [TDSMenuComponent,
        TDSMenuItemComponent,
        TDSMenuGroupInlineComponent,
        TDSMenuGroupPopupComponent], imports: [BidiModule,
        CommonModule,
        PlatformModule,
        OverlayModule,
        RouterModule,
        TDSOutletModule,
        TDSBadgeModule,
        TDSToolTipModule,
        TDSTagModule], exports: [TDSMenuComponent,
        TDSMenuItemComponent] }); })();
