// tds-ui/esm2020/carousel/carousel-content.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSCarouselContentDirective {
    constructor(elementRef, renderer) {
        this.renderer = renderer;
        this._active = false;
        this.el = elementRef.nativeElement;
        this.renderer.addClass(elementRef.nativeElement, 'slick-slide');
    }
    set isActive(value) {
        this._active = value;
        if (this.isActive) {
            this.renderer.addClass(this.el, 'slick-active');
        }
        else {
            this.renderer.removeClass(this.el, 'slick-active');
        }
    }
    get isActive() {
        return this._active;
    }
}
TDSCarouselContentDirective.ɵfac = function TDSCarouselContentDirective_Factory(t) { return new (t || TDSCarouselContentDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Renderer2)); };
TDSCarouselContentDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSCarouselContentDirective, selectors: [["", "tds-carousel-content", ""]], exportAs: ["tdsCarouselContent"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCarouselContentDirective, [{
        type: Directive,
        args: [{
                selector: '[tds-carousel-content]',
                exportAs: 'tdsCarouselContent'
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, null); })();
