// tds-ui/esm2020/carousel/carousel.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSCarouselContentDirective } from './carousel-content.directive';
import { TDSCarouselComponent } from './carousel.component';
import * as i0 from "@angular/core";
export class TDSCarouselModule {
}
TDSCarouselModule.ɵfac = function TDSCarouselModule_Factory(t) { return new (t || TDSCarouselModule)(); };
TDSCarouselModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCarouselModule });
TDSCarouselModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PlatformModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCarouselModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSCarouselComponent, TDSCarouselContentDirective],
                exports: [TDSCarouselComponent, TDSCarouselContentDirective],
                imports: [BidiModule, CommonModule, PlatformModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCarouselModule, { declarations: [TDSCarouselComponent, TDSCarouselContentDirective], imports: [BidiModule, CommonModule, PlatformModule], exports: [TDSCarouselComponent, TDSCarouselContentDirective] }); })();
