// tds-ui/esm2020/core/tree/tds-tree-base.mjs
import { TDSTreeNode } from './tds-tree-base-node';
export class TDSTreeBase {
    constructor(tdsTreeService) {
        this.tdsTreeService = tdsTreeService;
    }
    /**
     * Coerces a value({@link any[]}) to a TreeNodes({@link TDSTreeNode[]})
     */
    coerceTreeNodes(value) {
        let nodes = [];
        if (!this.tdsTreeService.isArrayOfTDSTreeNode(value)) {
            // has not been new TDSTreeNode
            nodes = value.map(item => new TDSTreeNode(item, null, this.tdsTreeService));
        }
        else {
            nodes = value.map((item) => {
                item.service = this.tdsTreeService;
                return item;
            });
        }
        return nodes;
    }
    /**
     * Get all nodes({@link TDSTreeNode})
     */
    getTreeNodes() {
        return this.tdsTreeService.rootNodes;
    }
    /**
     * Get {@link TDSTreeNode} with key
     */
    getTreeNodeByKey(key) {
        // flat tree nodes
        const nodes = [];
        const getNode = (node) => {
            nodes.push(node);
            node.getChildren().forEach(n => {
                getNode(n);
            });
        };
        this.getTreeNodes().forEach(n => {
            getNode(n);
        });
        return nodes.find(n => n.key === key) || null;
    }
    /**
     * Get checked nodes(merged)
     */
    getCheckedNodeList() {
        return this.tdsTreeService.getCheckedNodeList();
    }
    /**
     * Get selected nodes
     */
    getSelectedNodeList() {
        return this.tdsTreeService.getSelectedNodeList();
    }
    /**
     * Get half checked nodes
     */
    getHalfCheckedNodeList() {
        return this.tdsTreeService.getHalfCheckedNodeList();
    }
    /**
     * Get expanded nodes
     */
    getExpandedNodeList() {
        return this.tdsTreeService.getExpandedNodeList();
    }
    /**
     * Get matched nodes(if searchValue is not null)
     */
    getMatchedNodeList() {
        return this.tdsTreeService.getMatchedNodeList();
    }
}
