// tds-ui/esm2020/tds-checkbox/checkbox.module.mjs
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { TDSCheckbox } from "./checkbox";
import * as i0 from "@angular/core";
export class TDSCheckBoxModule {
}
TDSCheckBoxModule.ɵfac = function TDSCheckBoxModule_Factory(t) { return new (t || TDSCheckBoxModule)(); };
TDSCheckBoxModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSCheckBoxModule });
TDSCheckBoxModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            ReactiveFormsModule,
            FormsModule,
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSCheckBoxModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSCheckbox],
                imports: [
                    ReactiveFormsModule,
                    FormsModule,
                    CommonModule
                ],
                exports: [
                    TDSCheckbox
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSCheckBoxModule, { declarations: [TDSCheckbox], imports: [ReactiveFormsModule,
        FormsModule,
        CommonModule], exports: [TDSCheckbox] }); })();
