// tds-ui/esm2020/header/header.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSHeaderComponent } from './header.component';
import { TDSAvatarModule } from 'tds-ui/avatar';
import { TDSFormFieldModule } from 'tds-ui/form-field';
import { TDSInputModule } from 'tds-ui/tds-input';
import { TDSButtonModule } from 'tds-ui/button';
import * as i0 from "@angular/core";
export class TDSHeaderModule {
}
TDSHeaderModule.ɵfac = function TDSHeaderModule_Factory(t) { return new (t || TDSHeaderModule)(); };
TDSHeaderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSHeaderModule });
TDSHeaderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            TDSButtonModule,
            TDSFormFieldModule,
            TDSAvatarModule,
            TDSInputModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSHeaderModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSHeaderComponent
                ],
                exports: [
                    TDSHeaderComponent
                ],
                imports: [
                    CommonModule,
                    TDSButtonModule,
                    TDSFormFieldModule,
                    TDSAvatarModule,
                    TDSInputModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSHeaderModule, { declarations: [TDSHeaderComponent], imports: [CommonModule,
        TDSButtonModule,
        TDSFormFieldModule,
        TDSAvatarModule,
        TDSInputModule], exports: [TDSHeaderComponent] }); })();
