// tds-ui/esm2020/slider/slider.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSSliderHandleComponent } from './handle.component';
import { TDSSliderMarksComponent } from './marks.component';
import { TDSSliderComponent } from './slider.component';
import { TDSSliderStepComponent } from './step.component';
import { TDSSliderTrackComponent } from './track.component';
import * as i0 from "@angular/core";
export class TDSSliderModule {
}
TDSSliderModule.ɵfac = function TDSSliderModule_Factory(t) { return new (t || TDSSliderModule)(); };
TDSSliderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSSliderModule });
TDSSliderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PlatformModule, TDSToolTipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSliderModule, [{
        type: NgModule,
        args: [{
                exports: [
                    TDSSliderComponent,
                    TDSSliderTrackComponent,
                    TDSSliderHandleComponent,
                    TDSSliderStepComponent,
                    TDSSliderMarksComponent
                ],
                declarations: [
                    TDSSliderComponent,
                    TDSSliderTrackComponent,
                    TDSSliderHandleComponent,
                    TDSSliderStepComponent,
                    TDSSliderMarksComponent
                ],
                imports: [BidiModule, CommonModule, PlatformModule, TDSToolTipModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSSliderModule, { declarations: [TDSSliderComponent,
        TDSSliderTrackComponent,
        TDSSliderHandleComponent,
        TDSSliderStepComponent,
        TDSSliderMarksComponent], imports: [BidiModule, CommonModule, PlatformModule, TDSToolTipModule], exports: [TDSSliderComponent,
        TDSSliderTrackComponent,
        TDSSliderHandleComponent,
        TDSSliderStepComponent,
        TDSSliderMarksComponent] }); })();
