// tds-ui/esm2020/slider/step.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = function (a0, a1) { return { "w-1.5 h-px left-5 -mb-0.5": a0, "h-1.5 w-px top-5 -ml-px": a1 }; };
function TDSSliderStepComponent_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "span", 2);
} if (rf & 2) {
    const mark_r1 = ctx.$implicit;
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("tds-slider-dot-active", mark_r1.active);
    i0.ɵɵproperty("ngStyle", mark_r1.style)("ngClass", i0.ɵɵpureFunction2(4, _c0, ctx_r0.vertical, !ctx_r0.vertical));
} }
export class TDSSliderStepComponent {
    constructor() {
        this.lowerBound = null;
        this.upperBound = null;
        this.marksArray = [];
        this.vertical = false;
        this.included = false;
        this.steps = [];
    }
    ngOnChanges(changes) {
        const { marksArray, lowerBound, upperBound, reverse } = changes;
        if (marksArray || reverse) {
            this.buildSteps();
        }
        if (marksArray || lowerBound || upperBound || reverse) {
            this.togglePointActive();
        }
    }
    trackById(_index, step) {
        return step.value;
    }
    buildSteps() {
        const orient = this.vertical ? 'bottom' : 'left';
        this.steps = this.marksArray.map(mark => {
            const { value, config } = mark;
            let offset = mark.offset;
            const range = this.max - this.min;
            if (this.reverse) {
                offset = ((this.max - value) / range) * 100;
            }
            return {
                value,
                offset,
                config,
                active: false,
                style: {
                    [orient]: `${offset}%`
                }
            };
        });
    }
    togglePointActive() {
        if (this.steps && this.lowerBound !== null && this.upperBound !== null) {
            this.steps.forEach(step => {
                const value = step.value;
                const isActive = (!this.included && value === this.upperBound) ||
                    (this.included && value <= this.upperBound && value >= this.lowerBound);
                step.active = isActive;
            });
        }
    }
}
TDSSliderStepComponent.ɵfac = function TDSSliderStepComponent_Factory(t) { return new (t || TDSSliderStepComponent)(); };
TDSSliderStepComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSSliderStepComponent, selectors: [["tds-slider-step"]], inputs: { lowerBound: "lowerBound", upperBound: "upperBound", marksArray: "marksArray", min: "min", max: "max", vertical: "vertical", included: "included", reverse: "reverse" }, exportAs: ["tdsSliderStep"], features: [i0.ɵɵNgOnChangesFeature], decls: 2, vars: 2, consts: [[1, "tds-slider-step"], ["class", "tds-slider-dot absolute  bg-neutral-1-600 dark:bg-d-neutral-1-600", 3, "tds-slider-dot-active", "ngStyle", "ngClass", 4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "tds-slider-dot", "absolute", "bg-neutral-1-600", "dark:bg-d-neutral-1-600", 3, "ngStyle", "ngClass"]], template: function TDSSliderStepComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, TDSSliderStepComponent_span_1_Template, 1, 7, "span", 1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngForOf", ctx.steps)("ngForTrackBy", ctx.trackById);
    } }, directives: [i1.NgForOf, i1.NgStyle, i1.NgClass], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSSliderStepComponent.prototype, "vertical", void 0);
__decorate([
    InputBoolean()
], TDSSliderStepComponent.prototype, "included", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSliderStepComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-slider-step',
                exportAs: 'tdsSliderStep',
                preserveWhitespaces: false,
                template: `
    <div class="tds-slider-step">
      <span
        class="tds-slider-dot absolute  bg-neutral-1-600 dark:bg-d-neutral-1-600"
        *ngFor="let mark of steps; trackBy: trackById"
        [class.tds-slider-dot-active]="mark.active"
        [ngStyle]="mark.style!"
        [ngClass]="{'w-1.5 h-px left-5 -mb-0.5':vertical,'h-1.5 w-px top-5 -ml-px':!vertical}"
      ></span>
    </div>
  `
            }]
    }], null, { lowerBound: [{
            type: Input
        }], upperBound: [{
            type: Input
        }], marksArray: [{
            type: Input
        }], min: [{
            type: Input
        }], max: [{
            type: Input
        }], vertical: [{
            type: Input
        }], included: [{
            type: Input
        }], reverse: [{
            type: Input
        }] }); })();
