// tds-ui/esm2020/slider/track.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { InputBoolean, InputNumber } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = function (a0, a1) { return { "h-2": a0, "w-2": a1 }; };
export class TDSSliderTrackComponent {
    constructor() {
        this.offset = 0;
        this.reverse = false;
        this.dir = 'ltr';
        this.length = 0;
        this.vertical = false;
        this.included = false;
        this.style = {};
    }
    ngOnChanges() {
        const vertical = this.vertical;
        const reverse = this.reverse;
        const visibility = this.included ? 'visible' : 'hidden';
        const offset = this.offset;
        const length = this.length;
        const positonStyle = vertical
            ? {
                [reverse ? 'top' : 'bottom']: `${offset}%`,
                [reverse ? 'bottom' : 'top']: 'auto',
                height: `${length}%`,
                visibility
            }
            : {
                ...this.getHorizontalStylePosition(),
                width: `${length}%`,
                visibility
            };
        this.style = positonStyle;
    }
    getHorizontalStylePosition() {
        let left = this.reverse ? 'auto' : `${this.offset}%`;
        let right = this.reverse ? `${this.offset}%` : 'auto';
        if (this.dir === 'rtl') {
            const tmp = left;
            left = right;
            right = tmp;
        }
        return { left, right };
    }
}
TDSSliderTrackComponent.ɵfac = function TDSSliderTrackComponent_Factory(t) { return new (t || TDSSliderTrackComponent)(); };
TDSSliderTrackComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSSliderTrackComponent, selectors: [["tds-slider-track"]], inputs: { offset: "offset", reverse: "reverse", dir: "dir", length: "length", vertical: "vertical", included: "included" }, exportAs: ["tdsSliderTrack"], features: [i0.ɵɵNgOnChangesFeature], decls: 1, vars: 5, consts: [[1, "tds-slider-track", "absolute", "rounded-full", "transition-colors", "duration-300", "bg-primary-1", "dark:bg-d-primary-1", 3, "ngStyle", "ngClass"]], template: function TDSSliderTrackComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelement(0, "div", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("ngStyle", ctx.style)("ngClass", i0.ɵɵpureFunction2(2, _c0, !ctx.vertical, ctx.vertical));
    } }, directives: [i1.NgStyle, i1.NgClass], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputNumber()
], TDSSliderTrackComponent.prototype, "offset", void 0);
__decorate([
    InputBoolean()
], TDSSliderTrackComponent.prototype, "reverse", void 0);
__decorate([
    InputNumber()
], TDSSliderTrackComponent.prototype, "length", void 0);
__decorate([
    InputBoolean()
], TDSSliderTrackComponent.prototype, "vertical", void 0);
__decorate([
    InputBoolean()
], TDSSliderTrackComponent.prototype, "included", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSliderTrackComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-slider-track',
                exportAs: 'tdsSliderTrack',
                preserveWhitespaces: false,
                template: ` 
  <div class="tds-slider-track absolute  rounded-full transition-colors duration-300 bg-primary-1 dark:bg-d-primary-1"
   [ngStyle]="style"
   [ngClass]="{'h-2':!vertical, 'w-2':vertical}"
   ></div> 
  `
            }]
    }], null, { offset: [{
            type: Input
        }], reverse: [{
            type: Input
        }], dir: [{
            type: Input
        }], length: [{
            type: Input
        }], vertical: [{
            type: Input
        }], included: [{
            type: Input
        }] }); })();
