// tds-ui/esm2020/timeline/timeline.component.mjs
import { ChangeDetectionStrategy, Component, ContentChildren, Input, Optional, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TDSTimelineItemComponent } from './timeline-item.component';
import { TimelineService } from './timeline.service';
import { TDSHelperObject } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "./timeline.service";
import * as i2 from "@angular/cdk/bidi";
import * as i3 from "@angular/common";
import * as i4 from "tds-ui/progress-spinner";
import * as i5 from "tds-ui/core/outlet";
function TDSTimelineComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0, 4);
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r3 = i0.ɵɵreference(5);
    i0.ɵɵproperty("ngTemplateOutlet", _r3);
} }
function TDSTimelineComponent_ng_container_2_ng_template_1_Template(rf, ctx) { }
function TDSTimelineComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTimelineComponent_ng_container_2_ng_template_1_Template, 0, 0, "ng-template", 4);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", item_r5.template);
} }
function TDSTimelineComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0, 4);
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r3 = i0.ɵɵreference(5);
    i0.ɵɵproperty("ngTemplateOutlet", _r3);
} }
function TDSTimelineComponent_ng_template_4_li_0_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 18);
} }
function TDSTimelineComponent_ng_template_4_li_0_tds_spinner_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-spinner", 19);
} }
function TDSTimelineComponent_ng_template_4_li_0_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r13.tdsPendingDot, " ");
} }
function TDSTimelineComponent_ng_template_4_li_0_div_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 20);
    i0.ɵɵelement(1, "div", 18);
    i0.ɵɵelementEnd();
} }
function TDSTimelineComponent_ng_template_4_li_0_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r15 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r15.isPendingBoolean ? "" : ctx_r15.tdsPending, " ");
} }
function TDSTimelineComponent_ng_template_4_li_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 8)(1, "div", 9)(2, "div", 10)(3, "div", 11);
    i0.ɵɵtemplate(4, TDSTimelineComponent_ng_template_4_li_0_div_4_Template, 1, 0, "div", 12);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(5, "div", 13);
    i0.ɵɵtemplate(6, TDSTimelineComponent_ng_template_4_li_0_tds_spinner_6_Template, 1, 0, "tds-spinner", 14);
    i0.ɵɵtemplate(7, TDSTimelineComponent_ng_template_4_li_0_ng_container_7_Template, 2, 1, "ng-container", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(8, TDSTimelineComponent_ng_template_4_li_0_div_8_Template, 2, 0, "div", 16);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(9, "div", 17);
    i0.ɵɵtemplate(10, TDSTimelineComponent_ng_template_4_li_0_ng_container_10_Template, 2, 1, "ng-container", 15);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const isLast_r7 = i0.ɵɵnextContext().isLast;
    const ctx_r8 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("tds-timeline-item-last", isLast_r7);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", !ctx_r8.tdsReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r8.tdsPendingDot);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r8.tdsPendingDot);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r8.tdsReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r8.tdsPending);
} }
function TDSTimelineComponent_ng_template_4_li_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r17 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r17.isPendingBoolean ? "" : ctx_r17.tdsPending, " ");
} }
function TDSTimelineComponent_ng_template_4_li_1_div_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 18);
} }
function TDSTimelineComponent_ng_template_4_li_1_ng_container_9_tds_spinner_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-spinner", 19);
} }
function TDSTimelineComponent_ng_template_4_li_1_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵtemplate(2, TDSTimelineComponent_ng_template_4_li_1_ng_container_9_tds_spinner_2_Template, 1, 0, "tds-spinner", 14);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r19 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r19.tdsPendingDot, " ");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r19.tdsPendingDot);
} }
function TDSTimelineComponent_ng_template_4_li_1_div_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 20);
    i0.ɵɵelement(1, "div", 18);
    i0.ɵɵelementEnd();
} }
function TDSTimelineComponent_ng_template_4_li_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 21)(1, "div", 9)(2, "div", 22);
    i0.ɵɵtemplate(3, TDSTimelineComponent_ng_template_4_li_1_ng_container_3_Template, 2, 1, "ng-container", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 10)(5, "div", 11);
    i0.ɵɵtemplate(6, TDSTimelineComponent_ng_template_4_li_1_div_6_Template, 1, 0, "div", 12);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(7, "div", 13)(8, "div", 19);
    i0.ɵɵtemplate(9, TDSTimelineComponent_ng_template_4_li_1_ng_container_9_Template, 3, 2, "ng-container", 15);
    i0.ɵɵelementEnd()();
    i0.ɵɵtemplate(10, TDSTimelineComponent_ng_template_4_li_1_div_10_Template, 2, 0, "div", 23);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const isLast_r7 = i0.ɵɵnextContext().isLast;
    const ctx_r9 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("tds-timeline-item-last", isLast_r7);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r9.tdsPending);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !ctx_r9.tdsReverse);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r9.tdsPendingDot);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r9.tdsReverse);
} }
function TDSTimelineComponent_ng_template_4_li_2_div_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 18);
} }
function TDSTimelineComponent_ng_template_4_li_2_ng_container_8_tds_spinner_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-spinner", 19);
} }
function TDSTimelineComponent_ng_template_4_li_2_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵtemplate(2, TDSTimelineComponent_ng_template_4_li_2_ng_container_8_tds_spinner_2_Template, 1, 0, "tds-spinner", 14);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r24 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r24.tdsPendingDot, " ");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r24.tdsPendingDot);
} }
function TDSTimelineComponent_ng_template_4_li_2_div_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 20);
    i0.ɵɵelement(1, "div", 18);
    i0.ɵɵelementEnd();
} }
function TDSTimelineComponent_ng_template_4_li_2_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r26 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r26.isPendingBoolean ? "" : ctx_r26.tdsPending, " ");
} }
function TDSTimelineComponent_ng_template_4_li_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 24)(1, "div", 25);
    i0.ɵɵelement(2, "div", 26);
    i0.ɵɵelementStart(3, "div", 27)(4, "div", 11);
    i0.ɵɵtemplate(5, TDSTimelineComponent_ng_template_4_li_2_div_5_Template, 1, 0, "div", 12);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(6, "div", 13)(7, "div", 19);
    i0.ɵɵtemplate(8, TDSTimelineComponent_ng_template_4_li_2_ng_container_8_Template, 3, 2, "ng-container", 15);
    i0.ɵɵelementEnd()();
    i0.ɵɵtemplate(9, TDSTimelineComponent_ng_template_4_li_2_div_9_Template, 2, 0, "div", 23);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(10, "div", 28);
    i0.ɵɵtemplate(11, TDSTimelineComponent_ng_template_4_li_2_ng_container_11_Template, 2, 1, "ng-container", 15);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngIf", !ctx_r10.tdsReverse);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r10.tdsPendingDot);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r10.tdsReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r10.tdsPending);
} }
function TDSTimelineComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSTimelineComponent_ng_template_4_li_0_Template, 11, 7, "li", 5);
    i0.ɵɵtemplate(1, TDSTimelineComponent_ng_template_4_li_1_Template, 11, 6, "li", 6);
    i0.ɵɵtemplate(2, TDSTimelineComponent_ng_template_4_li_2_Template, 12, 4, "li", 7);
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngIf", !ctx_r4.hasLabelItem && ctx_r4.tdsMode == "left");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r4.hasLabelItem && ctx_r4.tdsMode == "right");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r4.hasLabelItem || ctx_r4.tdsMode === "alternate" || ctx_r4.tdsMode === "custom");
} }
const _c0 = ["*"];
export class TDSTimelineComponent {
    constructor(cdr, timelineService, directionality) {
        this.cdr = cdr;
        this.timelineService = timelineService;
        this.directionality = directionality;
        this.tdsMode = 'left';
        this.tdsReverse = false;
        this.isPendingBoolean = false;
        this.timelineItems = [];
        this.dir = 'ltr';
        this.hasLabelItem = false;
        this.destroy$ = new Subject();
    }
    ngOnChanges(changes) {
        const { tdsMode, tdsReverse, tdsPending } = changes;
        if (simpleChangeActivated(tdsMode) || simpleChangeActivated(tdsReverse)) {
            this.updateChildren();
        }
        if (tdsPending) {
            this.isPendingBoolean = tdsPending.currentValue === true;
        }
    }
    ngOnInit() {
        this.timelineService.check$.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.cdr.markForCheck();
        });
        this.directionality.change?.pipe(takeUntil(this.destroy$)).subscribe((direction) => {
            this.dir = direction;
            this.cdr.detectChanges();
        });
        this.dir = this.directionality.value;
    }
    ngAfterContentInit() {
        this.updateChildren();
        this.listOfItems.changes.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.updateChildren();
        });
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    updateChildren() {
        if (this.listOfItems && this.listOfItems.length) {
            const length = this.listOfItems.length;
            let hasLabelItem = false;
            //check has lable item
            hasLabelItem = this.listOfItems.find(item => TDSHelperObject.hasValue(item?.tdsLabel)) ? true : false;
            // for (let index = 0; index < this.listOfItems.length; index++) {
            //   const item = this.listOfItems.get(index);
            //   if (!hasLabelItem && item?.tdsLabel) {
            //     hasLabelItem = true;
            //     break;
            //   }
            // }
            this.listOfItems.forEach((item, index) => {
                item.isLast = !this.tdsReverse ? index === length - 1 : index === 0;
                item.isPending = this.tdsPending != undefined && this.tdsPending !== false;
                item.isReverse = this.tdsReverse;
                item.isFirst = this.tdsReverse ? index === length - 1 : index === 0;
                item.position = getInferredTimelineItemPosition(index, this.tdsMode);
                item.mode = (hasLabelItem || this.tdsMode === 'alternate' || this.tdsMode === 'custom') ? 'alternate' : this.tdsMode;
                if (!hasLabelItem && item?.tdsLabel) {
                    hasLabelItem = true;
                }
                item.detectChanges();
            });
            this.timelineItems = this.tdsReverse ? this.listOfItems.toArray().reverse() : this.listOfItems.toArray();
            this.hasLabelItem = hasLabelItem;
        }
        else {
            this.timelineItems = [];
            this.hasLabelItem = false;
        }
        this.cdr.markForCheck();
    }
}
TDSTimelineComponent.ɵfac = function TDSTimelineComponent_Factory(t) { return new (t || TDSTimelineComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TimelineService), i0.ɵɵdirectiveInject(i2.Directionality, 8)); };
TDSTimelineComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTimelineComponent, selectors: [["tds-timeline"]], contentQueries: function TDSTimelineComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSTimelineItemComponent, 4);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.listOfItems = _t);
    } }, inputs: { tdsMode: "tdsMode", tdsPending: "tdsPending", tdsPendingDot: "tdsPendingDot", tdsReverse: "tdsReverse" }, exportAs: ["tdsTimeline"], features: [i0.ɵɵProvidersFeature([TimelineService]), i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 7, vars: 15, consts: [[1, "tds-timeline", "box-border", "text-neutral-1-900", "dark:text-d-neutral-1-900", "m-0", "p-0"], [3, "ngTemplateOutlet", 4, "ngIf"], [4, "ngFor", "ngForOf"], ["pendingTemplate", ""], [3, "ngTemplateOutlet"], ["class", "tds-timeline-item relative  tds-timeline-item-left w-full", 3, "tds-timeline-item-last", 4, "ngIf"], ["class", "tds-timeline-item relative tds-timeline-item-right w-full text-right", 3, "tds-timeline-item-last", 4, "ngIf"], ["class", "tds-timeline-item-alternate relative w-full", 4, "ngIf"], [1, "tds-timeline-item", "relative", "tds-timeline-item-left", "w-full"], [1, "w-full", "flex", "min-h-[20px]"], [1, "flex-shrink-0", "flex", "flex-col", "min-w-[20px]", "items-center"], [1, "tds-timeline-item-tail", "flex", "w-full", "items-center", "justify-center", "h-1.5"], ["class", "w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200", 4, "ngIf"], [1, "tds-timeline-item-head", "flex", "items-center", "justify-center"], ["class", "w-3.5 h-3.5", 4, "ngIf"], [4, "tdsStringTemplateOutlet"], ["class", "tds-timeline-item-tail flex flex-auto  w-full items-center justify-center ", 4, "ngIf"], [1, "tds-timeline-item-content", "relative", "flex-1", "pl-4", "pb-4"], [1, "w-px", "h-full", "bg-neutral-2-200", "dark:bg-d-neutral-2-200"], [1, "w-3.5", "h-3.5"], [1, "tds-timeline-item-tail", "flex", "flex-auto", "w-full", "items-center", "justify-center"], [1, "tds-timeline-item", "relative", "tds-timeline-item-right", "w-full", "text-right"], [1, "tds-timeline-item-content", "relative", "flex-1", "pr-4", "pb-4"], ["class", "tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  ", 4, "ngIf"], [1, "tds-timeline-item-alternate", "relative", "w-full"], [1, "w-full", "flex", "items-stretch", "min-h-[20px]"], [1, "text-right", "flex-1"], [1, "flex-shrink-0", "flex", "flex-col", "px-4"], [1, "align-top", "flex-1"]], template: function TDSTimelineComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "ul", 0);
        i0.ɵɵtemplate(1, TDSTimelineComponent_ng_container_1_Template, 1, 1, "ng-container", 1);
        i0.ɵɵtemplate(2, TDSTimelineComponent_ng_container_2_Template, 2, 1, "ng-container", 2);
        i0.ɵɵtemplate(3, TDSTimelineComponent_ng_container_3_Template, 1, 1, "ng-container", 1);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(4, TDSTimelineComponent_ng_template_4_Template, 3, 3, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵprojection(6);
    } if (rf & 2) {
        i0.ɵɵclassProp("tds-timeline-label", ctx.hasLabelItem)("tds-timeline-right", !ctx.hasLabelItem && ctx.tdsMode === "right")("tds-timeline-alternate", ctx.tdsMode === "alternate" || ctx.tdsMode === "custom")("tds-timeline-pending", !!ctx.tdsPending)("tds-timeline-reverse", ctx.tdsReverse)("tds-timeline-rtl", ctx.dir === "rtl");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.tdsPending && ctx.tdsReverse);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngForOf", ctx.timelineItems);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.tdsPending && !ctx.tdsReverse);
    } }, directives: [i3.NgIf, i3.NgTemplateOutlet, i3.NgForOf, i4.TDSProgressSpinnerComponent, i5.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTimelineComponent, [{
        type: Component,
        args: [{ changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, preserveWhitespaces: false, selector: 'tds-timeline', providers: [TimelineService], exportAs: 'tdsTimeline', template: "<ul class=\"tds-timeline box-border text-neutral-1-900 dark:text-d-neutral-1-900 m-0 p-0\"\r\n  [class.tds-timeline-label]=\"hasLabelItem\" [class.tds-timeline-right]=\"!hasLabelItem && tdsMode === 'right'\"\r\n  [class.tds-timeline-alternate]=\"tdsMode === 'alternate' || tdsMode === 'custom'\"\r\n  [class.tds-timeline-pending]=\"!!tdsPending\" [class.tds-timeline-reverse]=\"tdsReverse\"\r\n  [class.tds-timeline-rtl]=\"dir === 'rtl'\">\r\n  <!-- pending dot (reversed) -->\r\n  <ng-container *ngIf=\"tdsPending && tdsReverse\" [ngTemplateOutlet]=\"pendingTemplate\"></ng-container>\r\n  <!-- timeline items -->\r\n  <ng-container *ngFor=\"let item of timelineItems\">\r\n    <ng-template [ngTemplateOutlet]=\"item.template\"></ng-template>\r\n  </ng-container>\r\n  <ng-container *ngIf=\"tdsPending && !tdsReverse\" [ngTemplateOutlet]=\"pendingTemplate\"></ng-container>\r\n  <!-- pending dot -->\r\n</ul>\r\n<ng-template #pendingTemplate let-isLast=\"isLast\">\r\n  <!-- tdsMode  left-->\r\n  <li *ngIf=\" !hasLabelItem && tdsMode =='left'\" class=\"tds-timeline-item relative  tds-timeline-item-left w-full\"\r\n    [class.tds-timeline-item-last]=\"isLast\">\r\n    <div class=\"w-full flex min-h-[20px]\">\r\n      <div class=\"flex-shrink-0 flex flex-col min-w-[20px] items-center  \">\r\n        <div class=\"tds-timeline-item-tail flex  w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!tdsReverse\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <tds-spinner *ngIf=\"!tdsPendingDot\" class=\"w-3.5 h-3.5\"></tds-spinner>\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsPendingDot\">\r\n            {{ tdsPendingDot }}\r\n          </ng-container>\r\n        </div>\r\n        <div *ngIf=\"tdsReverse\" class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"tds-timeline-item-content relative flex-1 pl-4 pb-4\">\r\n        <ng-container *tdsStringTemplateOutlet=\"tdsPending\">\r\n          {{ isPendingBoolean ? '' : tdsPending }}\r\n        </ng-container>\r\n      </div>\r\n    </div>\r\n  </li>\r\n  <!-- tdsMode right -->\r\n  <li *ngIf=\"hasLabelItem && tdsMode =='right'\"\r\n    class=\"tds-timeline-item relative tds-timeline-item-right w-full text-right\"\r\n    [class.tds-timeline-item-last]=\"isLast\">\r\n    <div class=\"w-full flex min-h-[20px]\">\r\n      <div class=\"tds-timeline-item-content relative flex-1 pr-4 pb-4\">\r\n        <ng-container *tdsStringTemplateOutlet=\"tdsPending\">\r\n          {{ isPendingBoolean ? '' : tdsPending }}\r\n        </ng-container>\r\n      </div>\r\n      <div class=\"flex-shrink-0 flex flex-col min-w-[20px] items-center \">\r\n        <div class=\"tds-timeline-item-tail flex  w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!tdsReverse\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <div class=\"w-3.5 h-3.5\">\r\n            <ng-container *tdsStringTemplateOutlet=\"tdsPendingDot\">\r\n              {{ tdsPendingDot }}\r\n              <tds-spinner *ngIf=\"!tdsPendingDot\" class=\"w-3.5 h-3.5\"></tds-spinner>\r\n            </ng-container>\r\n          </div>\r\n        </div>\r\n        <div *ngIf=\"tdsReverse\" class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n\r\n  </li>\r\n\r\n  <li *ngIf=\" (hasLabelItem || tdsMode === 'alternate' || tdsMode === 'custom')\"\r\n    class=\"tds-timeline-item-alternate relative w-full\">\r\n    <div class=\"w-full flex items-stretch min-h-[20px]\">\r\n      <!-- left -->\r\n      <div class=\"text-right flex-1\">\r\n        <!-- <div *ngIf=\"tdsLabel\" class=\"tds-timeline-item-label \">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsLabel\">{{ tdsLabel }}</ng-container>\r\n        </div> -->\r\n      </div>\r\n      <!-- icon -->\r\n      <div class=\"flex-shrink-0 flex flex-col  px-4 \">\r\n        <div class=\"tds-timeline-item-tail flex  w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!tdsReverse\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <div class=\"w-3.5 h-3.5\">\r\n            <ng-container *tdsStringTemplateOutlet=\"tdsPendingDot\">\r\n              {{ tdsPendingDot }}\r\n              <tds-spinner *ngIf=\"!tdsPendingDot\" class=\"w-3.5 h-3.5\"></tds-spinner>\r\n            </ng-container>\r\n          </div>\r\n\r\n        </div>\r\n        <div *ngIf=\"tdsReverse\" class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <!-- right -->\r\n      <div class=\"align-top flex-1\">\r\n        <ng-container *tdsStringTemplateOutlet=\"tdsPending\">\r\n          {{ isPendingBoolean ? '' : tdsPending }}\r\n        </ng-container>\r\n      </div>\r\n    </div>\r\n\r\n  </li>\r\n</ng-template>\r\n<!-- Grasp items -->\r\n<ng-content></ng-content>" }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.TimelineService }, { type: i2.Directionality, decorators: [{
                type: Optional
            }] }]; }, { listOfItems: [{
            type: ContentChildren,
            args: [TDSTimelineItemComponent]
        }], tdsMode: [{
            type: Input
        }], tdsPending: [{
            type: Input
        }], tdsPendingDot: [{
            type: Input
        }], tdsReverse: [{
            type: Input
        }] }); })();
function simpleChangeActivated(simpleChange) {
    return !!(simpleChange && (simpleChange.previousValue !== simpleChange.currentValue || simpleChange.isFirstChange()));
}
function getInferredTimelineItemPosition(index, mode) {
    return mode === 'custom'
        ? undefined
        : mode === 'left'
            ? 'left'
            : mode === 'right'
                ? 'right'
                : mode === 'alternate' && index % 2 === 0
                    ? 'left'
                    : 'right';
}
