// tds-ui/esm2020/timeline/timeline-item.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewChild, ViewEncapsulation } from '@angular/core';
import { TimelineTimeDefaultColors } from './typings';
import * as i0 from "@angular/core";
import * as i1 from "./timeline.service";
import * as i2 from "@angular/common";
import * as i3 from "tds-ui/core/outlet";
const _c0 = ["template"];
function TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    i0.ɵɵnextContext(3);
    const _r2 = i0.ɵɵreference(3);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_3_ng_container_1_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    i0.ɵɵnextContext(3);
    const _r4 = i0.ɵɵreference(5);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r4);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0)(1, 8);
    i0.ɵɵtemplate(2, TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_2_Template, 2, 1, "ng-container", 9);
    i0.ɵɵtemplate(3, TDSTimelineItemComponent_ng_template_0_ng_container_0_ng_container_3_Template, 2, 1, "ng-container", 9);
    i0.ɵɵelementContainerEnd()();
} if (rf & 2) {
    const ctx_r14 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitch", ctx_r14.getPositon());
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "left");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "right");
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    i0.ɵɵnextContext(3);
    const _r8 = i0.ɵɵreference(9);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r8);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_3_ng_container_1_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    i0.ɵɵnextContext(3);
    const _r6 = i0.ɵɵreference(7);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r6);
} }
function TDSTimelineItemComponent_ng_template_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0)(1, 8);
    i0.ɵɵtemplate(2, TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_2_Template, 2, 1, "ng-container", 9);
    i0.ɵɵtemplate(3, TDSTimelineItemComponent_ng_template_0_ng_container_1_ng_container_3_Template, 2, 1, "ng-container", 9);
    i0.ɵɵelementContainerEnd()();
} if (rf & 2) {
    const ctx_r15 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitch", ctx_r15.getPositon());
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "left");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", "right");
} }
function TDSTimelineItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSTimelineItemComponent_ng_template_0_ng_container_0_Template, 4, 3, "ng-container", 7);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_0_ng_container_1_Template, 4, 3, "ng-container", 7);
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngIf", ctx_r1.mode != "alternate");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.mode == "alternate");
} }
function TDSTimelineItemComponent_ng_template_2_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 20);
} }
function TDSTimelineItemComponent_ng_template_2_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r25 = i0.ɵɵnextContext(2);
    const _r12 = i0.ɵɵreference(13);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r25.tdsDot || _r12);
} }
function TDSTimelineItemComponent_ng_template_2_div_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 21);
    i0.ɵɵelement(1, "div", 20);
    i0.ɵɵelementEnd();
} }
function TDSTimelineItemComponent_ng_template_2_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 11)(1, "div", 12)(2, "div", 13)(3, "div", 14);
    i0.ɵɵtemplate(4, TDSTimelineItemComponent_ng_template_2_div_4_Template, 1, 0, "div", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(5, "div", 16);
    i0.ɵɵtemplate(6, TDSTimelineItemComponent_ng_template_2_ng_container_6_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(7, TDSTimelineItemComponent_ng_template_2_div_7_Template, 2, 0, "div", 18);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(8, "div", 19);
    i0.ɵɵtemplate(9, TDSTimelineItemComponent_ng_template_2_ng_container_9_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    const _r12 = i0.ɵɵreference(13);
    const _r10 = i0.ɵɵreference(11);
    i0.ɵɵclassProp("tds-timeline-item-last", ctx_r3.isLast);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", !ctx_r3.isFirst || ctx_r3.isPending && ctx_r3.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r3.tdsDot || _r12);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r3.isLast || ctx_r3.isPending && !ctx_r3.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", _r10);
} }
function TDSTimelineItemComponent_ng_template_4_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_4_div_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 20);
} }
function TDSTimelineItemComponent_ng_template_4_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r30 = i0.ɵɵnextContext(2);
    const _r12 = i0.ɵɵreference(13);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r30.tdsDot || _r12);
} }
function TDSTimelineItemComponent_ng_template_4_div_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 21);
    i0.ɵɵelement(1, "div", 20);
    i0.ɵɵelementEnd();
} }
function TDSTimelineItemComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 22)(1, "div", 12)(2, "div", 23);
    i0.ɵɵtemplate(3, TDSTimelineItemComponent_ng_template_4_ng_container_3_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 13)(5, "div", 14);
    i0.ɵɵtemplate(6, TDSTimelineItemComponent_ng_template_4_div_6_Template, 1, 0, "div", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(7, "div", 16);
    i0.ɵɵtemplate(8, TDSTimelineItemComponent_ng_template_4_ng_container_8_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(9, TDSTimelineItemComponent_ng_template_4_div_9_Template, 2, 0, "div", 24);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext();
    const _r10 = i0.ɵɵreference(11);
    const _r12 = i0.ɵɵreference(13);
    i0.ɵɵclassProp("tds-timeline-item-last", ctx_r5.isLast);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", _r10);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !ctx_r5.isFirst || ctx_r5.isPending && ctx_r5.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r5.tdsDot || _r12);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r5.isLast || ctx_r5.isPending && !ctx_r5.isReverse);
} }
function TDSTimelineItemComponent_ng_template_6_div_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r37 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r37.tdsLabel);
} }
function TDSTimelineItemComponent_ng_template_6_div_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 31);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_6_div_3_ng_container_1_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r32 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r32.tdsLabel);
} }
function TDSTimelineItemComponent_ng_template_6_div_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 20);
} }
function TDSTimelineItemComponent_ng_template_6_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r34 = i0.ɵɵnextContext(2);
    const _r12 = i0.ɵɵreference(13);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r34.tdsDot || _r12);
} }
function TDSTimelineItemComponent_ng_template_6_div_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 21);
    i0.ɵɵelement(1, "div", 20);
    i0.ɵɵelementEnd();
} }
function TDSTimelineItemComponent_ng_template_6_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 25)(1, "div", 26)(2, "div", 27);
    i0.ɵɵtemplate(3, TDSTimelineItemComponent_ng_template_6_div_3_Template, 2, 1, "div", 28);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 29)(5, "div", 14);
    i0.ɵɵtemplate(6, TDSTimelineItemComponent_ng_template_6_div_6_Template, 1, 0, "div", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(7, "div", 16);
    i0.ɵɵtemplate(8, TDSTimelineItemComponent_ng_template_6_ng_container_8_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(9, TDSTimelineItemComponent_ng_template_6_div_9_Template, 2, 0, "div", 24);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(10, "div", 30);
    i0.ɵɵtemplate(11, TDSTimelineItemComponent_ng_template_6_ng_container_11_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext();
    const _r12 = i0.ɵɵreference(13);
    const _r10 = i0.ɵɵreference(11);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", ctx_r7.tdsLabel);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !ctx_r7.isFirst || ctx_r7.isPending && ctx_r7.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r7.tdsDot || _r12);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r7.isLast || ctx_r7.isPending && !ctx_r7.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", _r10);
} }
function TDSTimelineItemComponent_ng_template_8_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSTimelineItemComponent_ng_template_8_div_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 20);
} }
function TDSTimelineItemComponent_ng_template_8_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r40 = i0.ɵɵnextContext(2);
    const _r12 = i0.ɵɵreference(13);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r40.tdsDot || _r12);
} }
function TDSTimelineItemComponent_ng_template_8_div_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 21);
    i0.ɵɵelement(1, "div", 20);
    i0.ɵɵelementEnd();
} }
function TDSTimelineItemComponent_ng_template_8_div_11_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r43 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r43.tdsLabel);
} }
function TDSTimelineItemComponent_ng_template_8_div_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 31);
    i0.ɵɵtemplate(1, TDSTimelineItemComponent_ng_template_8_div_11_ng_container_1_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r42 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r42.tdsLabel);
} }
function TDSTimelineItemComponent_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 25)(1, "div", 26)(2, "div", 32);
    i0.ɵɵtemplate(3, TDSTimelineItemComponent_ng_template_8_ng_container_3_Template, 1, 0, "ng-container", 10);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 29)(5, "div", 14);
    i0.ɵɵtemplate(6, TDSTimelineItemComponent_ng_template_8_div_6_Template, 1, 0, "div", 15);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(7, "div", 16);
    i0.ɵɵtemplate(8, TDSTimelineItemComponent_ng_template_8_ng_container_8_Template, 2, 1, "ng-container", 17);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(9, TDSTimelineItemComponent_ng_template_8_div_9_Template, 2, 0, "div", 24);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(10, "div", 33);
    i0.ɵɵtemplate(11, TDSTimelineItemComponent_ng_template_8_div_11_Template, 2, 1, "div", 34);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext();
    const _r10 = i0.ɵɵreference(11);
    const _r12 = i0.ɵɵreference(13);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", _r10);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !ctx_r9.isFirst || ctx_r9.isPending && ctx_r9.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r9.tdsDot || _r12);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r9.isLast || ctx_r9.isPending && !ctx_r9.isReverse);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r9.tdsLabel);
} }
function TDSTimelineItemComponent_ng_template_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0);
} }
const _c1 = function (a0, a1, a2, a3, a4, a5) { return { "border-primary-1 dark:border-d-primary-1": a0, "border-info-400 dark:border-d-info-400": a1, "border-error-400 dark:border-d-error-400": a2, "border-success-400 dark:border-d-success-400": a3, "border-neutral-2-200 dark:border-d-neutral-2-200": a4, "border-warning-400 dark:border-d-warning-400": a5 }; };
function TDSTimelineItemComponent_ng_template_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 35);
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("border-color", ctx_r13.borderColor);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction6(3, _c1, ctx_r13.tdsColor == "primary", ctx_r13.tdsColor == "info", ctx_r13.tdsColor == "error", ctx_r13.tdsColor == "success", ctx_r13.tdsColor == "secondary", ctx_r13.tdsColor == "warning"));
} }
const _c2 = ["*"];
function isDefaultColor(color) {
    return TimelineTimeDefaultColors.findIndex(i => i === color) !== -1;
}
export class TDSTimelineItemComponent {
    constructor(cdr, timelineService) {
        this.cdr = cdr;
        this.timelineService = timelineService;
        this.tdsColor = 'primary';
        this.isLast = false;
        this.isPending = false;
        this.isReverse = false;
        this.borderColor = null;
        this.position = 'left';
        this.mode = 'alternate';
        this.isFirst = false;
    }
    ngOnChanges(changes) {
        this.timelineService.markForCheck();
        if (changes.tdsColor) {
            this.updateCustomColor();
        }
    }
    detectChanges() {
        this.cdr.detectChanges();
    }
    updateCustomColor() {
        this.borderColor = isDefaultColor(this.tdsColor) ? null : this.tdsColor;
    }
    getPositon() {
        return this.tdsPosition || this.position || 'left';
    }
}
TDSTimelineItemComponent.ɵfac = function TDSTimelineItemComponent_Factory(t) { return new (t || TDSTimelineItemComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TimelineService)); };
TDSTimelineItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTimelineItemComponent, selectors: [["tds-timeline-item"], ["", "tds-timeline-item", ""]], viewQuery: function TDSTimelineItemComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { tdsPosition: "tdsPosition", tdsColor: "tdsColor", tdsDot: "tdsDot", tdsLabel: "tdsLabel" }, exportAs: ["tdsTimelineItem"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c2, decls: 14, vars: 0, consts: [["template", ""], ["leftTemplate", ""], ["rightTemplate", ""], ["rightAlternate", ""], ["leftAlternate", ""], ["content", ""], ["defaultDot", ""], [4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngTemplateOutlet"], [1, "tds-timeline-item", "relative", "tds-timeline-item-left", "w-full"], [1, "w-full", "flex", "min-h-[20px]"], [1, "flex-shrink-0", "flex", "flex-col", "min-w-[20px]", "items-center"], [1, "tds-timeline-item-tail", "flex", "w-full", "items-center", "justify-center", "h-1.5"], ["class", "w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200", 4, "ngIf"], [1, "tds-timeline-item-head", "flex", "items-center", "justify-center"], [4, "tdsStringTemplateOutlet"], ["class", "tds-timeline-item-tail flex flex-auto  w-full items-center justify-center ", 4, "ngIf"], [1, "tds-timeline-item-content", "relative", "flex-1", "pl-4", "pb-4"], [1, "w-px", "h-full", "bg-neutral-2-200", "dark:bg-d-neutral-2-200"], [1, "tds-timeline-item-tail", "flex", "flex-auto", "w-full", "items-center", "justify-center"], [1, "tds-timeline-item", "relative", "tds-timeline-item-right", "w-full", "text-right"], [1, "tds-timeline-item-content", "relative", "flex-1", "pr-4", "pb-4"], ["class", "tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  ", 4, "ngIf"], [1, "tds-timeline-item-alternate", "relative", "w-full"], [1, "w-full", "flex", "items-stretch", "min-h-[20px]"], [1, "text-right", "flex-1"], ["class", "tds-timeline-item-label ", 4, "ngIf"], [1, "flex-shrink-0", "flex", "flex-col", "px-4"], [1, "align-top", "flex-1", "pb-4"], [1, "tds-timeline-item-label"], [1, "flex-1", "text-right", "pb-4"], [1, "align-top", "flex-1"], ["class", "tds-timeline-item-label", 4, "ngIf"], [1, "rounded-full", "border-2", "w-3.5", "h-3.5", 3, "ngClass"]], template: function TDSTimelineItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtemplate(0, TDSTimelineItemComponent_ng_template_0_Template, 2, 2, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(2, TDSTimelineItemComponent_ng_template_2_Template, 10, 6, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(4, TDSTimelineItemComponent_ng_template_4_Template, 10, 6, "ng-template", null, 2, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(6, TDSTimelineItemComponent_ng_template_6_Template, 12, 5, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(8, TDSTimelineItemComponent_ng_template_8_Template, 12, 5, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(10, TDSTimelineItemComponent_ng_template_10_Template, 1, 0, "ng-template", null, 5, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(12, TDSTimelineItemComponent_ng_template_12_Template, 1, 10, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
    } }, directives: [i2.NgIf, i2.NgSwitch, i2.NgSwitchCase, i2.NgTemplateOutlet, i3.TDSStringTemplateOutletDirective, i2.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTimelineItemComponent, [{
        type: Component,
        args: [{ changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, preserveWhitespaces: false, selector: 'tds-timeline-item, [tds-timeline-item]', exportAs: 'tdsTimelineItem', template: "<ng-template #template>\r\n  <ng-container *ngIf=\"mode != 'alternate'\">\r\n    <ng-container [ngSwitch]=\"getPositon()\">\r\n      <ng-container *ngSwitchCase=\"'left'\">\r\n        <ng-container *ngTemplateOutlet=\"leftTemplate\"></ng-container>\r\n      </ng-container>\r\n      <ng-container *ngSwitchCase=\"'right'\">\r\n        <ng-container *ngTemplateOutlet=\"rightTemplate\"></ng-container>\r\n      </ng-container>\r\n\r\n    </ng-container>\r\n  </ng-container>\r\n  <ng-container *ngIf=\"mode == 'alternate'\">\r\n    <ng-container [ngSwitch]=\"getPositon()\">\r\n      <ng-container *ngSwitchCase=\"'left'\">\r\n        <ng-container *ngTemplateOutlet=\"leftAlternate\"></ng-container>\r\n      </ng-container>\r\n      <ng-container *ngSwitchCase=\"'right'\">\r\n        <ng-container *ngTemplateOutlet=\"rightAlternate\"></ng-container>\r\n      </ng-container>\r\n    </ng-container>\r\n  </ng-container>\r\n</ng-template>\r\n<ng-template #leftTemplate>\r\n\r\n  <li class=\"tds-timeline-item relative  tds-timeline-item-left w-full\" [class.tds-timeline-item-last]=\"isLast\">\r\n    <div class=\"w-full flex min-h-[20px]\">\r\n      <div class=\"flex-shrink-0 flex flex-col min-w-[20px] items-center  \">\r\n        <div class=\"tds-timeline-item-tail flex  w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!isFirst || (isPending && isReverse)\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsDot || defaultDot\">{{ tdsDot || defaultDot }}</ng-container>\r\n        </div>\r\n        <div *ngIf=\"!isLast || (isPending && !isReverse)\"\r\n          class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"tds-timeline-item-content relative flex-1 pl-4 pb-4\">\r\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n      </div>\r\n    </div>\r\n  </li>\r\n\r\n</ng-template>\r\n<ng-template #rightTemplate>\r\n  <li class=\"tds-timeline-item relative tds-timeline-item-right w-full text-right\"\r\n    [class.tds-timeline-item-last]=\"isLast\">\r\n    <div class=\"w-full flex min-h-[20px]\">\r\n      <div class=\"tds-timeline-item-content relative flex-1 pr-4 pb-4\">\r\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n      </div>\r\n      <div class=\"flex-shrink-0 flex flex-col min-w-[20px] items-center \">\r\n        <div class=\"tds-timeline-item-tail flex w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!isFirst || (isPending && isReverse)\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsDot || defaultDot\">{{ tdsDot || defaultDot }}</ng-container>\r\n        </div>\r\n        <div *ngIf=\"!isLast || (isPending && !isReverse)\"\r\n          class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n\r\n  </li>\r\n</ng-template>\r\n<ng-template #rightAlternate>\r\n  <li class=\"tds-timeline-item-alternate relative w-full\">\r\n    <div class=\"w-full flex items-stretch min-h-[20px]\">\r\n      <!-- left -->\r\n      <div class=\"text-right flex-1\">\r\n        <div *ngIf=\"tdsLabel\" class=\"tds-timeline-item-label \">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsLabel\">{{ tdsLabel }}</ng-container>\r\n        </div>\r\n      </div>\r\n      <!-- icon -->\r\n      <div class=\"flex-shrink-0 flex flex-col px-4\">\r\n        <div class=\"tds-timeline-item-tail flex   w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!isFirst || (isPending && isReverse)\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsDot || defaultDot\">{{ tdsDot || defaultDot }}</ng-container>\r\n        </div>\r\n        <div *ngIf=\"!isLast || (isPending && !isReverse)\"\r\n          class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <!-- right -->\r\n      <div class=\"align-top flex-1 pb-4\">\r\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n\r\n      </div>\r\n    </div>\r\n\r\n  </li>\r\n</ng-template>\r\n<ng-template #leftAlternate>\r\n  <li class=\"tds-timeline-item-alternate relative w-full\">\r\n    <div class=\"w-full flex items-stretch min-h-[20px]\">\r\n\r\n      <!-- left -->\r\n      <div class=\"flex-1 text-right pb-4\">\r\n        <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n      </div>\r\n      <!-- icon -->\r\n      <div class=\"flex-shrink-0 flex flex-col  px-4 \">\r\n        <div class=\"tds-timeline-item-tail flex   w-full items-center justify-center h-1.5\">\r\n          <div *ngIf=\"!isFirst || (isPending && isReverse)\" class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n        <div class=\"tds-timeline-item-head  flex items-center justify-center\">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsDot || defaultDot\">{{ tdsDot || defaultDot }}</ng-container>\r\n        </div>\r\n        <div *ngIf=\"!isLast || (isPending && !isReverse)\"\r\n          class=\"tds-timeline-item-tail flex flex-auto  w-full items-center justify-center  \">\r\n          <div class=\"w-px h-full bg-neutral-2-200 dark:bg-d-neutral-2-200\">\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <!-- right -->\r\n      <div class=\"align-top flex-1\">\r\n        <div *ngIf=\"tdsLabel\" class=\"tds-timeline-item-label\">\r\n          <ng-container *tdsStringTemplateOutlet=\"tdsLabel\">{{ tdsLabel }}</ng-container>\r\n        </div>\r\n\r\n      </div>\r\n    </div>\r\n\r\n  </li>\r\n</ng-template>\r\n\r\n<ng-template #content>\r\n  <ng-content></ng-content>\r\n</ng-template>\r\n\r\n<ng-template #defaultDot>\r\n  <div class=\"rounded-full border-2 w-3.5 h-3.5\" [ngClass]=\"{\r\n    'border-primary-1 dark:border-d-primary-1':tdsColor == 'primary',\r\n    'border-info-400 dark:border-d-info-400':tdsColor == 'info',\r\n    'border-error-400 dark:border-d-error-400':tdsColor == 'error',\r\n    'border-success-400 dark:border-d-success-400':tdsColor == 'success',\r\n    'border-neutral-2-200 dark:border-d-neutral-2-200':tdsColor == 'secondary',\r\n    'border-warning-400 dark:border-d-warning-400':tdsColor == 'warning'\r\n  }\"  [style.border-color]=\"borderColor\"></div>\r\n</ng-template>" }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.TimelineService }]; }, { template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }], tdsPosition: [{
            type: Input
        }], tdsColor: [{
            type: Input
        }], tdsDot: [{
            type: Input
        }], tdsLabel: [{
            type: Input
        }] }); })();
