// tds-ui/esm2020/timeline/timeline.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSTimelineItemComponent } from './timeline-item.component';
import { TDSTimelineComponent } from './timeline.component';
import * as i0 from "@angular/core";
export class TDSTimelineModule {
}
TDSTimelineModule.ɵfac = function TDSTimelineModule_Factory(t) { return new (t || TDSTimelineModule)(); };
TDSTimelineModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSTimelineModule });
TDSTimelineModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PlatformModule, TDSOutletModule, TDSSpinnerModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTimelineModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSTimelineItemComponent, TDSTimelineComponent,],
                exports: [TDSTimelineItemComponent, TDSTimelineComponent,],
                imports: [BidiModule, CommonModule, PlatformModule, TDSOutletModule, TDSSpinnerModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSTimelineModule, { declarations: [TDSTimelineItemComponent, TDSTimelineComponent], imports: [BidiModule, CommonModule, PlatformModule, TDSOutletModule, TDSSpinnerModule], exports: [TDSTimelineItemComponent, TDSTimelineComponent] }); })();
