// tds-ui/esm2020/buttton-split/button-split-module.mjs
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { TDSButtonMenuModule } from "tds-ui/button-menu";
import { TDSButtonModule } from "tds-ui/button";
import { TDSButtonSplit } from "./button-split";
import * as i0 from "@angular/core";
export class TDSButtonSpitModule {
}
TDSButtonSpitModule.ɵfac = function TDSButtonSpitModule_Factory(t) { return new (t || TDSButtonSpitModule)(); };
TDSButtonSpitModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSButtonSpitModule });
TDSButtonSpitModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            TDSButtonModule,
            TDSButtonMenuModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonSpitModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    TDSButtonModule,
                    TDSButtonMenuModule
                ],
                exports: [TDSButtonSplit],
                declarations: [TDSButtonSplit],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSButtonSpitModule, { declarations: [TDSButtonSplit], imports: [CommonModule,
        TDSButtonModule,
        TDSButtonMenuModule], exports: [TDSButtonSplit] }); })();
