// tds-ui/esm2020/tag/tag.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSTagComponent } from './tag.component';
import * as i0 from "@angular/core";
export class TDSTagModule {
}
TDSTagModule.ɵfac = function TDSTagModule_Factory(t) { return new (t || TDSTagModule)(); };
TDSTagModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSTagModule });
TDSTagModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTagModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, FormsModule],
                declarations: [TDSTagComponent],
                exports: [TDSTagComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSTagModule, { declarations: [TDSTagComponent], imports: [BidiModule, CommonModule, FormsModule], exports: [TDSTagComponent] }); })();
