// tds-ui/esm2020/core/outlet/outlet.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSStringTemplateOutletDirective } from './string_template_outlet.directive';
import * as i0 from "@angular/core";
export class TDSOutletModule {
}
TDSOutletModule.ɵfac = function TDSOutletModule_Factory(t) { return new (t || TDSOutletModule)(); };
TDSOutletModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSOutletModule });
TDSOutletModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSOutletModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [TDSStringTemplateOutletDirective],
                declarations: [TDSStringTemplateOutletDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSOutletModule, { declarations: [TDSStringTemplateOutletDirective], imports: [CommonModule], exports: [TDSStringTemplateOutletDirective] }); })();
