// tds-ui/esm2020/typography/typography.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { ClipboardModule } from '@angular/cdk/clipboard';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSInputModule } from 'tds-ui/tds-input';
import { TDSToolTipModule } from 'tds-ui/tooltip';
import { TDSTextCopyComponent } from './text-copy.component';
import { TDSTextEditComponent } from './text-edit.component';
import { TDSTypographyComponent } from './typography.component';
import * as i0 from "@angular/core";
export class TDSTypographyModule {
}
TDSTypographyModule.ɵfac = function TDSTypographyModule_Factory(t) { return new (t || TDSTypographyModule)(); };
TDSTypographyModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSTypographyModule });
TDSTypographyModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            TDSToolTipModule,
            TDSInputModule,
            TDSI18nModule,
            ClipboardModule,
            TDSOutletModule,
            TDSButtonModule
        ], PlatformModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTypographyModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    TDSToolTipModule,
                    TDSInputModule,
                    TDSI18nModule,
                    ClipboardModule,
                    TDSOutletModule,
                    TDSButtonModule
                ],
                exports: [TDSTypographyComponent, TDSTextCopyComponent, TDSTextEditComponent, PlatformModule],
                declarations: [TDSTypographyComponent, TDSTextCopyComponent, TDSTextEditComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSTypographyModule, { declarations: [TDSTypographyComponent, TDSTextCopyComponent, TDSTextEditComponent], imports: [BidiModule,
        CommonModule,
        TDSToolTipModule,
        TDSInputModule,
        TDSI18nModule,
        ClipboardModule,
        TDSOutletModule,
        TDSButtonModule], exports: [TDSTypographyComponent, TDSTextCopyComponent, TDSTextEditComponent, PlatformModule] }); })();
