// tds-ui/esm2020/table/src/addon/filter-trigger.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/dropdown";
const _c0 = ["*"];
export class TDSFilterTriggerComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.active = false;
        this.visible = false;
        /**
         * @deprecated Not supported, use `backdrop` instead.
         * @breaking-change 13.0.0
         */
        this.hasBackdrop = false;
        this.backdrop = false;
        this.visibleChange = new EventEmitter();
    }
    onVisibleChange(visible) {
        this.visible = visible;
        this.visibleChange.next(visible);
    }
    onFilterClick($event) {
        $event.stopPropagation();
    }
    hide() {
        this.visible = false;
        this.cdr.markForCheck();
    }
    show() {
        this.visible = true;
        this.cdr.markForCheck();
    }
    ngOnChanges(changes) {
        const { hasBackdrop } = changes;
        // if (hasBackdrop) {
        //   console.warn(
        //     '`hasBackdrop` in tds-filter-trigger component will be removed in 13.0.0, please use `backdrop` instead.'
        //   );
        // }
    }
}
TDSFilterTriggerComponent.ɵfac = function TDSFilterTriggerComponent_Factory(t) { return new (t || TDSFilterTriggerComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSFilterTriggerComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSFilterTriggerComponent, selectors: [["tds-filter-trigger"]], inputs: { active: "active", dropdownMenu: "dropdownMenu", visible: "visible", hasBackdrop: "hasBackdrop", backdrop: "backdrop" }, outputs: { visibleChange: "visibleChange" }, exportAs: ["tdsFilterTrigger"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 2, vars: 8, consts: [["tds-dropdown", "", "trigger", "click", "placement", "bottomRight", 1, "tds-table-filter-trigger", 3, "backdrop", "clickHide", "tdsDropdownMenu", "visible", "visibleChange", "click"]], template: function TDSFilterTriggerComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "span", 0);
        i0.ɵɵlistener("visibleChange", function TDSFilterTriggerComponent_Template_span_visibleChange_0_listener($event) { return ctx.onVisibleChange($event); })("click", function TDSFilterTriggerComponent_Template_span_click_0_listener($event) { return ctx.onFilterClick($event); });
        i0.ɵɵprojection(1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵclassProp("active", ctx.active)("tds-table-filter-open", ctx.visible);
        i0.ɵɵproperty("backdrop", ctx.backdrop || ctx.hasBackdrop)("clickHide", false)("tdsDropdownMenu", ctx.dropdownMenu)("visible", ctx.visible);
    } }, directives: [i1.TDSDropDownDirective], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSFilterTriggerComponent.prototype, "hasBackdrop", void 0);
__decorate([
    InputBoolean()
], TDSFilterTriggerComponent.prototype, "backdrop", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFilterTriggerComponent, [{
        type: Component,
        args: [{
                selector: 'tds-filter-trigger',
                exportAs: `tdsFilterTrigger`,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                template: `
    <span
      tds-dropdown
      class="tds-table-filter-trigger"
      trigger="click"
      placement="bottomRight"
      [backdrop]="backdrop || hasBackdrop"
      [clickHide]="false"
      [tdsDropdownMenu]="dropdownMenu"
      [class.active]="active"
      [class.tds-table-filter-open]="visible"
      [visible]="visible"
      (visibleChange)="onVisibleChange($event)"
      (click)="onFilterClick($event)"
    >
      <ng-content></ng-content>
    </span>
  `
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { active: [{
            type: Input
        }], dropdownMenu: [{
            type: Input
        }], visible: [{
            type: Input
        }], hasBackdrop: [{
            type: Input
        }], backdrop: [{
            type: Input
        }], visibleChange: [{
            type: Output
        }] }); })();
