// tds-ui/esm2020/table/src/addon/row-expand-button.directive.mjs
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSRowExpandButtonDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.expand = false;
        this.spaceMode = false;
        this.expandChange = new EventEmitter();
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-table-row-expand-icon');
    }
    onHostClick() {
        if (!this.spaceMode) {
            this.expand = !this.expand;
            this.expandChange.next(this.expand);
        }
    }
}
TDSRowExpandButtonDirective.ɵfac = function TDSRowExpandButtonDirective_Factory(t) { return new (t || TDSRowExpandButtonDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSRowExpandButtonDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSRowExpandButtonDirective, selectors: [["button", "tds-row-expand-button", ""]], hostVars: 11, hostBindings: function TDSRowExpandButtonDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function TDSRowExpandButtonDirective_click_HostBindingHandler() { return ctx.onHostClick(); });
    } if (rf & 2) {
        i0.ɵɵhostProperty("type", "button");
        i0.ɵɵclassProp("tds-table-row-expand-icon-expanded", !ctx.spaceMode && ctx.expand === true)("tds-table-row-expand-icon-collapsed", !ctx.spaceMode && ctx.expand === false)("bg-transparent", ctx.spaceMode)("border-0", ctx.spaceMode)("invisible", ctx.spaceMode);
    } }, inputs: { expand: "expand", spaceMode: "spaceMode" }, outputs: { expandChange: "expandChange" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRowExpandButtonDirective, [{
        type: Directive,
        args: [{
                selector: 'button[tds-row-expand-button]',
                host: {
                    '[type]': `'button'`,
                    '[class.tds-table-row-expand-icon-expanded]': `!spaceMode && expand === true`,
                    '[class.tds-table-row-expand-icon-collapsed]': `!spaceMode && expand === false`,
                    '[class.bg-transparent]': 'spaceMode',
                    '[class.border-0]': 'spaceMode',
                    '[class.invisible]': 'spaceMode',
                    '(click)': 'onHostClick()',
                }
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { expand: [{
            type: Input
        }], spaceMode: [{
            type: Input
        }], expandChange: [{
            type: Output
        }] }); })();
