// tds-ui/esm2020/table/src/addon/row-indent.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSRowIndentDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.indentSize = 0;
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-table-row-indent');
    }
}
TDSRowIndentDirective.ɵfac = function TDSRowIndentDirective_Factory(t) { return new (t || TDSRowIndentDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSRowIndentDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSRowIndentDirective, selectors: [["tds-row-indent"]], hostVars: 2, hostBindings: function TDSRowIndentDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵstyleProp("padding-left", ctx.indentSize, "px");
    } }, inputs: { indentSize: "indentSize" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRowIndentDirective, [{
        type: Directive,
        args: [{
                selector: 'tds-row-indent',
                host: {
                    '[style.padding-left.px]': 'indentSize'
                }
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { indentSize: [{
            type: Input
        }] }); })();
