// tds-ui/esm2020/table/src/addon/selection.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/tds-checkbox";
import * as i3 from "@angular/forms";
import * as i4 from "tds-ui/dropdown";
function TDSTableSelectionComponent_tds_checkbox_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "tds-checkbox", 2);
    i0.ɵɵlistener("ngModelChange", function TDSTableSelectionComponent_tds_checkbox_0_Template_tds_checkbox_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.onCheckedChange($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("tds-table-selection-select-all-custom", ctx_r0.showRowSelection);
    i0.ɵɵproperty("ngModel", ctx_r0.checked)("disabled", ctx_r0.disabled)("indeterminate", ctx_r0.indeterminate);
} }
function TDSTableSelectionComponent_div_1_div_6_Template(rf, ctx) { if (rf & 1) {
    const _r8 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 9);
    i0.ɵɵlistener("click", function TDSTableSelectionComponent_div_1_div_6_Template_div_click_0_listener() { const restoredCtx = i0.ɵɵrestoreView(_r8); const selection_r6 = restoredCtx.$implicit; return selection_r6.onSelect(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const selection_r6 = ctx.$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", selection_r6.text, " ");
} }
function TDSTableSelectionComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 3)(1, "span", 4);
    i0.ɵɵelement(2, "i", 5);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "tds-dropdown-menu", null, 6)(5, "div", 7);
    i0.ɵɵtemplate(6, TDSTableSelectionComponent_div_1_div_6_Template, 2, 1, "div", 8);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const _r4 = i0.ɵɵreference(4);
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsDropdownMenu", _r4);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngForOf", ctx_r1.listOfSelections);
} }
export class TDSTableSelectionComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.listOfSelections = [];
        this.checked = false;
        this.disabled = false;
        this.indeterminate = false;
        this.showCheckbox = false;
        this.showRowSelection = false;
        this.checkedChange = new EventEmitter();
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-table-selection');
    }
    onCheckedChange(checked) {
        this.checked = checked;
        this.checkedChange.emit(checked);
    }
}
TDSTableSelectionComponent.ɵfac = function TDSTableSelectionComponent_Factory(t) { return new (t || TDSTableSelectionComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSTableSelectionComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTableSelectionComponent, selectors: [["tds-table-selection"]], inputs: { listOfSelections: "listOfSelections", checked: "checked", disabled: "disabled", indeterminate: "indeterminate", showCheckbox: "showCheckbox", showRowSelection: "showRowSelection" }, outputs: { checkedChange: "checkedChange" }, decls: 2, vars: 2, consts: [[3, "tds-table-selection-select-all-custom", "ngModel", "disabled", "indeterminate", "ngModelChange", 4, "ngIf"], ["class", "tds-table-selection-extra inline", 4, "ngIf"], [3, "ngModel", "disabled", "indeterminate", "ngModelChange"], [1, "tds-table-selection-extra", "inline"], ["tds-dropdown", "", "placement", "bottomLeft", 1, "tds-table-selection-down", 3, "tdsDropdownMenu"], [1, "tdsi-arrow-down-fill"], ["selectionMenu", "tdsDropdownMenu"], [1, "w-full"], ["tds-dropdown-item", "", 3, "click", 4, "ngFor", "ngForOf"], ["tds-dropdown-item", "", 3, "click"]], template: function TDSTableSelectionComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSTableSelectionComponent_tds_checkbox_0_Template, 1, 5, "tds-checkbox", 0);
        i0.ɵɵtemplate(1, TDSTableSelectionComponent_div_1_Template, 7, 2, "div", 1);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.showCheckbox);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.showRowSelection);
    } }, directives: [i1.NgIf, i2.TDSCheckbox, i3.NgControlStatus, i3.NgModel, i4.TDSDropDownDirective, i4.TDSDropdownMenuComponent, i1.NgForOf, i4.TDSDropDownItemDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableSelectionComponent, [{
        type: Component,
        args: [{
                selector: 'tds-table-selection',
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
  <tds-checkbox 
  *ngIf="showCheckbox"
  [class.tds-table-selection-select-all-custom]="showRowSelection"
  [ngModel]="checked"
  [disabled]="disabled"
  [indeterminate]="indeterminate"
  (ngModelChange)="onCheckedChange($event)"
  >
  </tds-checkbox>
   
    <div class="tds-table-selection-extra inline" *ngIf="showRowSelection">
      <span tds-dropdown class="tds-table-selection-down" placement="bottomLeft" [tdsDropdownMenu]="selectionMenu">
          <i class="tdsi-arrow-down-fill"></i>
      </span>
      <tds-dropdown-menu #selectionMenu="tdsDropdownMenu">
        <div class="w-full" >
            <div  tds-dropdown-item *ngFor="let selection of listOfSelections" (click)="selection.onSelect()">
            {{ selection.text }}
            </div>            
        </div>
      </tds-dropdown-menu>    
    </div>
  `
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { listOfSelections: [{
            type: Input
        }], checked: [{
            type: Input
        }], disabled: [{
            type: Input
        }], indeterminate: [{
            type: Input
        }], showCheckbox: [{
            type: Input
        }], showRowSelection: [{
            type: Input
        }], checkedChange: [{
            type: Output
        }] }); })();
