// tds-ui/esm2020/table/src/addon/sorters.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
function TDSTableSortersComponent_ng_template_1_Template(rf, ctx) { }
function TDSTableSortersComponent_span_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 5);
    i0.ɵɵelement(1, "i", 6);
    i0.ɵɵelementEnd();
} }
function TDSTableSortersComponent_span_4_i_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 9);
} }
function TDSTableSortersComponent_span_4_i_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "i", 10);
} }
function TDSTableSortersComponent_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 5);
    i0.ɵɵtemplate(1, TDSTableSortersComponent_span_4_i_1_Template, 1, 0, "i", 7);
    i0.ɵɵtemplate(2, TDSTableSortersComponent_span_4_i_2_Template, 1, 0, "i", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.sortOrder === "ascend");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.sortOrder === "descend");
} }
export class TDSTableSortersComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this._sortDirections = ['ascend', 'descend', null];
        this.sortOrder = null;
        this.contentTemplate = null;
        this.isUp = false;
        this.isDown = false;
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-table-column-sorters');
    }
    ngOnChanges(changes) {
        const { _sortDirections } = changes;
        if (_sortDirections) {
            this.isUp = this._sortDirections.indexOf('ascend') !== -1;
            this.isDown = this._sortDirections.indexOf('descend') !== -1;
        }
    }
}
TDSTableSortersComponent.ɵfac = function TDSTableSortersComponent_Factory(t) { return new (t || TDSTableSortersComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSTableSortersComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTableSortersComponent, selectors: [["tds-table-sorters"]], hostAttrs: [1, "w-full", "flex", "items-center"], inputs: { _sortDirections: "_sortDirections", sortOrder: "sortOrder", contentTemplate: "contentTemplate" }, features: [i0.ɵɵNgOnChangesFeature], decls: 5, vars: 3, consts: [[1, "tds-table-column-title"], [3, "ngTemplateOutlet"], [1, "tds-table-column-sorter", "text-base", "ml-2"], ["class", "tds-table-column-sorter-inner flex flex-col p-0.5", 4, "ngIf"], ["class", "tds-table-column-sorter-inner flex flex-col  p-0.5", 4, "ngIf"], [1, "tds-table-column-sorter-inner", "flex", "flex-col", "p-0.5"], [1, "tds-table-column-sorter", "tdsi-sort-up-down-fill", "text-neutral-1-500", "active", "text-base", "leading-4", "hidden", "group-hover:inline-block"], ["class", "tds-table-column-sorter-up tdsi-sort-arrow-up-fill text-primary-1 active text-base leading-4", 4, "ngIf"], ["class", "tds-table-column-sorter-down tdsi-sort-arrow-down-fill text-primary-1 active text-base leading-4", 4, "ngIf"], [1, "tds-table-column-sorter-up", "tdsi-sort-arrow-up-fill", "text-primary-1", "active", "text-base", "leading-4"], [1, "tds-table-column-sorter-down", "tdsi-sort-arrow-down-fill", "text-primary-1", "active", "text-base", "leading-4"]], template: function TDSTableSortersComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "span", 0);
        i0.ɵɵtemplate(1, TDSTableSortersComponent_ng_template_1_Template, 0, 0, "ng-template", 1);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(2, "span", 2);
        i0.ɵɵtemplate(3, TDSTableSortersComponent_span_3_Template, 2, 0, "span", 3);
        i0.ɵɵtemplate(4, TDSTableSortersComponent_span_4_Template, 3, 2, "span", 4);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.contentTemplate);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.sortOrder !== "ascend" && ctx.sortOrder !== "descend");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.sortOrder === "ascend" || ctx.sortOrder === "descend");
    } }, directives: [i1.NgTemplateOutlet, i1.NgIf], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableSortersComponent, [{
        type: Component,
        args: [{
                selector: 'tds-table-sorters',
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
    <span class="tds-table-column-title">   
      <ng-template [ngTemplateOutlet]="contentTemplate"></ng-template>
    </span>
    <span class="tds-table-column-sorter  text-base ml-2">
      <span class="tds-table-column-sorter-inner flex flex-col p-0.5" *ngIf="sortOrder !== 'ascend' && sortOrder !== 'descend'">    
        <i  class="tds-table-column-sorter  tdsi-sort-up-down-fill text-neutral-1-500 active text-base leading-4 hidden group-hover:inline-block"
        ></i>
      </span>
      <span class="tds-table-column-sorter-inner flex flex-col  p-0.5" *ngIf="sortOrder === 'ascend' || sortOrder === 'descend'">
        <i         
          *ngIf="sortOrder === 'ascend'"
          class="tds-table-column-sorter-up tdsi-sort-arrow-up-fill text-primary-1 active text-base leading-4"
        ></i>
        <i         
          *ngIf="sortOrder === 'descend'"
          class="tds-table-column-sorter-down tdsi-sort-arrow-down-fill text-primary-1 active text-base leading-4"></i>
      </span>
    </span>
  `,
                host: {
                    class: "w-full flex  items-center "
                }
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { _sortDirections: [{
            type: Input
        }], sortOrder: [{
            type: Input
        }], contentTemplate: [{
            type: Input
        }] }); })();
