// tds-ui/esm2020/scroll-into-view/scroll-into-view.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSScrollIntoView, TDSScrollIntoViewComponent } from './scroll-into-view.component';
import * as i0 from "@angular/core";
export class TDSScrollIntoViewModule {
}
TDSScrollIntoViewModule.ɵfac = function TDSScrollIntoViewModule_Factory(t) { return new (t || TDSScrollIntoViewModule)(); };
TDSScrollIntoViewModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSScrollIntoViewModule });
TDSScrollIntoViewModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSScrollIntoViewModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSScrollIntoViewComponent,
                    TDSScrollIntoView
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    TDSScrollIntoViewComponent,
                    TDSScrollIntoView
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSScrollIntoViewModule, { declarations: [TDSScrollIntoViewComponent,
        TDSScrollIntoView], imports: [CommonModule], exports: [TDSScrollIntoViewComponent,
        TDSScrollIntoView] }); })();
