// tds-ui/esm2020/input-number/input-number.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSInputNumberComponent } from './input-number.component';
import * as i0 from "@angular/core";
export class TDSInputNumberModule {
}
TDSInputNumberModule.ɵfac = function TDSInputNumberModule_Factory(t) { return new (t || TDSInputNumberModule)(); };
TDSInputNumberModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSInputNumberModule });
TDSInputNumberModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSInputNumberModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, FormsModule],
                declarations: [TDSInputNumberComponent],
                exports: [TDSInputNumberComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSInputNumberModule, { declarations: [TDSInputNumberComponent], imports: [BidiModule, CommonModule, FormsModule], exports: [TDSInputNumberComponent] }); })();
