// tds-ui/esm2020/core/services/breakpoint.mjs
import { Injectable } from "@angular/core";
import { distinctUntilChanged, map, startWith } from "rxjs/operators";
import * as i0 from "@angular/core";
import * as i1 from "./resize";
import * as i2 from "@angular/cdk/layout";
export var TDSBreakpointEnum;
(function (TDSBreakpointEnum) {
    TDSBreakpointEnum["xxl"] = "xxl";
    TDSBreakpointEnum["xl"] = "xl";
    TDSBreakpointEnum["lg"] = "lg";
    TDSBreakpointEnum["md"] = "md";
    TDSBreakpointEnum["sm"] = "sm";
    TDSBreakpointEnum["xs"] = "xs";
})(TDSBreakpointEnum || (TDSBreakpointEnum = {}));
export const gridResponsiveMap = {
    xs: '(max-width: 575px)',
    sm: '(min-width: 576px)',
    md: '(min-width: 768px)',
    lg: '(min-width: 992px)',
    xl: '(min-width: 1200px)',
    xxl: '(min-width: 1600px)'
};
export const siderResponsiveMap = {
    xs: '(max-width: 479.98px)',
    sm: '(max-width: 575.98px)',
    md: '(max-width: 767.98px)',
    lg: '(max-width: 991.98px)',
    xl: '(max-width: 1199.98px)',
    xxl: '(max-width: 1599.98px)'
};
export class TDSBreakpointService {
    constructor(resizeService, mediaMatcher) {
        this.resizeService = resizeService;
        this.mediaMatcher = mediaMatcher;
        this.resizeService.subscribe().subscribe(() => { });
    }
    subscribe(breakpointMap, fullMap) {
        if (fullMap) {
            const get = () => this.matchMedia(breakpointMap, true);
            return this.resizeService.subscribe().pipe(map(get), startWith(get()), distinctUntilChanged((x, y) => x[0] === y[0]), map(x => x[1]));
        }
        else {
            const get = () => this.matchMedia(breakpointMap);
            return this.resizeService.subscribe().pipe(map(get), startWith(get()), distinctUntilChanged());
        }
    }
    matchMedia(breakpointMap, fullMap) {
        let bp = TDSBreakpointEnum.md;
        const breakpointBooleanMap = {};
        Object.keys(breakpointMap).map(breakpoint => {
            const castBP = breakpoint;
            const matched = this.mediaMatcher.matchMedia(gridResponsiveMap[castBP]).matches;
            breakpointBooleanMap[breakpoint] = matched;
            if (matched) {
                bp = castBP;
            }
        });
        if (fullMap) {
            return [bp, breakpointBooleanMap];
        }
        else {
            return bp;
        }
    }
}
TDSBreakpointService.ɵfac = function TDSBreakpointService_Factory(t) { return new (t || TDSBreakpointService)(i0.ɵɵinject(i1.TDSResizeService), i0.ɵɵinject(i2.MediaMatcher)); };
TDSBreakpointService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TDSBreakpointService, factory: TDSBreakpointService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSBreakpointService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.TDSResizeService }, { type: i2.MediaMatcher }]; }, null); })();
