// tds-ui/esm2020/skeleton/skeleton.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSSkeletonElementAvatarComponent, TDSSkeletonElementButtonComponent, TDSSkeletonElementDirective, TDSSkeletonElementImageComponent, TDSSkeletonElementInputComponent } from './skeleton-element.component';
import { TDSSkeletonComponent } from './skeleton.component';
import * as i0 from "@angular/core";
export class TDSSkeletonModule {
}
TDSSkeletonModule.ɵfac = function TDSSkeletonModule_Factory(t) { return new (t || TDSSkeletonModule)(); };
TDSSkeletonModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSSkeletonModule });
TDSSkeletonModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSkeletonModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSSkeletonComponent,
                    TDSSkeletonElementDirective,
                    TDSSkeletonElementButtonComponent,
                    TDSSkeletonElementAvatarComponent,
                    TDSSkeletonElementImageComponent,
                    TDSSkeletonElementInputComponent
                ],
                imports: [BidiModule, CommonModule],
                exports: [
                    TDSSkeletonComponent,
                    TDSSkeletonElementDirective,
                    TDSSkeletonElementButtonComponent,
                    TDSSkeletonElementAvatarComponent,
                    TDSSkeletonElementImageComponent,
                    TDSSkeletonElementInputComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSSkeletonModule, { declarations: [TDSSkeletonComponent,
        TDSSkeletonElementDirective,
        TDSSkeletonElementButtonComponent,
        TDSSkeletonElementAvatarComponent,
        TDSSkeletonElementImageComponent,
        TDSSkeletonElementInputComponent], imports: [BidiModule, CommonModule], exports: [TDSSkeletonComponent,
        TDSSkeletonElementDirective,
        TDSSkeletonElementButtonComponent,
        TDSSkeletonElementAvatarComponent,
        TDSSkeletonElementImageComponent,
        TDSSkeletonElementInputComponent] }); })();
