// tds-ui/esm2020/core/highlight/highlight.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSHighlightPipe } from './highlight.pipe';
import * as i0 from "@angular/core";
export class TDSHighlightModule {
}
TDSHighlightModule.ɵfac = function TDSHighlightModule_Factory(t) { return new (t || TDSHighlightModule)(); };
TDSHighlightModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSHighlightModule });
TDSHighlightModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSHighlightModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [TDSHighlightPipe],
                declarations: [TDSHighlightPipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSHighlightModule, { declarations: [TDSHighlightPipe], imports: [CommonModule], exports: [TDSHighlightPipe] }); })();
