// tds-ui/esm2020/image/image.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSPipesModule } from 'tds-ui/core/pipes';
import { TDSImageGroupComponent } from './image-group.component';
import { TDSImagePreviewComponent } from './image-preview.component';
import { TDSImageDirective } from './image.directive';
import { TDSImageService } from './image.service';
import * as i0 from "@angular/core";
export class TDSImageModule {
}
TDSImageModule.ɵfac = function TDSImageModule_Factory(t) { return new (t || TDSImageModule)(); };
TDSImageModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSImageModule });
TDSImageModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [TDSImageService], imports: [[BidiModule, OverlayModule, PortalModule, DragDropModule, CommonModule, TDSPipesModule, TDSButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSImageModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, OverlayModule, PortalModule, DragDropModule, CommonModule, TDSPipesModule, TDSButtonModule],
                exports: [TDSImageDirective, TDSImagePreviewComponent, TDSImageGroupComponent],
                providers: [TDSImageService],
                declarations: [TDSImageDirective, TDSImagePreviewComponent, TDSImageGroupComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSImageModule, { declarations: [TDSImageDirective, TDSImagePreviewComponent, TDSImageGroupComponent], imports: [BidiModule, OverlayModule, PortalModule, DragDropModule, CommonModule, TDSPipesModule, TDSButtonModule], exports: [TDSImageDirective, TDSImagePreviewComponent, TDSImageGroupComponent] }); })();
