// tds-ui/esm2020/image/image-preview.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, ViewChild, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { fadeMotion } from 'tds-ui/core/animation';
import { isNotNil } from 'tds-ui/core/util';
import { TDS_IMAGE_FADE_CLASS_NAME_MAP } from './image-config';
import { getClientSize, getFitContentPosition, getOffset } from './utils';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/config";
import * as i2 from "./image-preview-options";
import * as i3 from "@angular/cdk/overlay";
import * as i4 from "tds-ui/button";
import * as i5 from "@angular/cdk/drag-drop";
import * as i6 from "@angular/common";
const _c0 = ["imgRef"];
function TDSImagePreviewComponent_ng_container_8_img_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "img", 13, 14);
} if (rf & 2) {
    const image_r3 = i0.ɵɵnextContext().$implicit;
    const ctx_r5 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("width", image_r3.width)("height", image_r3.height)("transform", ctx_r5.previewImageTransform);
    i0.ɵɵattribute("src", image_r3.src, i0.ɵɵsanitizeUrl)("srcset", image_r3.srcset, i0.ɵɵsanitizeUrl)("alt", image_r3.alt);
} }
function TDSImagePreviewComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSImagePreviewComponent_ng_container_8_img_1_Template, 2, 9, "img", 12);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const imageIndex_r4 = ctx.index;
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.index === imageIndex_r4);
} }
function TDSImagePreviewComponent_li_10_Template(rf, ctx) { if (rf & 1) {
    const _r10 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "li", 15);
    i0.ɵɵlistener("click", function TDSImagePreviewComponent_li_10_Template_li_click_0_listener() { const restoredCtx = i0.ɵɵrestoreView(_r10); const option_r8 = restoredCtx.$implicit; return option_r8.onClick(); });
    i0.ɵɵelement(1, "span", 16);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r8 = ctx.$implicit;
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("tds-image-preview-operations-operation-disabled", ctx_r1.zoomOutDisabled && option_r8.type === "zoomOut");
    i0.ɵɵproperty("ngClass", ctx_r1.zoomOutDisabled && option_r8.type === "zoomOut" ? "text-neutral-1-900" : "text-white");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", option_r8.icon);
} }
function TDSImagePreviewComponent_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    const _r12 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "div", 17);
    i0.ɵɵlistener("click", function TDSImagePreviewComponent_ng_container_11_Template_div_click_1_listener($event) { i0.ɵɵrestoreView(_r12); const ctx_r11 = i0.ɵɵnextContext(); return ctx_r11.onSwitchLeft($event); });
    i0.ɵɵelement(2, "span", 18);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(3, "div", 19);
    i0.ɵɵlistener("click", function TDSImagePreviewComponent_ng_container_11_Template_div_click_3_listener($event) { i0.ɵɵrestoreView(_r12); const ctx_r13 = i0.ɵɵnextContext(); return ctx_r13.onSwitchRight($event); });
    i0.ɵɵelement(4, "span", 20);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("tds-image-preview-switch-left-disabled", ctx_r2.index <= 0);
    i0.ɵɵproperty("ngClass", ctx_r2.index <= 0 ? "text-neutral-1-900" : "text-white");
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("tds-image-preview-switch-right-disabled", ctx_r2.index >= ctx_r2.images.length - 1);
    i0.ɵɵproperty("ngClass", ctx_r2.index >= ctx_r2.images.length - 1 ? "text-neutral-1-900" : "text-white");
} }
const TDS_CONFIG_MODULE_NAME = 'image';
const initialPosition = {
    x: 0,
    y: 0
};
export class TDSImagePreviewComponent {
    constructor(cdr, tdsConfigService, config, overlayRef) {
        this.cdr = cdr;
        this.tdsConfigService = tdsConfigService;
        this.config = config;
        this.overlayRef = overlayRef;
        this.images = [];
        this.index = 0;
        this.isDragging = false;
        this.visible = true;
        this.animationState = 'enter';
        this.animationStateChanged = new EventEmitter();
        this.previewImageTransform = '';
        this.previewImageWrapperTransform = '';
        this.operations = [
            {
                icon: 'tdsi-rotate-left-line',
                onClick: () => {
                    this.onRotateLeft();
                },
                type: 'rotateLeft'
            },
            {
                icon: 'tdsi-rotate-right-line',
                onClick: () => {
                    this.onRotateRight();
                },
                type: 'rotateRight'
            },
            {
                icon: 'tdsi-zoom-out-line',
                onClick: () => {
                    this.onZoomOut();
                },
                type: 'zoomOut'
            },
            {
                icon: 'tdsi-zoom-in-line',
                onClick: () => {
                    this.onZoomIn();
                },
                type: 'zoomIn'
            },
        ];
        this.zoomOutDisabled = false;
        this.position = { ...initialPosition };
        this.containerClick = new EventEmitter();
        this.closeClick = new EventEmitter();
        this.destroy$ = new Subject();
        // TODO: move to host after View Engine deprecation
        this.zoom = this.config.zoom ?? 1;
        this.rotate = this.config.rotate ?? 0;
        this.updateZoomOutDisabled();
        this.updatePreviewImageTransform();
        this.updatePreviewImageWrapperTransform();
    }
    get animationDisabled() {
        return this.config.noAnimation ?? false;
    }
    get maskClosable() {
        const defaultConfig = this.tdsConfigService.getConfigForComponent(TDS_CONFIG_MODULE_NAME) || {};
        return this.config.maskClosable ?? defaultConfig.maskClosable ?? true;
    }
    setImages(images) {
        this.images = images;
        this.cdr.markForCheck();
    }
    switchTo(index) {
        this.index = index;
        this.cdr.markForCheck();
    }
    next() {
        if (this.index < this.images.length - 1) {
            this.reset();
            this.index++;
            this.updatePreviewImageTransform();
            this.updatePreviewImageWrapperTransform();
            this.updateZoomOutDisabled();
            this.cdr.markForCheck();
        }
    }
    prev() {
        if (this.index > 0) {
            this.reset();
            this.index--;
            this.updatePreviewImageTransform();
            this.updatePreviewImageWrapperTransform();
            this.updateZoomOutDisabled();
            this.cdr.markForCheck();
        }
    }
    markForCheck() {
        this.cdr.markForCheck();
    }
    onClose() {
        this.closeClick.emit();
    }
    onZoomIn() {
        this.zoom += 1;
        this.updatePreviewImageTransform();
        this.updateZoomOutDisabled();
        this.position = { ...initialPosition };
    }
    onZoomOut() {
        if (this.zoom > 1) {
            this.zoom -= 1;
            this.updatePreviewImageTransform();
            this.updateZoomOutDisabled();
            this.position = { ...initialPosition };
        }
    }
    onRotateRight() {
        this.rotate += 90;
        this.updatePreviewImageTransform();
    }
    onRotateLeft() {
        this.rotate -= 90;
        this.updatePreviewImageTransform();
    }
    onSwitchLeft(event) {
        event.preventDefault();
        event.stopPropagation();
        this.prev();
    }
    onSwitchRight(event) {
        event.preventDefault();
        event.stopPropagation();
        this.next();
    }
    onContainerClick(e) {
        if (e.target === e.currentTarget && this.maskClosable) {
            this.containerClick.emit();
        }
    }
    onAnimationStart(event) {
        if (event.toState === 'enter') {
            this.setEnterAnimationClass();
        }
        else if (event.toState === 'leave') {
            this.setLeaveAnimationClass();
        }
        this.animationStateChanged.emit(event);
    }
    onAnimationDone(event) {
        if (event.toState === 'enter') {
            this.setEnterAnimationClass();
        }
        else if (event.toState === 'leave') {
            this.setLeaveAnimationClass();
        }
        this.animationStateChanged.emit(event);
    }
    startLeaveAnimation() {
        this.animationState = 'leave';
        this.cdr.markForCheck();
    }
    onDragStarted() {
        this.isDragging = true;
    }
    onDragReleased() {
        this.isDragging = false;
        const width = this.imageRef.nativeElement.offsetWidth * this.zoom;
        const height = this.imageRef.nativeElement.offsetHeight * this.zoom;
        const { left, top } = getOffset(this.imageRef.nativeElement);
        const { width: clientWidth, height: clientHeight } = getClientSize();
        const isRotate = this.rotate % 180 !== 0;
        const fitContentParams = {
            width: isRotate ? height : width,
            height: isRotate ? width : height,
            left,
            top,
            clientWidth,
            clientHeight
        };
        const fitContentPos = getFitContentPosition(fitContentParams);
        if (isNotNil(fitContentPos.x) || isNotNil(fitContentPos.y)) {
            this.position = { ...this.position, ...fitContentPos };
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    updatePreviewImageTransform() {
        this.previewImageTransform = `scale3d(${this.zoom}, ${this.zoom}, 1) rotate(${this.rotate}deg)`;
    }
    updatePreviewImageWrapperTransform() {
        this.previewImageWrapperTransform = `translate3d(${this.position.x}px, ${this.position.y}px, 0)`;
    }
    updateZoomOutDisabled() {
        this.zoomOutDisabled = this.zoom <= 1;
    }
    setEnterAnimationClass() {
        if (this.animationDisabled) {
            return;
        }
        const backdropElement = this.overlayRef.backdropElement;
        if (backdropElement) {
            backdropElement.classList.add(TDS_IMAGE_FADE_CLASS_NAME_MAP.enter);
            backdropElement.classList.add(TDS_IMAGE_FADE_CLASS_NAME_MAP.enterActive);
        }
    }
    setLeaveAnimationClass() {
        if (this.animationDisabled) {
            return;
        }
        const backdropElement = this.overlayRef.backdropElement;
        if (backdropElement) {
            backdropElement.classList.add(TDS_IMAGE_FADE_CLASS_NAME_MAP.leave);
            backdropElement.classList.add(TDS_IMAGE_FADE_CLASS_NAME_MAP.leaveActive);
        }
    }
    reset() {
        this.zoom = 1;
        this.rotate = 0;
        this.position = { ...initialPosition };
    }
}
TDSImagePreviewComponent.ɵfac = function TDSImagePreviewComponent_Factory(t) { return new (t || TDSImagePreviewComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSConfigService), i0.ɵɵdirectiveInject(i2.TDSImagePreviewOptions), i0.ɵɵdirectiveInject(i3.OverlayRef)); };
TDSImagePreviewComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSImagePreviewComponent, selectors: [["tds-image-preview"]], viewQuery: function TDSImagePreviewComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.imageRef = _t.first);
    } }, hostAttrs: ["tabindex", "-1", "role", "document", 1, "fixed", "inset-0", "overflow-auto", "bg-black", "bg-opacity-40"], hostVars: 8, hostBindings: function TDSImagePreviewComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵsyntheticHostListener("@fadeMotion.start", function TDSImagePreviewComponent_animation_fadeMotion_start_HostBindingHandler($event) { return ctx.onAnimationStart($event); })("@fadeMotion.done", function TDSImagePreviewComponent_animation_fadeMotion_done_HostBindingHandler($event) { return ctx.onAnimationDone($event); });
        i0.ɵɵlistener("click", function TDSImagePreviewComponent_click_HostBindingHandler($event) { return ctx.onContainerClick($event); });
    } if (rf & 2) {
        i0.ɵɵsyntheticHostProperty("@.disabled", ctx.config.noAnimation)("@fadeMotion", ctx.animationState);
        i0.ɵɵstyleProp("z-index", ctx.config.zIndex);
        i0.ɵɵclassProp("tds-image-preview-moving", ctx.isDragging)("tds-image-preview-wrap", true);
    } }, exportAs: ["tdsImagePreview"], decls: 13, vars: 6, consts: [[1, "tds-image-preview", "pointer-events-none", "h-full", "text-center", "m-0", "p-0"], ["tabindex", "0", "aria-hidden", "true", 2, "width", "0", "height", "0", "overflow", "hidden", "outline", "none"], [1, "tds-image-preview-content", "m-0", "p-0", "w-full", "h-full"], [1, "tds-image-preview-body", "absolute", "inset-0", "overflow-hidden"], [1, "tds-image-preview-operations", "box-border", "absolute", "m-0", "p-0", "top-6", "right-6", "pointer-events-auto", 2, "z-index", "1"], [1, "tds-image-preview-operations-operation", "flex", "items-center", "justify-center", "cursor-pointer", 3, "click"], ["tds-button-close", "", 1, "tds-image-preview-operations-icon", "text-white", "dark:text-white"], ["cdkDrag", "", 1, "tds-image-preview-img-wrapper", "absolute", "inset-0", "flex", "items-center", "justify-center", 3, "cdkDragFreeDragPosition", "mousedown", "cdkDragReleased"], [4, "ngFor", "ngForOf"], [1, "tds-image-preview-operations", "pointer-events-auto", "box-border", "absolute", "m-0", "cursor-pointer", "bottom-0", "right-0", "flex", "justify-center", "items-center", "w-full", "bg-black", "bg-opacity-40", "py-3", 2, "z-index", "1"], ["class", "tds-image-preview-operations-operation mr-5  cursor-pointer  flex ", 3, "ngClass", "tds-image-preview-operations-operation-disabled", "click", 4, "ngFor", "ngForOf"], [4, "ngIf"], ["cdkDragHandle", "", "class", "tds-image-preview-img text-center pointer-events-auto", 3, "width", "height", "transform", 4, "ngIf"], ["cdkDragHandle", "", 1, "tds-image-preview-img", "text-center", "pointer-events-auto"], ["imgRef", ""], [1, "tds-image-preview-operations-operation", "mr-5", "cursor-pointer", "flex", 3, "ngClass", "click"], [1, "tds-image-preview-operations-icon", "text-xl", "leading-none", 3, "ngClass"], [1, "tds-image-preview-switch-left", "pointer-events-auto", "-mt-5", "cursor-pointer", "rounded-full", "bg-black", "bg-opacity-40", "text-white", "left-5", "absolute", "top-1/2", "flex", "items-censter", "z-10", "p-5", 3, "ngClass", "click"], [1, "tdsi-arrow-left-line", "text-xl", "leading-none"], [1, "tds-image-preview-switch-right", "rounded-full", "-mt-5", "pointer-events-auto", "cursor-pointer", "bg-black", "bg-opacity-40", "text-white", "absolute", "top-1/2", "right-5", "flex", "items-censter", "z-10", "p-5", 3, "ngClass", "click"], [1, "tdsi-arrow-right-line", "text-xl", "leading-none"]], template: function TDSImagePreviewComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵelement(1, "div", 1);
        i0.ɵɵelementStart(2, "div", 2)(3, "div", 3)(4, "div", 4)(5, "div", 5);
        i0.ɵɵlistener("click", function TDSImagePreviewComponent_Template_div_click_5_listener() { return ctx.onClose(); });
        i0.ɵɵelement(6, "button", 6);
        i0.ɵɵelementEnd()();
        i0.ɵɵelementStart(7, "div", 7);
        i0.ɵɵlistener("mousedown", function TDSImagePreviewComponent_Template_div_mousedown_7_listener() { return ctx.onDragStarted(); })("cdkDragReleased", function TDSImagePreviewComponent_Template_div_cdkDragReleased_7_listener() { return ctx.onDragReleased(); });
        i0.ɵɵtemplate(8, TDSImagePreviewComponent_ng_container_8_Template, 2, 1, "ng-container", 8);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(9, "ul", 9);
        i0.ɵɵtemplate(10, TDSImagePreviewComponent_li_10_Template, 2, 4, "li", 10);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(11, TDSImagePreviewComponent_ng_container_11_Template, 5, 6, "ng-container", 11);
        i0.ɵɵelementEnd()();
        i0.ɵɵelement(12, "div", 1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(7);
        i0.ɵɵstyleProp("transform", ctx.previewImageWrapperTransform);
        i0.ɵɵproperty("cdkDragFreeDragPosition", ctx.position);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngForOf", ctx.images);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngForOf", ctx.operations);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.images.length > 1);
    } }, directives: [i4.TDSButtonCloseComponent, i5.CdkDrag, i6.NgForOf, i6.NgIf, i5.CdkDragHandle, i6.NgClass], encapsulation: 2, data: { animation: [fadeMotion] }, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSImagePreviewComponent, [{
        type: Component,
        args: [{
                selector: 'tds-image-preview',
                exportAs: 'tdsImagePreview',
                animations: [fadeMotion],
                template: `
    <div class="tds-image-preview pointer-events-none h-full	text-center m-0 p-0">
      <div tabindex="0" aria-hidden="true" style="width: 0; height: 0; overflow: hidden; outline: none;"></div>
      <div class="tds-image-preview-content m-0 p-0 w-full h-full">
        <div class="tds-image-preview-body absolute inset-0 overflow-hidden">
          <div class="tds-image-preview-operations box-border absolute m-0 p-0 top-6 right-6  pointer-events-auto" style="zIndex:1">
            <div
              class="tds-image-preview-operations-operation flex items-center justify-center cursor-pointer"             
              (click)="onClose()"
            >
            <button tds-button-close class="tds-image-preview-operations-icon text-white dark:text-white"></button>
             </div>
          </div>
          <div
            class="tds-image-preview-img-wrapper absolute inset-0 flex items-center justify-center"
            cdkDrag
            [style.transform]="previewImageWrapperTransform"
            [cdkDragFreeDragPosition]="position"
            (mousedown)="onDragStarted()"
            (cdkDragReleased)="onDragReleased()"
          >
            <ng-container *ngFor="let image of images; index as imageIndex">
              <img
                cdkDragHandle
                class="tds-image-preview-img text-center pointer-events-auto"
                #imgRef
                *ngIf="index === imageIndex"
                [attr.src]="image.src"
                [attr.srcset]="image.srcset"
                [attr.alt]="image.alt"
                [style.width]="image.width"
                [style.height]="image.height"
                [style.transform]="previewImageTransform"
              />
            </ng-container>
          </div>
          <ul class="tds-image-preview-operations pointer-events-auto box-border absolute m-0  cursor-pointer bottom-0 right-0 flex justify-center items-center w-full bg-black bg-opacity-40 py-3" style="zIndex:1">
            <li
              class="tds-image-preview-operations-operation mr-5  cursor-pointer  flex "
              [ngClass]="zoomOutDisabled && option.type === 'zoomOut' ?'text-neutral-1-900':'text-white'"
              [class.tds-image-preview-operations-operation-disabled]="zoomOutDisabled && option.type === 'zoomOut'"
              (click)="option.onClick()"
              *ngFor="let option of operations"
            >
              <span class="tds-image-preview-operations-icon text-xl leading-none " [ngClass]="option.icon"></span>
            </li>
          </ul>
          <ng-container *ngIf="images.length > 1">
            <div
              class="tds-image-preview-switch-left pointer-events-auto -mt-5  cursor-pointer rounded-full bg-black bg-opacity-40  text-white left-5 absolute top-1/2  flex items-censter  z-10 p-5"
              [class.tds-image-preview-switch-left-disabled]="index <= 0"
              [ngClass]="index <= 0 ?'text-neutral-1-900':'text-white'"
              (click)="onSwitchLeft($event)"
            >
              <span class="tdsi-arrow-left-line text-xl leading-none"></span>
            </div>
            <div
              class="tds-image-preview-switch-right rounded-full -mt-5 pointer-events-auto cursor-pointer bg-black bg-opacity-40 text-white absolute top-1/2 right-5 flex items-censter z-10 p-5"
              [class.tds-image-preview-switch-right-disabled]="index >= images.length - 1"
              [ngClass]="index >= (images.length - 1) ?'text-neutral-1-900':'text-white'"
              (click)="onSwitchRight($event)"
            >
              <span class="tdsi-arrow-right-line text-xl leading-none"></span>
            </div>
          </ng-container>
        </div>
      </div>
      <div tabindex="0" aria-hidden="true" style="width: 0; height: 0; overflow: hidden; outline: none;"></div>
    </div>
  `,
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.tds-image-preview-moving]': 'isDragging',
                    '[style.zIndex]': 'config.zIndex',
                    '[class.tds-image-preview-wrap]': 'true',
                    '[@.disabled]': 'config.noAnimation',
                    '[@fadeMotion]': 'animationState',
                    '(@fadeMotion.start)': 'onAnimationStart($event)',
                    '(@fadeMotion.done)': 'onAnimationDone($event)',
                    '(click)': 'onContainerClick($event)',
                    'class': "fixed inset-0 overflow-auto bg-black bg-opacity-40",
                    tabindex: '-1',
                    role: 'document'
                }
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.TDSConfigService }, { type: i2.TDSImagePreviewOptions }, { type: i3.OverlayRef }]; }, { imageRef: [{
            type: ViewChild,
            args: ['imgRef']
        }] }); })();
