// tds-ui/esm2020/image/image.service.mjs
import { OverlayConfig, OverlayRef } from '@angular/cdk/overlay';
import { ComponentPortal } from '@angular/cdk/portal';
import { Injectable, Injector, Optional } from '@angular/core';
import { TDS_IMAGE_PREVIEW_MASK_CLASS_NAME } from './image-config';
import { TDSImagePreviewOptions } from './image-preview-options';
import { TDSImagePreviewRef } from './image-preview-ref';
import { TDSImagePreviewComponent } from './image-preview.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
import * as i2 from "tds-ui/core/config";
import * as i3 from "@angular/cdk/bidi";
const TDS_CONFIG_MODULE_NAME = 'image';
export class TDSImageService {
    constructor(overlay, injector, tdsConfigService, directionality) {
        this.overlay = overlay;
        this.injector = injector;
        this.tdsConfigService = tdsConfigService;
        this.directionality = directionality;
    }
    preview(images, options) {
        return this.display(images, options);
    }
    display(images, config) {
        const configMerged = { ...new TDSImagePreviewOptions(), ...(config ?? {}) };
        const overlayRef = this.createOverlay(configMerged);
        const previewComponent = this.attachPreviewComponent(overlayRef, configMerged);
        previewComponent.setImages(images);
        const previewRef = new TDSImagePreviewRef(previewComponent, configMerged, overlayRef);
        previewComponent.previewRef = previewRef;
        return previewRef;
    }
    attachPreviewComponent(overlayRef, config) {
        const injector = Injector.create({
            parent: this.injector,
            providers: [
                { provide: OverlayRef, useValue: overlayRef },
                { provide: TDSImagePreviewOptions, useValue: config }
            ]
        });
        const containerPortal = new ComponentPortal(TDSImagePreviewComponent, null, injector);
        const containerRef = overlayRef.attach(containerPortal);
        return containerRef.instance;
    }
    createOverlay(config) {
        const globalConfig = this.tdsConfigService.getConfigForComponent(TDS_CONFIG_MODULE_NAME) || {};
        const overLayConfig = new OverlayConfig({
            hasBackdrop: true,
            scrollStrategy: this.overlay.scrollStrategies.block(),
            positionStrategy: this.overlay.position().global(),
            disposeOnNavigation: config.closeOnNavigation ?? globalConfig.closeOnNavigation ?? true,
            backdropClass: TDS_IMAGE_PREVIEW_MASK_CLASS_NAME,
            direction: config.direction || globalConfig.direction || this.directionality.value
        });
        return this.overlay.create(overLayConfig);
    }
}
TDSImageService.ɵfac = function TDSImageService_Factory(t) { return new (t || TDSImageService)(i0.ɵɵinject(i1.Overlay), i0.ɵɵinject(i0.Injector), i0.ɵɵinject(i2.TDSConfigService), i0.ɵɵinject(i3.Directionality, 8)); };
TDSImageService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TDSImageService, factory: TDSImageService.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSImageService, [{
        type: Injectable
    }], function () { return [{ type: i1.Overlay }, { type: i0.Injector }, { type: i2.TDSConfigService }, { type: i3.Directionality, decorators: [{
                type: Optional
            }] }]; }, null); })();
