// tds-ui/esm2020/image/image-group.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSImageGroupComponent {
    constructor() {
        this.images = [];
    }
    addImage(image) {
        this.images.push(image);
    }
}
TDSImageGroupComponent.ɵfac = function TDSImageGroupComponent_Factory(t) { return new (t || TDSImageGroupComponent)(); };
TDSImageGroupComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSImageGroupComponent, selectors: [["tds-image-group"]], exportAs: ["tdsImageGroup"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSImageGroupComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSImageGroupComponent, [{
        type: Component,
        args: [{
                selector: 'tds-image-group',
                exportAs: 'tdsImageGroup',
                template: '<ng-content></ng-content>',
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }]
    }], null, null); })();
