// tds-ui/esm2020/radio/radio.service.mjs
import { Injectable } from '@angular/core';
import { ReplaySubject, Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class TDSRadioService {
    constructor() {
        this.selected$ = new ReplaySubject(1);
        this.touched$ = new Subject();
        this.disabled$ = new ReplaySubject(1);
        this.name$ = new ReplaySubject(1);
        this.size$ = new ReplaySubject(1);
    }
    touch() {
        this.touched$.next();
    }
    select(value) {
        this.selected$.next(value);
    }
    setDisabled(value) {
        this.disabled$.next(value);
    }
    setName(value) {
        this.name$.next(value);
    }
    setSize(value) {
        this.size$.next(value);
    }
}
TDSRadioService.ɵfac = function TDSRadioService_Factory(t) { return new (t || TDSRadioService)(); };
TDSRadioService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TDSRadioService, factory: TDSRadioService.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRadioService, [{
        type: Injectable
    }], null, null); })();
