// tds-ui/esm2020/radio/radio.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSRadioButtonDirective } from './radio-button.directive';
import { TDSRadioGroupComponent } from './radio-group.component';
import { TDSRadioComponent } from './radio.component';
import * as i0 from "@angular/core";
export class TDSRadioModule {
}
TDSRadioModule.ɵfac = function TDSRadioModule_Factory(t) { return new (t || TDSRadioModule)(); };
TDSRadioModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSRadioModule });
TDSRadioModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRadioModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, FormsModule],
                exports: [TDSRadioComponent, TDSRadioButtonDirective, TDSRadioGroupComponent],
                declarations: [TDSRadioComponent, TDSRadioButtonDirective, TDSRadioGroupComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSRadioModule, { declarations: [TDSRadioComponent, TDSRadioButtonDirective, TDSRadioGroupComponent], imports: [BidiModule, CommonModule, FormsModule], exports: [TDSRadioComponent, TDSRadioButtonDirective, TDSRadioGroupComponent] }); })();
