// tds-ui/esm2020/menu/tds-menu/tds-menu.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChild, EventEmitter, Input, Output, ViewChildren, ViewEncapsulation } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { InputBoolean, TDSHelperArray, TDSHelperString } from 'tds-ui/shared/utility';
import { TDSMenuService } from '../services/menu.service';
import { TDSMenuGroupInlineComponent } from '../tds-menu-group-inline/tds-menu-group-inline.component';
import { TDSMenuGroupPopupComponent } from '../tds-menu-group-popup/tds-menu-group-popup.component';
import * as i0 from "@angular/core";
import * as i1 from "../services/menu.service";
import * as i2 from "@angular/common";
import * as i3 from "@angular/cdk/scrolling";
import * as i4 from "../tds-menu-group-popup/tds-menu-group-popup.component";
import * as i5 from "../tds-menu-item/tds-menu-item.component";
import * as i6 from "../tds-menu-group-inline/tds-menu-group-inline.component";
const _c0 = ["logo-text"];
function TDSMenuComponent_div_1_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 8);
    i0.ɵɵprojection(1);
    i0.ɵɵelementEnd();
} }
function TDSMenuComponent_div_1_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 8);
    i0.ɵɵprojection(1, 1);
    i0.ɵɵelementEnd();
} }
function TDSMenuComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtemplate(1, TDSMenuComponent_div_1_div_1_Template, 2, 0, "div", 6);
    i0.ɵɵtemplate(2, TDSMenuComponent_div_1_div_2_Template, 2, 0, "div", 7);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r0.options.background);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r0.inlineCollapsed);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.inlineCollapsed);
} }
function TDSMenuComponent_ng_container_3_ng_template_1_tds_menu_group_popup_0_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-menu-item", 12);
} if (rf & 2) {
    const children_r10 = ctx.$implicit;
    const ctx_r9 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("options", ctx_r9.options)("item", children_r10)("isSelected", children_r10.isSelected)("parentIsGroup", true);
} }
function TDSMenuComponent_ng_container_3_ng_template_1_tds_menu_group_popup_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "tds-menu-group-popup", 11);
    i0.ɵɵtemplate(1, TDSMenuComponent_ng_container_3_ng_template_1_tds_menu_group_popup_0_ng_template_1_Template, 1, 4, "ng-template", 9);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r7 = i0.ɵɵnextContext().$implicit;
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("options", ctx_r8.options)("item", item_r7);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", item_r7.listChild);
} }
function TDSMenuComponent_ng_container_3_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSMenuComponent_ng_container_3_ng_template_1_tds_menu_group_popup_0_Template, 2, 3, "tds-menu-group-popup", 10);
} if (rf & 2) {
    const item_r7 = ctx.$implicit;
    i0.ɵɵproperty("ngIf", !item_r7.hidden);
} }
function TDSMenuComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSMenuComponent_ng_container_3_ng_template_1_Template, 1, 1, "ng-template", 9);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", ctx_r1.listMenu);
} }
function TDSMenuComponent_ng_container_4_ng_template_1_tds_menu_group_inline_0_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-menu-item", 15);
} if (rf & 2) {
    const children_r16 = ctx.$implicit;
    const ctx_r15 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("options", ctx_r15.options)("showIcon", ctx_r15.hasIcon)("item", children_r16)("isSelected", children_r16.isSelected)("parentIsGroup", true);
} }
function TDSMenuComponent_ng_container_4_ng_template_1_tds_menu_group_inline_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "tds-menu-group-inline", 14);
    i0.ɵɵtemplate(1, TDSMenuComponent_ng_container_4_ng_template_1_tds_menu_group_inline_0_ng_template_1_Template, 1, 5, "ng-template", 9);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r13 = i0.ɵɵnextContext().$implicit;
    const ctx_r14 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("options", ctx_r14.options)("showIcon", ctx_r14.hasIcon)("item", item_r13);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", item_r13.listChild);
} }
function TDSMenuComponent_ng_container_4_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSMenuComponent_ng_container_4_ng_template_1_tds_menu_group_inline_0_Template, 2, 4, "tds-menu-group-inline", 13);
} if (rf & 2) {
    const item_r13 = ctx.$implicit;
    i0.ɵɵproperty("ngIf", !item_r13.hidden);
} }
function TDSMenuComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSMenuComponent_ng_container_4_ng_template_1_Template, 1, 1, "ng-template", 9);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", ctx_r2.listMenu);
} }
const _c1 = function (a0) { return { "flex-col": a0 }; };
const _c2 = function (a0) { return { "mt-6": a0 }; };
const _c3 = function (a0, a1) { return { "tdsi-arrow-right-line": a0, "tdsi-arrow-left-line": a1 }; };
function TDSMenuComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    const _r19 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 5)(1, "div", 16)(2, "div");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(3, "svg", 17);
    i0.ɵɵelement(4, "path", 18)(5, "path", 19)(6, "path", 20)(7, "path", 21);
    i0.ɵɵelementEnd()();
    i0.ɵɵnamespaceHTML();
    i0.ɵɵelementStart(8, "div", 22);
    i0.ɵɵlistener("click", function TDSMenuComponent_div_5_Template_div_click_8_listener() { i0.ɵɵrestoreView(_r19); const ctx_r18 = i0.ɵɵnextContext(); return ctx_r18.onClickInlineCollapsed(); });
    i0.ɵɵelement(9, "span", 23);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r3.options.background);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(5, _c1, ctx_r3.inlineCollapsed));
    i0.ɵɵadvance(7);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(7, _c2, ctx_r3.inlineCollapsed));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c3, ctx_r3.inlineCollapsed, !ctx_r3.inlineCollapsed));
} }
const _c4 = [[["", "logo-text", ""]], [["", "logo", ""]]];
const _c5 = function (a0) { return { " overflow-x-hidden": a0 }; };
const _c6 = ["[logo-text]", "[logo]"];
export function TDSMenuServiceFactory(serviceInsideDropDown, serviceOutsideDropDown) {
    return serviceInsideDropDown ? serviceInsideDropDown : serviceOutsideDropDown;
}
export class TDSMenuComponent {
    constructor(_cdr, tdsMenuService) {
        this._cdr = _cdr;
        this.tdsMenuService = tdsMenuService;
        this.destroy$ = new Subject();
        this.listMenu = [];
        this.showLogo = false;
        this.data = [];
        this.mode = 'dark';
        this.showIcon = true;
        this.inlineCollapsed = false;
        this.options = {
            background: 'bg-white dark:bg-d-neutral-3-200',
            backgroundItem: 'bg-white dark:bg-d-neutral-3-200',
            backgroundItemSelected: 'bg-neutral-3-50 dark:bg-d-neutral-3-300',
            backgroundItemHover: 'dark:hover:bg-d-neutral-3-300  hover:bg-neutral-3-50'
        };
        this.onClickItem = new EventEmitter();
        this.onOpenChange = new EventEmitter();
        this.inlineCollapsed$ = new BehaviorSubject(this.inlineCollapsed);
    }
    ngOnInit() {
        this.tdsMenuService.descendantMenuItemClick$.pipe(takeUntil(this.destroy$)).subscribe(menu => {
            this.onClickItem.emit(menu.item);
            if (TDSHelperArray.hasListValue(this.listOfTDSMenuGroupInlineComponent)) {
                this.listOfTDSMenuGroupInlineComponent.forEach(group => {
                    group.setSelectedStateListChildren(menu.uid);
                });
            }
            if (TDSHelperArray.hasListValue(this.listOfTDSMenuGroupPopupComponent)) {
                this.listOfTDSMenuGroupPopupComponent.forEach(group => {
                    group.setSelectedStateListChildren(menu.uid);
                });
            }
        });
        this.tdsMenuService.groupMenuOpen$.pipe(takeUntil(this.destroy$)).subscribe(menu => {
            if (TDSHelperArray.hasListValue(this.listOfTDSMenuGroupInlineComponent)) {
                this.listOfTDSMenuGroupInlineComponent.forEach(group => {
                    if (group.hasListChild)
                        group.setOpenStateListChildren(group == menu && !group.item.isOpen);
                });
            }
        });
    }
    ngOnChanges(changes) {
        const { data, inlineCollapsed, mode } = changes;
        if (data) {
            this.p_mapData(this.data);
        }
        if (inlineCollapsed) {
            this.inlineCollapsed$.next(this.inlineCollapsed);
        }
        if (mode) {
            this.tdsMenuService.onModeChange(this.mode);
        }
        this._cdr.markForCheck();
    }
    p_mapData(data) {
        if (!TDSHelperArray.hasListValue(data)) {
            this.listMenu = [];
        }
        else {
            this.listMenu = data.map(item => {
                return this.p_mapItem(item);
            });
        }
    }
    p_mapItem(data) {
        let listChild = [];
        if (TDSHelperArray.hasListValue(data.listChild)) {
            listChild = data.listChild.map((item) => {
                return this.p_mapItem(item);
            });
        }
        return {
            name: data.name,
            icon: data.icon,
            disabled: data.disabled,
            isSelected: data.isSelected,
            isOpen: data.isOpen,
            link: data.link,
            listChild: listChild,
            uid: TDSHelperString.genid(),
            hidden: data.hidden,
            groupTitle: data.groupTitle,
            htmlIcon: data.htmlIcon,
            linkProps: data.linkProps,
            badge: data.badge,
            tag: data.tag,
            dot: data.dot,
        };
    }
    ngAfterViewInit() {
        this.inlineCollapsed$.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.updateInlineCollapse();
            this._cdr.markForCheck();
        });
    }
    updateInlineCollapse() {
        if (this.inlineCollapsed) {
            if (this.listOfTDSMenuGroupInlineComponent.length) {
                this.listOfTDSMenuGroupInlineComponent.forEach(group => {
                    group.setOpenStateListChildren(false);
                });
            }
        }
        else {
            if (this.listOfTDSMenuGroupPopupComponent.length) {
                this.listOfTDSMenuGroupPopupComponent.forEach(group => {
                    group.setOpenStateListChildren(false);
                });
            }
        }
    }
    setInlineCollapsed(inlineCollapsed) {
        this.inlineCollapsed = inlineCollapsed;
        this.inlineCollapsed$.next(inlineCollapsed);
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    onClickInlineCollapsed() {
        this.inlineCollapsed = !this.inlineCollapsed;
        this.onOpenChange.emit(this.inlineCollapsed);
        this.inlineCollapsed$.next(this.inlineCollapsed);
    }
    get hasIcon() {
        if (!this.showIcon) {
            return false;
        }
        const lstIcon = this.listMenu.filter(f => TDSHelperString.hasValueString(f.icon));
        return lstIcon.length > 0;
    }
}
TDSMenuComponent.ɵfac = function TDSMenuComponent_Factory(t) { return new (t || TDSMenuComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSMenuService)); };
TDSMenuComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSMenuComponent, selectors: [["tds-menu"]], contentQueries: function TDSMenuComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, _c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.logoText = _t.first);
    } }, viewQuery: function TDSMenuComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(TDSMenuGroupInlineComponent, 5);
        i0.ɵɵviewQuery(TDSMenuGroupPopupComponent, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.listOfTDSMenuGroupInlineComponent = _t);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.listOfTDSMenuGroupPopupComponent = _t);
    } }, hostAttrs: [1, "h-full", "flex", "flex-col", "box-border"], hostVars: 4, hostBindings: function TDSMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("w-full", !ctx.inlineCollapsed)("w-16", ctx.inlineCollapsed);
    } }, inputs: { showLogo: "showLogo", data: "data", mode: "mode", showIcon: "showIcon", inlineCollapsed: "inlineCollapsed", options: "options" }, outputs: { onClickItem: "onClickItem", onOpenChange: "onOpenChange" }, features: [i0.ɵɵProvidersFeature([
            /** use the top level service **/
            TDSMenuService
        ]), i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c6, decls: 6, vars: 10, consts: [[1, "h-full", "w-full", "flex", "flex-col", 3, "ngClass"], ["class", "w-full", 3, "class", 4, "ngIf"], ["cdkScrollable", "", 1, "w-full", "flex-auto", "p-2", "overflow-y-hidden", "no-scrollbar", "hover:overflow-y-auto", 3, "ngClass"], [4, "ngIf"], ["class", "w-full ", 3, "class", 4, "ngIf"], [1, "w-full"], ["class", "w-full flex items-center justify-center  py-2", 4, "ngIf"], ["class", "w-full flex items-center justify-center py-2", 4, "ngIf"], [1, "w-full", "flex", "items-center", "justify-center", "py-2"], ["ngFor", "", 3, "ngForOf"], [3, "options", "item", 4, "ngIf"], [3, "options", "item"], [3, "options", "item", "isSelected", "parentIsGroup"], [3, "options", "showIcon", "item", 4, "ngIf"], [3, "options", "showIcon", "item"], [3, "options", "showIcon", "item", "isSelected", "parentIsGroup"], [1, "w-full", "flex", "justify-between", "items-center", "p-4", 3, "ngClass"], ["width", "45", "height", "45", "viewBox", "0 0 45 45", "fill", "none", "xmlns", "http://www.w3.org/2000/svg"], ["d", "M22.5 45C34.9264 45 45 34.9264 45 22.5C45 10.0736 34.9264 0 22.5 0C10.0736 0 3.60916e-05 10.0736 3.60916e-05 22.5L0 40C0 43 2 45 4.99999 45H22.5Z", 1, "fill-current", "text-primary-1"], ["d", "M37.3261 15.5724C37.1526 14.4349 36.6617 13.3653 35.9067 12.4797C35.1518 11.594 34.1615 10.9262 33.0435 10.5488C31.9255 10.1713 30.7224 10.0986 29.5649 10.3386C28.4073 10.5785 27.3395 11.1219 26.4774 11.9098C25.8672 12.4396 25.3735 13.0838 25.0261 13.8037C24.6787 14.5236 24.4848 15.3043 24.4559 16.0989C24.3652 18.9134 24.364 21.7302 24.3381 24.5458C24.3359 24.6257 24.3517 24.705 24.3842 24.7783C24.4168 24.8516 24.4653 24.9172 24.5266 24.9705C24.6052 24.9896 24.6872 24.9918 24.7668 24.9767C24.8464 24.9617 24.9217 24.9299 24.9872 24.8835C26.1362 24.1121 27.2789 23.3326 28.4153 22.5451C28.4806 22.4941 28.5589 22.4613 28.6418 22.45C28.7248 22.4387 28.8093 22.4494 28.8865 22.481C29.9008 22.8233 30.9895 22.9022 32.0449 22.71C33.2011 22.5252 34.2814 22.0311 35.1648 21.283C36.0481 20.535 36.6994 19.5627 37.0457 18.4751C37.1952 17.9712 37.3132 17.459 37.3991 16.9413V16.0371C37.3779 15.8838 37.3567 15.7258 37.3261 15.5724ZM31.4358 19.0599C31.4218 19.1922 31.3545 19.3137 31.2485 19.3978C31.1425 19.4819 31.0065 19.5217 30.8704 19.5086C30.7981 19.5079 30.7267 19.4933 30.6603 19.4657C30.5938 19.4382 30.5336 19.3982 30.4831 19.348C30.4325 19.2979 30.3927 19.2385 30.3657 19.1734C30.3388 19.1082 30.3253 19.0386 30.3261 18.9684C30.3323 18.8268 30.3923 18.6924 30.4946 18.5913C30.5969 18.4902 30.7341 18.4297 30.8798 18.4213H30.921C30.9924 18.4245 31.0625 18.4414 31.1271 18.4709C31.1918 18.5004 31.2499 18.542 31.298 18.5934C31.3462 18.6447 31.3834 18.7048 31.4077 18.7701C31.4319 18.8354 31.4427 18.9047 31.4394 18.9741C31.4431 19.0014 31.4451 19.0289 31.4453 19.0565L31.4358 19.0599ZM32.456 15.7579C32.2568 15.9773 32.041 16.182 31.8104 16.3702C31.6478 16.4831 31.5205 16.6374 31.4423 16.8161C31.3641 16.9949 31.3383 17.1911 31.3675 17.3831C31.4028 17.6304 31.3004 17.7265 31.0424 17.7059C30.8327 17.6922 30.6206 17.7059 30.3968 17.7059C30.299 17.3587 30.3155 16.9905 30.444 16.6529C30.5724 16.3152 30.8064 16.0251 31.113 15.8231C31.3544 15.6555 31.541 15.4238 31.6502 15.1558C31.6967 15.0338 31.6922 14.899 31.6379 14.7802C31.5835 14.6613 31.4835 14.5678 31.3593 14.5194C31.2063 14.4263 31.0273 14.3816 30.847 14.3917C30.6668 14.4017 30.4942 14.466 30.3532 14.5755C30.2202 14.6968 30.1153 14.8443 30.0457 15.0082C29.9868 15.149 29.9279 15.2085 29.7736 15.189L28.9737 15.0826C28.9931 14.851 29.0655 14.6265 29.1858 14.4257C29.3061 14.2249 29.4711 14.0528 29.6688 13.922C30.078 13.6473 30.5702 13.5138 31.0667 13.5428C31.5632 13.5719 32.0354 13.7618 32.4077 14.0822C32.4502 14.1171 32.4896 14.1554 32.5255 14.1967C32.7293 14.4114 32.8374 14.6957 32.8261 14.9876C32.8149 15.2795 32.6852 15.5551 32.4654 15.7544L32.456 15.7579Z", "fill", "white"], ["d", "M25.7737 30.532C25.6559 29.9849 25.5063 29.4836 24.8678 29.273C23.751 28.9102 22.6436 28.5176 21.5351 28.1284C21.4628 28.096 21.3814 28.0881 21.304 28.1062C21.2266 28.1243 21.1578 28.1672 21.1086 28.228C20.1944 29.1299 19.2661 30.0193 18.3496 30.92C18.1576 31.1123 17.9856 31.3309 17.8042 31.5381L17.6616 31.5324C17.5981 31.425 17.5252 31.3232 17.4437 31.2279C16.4282 30.2287 15.401 29.2352 14.3808 28.2349C14.3329 28.1735 14.2652 28.1295 14.1885 28.1099C14.1118 28.0904 14.0306 28.0965 13.9579 28.1273C12.8332 28.5256 11.7054 28.9182 10.5745 29.3051C10.3839 29.3615 10.2135 29.4689 10.0829 29.6151C9.95224 29.7612 9.86663 29.9402 9.83587 30.1314C9.55785 31.4271 9.26569 32.7204 8.98179 34.015C8.85338 34.601 8.73086 35.1881 8.6001 35.8005L26.8999 35.7913C26.8787 35.6769 26.8622 35.5853 26.8433 35.496C26.4868 33.841 26.1302 32.1863 25.7737 30.532Z", "fill", "white"], ["d", "M14.2763 24.0218C14.377 24.8267 14.7715 25.5697 15.3881 26.116C16.0046 26.6624 16.8024 26.9758 17.6365 26.9993C18.4707 27.0229 19.2858 26.755 19.934 26.2444C20.5822 25.7337 21.0204 25.0142 21.1691 24.2164C21.2975 23.4564 21.46 22.7021 21.612 21.9467C21.7537 21.3605 21.7358 20.7486 21.56 20.1711C21.3842 19.5937 21.0566 19.0708 20.6095 18.6538C20.2307 18.2975 19.7812 18.0199 19.2885 17.8377C18.7958 17.6555 18.27 17.5725 17.7433 17.5939C17.3498 17.6431 16.9493 17.6638 16.5652 17.7462C15.6713 17.9579 14.8937 18.4923 14.3905 19.2408C13.8873 19.9892 13.6961 20.8957 13.8557 21.7762C14.0053 22.5224 14.1585 23.2698 14.2763 24.0218Z", "fill", "white"], [1, "p-2", "bg-neutral-3-50", "dark:bg-d-neutral-3-300", "border", "border-transparent", "rounded-full", "flex", 3, "ngClass", "click"], [1, "text-neutral-1-300", "dark:text-d-neutral-1-500", "text-base", "leading-4", "cursor-pointer", 3, "ngClass"]], template: function TDSMenuComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c4);
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, TDSMenuComponent_div_1_Template, 3, 4, "div", 1);
        i0.ɵɵelementStart(2, "div", 2);
        i0.ɵɵtemplate(3, TDSMenuComponent_ng_container_3_Template, 2, 1, "ng-container", 3);
        i0.ɵɵtemplate(4, TDSMenuComponent_ng_container_4_Template, 2, 1, "ng-container", 3);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(5, TDSMenuComponent_div_5_Template, 10, 12, "div", 4);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵproperty("ngClass", ctx.mode);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.showLogo);
        i0.ɵɵadvance(1);
        i0.ɵɵclassMap(ctx.options.background);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c5, ctx.inlineCollapsed));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.inlineCollapsed);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.inlineCollapsed);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.hasIcon);
    } }, directives: [i2.NgClass, i2.NgIf, i3.CdkScrollable, i2.NgForOf, i4.TDSMenuGroupPopupComponent, i5.TDSMenuItemComponent, i6.TDSMenuGroupInlineComponent], styles: [""], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSMenuComponent.prototype, "showLogo", void 0);
__decorate([
    InputBoolean()
], TDSMenuComponent.prototype, "inlineCollapsed", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMenuComponent, [{
        type: Component,
        args: [{ selector: 'tds-menu', changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, host: {
                    class: "h-full flex flex-col box-border",
                    '[class.w-full]': `!inlineCollapsed`,
                    '[class.w-16]': `inlineCollapsed`,
                }, providers: [
                    /** use the top level service **/
                    TDSMenuService
                ], template: "<div class=\"h-full w-full flex flex-col\" [ngClass]=\"mode\">\r\n    <!-- header -->\r\n    <div class=\"w-full\" [class]=\"options.background\" *ngIf=\"showLogo\">\r\n        <div *ngIf=\"!inlineCollapsed\" class=\"w-full flex items-center justify-center  py-2\">\r\n            <ng-content select=\"[logo-text]\">\r\n            </ng-content>\r\n        </div>\r\n        <div *ngIf=\"inlineCollapsed\" class=\"w-full flex items-center justify-center py-2\">\r\n            <ng-content select=\"[logo]\">\r\n            </ng-content>\r\n        </div>\r\n    </div>\r\n    <!-- body -->\r\n    <div class=\"w-full flex-auto p-2 overflow-y-hidden no-scrollbar hover:overflow-y-auto\" [class]=\"options.background\"\r\n        [ngClass]=\"{' overflow-x-hidden':inlineCollapsed}\" cdkScrollable>\r\n        <!-- menu item  -->\r\n        <ng-container *ngIf=\"inlineCollapsed\">\r\n            <ng-template ngFor let-item [ngForOf]=\"listMenu\">\r\n                <tds-menu-group-popup [options]=\"options\"  [item]=\"item\" *ngIf=\"!item.hidden\">\r\n                    <ng-template ngFor let-children [ngForOf]=\"item.listChild\">\r\n                        <tds-menu-item  [options]=\"options\" [item]=\"children\" [isSelected]=\"children.isSelected!\" [parentIsGroup]=\"true\">\r\n                        </tds-menu-item>\r\n                    </ng-template>\r\n                </tds-menu-group-popup>\r\n            </ng-template>\r\n        </ng-container>\r\n        <ng-container *ngIf=\"!inlineCollapsed\">\r\n            <ng-template ngFor let-item [ngForOf]=\"listMenu\">\r\n                <tds-menu-group-inline [options]=\"options\" [showIcon]=\"hasIcon\" [item]=\"item\" *ngIf=\"!item.hidden\">\r\n                    <ng-template ngFor let-children [ngForOf]=\"item.listChild\">\r\n                        <tds-menu-item [options]=\"options\"  [showIcon]=\"hasIcon\" [item]=\"children\" [isSelected]=\"children.isSelected!\"\r\n                            [parentIsGroup]=\"true\">\r\n                        </tds-menu-item>\r\n                    </ng-template>\r\n                </tds-menu-group-inline>\r\n            </ng-template>\r\n        </ng-container>\r\n    </div>\r\n    <!-- footer -->\r\n    <div class=\"w-full \" [class]=\"options.background\" *ngIf='hasIcon'>\r\n        <div class=\"w-full flex justify-between items-center p-4\" [ngClass]=\"{'flex-col':inlineCollapsed}\">\r\n            <div>\r\n                <svg width=\"45\" height=\"45\" viewBox=\"0 0 45 45\" fill=\"none\" xmlns=\"http://www.w3.org/2000/svg\">\r\n                    <path\r\n                        d=\"M22.5 45C34.9264 45 45 34.9264 45 22.5C45 10.0736 34.9264 0 22.5 0C10.0736 0 3.60916e-05 10.0736 3.60916e-05 22.5L0 40C0 43 2 45 4.99999 45H22.5Z\"\r\n                        class=\"fill-current text-primary-1\" />\r\n                    <path\r\n                        d=\"M37.3261 15.5724C37.1526 14.4349 36.6617 13.3653 35.9067 12.4797C35.1518 11.594 34.1615 10.9262 33.0435 10.5488C31.9255 10.1713 30.7224 10.0986 29.5649 10.3386C28.4073 10.5785 27.3395 11.1219 26.4774 11.9098C25.8672 12.4396 25.3735 13.0838 25.0261 13.8037C24.6787 14.5236 24.4848 15.3043 24.4559 16.0989C24.3652 18.9134 24.364 21.7302 24.3381 24.5458C24.3359 24.6257 24.3517 24.705 24.3842 24.7783C24.4168 24.8516 24.4653 24.9172 24.5266 24.9705C24.6052 24.9896 24.6872 24.9918 24.7668 24.9767C24.8464 24.9617 24.9217 24.9299 24.9872 24.8835C26.1362 24.1121 27.2789 23.3326 28.4153 22.5451C28.4806 22.4941 28.5589 22.4613 28.6418 22.45C28.7248 22.4387 28.8093 22.4494 28.8865 22.481C29.9008 22.8233 30.9895 22.9022 32.0449 22.71C33.2011 22.5252 34.2814 22.0311 35.1648 21.283C36.0481 20.535 36.6994 19.5627 37.0457 18.4751C37.1952 17.9712 37.3132 17.459 37.3991 16.9413V16.0371C37.3779 15.8838 37.3567 15.7258 37.3261 15.5724ZM31.4358 19.0599C31.4218 19.1922 31.3545 19.3137 31.2485 19.3978C31.1425 19.4819 31.0065 19.5217 30.8704 19.5086C30.7981 19.5079 30.7267 19.4933 30.6603 19.4657C30.5938 19.4382 30.5336 19.3982 30.4831 19.348C30.4325 19.2979 30.3927 19.2385 30.3657 19.1734C30.3388 19.1082 30.3253 19.0386 30.3261 18.9684C30.3323 18.8268 30.3923 18.6924 30.4946 18.5913C30.5969 18.4902 30.7341 18.4297 30.8798 18.4213H30.921C30.9924 18.4245 31.0625 18.4414 31.1271 18.4709C31.1918 18.5004 31.2499 18.542 31.298 18.5934C31.3462 18.6447 31.3834 18.7048 31.4077 18.7701C31.4319 18.8354 31.4427 18.9047 31.4394 18.9741C31.4431 19.0014 31.4451 19.0289 31.4453 19.0565L31.4358 19.0599ZM32.456 15.7579C32.2568 15.9773 32.041 16.182 31.8104 16.3702C31.6478 16.4831 31.5205 16.6374 31.4423 16.8161C31.3641 16.9949 31.3383 17.1911 31.3675 17.3831C31.4028 17.6304 31.3004 17.7265 31.0424 17.7059C30.8327 17.6922 30.6206 17.7059 30.3968 17.7059C30.299 17.3587 30.3155 16.9905 30.444 16.6529C30.5724 16.3152 30.8064 16.0251 31.113 15.8231C31.3544 15.6555 31.541 15.4238 31.6502 15.1558C31.6967 15.0338 31.6922 14.899 31.6379 14.7802C31.5835 14.6613 31.4835 14.5678 31.3593 14.5194C31.2063 14.4263 31.0273 14.3816 30.847 14.3917C30.6668 14.4017 30.4942 14.466 30.3532 14.5755C30.2202 14.6968 30.1153 14.8443 30.0457 15.0082C29.9868 15.149 29.9279 15.2085 29.7736 15.189L28.9737 15.0826C28.9931 14.851 29.0655 14.6265 29.1858 14.4257C29.3061 14.2249 29.4711 14.0528 29.6688 13.922C30.078 13.6473 30.5702 13.5138 31.0667 13.5428C31.5632 13.5719 32.0354 13.7618 32.4077 14.0822C32.4502 14.1171 32.4896 14.1554 32.5255 14.1967C32.7293 14.4114 32.8374 14.6957 32.8261 14.9876C32.8149 15.2795 32.6852 15.5551 32.4654 15.7544L32.456 15.7579Z\"\r\n                        fill=\"white\" />\r\n                    <path\r\n                        d=\"M25.7737 30.532C25.6559 29.9849 25.5063 29.4836 24.8678 29.273C23.751 28.9102 22.6436 28.5176 21.5351 28.1284C21.4628 28.096 21.3814 28.0881 21.304 28.1062C21.2266 28.1243 21.1578 28.1672 21.1086 28.228C20.1944 29.1299 19.2661 30.0193 18.3496 30.92C18.1576 31.1123 17.9856 31.3309 17.8042 31.5381L17.6616 31.5324C17.5981 31.425 17.5252 31.3232 17.4437 31.2279C16.4282 30.2287 15.401 29.2352 14.3808 28.2349C14.3329 28.1735 14.2652 28.1295 14.1885 28.1099C14.1118 28.0904 14.0306 28.0965 13.9579 28.1273C12.8332 28.5256 11.7054 28.9182 10.5745 29.3051C10.3839 29.3615 10.2135 29.4689 10.0829 29.6151C9.95224 29.7612 9.86663 29.9402 9.83587 30.1314C9.55785 31.4271 9.26569 32.7204 8.98179 34.015C8.85338 34.601 8.73086 35.1881 8.6001 35.8005L26.8999 35.7913C26.8787 35.6769 26.8622 35.5853 26.8433 35.496C26.4868 33.841 26.1302 32.1863 25.7737 30.532Z\"\r\n                        fill=\"white\" />\r\n                    <path\r\n                        d=\"M14.2763 24.0218C14.377 24.8267 14.7715 25.5697 15.3881 26.116C16.0046 26.6624 16.8024 26.9758 17.6365 26.9993C18.4707 27.0229 19.2858 26.755 19.934 26.2444C20.5822 25.7337 21.0204 25.0142 21.1691 24.2164C21.2975 23.4564 21.46 22.7021 21.612 21.9467C21.7537 21.3605 21.7358 20.7486 21.56 20.1711C21.3842 19.5937 21.0566 19.0708 20.6095 18.6538C20.2307 18.2975 19.7812 18.0199 19.2885 17.8377C18.7958 17.6555 18.27 17.5725 17.7433 17.5939C17.3498 17.6431 16.9493 17.6638 16.5652 17.7462C15.6713 17.9579 14.8937 18.4923 14.3905 19.2408C13.8873 19.9892 13.6961 20.8957 13.8557 21.7762C14.0053 22.5224 14.1585 23.2698 14.2763 24.0218Z\"\r\n                        fill=\"white\" />\r\n                </svg>\r\n\r\n            </div>\r\n            <div class=\"p-2 bg-neutral-3-50 dark:bg-d-neutral-3-300 border border-transparent rounded-full flex \"\r\n                [ngClass]=\"{'mt-6':inlineCollapsed}\" (click)=\"onClickInlineCollapsed()\">\r\n                <span [ngClass]=\"{'tdsi-arrow-right-line':inlineCollapsed, 'tdsi-arrow-left-line':!inlineCollapsed }\"\r\n                    class=\"text-neutral-1-300 dark:text-d-neutral-1-500 text-base leading-4 cursor-pointer\"></span>\r\n            </div>\r\n        </div>\r\n    </div>\r\n</div>", styles: [""] }]
    }], function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.TDSMenuService }]; }, { showLogo: [{
            type: Input
        }], data: [{
            type: Input
        }], mode: [{
            type: Input
        }], showIcon: [{
            type: Input
        }], inlineCollapsed: [{
            type: Input
        }], options: [{
            type: Input
        }], onClickItem: [{
            type: Output
        }], onOpenChange: [{
            type: Output
        }], listOfTDSMenuGroupInlineComponent: [{
            type: ViewChildren,
            args: [TDSMenuGroupInlineComponent]
        }], listOfTDSMenuGroupPopupComponent: [{
            type: ViewChildren,
            args: [TDSMenuGroupPopupComponent]
        }], logoText: [{
            type: ContentChild,
            args: ['logo-text']
        }] }); })();
