// tds-ui/esm2020/avatar/avatar.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSAvatarGroupComponent } from './avatar-group.component';
import { TDSAvatarComponent } from './avatar.component';
import * as i0 from "@angular/core";
export class TDSAvatarModule {
}
TDSAvatarModule.ɵfac = function TDSAvatarModule_Factory(t) { return new (t || TDSAvatarModule)(); };
TDSAvatarModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSAvatarModule });
TDSAvatarModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PlatformModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAvatarModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSAvatarComponent, TDSAvatarGroupComponent],
                exports: [TDSAvatarComponent, TDSAvatarGroupComponent],
                imports: [BidiModule, CommonModule, PlatformModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSAvatarModule, { declarations: [TDSAvatarComponent, TDSAvatarGroupComponent], imports: [BidiModule, CommonModule, PlatformModule], exports: [TDSAvatarComponent, TDSAvatarGroupComponent] }); })();
