// tds-ui/esm2020/avatar/avatar-group.component.mjs
import { ChangeDetectionStrategy, Component, ContentChildren, Input } from '@angular/core';
import { TDSAvatarComponent } from './avatar.component';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSAvatarGroupComponent {
    constructor() {
        this.tdsClass = ['-ml-3', 'border', 'border-white', 'rounded-full'];
    }
    ngAfterViewInit() {
        if (this.lstAvatar) {
            this.lstAvatar.forEach((avatar, index) => {
                if (index != 0) {
                    this.tdsClass.forEach(cl => {
                        avatar.element.classList.add(cl);
                    });
                }
            });
        }
    }
}
TDSAvatarGroupComponent.ɵfac = function TDSAvatarGroupComponent_Factory(t) { return new (t || TDSAvatarGroupComponent)(); };
TDSAvatarGroupComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSAvatarGroupComponent, selectors: [["tds-avatar-group"]], contentQueries: function TDSAvatarGroupComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSAvatarComponent, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.lstAvatar = _t);
    } }, hostAttrs: [1, "tds-avatar-group", "inline-flex"], inputs: { tdsClass: "tdsClass" }, exportAs: ["tdsAvatarGroup"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSAvatarGroupComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAvatarGroupComponent, [{
        type: Component,
        args: [{ selector: 'tds-avatar-group', exportAs: 'tdsAvatarGroup', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                    class: 'tds-avatar-group inline-flex'
                }, template: "<ng-content></ng-content>" }]
    }], null, { lstAvatar: [{
            type: ContentChildren,
            args: [TDSAvatarComponent, { descendants: true }]
        }], tdsClass: [{
            type: Input
        }] }); })();
