// tds-ui/esm2020/select/select.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TDSSelect } from './select';
import { TDSSelectSearchComponent } from './select-search';
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { PlatformModule } from '@angular/cdk/platform';
import { FormsModule } from '@angular/forms';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { A11yModule } from '@angular/cdk/a11y';
import { TDSFooterTemplateDirective, TDSHeaderTemplateDirective, TDSItemLabelDirective, TDSLabelTemplateDirective, TDSLoadingSpinnerTemplateDirective, TDSLoadingTextTemplateDirective, TDSMultiLabelTemplateDirective, TDSNotFoundTemplateDirective, TDSOptgroupTemplateDirective, TDSOptionTemplateDirective, TDSTagTemplateDirective, TDSTypeToSearchTemplateDirective } from './select.directive';
import { TDSSelectOptionGroup } from './select-option-group';
import { TDSOptionItem } from './select-option-item';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSSelectPlaceholderComponent } from './select-placeholder';
import { TDSSelectValueControlComponent } from './select-value';
import { TDSEmptyModule } from 'tds-ui/empty';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import * as i0 from "@angular/core";
const tmpDirective = [
    TDSOptgroupTemplateDirective,
    TDSOptionTemplateDirective,
    TDSLabelTemplateDirective,
    TDSMultiLabelTemplateDirective,
    TDSHeaderTemplateDirective,
    TDSFooterTemplateDirective,
    TDSNotFoundTemplateDirective,
    TDSTypeToSearchTemplateDirective,
    TDSLoadingTextTemplateDirective,
    TDSTagTemplateDirective,
    TDSLoadingSpinnerTemplateDirective,
    TDSItemLabelDirective
];
const Cpm = [
    TDSSelectOptionGroup,
    TDSOptionItem,
    TDSSelectPlaceholderComponent,
    TDSSelectValueControlComponent
];
export class TDSSelectModule {
}
TDSSelectModule.ɵfac = function TDSSelectModule_Factory(t) { return new (t || TDSSelectModule)(); };
TDSSelectModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSSelectModule });
TDSSelectModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            PlatformModule,
            OverlayModule,
            BidiModule,
            CommonModule,
            ScrollingModule,
            A11yModule,
            TDSOutletModule,
            TDSEmptyModule,
            TDSSpinnerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSelectModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TDSSelect,
                    TDSSelectSearchComponent,
                    tmpDirective,
                    Cpm
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    PlatformModule,
                    OverlayModule,
                    BidiModule,
                    CommonModule,
                    ScrollingModule,
                    A11yModule,
                    TDSOutletModule,
                    TDSEmptyModule,
                    TDSSpinnerModule
                ],
                exports: [
                    TDSSelect,
                    TDSSelectSearchComponent,
                    tmpDirective,
                    Cpm
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSSelectModule, { declarations: [TDSSelect,
        TDSSelectSearchComponent, TDSOptgroupTemplateDirective,
        TDSOptionTemplateDirective,
        TDSLabelTemplateDirective,
        TDSMultiLabelTemplateDirective,
        TDSHeaderTemplateDirective,
        TDSFooterTemplateDirective,
        TDSNotFoundTemplateDirective,
        TDSTypeToSearchTemplateDirective,
        TDSLoadingTextTemplateDirective,
        TDSTagTemplateDirective,
        TDSLoadingSpinnerTemplateDirective,
        TDSItemLabelDirective, TDSSelectOptionGroup,
        TDSOptionItem,
        TDSSelectPlaceholderComponent,
        TDSSelectValueControlComponent], imports: [CommonModule,
        FormsModule,
        PlatformModule,
        OverlayModule,
        BidiModule,
        CommonModule,
        ScrollingModule,
        A11yModule,
        TDSOutletModule,
        TDSEmptyModule,
        TDSSpinnerModule], exports: [TDSSelect,
        TDSSelectSearchComponent, TDSOptgroupTemplateDirective,
        TDSOptionTemplateDirective,
        TDSLabelTemplateDirective,
        TDSMultiLabelTemplateDirective,
        TDSHeaderTemplateDirective,
        TDSFooterTemplateDirective,
        TDSNotFoundTemplateDirective,
        TDSTypeToSearchTemplateDirective,
        TDSLoadingTextTemplateDirective,
        TDSTagTemplateDirective,
        TDSLoadingSpinnerTemplateDirective,
        TDSItemLabelDirective, TDSSelectOptionGroup,
        TDSOptionItem,
        TDSSelectPlaceholderComponent,
        TDSSelectValueControlComponent] }); })();
