// tds-ui/esm2020/select/select.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSItemLabelDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSItemLabelDirective.ɵfac = function TDSItemLabelDirective_Factory(t) { return new (t || TDSItemLabelDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSItemLabelDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSItemLabelDirective, selectors: [["", "tds-item-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSItemLabelDirective, [{
        type: Directive,
        args: [{ selector: '[tds-item-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSOptionTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSOptionTemplateDirective.ɵfac = function TDSOptionTemplateDirective_Factory(t) { return new (t || TDSOptionTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSOptionTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSOptionTemplateDirective, selectors: [["", "tds-option-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSOptionTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-option-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSOptgroupTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSOptgroupTemplateDirective.ɵfac = function TDSOptgroupTemplateDirective_Factory(t) { return new (t || TDSOptgroupTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSOptgroupTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSOptgroupTemplateDirective, selectors: [["", "tds-optgroup-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSOptgroupTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-optgroup-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSLabelTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSLabelTemplateDirective.ɵfac = function TDSLabelTemplateDirective_Factory(t) { return new (t || TDSLabelTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSLabelTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSLabelTemplateDirective, selectors: [["", "tds-label-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSLabelTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-label-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSMultiLabelTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSMultiLabelTemplateDirective.ɵfac = function TDSMultiLabelTemplateDirective_Factory(t) { return new (t || TDSMultiLabelTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSMultiLabelTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSMultiLabelTemplateDirective, selectors: [["", "tds-multi-label-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMultiLabelTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-multi-label-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSHeaderTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSHeaderTemplateDirective.ɵfac = function TDSHeaderTemplateDirective_Factory(t) { return new (t || TDSHeaderTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSHeaderTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSHeaderTemplateDirective, selectors: [["", "tds-header-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSHeaderTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-header-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSFooterTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSFooterTemplateDirective.ɵfac = function TDSFooterTemplateDirective_Factory(t) { return new (t || TDSFooterTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSFooterTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSFooterTemplateDirective, selectors: [["", "tds-footer-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFooterTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-footer-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSNotFoundTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSNotFoundTemplateDirective.ɵfac = function TDSNotFoundTemplateDirective_Factory(t) { return new (t || TDSNotFoundTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSNotFoundTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSNotFoundTemplateDirective, selectors: [["", "tds-notfound-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSNotFoundTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-notfound-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSTypeToSearchTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSTypeToSearchTemplateDirective.ɵfac = function TDSTypeToSearchTemplateDirective_Factory(t) { return new (t || TDSTypeToSearchTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSTypeToSearchTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTypeToSearchTemplateDirective, selectors: [["", "tds-typetosearch-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTypeToSearchTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-typetosearch-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSLoadingTextTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSLoadingTextTemplateDirective.ɵfac = function TDSLoadingTextTemplateDirective_Factory(t) { return new (t || TDSLoadingTextTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSLoadingTextTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSLoadingTextTemplateDirective, selectors: [["", "tds-loadingtext-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSLoadingTextTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-loadingtext-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSTagTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSTagTemplateDirective.ɵfac = function TDSTagTemplateDirective_Factory(t) { return new (t || TDSTagTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSTagTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTagTemplateDirective, selectors: [["", "tds-tag-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTagTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-tag-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
export class TDSLoadingSpinnerTemplateDirective {
    constructor(template) {
        this.template = template;
    }
}
TDSLoadingSpinnerTemplateDirective.ɵfac = function TDSLoadingSpinnerTemplateDirective_Factory(t) { return new (t || TDSLoadingSpinnerTemplateDirective)(i0.ɵɵdirectiveInject(i0.TemplateRef)); };
TDSLoadingSpinnerTemplateDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSLoadingSpinnerTemplateDirective, selectors: [["", "tds-loadingspinner-tmp", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSLoadingSpinnerTemplateDirective, [{
        type: Directive,
        args: [{ selector: '[tds-loadingspinner-tmp]' }]
    }], function () { return [{ type: i0.TemplateRef }]; }, null); })();
