// tds-ui/esm2020/select/select-option-group.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSSelectOptionGroup {
    constructor() {
        this.label = null;
        this.changes = new Subject();
    }
    ngOnChanges() {
        this.changes.next();
    }
}
TDSSelectOptionGroup.ɵfac = function TDSSelectOptionGroup_Factory(t) { return new (t || TDSSelectOptionGroup)(); };
TDSSelectOptionGroup.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSSelectOptionGroup, selectors: [["tds-option-group"]], inputs: { label: "label" }, exportAs: ["tdsOptionGroup"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSSelectOptionGroup_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSelectOptionGroup, [{
        type: Component,
        args: [{
                selector: 'tds-option-group',
                exportAs: 'tdsOptionGroup',
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <ng-content></ng-content>
  `
            }]
    }], null, { label: [{
            type: Input
        }] }); })();
