// tds-ui/esm2020/select/select-placeholder.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSSelectPlaceholderComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.placeholder = null;
    }
}
TDSSelectPlaceholderComponent.ɵfac = function TDSSelectPlaceholderComponent_Factory(t) { return new (t || TDSSelectPlaceholderComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSSelectPlaceholderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSSelectPlaceholderComponent, selectors: [["tds-select-placeholder"]], inputs: { placeholder: "placeholder" }, decls: 2, vars: 1, consts: [[1, "w-full", "flex", "items-center", "text-body-2", "font-regular", "text-neutral-1-400", "dark:text-neutral-1-400", "truncate"]], template: function TDSSelectPlaceholderComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtext(1);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx.placeholder, " ");
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSSelectPlaceholderComponent, [{
        type: Component,
        args: [{
                selector: 'tds-select-placeholder',
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
            <div class="w-full flex items-center text-body-2 font-regular  text-neutral-1-400 dark:text-neutral-1-400 truncate">
                 {{ placeholder }}
            </div>`,
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { placeholder: [{
            type: Input
        }] }); })();
