// tds-ui/esm2020/table/src/cell/cell.directive.mjs
import { Directive, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../table-style.service";
export class TDSTableCellDirective {
    constructor(tdsTableStyleService) {
        this.isInsideTable = false;
        this.isInsideTable = !!tdsTableStyleService;
    }
}
TDSTableCellDirective.ɵfac = function TDSTableCellDirective_Factory(t) { return new (t || TDSTableCellDirective)(i0.ɵɵdirectiveInject(i1.TDSTableStyleService, 8)); };
TDSTableCellDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSTableCellDirective, selectors: [["th", 9, "tds-disable-th", 3, "mat-cell", ""], ["td", 9, "tds-disable-td", 3, "mat-cell", ""]], hostVars: 2, hostBindings: function TDSTableCellDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-table-cell", ctx.isInsideTable);
    } } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTableCellDirective, [{
        type: Directive,
        args: [{
                selector: 'th:not(.tds-disable-th):not([mat-cell]), td:not(.tds-disable-td):not([mat-cell])',
                host: {
                    '[class.tds-table-cell]': 'isInsideTable',
                }
            }]
    }], function () { return [{ type: i1.TDSTableStyleService, decorators: [{
                type: Optional
            }] }]; }, null); })();
