// tds-ui/esm2020/table/src/cell/td-addon.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../addon/row-indent.directive";
import * as i3 from "../addon/row-expand-button.directive";
import * as i4 from "tds-ui/tds-checkbox";
import * as i5 from "@angular/forms";
const _c0 = ["checked", ""];
const _c1 = function (a0, a1) { return { "tdsi-arrow-right-line": a0, "tdsi-arrow-down-line": a1 }; };
function TDSTdAddOnComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "tds-row-indent", 2);
    i0.ɵɵelementStart(2, "button", 3);
    i0.ɵɵlistener("expandChange", function TDSTdAddOnComponent_ng_container_0_Template_button_expandChange_2_listener($event) { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.onExpandChange($event); });
    i0.ɵɵelementStart(3, "span", 4);
    i0.ɵɵelement(4, "i", 5);
    i0.ɵɵelementEnd()();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("indentSize", ctx_r0.indentSize);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("expand", ctx_r0.expand)("spaceMode", !ctx_r0.showExpand);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(4, _c1, !ctx_r0.expand, ctx_r0.expand));
} }
function TDSTdAddOnComponent_tds_checkbox_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "tds-checkbox", 6);
    i0.ɵɵlistener("ngModelChange", function TDSTdAddOnComponent_tds_checkbox_1_Template_tds_checkbox_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return ctx_r4.onCheckedChange($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("disabled", ctx_r1.disabled)("ngModel", ctx_r1.checked)("indeterminate", ctx_r1.indeterminate);
} }
const _c2 = ["*"];
export class TDSTdAddOnComponent {
    constructor() {
        this.checked = false;
        this.disabled = false;
        this.indeterminate = false;
        this.indentSize = 0;
        this.showExpand = false;
        this.showCheckbox = false;
        this.expand = false;
        this.checkedChange = new EventEmitter();
        this.expandChange = new EventEmitter();
        this.isNzShowExpandChanged = false;
        this.isNzShowCheckboxChanged = false;
    }
    onCheckedChange(checked) {
        this.checked = checked;
        this.checkedChange.emit(checked);
    }
    onExpandChange(expand) {
        this.expand = expand;
        this.expandChange.emit(expand);
    }
    ngOnChanges(changes) {
        const isFirstChange = (value) => value && value.firstChange && value.currentValue !== undefined;
        const { expand, checked, showExpand, showCheckbox } = changes;
        if (showExpand) {
            this.isNzShowExpandChanged = true;
        }
        if (showCheckbox) {
            this.isNzShowCheckboxChanged = true;
        }
        if (isFirstChange(expand) && !this.isNzShowExpandChanged) {
            this.showExpand = true;
        }
        if (isFirstChange(checked) && !this.isNzShowCheckboxChanged) {
            this.showCheckbox = true;
        }
    }
}
TDSTdAddOnComponent.ɵfac = function TDSTdAddOnComponent_Factory(t) { return new (t || TDSTdAddOnComponent)(); };
TDSTdAddOnComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSTdAddOnComponent, selectors: [["td", "checked", ""], ["td", "disabled", ""], ["td", "indeterminate", ""], ["td", "indentSize", ""], ["td", "expand", ""], ["td", "showExpand", ""], ["td", "showCheckbox", ""]], hostVars: 4, hostBindings: function TDSTdAddOnComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-table-cell-with-append", ctx.showExpand || ctx.indentSize > 0)("tds-table-selection-column", ctx.showCheckbox);
    } }, inputs: { checked: "checked", disabled: "disabled", indeterminate: "indeterminate", indentSize: "indentSize", showExpand: "showExpand", showCheckbox: "showCheckbox", expand: "expand" }, outputs: { checkedChange: "checkedChange", expandChange: "expandChange" }, features: [i0.ɵɵNgOnChangesFeature], attrs: _c0, ngContentSelectors: _c2, decls: 3, vars: 2, consts: [[4, "ngIf"], [3, "disabled", "ngModel", "indeterminate", "ngModelChange", 4, "ngIf"], [3, "indentSize"], ["tds-row-expand-button", "", 1, "p-2", "text-neutral-1-500", "hover:text-primary-1", 3, "expand", "spaceMode", "expandChange"], [1, "flex", "items-center"], [1, "text-base", "leading-4", 3, "ngClass"], [3, "disabled", "ngModel", "indeterminate", "ngModelChange"]], template: function TDSTdAddOnComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtemplate(0, TDSTdAddOnComponent_ng_container_0_Template, 5, 7, "ng-container", 0);
        i0.ɵɵtemplate(1, TDSTdAddOnComponent_tds_checkbox_1_Template, 1, 3, "tds-checkbox", 1);
        i0.ɵɵprojection(2);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.showExpand || ctx.indentSize > 0);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.showCheckbox);
    } }, directives: [i1.NgIf, i2.TDSRowIndentDirective, i3.TDSRowExpandButtonDirective, i1.NgClass, i4.TDSCheckbox, i5.NgControlStatus, i5.NgModel], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSTdAddOnComponent.prototype, "showExpand", void 0);
__decorate([
    InputBoolean()
], TDSTdAddOnComponent.prototype, "showCheckbox", void 0);
__decorate([
    InputBoolean()
], TDSTdAddOnComponent.prototype, "expand", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTdAddOnComponent, [{
        type: Component,
        args: [{
                selector: 'td[checked], td[disabled], td[indeterminate], td[indentSize], td[expand], td[showExpand], td[showCheckbox]',
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                template: `
    <ng-container *ngIf="showExpand || indentSize > 0">
      <tds-row-indent [indentSize]="indentSize"></tds-row-indent>
      <button
        tds-row-expand-button      
        [expand]="expand"
        (expandChange)="onExpandChange($event)"
        [spaceMode]="!showExpand"
        class="p-2 text-neutral-1-500 hover:text-primary-1"
      >
        <span class="flex items-center">
         <i [ngClass]="{'tdsi-arrow-right-line':!expand,'tdsi-arrow-down-line':expand}" class="text-base leading-4"></i>
        </span>
      </button>
    </ng-container>
    <tds-checkbox     
      *ngIf="showCheckbox"
      [disabled]="disabled"
      [ngModel]="checked"
      [indeterminate]="indeterminate"
      (ngModelChange)="onCheckedChange($event)"
    ></tds-checkbox>
    <ng-content></ng-content>
  `,
                host: {
                    '[class.tds-table-cell-with-append]': `showExpand || indentSize > 0`,
                    '[class.tds-table-selection-column]': `showCheckbox`
                }
            }]
    }], null, { checked: [{
            type: Input
        }], disabled: [{
            type: Input
        }], indeterminate: [{
            type: Input
        }], indentSize: [{
            type: Input
        }], showExpand: [{
            type: Input
        }], showCheckbox: [{
            type: Input
        }], expand: [{
            type: Input
        }], checkedChange: [{
            type: Output
        }], expandChange: [{
            type: Output
        }] }); })();
