// tds-ui/esm2020/table/src/cell/th-measure.directive.mjs
import { Directive, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { isNil } from 'tds-ui/core/util';
import * as i0 from "@angular/core";
export class TDSThMeasureDirective {
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.changes$ = new Subject();
        this.width = null;
        this.colspan = null;
        this.colSpan = null;
        this.rowspan = null;
        this.rowSpan = null;
    }
    ngOnChanges(changes) {
        const { width, colspan, rowspan, colSpan, rowSpan } = changes;
        if (colspan || colSpan) {
            const col = this.colspan || this.colSpan;
            if (!isNil(col)) {
                this.renderer.setAttribute(this.elementRef.nativeElement, 'colspan', `${col}`);
            }
            else {
                this.renderer.removeAttribute(this.elementRef.nativeElement, 'colspan');
            }
        }
        if (rowspan || rowSpan) {
            const row = this.rowspan || this.rowSpan;
            if (!isNil(row)) {
                this.renderer.setAttribute(this.elementRef.nativeElement, 'rowspan', `${row}`);
            }
            else {
                this.renderer.removeAttribute(this.elementRef.nativeElement, 'rowspan');
            }
        }
        if (width || colspan) {
            this.changes$.next();
        }
    }
}
TDSThMeasureDirective.ɵfac = function TDSThMeasureDirective_Factory(t) { return new (t || TDSThMeasureDirective)(i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSThMeasureDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSThMeasureDirective, selectors: [["th"]], hostAttrs: [1, "text-body-2", "font-semibold", "group"], inputs: { width: "width", colspan: "colspan", colSpan: "colSpan", rowspan: "rowspan", rowSpan: "rowSpan" }, features: [i0.ɵɵNgOnChangesFeature] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSThMeasureDirective, [{
        type: Directive,
        args: [{
                selector: 'th',
                host: {
                    class: "text-body-2 font-semibold group",
                }
            }]
    }], function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }]; }, { width: [{
            type: Input
        }], colspan: [{
            type: Input
        }], colSpan: [{
            type: Input
        }], rowspan: [{
            type: Input
        }], rowSpan: [{
            type: Input
        }] }); })();
