// tds-ui/esm2020/table/src/cell/th-selection.component.mjs
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "../addon/selection.component";
const _c0 = ["selections", ""];
const _c1 = ["*"];
export class TDSThSelectionComponent {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.selections = [];
        this.checked = false;
        this.disabled = false;
        this.indeterminate = false;
        this.showCheckbox = false;
        this.showRowSelection = false;
        this.checkedChange = new EventEmitter();
        this.isNzShowExpandChanged = false;
        this.isNzShowCheckboxChanged = false;
        // TODO: move to host after View Engine deprecation
        this.elementRef.nativeElement.classList.add('tds-table-selection-column');
    }
    onCheckedChange(checked) {
        this.checked = checked;
        this.checkedChange.emit(checked);
    }
    ngOnChanges(changes) {
        const isFirstChange = (value) => value && value.firstChange && value.currentValue !== undefined;
        const { checked, selections, showExpand, showCheckbox } = changes;
        if (showExpand) {
            this.isNzShowExpandChanged = true;
        }
        if (showCheckbox) {
            this.isNzShowCheckboxChanged = true;
        }
        if (isFirstChange(selections) && !this.isNzShowExpandChanged) {
            this.showRowSelection = true;
        }
        if (isFirstChange(checked) && !this.isNzShowCheckboxChanged) {
            this.showCheckbox = true;
        }
    }
}
TDSThSelectionComponent.ɵfac = function TDSThSelectionComponent_Factory(t) { return new (t || TDSThSelectionComponent)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSThSelectionComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSThSelectionComponent, selectors: [["th", "selections", ""], ["th", "checked", ""], ["th", "showCheckbox", ""], ["th", "showRowSelection", ""]], inputs: { selections: "selections", checked: "checked", disabled: "disabled", indeterminate: "indeterminate", showCheckbox: "showCheckbox", showRowSelection: "showRowSelection" }, outputs: { checkedChange: "checkedChange" }, features: [i0.ɵɵNgOnChangesFeature], attrs: _c0, ngContentSelectors: _c1, decls: 2, vars: 6, consts: [[3, "checked", "disabled", "indeterminate", "listOfSelections", "showCheckbox", "showRowSelection", "checkedChange"]], template: function TDSThSelectionComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "tds-table-selection", 0);
        i0.ɵɵlistener("checkedChange", function TDSThSelectionComponent_Template_tds_table_selection_checkedChange_0_listener($event) { return ctx.onCheckedChange($event); });
        i0.ɵɵelementEnd();
        i0.ɵɵprojection(1);
    } if (rf & 2) {
        i0.ɵɵproperty("checked", ctx.checked)("disabled", ctx.disabled)("indeterminate", ctx.indeterminate)("listOfSelections", ctx.selections)("showCheckbox", ctx.showCheckbox)("showRowSelection", ctx.showRowSelection);
    } }, directives: [i1.TDSTableSelectionComponent], encapsulation: 2, changeDetection: 0 });
__decorate([
    InputBoolean()
], TDSThSelectionComponent.prototype, "showCheckbox", void 0);
__decorate([
    InputBoolean()
], TDSThSelectionComponent.prototype, "showRowSelection", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSThSelectionComponent, [{
        type: Component,
        args: [{
                selector: 'th[selections],th[checked],th[showCheckbox],th[showRowSelection]',
                preserveWhitespaces: false,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <tds-table-selection
      [checked]="checked"
      [disabled]="disabled"
      [indeterminate]="indeterminate"
      [listOfSelections]="selections"
      [showCheckbox]="showCheckbox"
      [showRowSelection]="showRowSelection"
      (checkedChange)="onCheckedChange($event)"
    ></tds-table-selection>
    <ng-content></ng-content>
  `,
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { selections: [{
            type: Input
        }], checked: [{
            type: Input
        }], disabled: [{
            type: Input
        }], indeterminate: [{
            type: Input
        }], showCheckbox: [{
            type: Input
        }], showRowSelection: [{
            type: Input
        }], checkedChange: [{
            type: Output
        }] }); })();
