// tds-ui/esm2020/core/selection/tdspseudoCheckbox.module.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { TDSPseudoCheckbox } from './pseudo-checkbox/pseudo-checkbox';
import * as i0 from "@angular/core";
export class TDSPseudoCheckboxModule {
}
TDSPseudoCheckboxModule.ɵfac = function TDSPseudoCheckboxModule_Factory(t) { return new (t || TDSPseudoCheckboxModule)(); };
TDSPseudoCheckboxModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSPseudoCheckboxModule });
TDSPseudoCheckboxModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSPseudoCheckboxModule, [{
        type: NgModule,
        args: [{
                imports: [],
                exports: [TDSPseudoCheckbox],
                declarations: [TDSPseudoCheckbox]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSPseudoCheckboxModule, { declarations: [TDSPseudoCheckbox], exports: [TDSPseudoCheckbox] }); })();
