// tds-ui/esm2020/filter-status/filter-status-link.directive.mjs
import { Directive, Optional, Self } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * This component is for catching `routerLink` directive.
 */
export class TDSFilterStatusLinkDirective {
    constructor(elementRef, routerLink, routerLinkWithHref) {
        this.elementRef = elementRef;
        this.routerLink = routerLink;
        this.routerLinkWithHref = routerLinkWithHref;
    }
}
TDSFilterStatusLinkDirective.ɵfac = function TDSFilterStatusLinkDirective_Factory(t) { return new (t || TDSFilterStatusLinkDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.RouterLink, 10), i0.ɵɵdirectiveInject(i1.RouterLinkWithHref, 10)); };
TDSFilterStatusLinkDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSFilterStatusLinkDirective, selectors: [["a", "tds-filter-status-router-link", ""]], exportAs: ["tdsFilterStatusRouterLink"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFilterStatusLinkDirective, [{
        type: Directive,
        args: [{
                selector: 'a[tds-filter-status-router-link]',
                exportAs: 'tdsFilterStatusRouterLink'
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.RouterLink, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }, { type: i1.RouterLinkWithHref, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }]; }, null); })();
