// tds-ui/esm2020/filter-status/filter-status-link.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, Input, Optional, ViewEncapsulation } from "@angular/core";
import { Subject } from "rxjs";
import { takeUntil } from "rxjs/operators";
import { TDSFilterStatusLinkDirective } from "./filter-status-link.directive";
import * as i0 from "@angular/core";
import * as i1 from "./filter-status.service";
const _c0 = ["*"];
export class TDSFilterStatusLinkComponent {
    constructor(elementRef, cdr, tdsFilterStatusService) {
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.tdsFilterStatusService = tdsFilterStatusService;
        this.active = false;
        this.disabled = false;
        this.value = null;
        this.count = null;
        this.name = null;
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        if (this.tdsFilterStatusService) {
            this.tdsFilterStatusService.disabled$.pipe(takeUntil(this.destroy$)).subscribe(disabled => {
                this.disabled = disabled;
                this.cdr.markForCheck();
            });
            this.tdsFilterStatusService.selected$.pipe(takeUntil(this.destroy$)).subscribe(value => {
                this.active = this.value === value;
                this.cdr.markForCheck();
            });
        }
    }
    onHostClick(event) {
        /** prevent label click triggered twice. **/
        event.stopPropagation();
        event.preventDefault();
        if (!this.disabled && !this.active) {
            if (this.tdsFilterStatusService) {
                this.tdsFilterStatusService.select(this.value);
            }
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    ngOnChanges(changes) {
        this.cdr.markForCheck();
    }
}
TDSFilterStatusLinkComponent.ɵfac = function TDSFilterStatusLinkComponent_Factory(t) { return new (t || TDSFilterStatusLinkComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSFilterStatusService, 8)); };
TDSFilterStatusLinkComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSFilterStatusLinkComponent, selectors: [["tds-filter-status-link"]], contentQueries: function TDSFilterStatusLinkComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSFilterStatusLinkDirective, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.linkDirective = _t.first);
    } }, hostAttrs: [1, "inline-block"], hostVars: 4, hostBindings: function TDSFilterStatusLinkComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function TDSFilterStatusLinkComponent_click_HostBindingHandler($event) { return ctx.onHostClick($event); });
    } if (rf & 2) {
        i0.ɵɵclassProp("cursor-pointer", !ctx.disabled)("cursor-not-allowed", ctx.disabled);
    } }, inputs: { disabled: "disabled", value: "value", count: "count", name: "name" }, exportAs: ["tdsFilterLink"], features: [i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSFilterStatusLinkComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFilterStatusLinkComponent, [{
        type: Component,
        args: [{ selector: 'tds-filter-status-link', exportAs: 'tdsFilterLink', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                    '(click)': 'onHostClick($event)',
                    '[class.cursor-pointer]': '!disabled',
                    '[class.cursor-not-allowed]': 'disabled',
                    'class': "inline-block"
                }, template: "<ng-content></ng-content> " }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.ChangeDetectorRef }, { type: i1.TDSFilterStatusService, decorators: [{
                type: Optional
            }] }]; }, { disabled: [{
            type: Input
        }], value: [{
            type: Input
        }], count: [{
            type: Input
        }], name: [{
            type: Input
        }], linkDirective: [{
            type: ContentChild,
            args: [TDSFilterStatusLinkDirective, { static: false }]
        }] }); })();
