// tds-ui/esm2020/filter-status/filter-status-item.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, Input, Optional, TemplateRef, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TDSFilterStatusTemplateDirective } from './filter-status.directive';
import * as i0 from "@angular/core";
import * as i1 from "./filter-status.service";
import * as i2 from "@angular/common";
function TDSFilterStatusItemComponent_ng_template_0_Template(rf, ctx) { }
const _c0 = function (a0, a1) { return { "text-white bg-primary-1": a0, "text-neutral-1-500  dark:text-d-neutral-1-500 bg-neutral-3-100 dark:bg-d-neutral-1-400  group-hover:text-white group-hover:bg-primary-1": a1 }; };
function TDSFilterStatusItemComponent_ng_template_1_div_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext();
    const active_r3 = ctx_r9.active;
    const count_r5 = ctx_r9.count;
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(2, _c0, active_r3, !active_r3));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", count_r5, " ");
} }
const _c1 = function (a0, a1) { return { "border-b-3 border-primary-1": a0, "hover:border-b-3 hover:border-primary-1 group": a1 }; };
const _c2 = function (a0, a1) { return { "text-neutral-1-900 dark:text-d-neutral-1-900": a0, "text-neutral-1-400 dark:text-d-neutral-1-400 group-hover:text-neutral-1-900 dark:group-hover:text-d-neutral-1-900": a1 }; };
function TDSFilterStatusItemComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 2)(1, "div", 3);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(3, TDSFilterStatusItemComponent_ng_template_1_div_3_Template, 2, 5, "div", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const active_r3 = ctx.active;
    const disabled_r4 = ctx.disabled;
    const count_r5 = ctx.count;
    const name_r6 = ctx.name;
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(4, _c1, active_r3, !disabled_r4));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(7, _c2, active_r3, !active_r3));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(name_r6);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", count_r5);
} }
const _c3 = function (a0, a1, a2, a3, a4) { return { active: a0, disabled: a1, name: a2, count: a3, $implicit: a4 }; };
export class TDSFilterStatusItemComponent {
    constructor(elementRef, cdr, tdsFilterStatusService) {
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.tdsFilterStatusService = tdsFilterStatusService;
        this.active = false;
        this.disabled = false;
        this.value = null;
        this.count = null;
        this.name = null;
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        if (this.tdsFilterStatusService) {
            this.tdsFilterStatusService.disabled$.pipe(takeUntil(this.destroy$)).subscribe(disabled => {
                this.disabled = disabled;
                this.cdr.markForCheck();
            });
            this.tdsFilterStatusService.selected$.pipe(takeUntil(this.destroy$)).subscribe(value => {
                this.active = this.value === value;
                this.cdr.markForCheck();
            });
        }
    }
    onHostClick(event) {
        /** prevent label click triggered twice. **/
        event.stopPropagation();
        event.preventDefault();
        if (!this.disabled && !this.active) {
            if (this.tdsFilterStatusService) {
                this.tdsFilterStatusService.select(this.value);
            }
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    ngOnChanges(changes) {
        this.cdr.markForCheck();
    }
}
TDSFilterStatusItemComponent.ɵfac = function TDSFilterStatusItemComponent_Factory(t) { return new (t || TDSFilterStatusItemComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSFilterStatusService, 8)); };
TDSFilterStatusItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSFilterStatusItemComponent, selectors: [["tds-filter-status-item"]], contentQueries: function TDSFilterStatusItemComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, TDSFilterStatusTemplateDirective, 5, TemplateRef);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.customTemplate = _t.first);
    } }, hostAttrs: [1, "inline-block"], hostVars: 4, hostBindings: function TDSFilterStatusItemComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function TDSFilterStatusItemComponent_click_HostBindingHandler($event) { return ctx.onHostClick($event); });
    } if (rf & 2) {
        i0.ɵɵclassProp("cursor-pointer", !ctx.disabled)("cursor-not-allowed", ctx.disabled);
    } }, inputs: { disabled: "disabled", value: "value", count: "count", name: "name" }, exportAs: ["tdsFilterTab"], features: [i0.ɵɵNgOnChangesFeature], decls: 3, vars: 8, consts: [[3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["defaultTemplate", ""], [1, "p-4", "flex", "items-center", 3, "ngClass"], [1, "text-body-2", "mr-2", "font-semibold", 3, "ngClass"], ["class", "text-caption-2 font-semibold py-0.5 px-1.5 rounded-full flex", 3, "ngClass", 4, "ngIf"], [1, "text-caption-2", "font-semibold", "py-0.5", "px-1.5", "rounded-full", "flex", 3, "ngClass"]], template: function TDSFilterStatusItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSFilterStatusItemComponent_ng_template_0_Template, 0, 0, "ng-template", 0);
        i0.ɵɵtemplate(1, TDSFilterStatusItemComponent_ng_template_1_Template, 4, 10, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = i0.ɵɵreference(2);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.customTemplate || _r1)("ngTemplateOutletContext", i0.ɵɵpureFunction5(2, _c3, ctx.active, ctx.disabled, ctx.name, ctx.count, ctx.value));
    } }, directives: [i2.NgTemplateOutlet, i2.NgClass, i2.NgIf], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFilterStatusItemComponent, [{
        type: Component,
        args: [{ selector: 'tds-filter-status-item', exportAs: 'tdsFilterTab', preserveWhitespaces: false, encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, host: {
                    '(click)': 'onHostClick($event)',
                    '[class.cursor-pointer]': '!disabled',
                    '[class.cursor-not-allowed]': 'disabled',
                    'class': "inline-block"
                }, template: "<!-- <ng-content></ng-content> -->\r\n<ng-template [ngTemplateOutlet]=\"customTemplate || defaultTemplate\"\r\n    [ngTemplateOutletContext]=\"{active:active, disabled:disabled, name:name, count:count, $implicit:value }\">\r\n</ng-template>\r\n<ng-template #defaultTemplate let-active='active' let-disabled=\"disabled\" let-count='count' let-name='name' let-value>\r\n    <div class=\"p-4 flex items-center \" [ngClass]=\"{'border-b-3 border-primary-1':active,'hover:border-b-3 hover:border-primary-1 group':!disabled}\">\r\n        <div class=\"text-body-2 mr-2 font-semibold\" [ngClass]=\"{'text-neutral-1-900 dark:text-d-neutral-1-900':active,'text-neutral-1-400 dark:text-d-neutral-1-400 group-hover:text-neutral-1-900 dark:group-hover:text-d-neutral-1-900':!active}\">{{name}}</div>\r\n        <div *ngIf=\"count\" class=\"text-caption-2 font-semibold py-0.5 px-1.5 rounded-full flex\" [ngClass]=\"{'text-white bg-primary-1':active,'text-neutral-1-500  dark:text-d-neutral-1-500 bg-neutral-3-100 dark:bg-d-neutral-1-400  group-hover:text-white group-hover:bg-primary-1':!active}\"> \r\n            {{count}}\r\n        </div>\r\n    </div>\r\n</ng-template>" }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.ChangeDetectorRef }, { type: i1.TDSFilterStatusService, decorators: [{
                type: Optional
            }] }]; }, { disabled: [{
            type: Input
        }], value: [{
            type: Input
        }], count: [{
            type: Input
        }], name: [{
            type: Input
        }], customTemplate: [{
            type: ContentChild,
            args: [TDSFilterStatusTemplateDirective, { read: TemplateRef }]
        }] }); })();
