// tds-ui/esm2020/filter-status/filter-status.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSBadgeModule } from 'tds-ui/badges';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSFilterStatusItemComponent } from './filter-status-item.component';
import { TDSFilterStatusLinkComponent } from './filter-status-link.component';
import { TDSFilterStatusLinkDirective } from './filter-status-link.directive';
import { TDSFilterStatusComponent } from './filter-status.component';
import { TDSFilterStatusTemplateDirective } from './filter-status.directive';
import * as i0 from "@angular/core";
export class TDSFilterStatusModule {
}
TDSFilterStatusModule.ɵfac = function TDSFilterStatusModule_Factory(t) { return new (t || TDSFilterStatusModule)(); };
TDSFilterStatusModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSFilterStatusModule });
TDSFilterStatusModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule, TDSBadgeModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSFilterStatusModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, TDSOutletModule, TDSBadgeModule],
                declarations: [
                    TDSFilterStatusComponent,
                    TDSFilterStatusItemComponent,
                    TDSFilterStatusTemplateDirective,
                    TDSFilterStatusLinkComponent,
                    TDSFilterStatusLinkDirective
                ],
                exports: [
                    TDSFilterStatusComponent,
                    TDSFilterStatusItemComponent,
                    TDSFilterStatusTemplateDirective,
                    TDSFilterStatusLinkComponent,
                    TDSFilterStatusLinkDirective
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSFilterStatusModule, { declarations: [TDSFilterStatusComponent,
        TDSFilterStatusItemComponent,
        TDSFilterStatusTemplateDirective,
        TDSFilterStatusLinkComponent,
        TDSFilterStatusLinkDirective], imports: [BidiModule, CommonModule, TDSOutletModule, TDSBadgeModule], exports: [TDSFilterStatusComponent,
        TDSFilterStatusItemComponent,
        TDSFilterStatusTemplateDirective,
        TDSFilterStatusLinkComponent,
        TDSFilterStatusLinkDirective] }); })();
