// tds-ui/esm2020/date-picker/calendar-footer.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { TDSHelperObject } from 'tds-ui/shared/utility';
import { CandyDate } from 'tds-ui/core/time';
import { isNonEmptyString, isTemplateRef } from 'tds-ui/core/util';
import { transCompatFormat } from './lib/util';
import { PREFIX_CLASS } from './util';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/i18n";
import * as i2 from "./date-picker.service";
import * as i3 from "@angular/common";
import * as i4 from "tds-ui/button";
function CalendarFooterComponent_div_1_ng_container_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function CalendarFooterComponent_div_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, CalendarFooterComponent_div_1_ng_container_2_ng_container_1_Template, 1, 0, "ng-container", 5);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r3.extraFooter);
} }
function CalendarFooterComponent_div_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "span", 6);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("innerHTML", ctx_r4.extraFooter, i0.ɵɵsanitizeHtml);
} }
function CalendarFooterComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵelementContainerStart(1, 3);
    i0.ɵɵtemplate(2, CalendarFooterComponent_div_1_ng_container_2_Template, 2, 1, "ng-container", 4);
    i0.ɵɵtemplate(3, CalendarFooterComponent_div_1_ng_container_3_Template, 2, 1, "ng-container", 4);
    i0.ɵɵelementContainerEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassMapInterpolate1("", ctx_r0.prefixCls, "-footer-extra");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitch", true);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.isTemplateRef(ctx_r0.extraFooter));
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.isNonEmptyString(ctx_r0.extraFooter));
} }
function CalendarFooterComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 7)(1, "a", 8);
    i0.ɵɵlistener("click", function CalendarFooterComponent_div_2_Template_a_click_1_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return ctx_r6.isTodayDisabled ? null : ctx_r6.onClickToday(); });
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵclassMapInterpolate2("", ctx_r1.prefixCls, "-today-btn ", ctx_r1.isTodayDisabled ? ctx_r1.prefixCls + "-today-btn-disabled" : "text-primary-1", " text-body-2 font-semibold uppercase");
    i0.ɵɵpropertyInterpolate("title", ctx_r1.todayTitle);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r1.locale.today, " ");
} }
function CalendarFooterComponent_div_3_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 14)(1, "div", 15)(2, "div", 16);
    i0.ɵɵelement(3, "span", 17);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 18);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(6, "div", 19);
    i0.ɵɵtext(7, " - ");
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(8, "div", 18);
    i0.ɵɵtext(9);
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate1(" ", ctx_r8.startDate, " ");
    i0.ɵɵadvance(4);
    i0.ɵɵtextInterpolate1(" ", ctx_r8.endDate, " ");
} }
function CalendarFooterComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r10 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 9);
    i0.ɵɵtemplate(1, CalendarFooterComponent_div_3_div_1_Template, 10, 2, "div", 10);
    i0.ɵɵelementStart(2, "div", 11)(3, "button", 12);
    i0.ɵɵlistener("click", function CalendarFooterComponent_div_3_Template_button_click_3_listener() { i0.ɵɵrestoreView(_r10); const ctx_r9 = i0.ɵɵnextContext(); return ctx_r9.onClickCancel(); });
    i0.ɵɵtext(4, "Cancel");
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(5, "button", 13);
    i0.ɵɵlistener("click", function CalendarFooterComponent_div_3_Template_button_click_5_listener() { i0.ɵɵrestoreView(_r10); const ctx_r11 = i0.ɵɵnextContext(); return ctx_r11.okDisabled ? null : ctx_r11.clickOk.emit(); });
    i0.ɵɵtext(6, "Apply");
    i0.ɵɵelementEnd()()();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r2.hasTimePicker && (ctx_r2.startDate || ctx_r2.endDate));
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("disabled", ctx_r2.okDisabled);
} }
export class CalendarFooterComponent {
    constructor(dateHelper, datePickerService) {
        this.dateHelper = dateHelper;
        this.datePickerService = datePickerService;
        this.showToday = false;
        this.showNow = false;
        this.hasTimePicker = false;
        this.isRange = false;
        this.okDisabled = false;
        this.rangeQuickSelector = null;
        this.clickOk = new EventEmitter();
        this.clickCancel = new EventEmitter();
        this.clickToday = new EventEmitter();
        this.prefixCls = PREFIX_CLASS;
        this.isTemplateRef = isTemplateRef;
        this.isNonEmptyString = isNonEmptyString;
        this.isTodayDisabled = false;
        this.todayTitle = '';
        this.destroy$ = new Subject();
        this.startDate = "";
        this.endDate = "";
    }
    ngOnChanges(changes) {
        const now = new Date();
        if (changes.disabledDate) {
            this.isTodayDisabled = !!(this.disabledDate && this.disabledDate(now));
        }
        if (changes.locale) {
            // NOTE: Compat for DatePipe formatting rules
            const dateFormat = transCompatFormat(this.locale.dateFormat);
            this.todayTitle = this.dateHelper.format(now, dateFormat);
        }
        if (changes.listDate) {
            if (this.isRange) {
                const dateFormat = transCompatFormat(this.locale.dateFormat);
                if (TDSHelperObject.hasValue(this.listDate)) {
                    let startDate = this.listDate[0];
                    let endDate = this.listDate[1];
                    if (TDSHelperObject.hasValue(startDate)) {
                        this.startDate = this.dateHelper.format(startDate.nativeDate, dateFormat);
                    }
                    else {
                        this.startDate = "";
                    }
                    if (TDSHelperObject.hasValue(endDate)) {
                        this.endDate = this.dateHelper.format(endDate.nativeDate, dateFormat);
                    }
                    else {
                        this.endDate = "";
                    }
                }
                else {
                    this.startDate = "";
                    this.endDate = "";
                }
            }
        }
    }
    onClickToday() {
        const now = new CandyDate();
        this.clickToday.emit(now.clone()); // To prevent the "now" being modified from outside, we use clone
    }
    onClickCancel() {
        this.clickCancel.emit(); // To prevent the "now" being modified from outside, we use clone
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
CalendarFooterComponent.ɵfac = function CalendarFooterComponent_Factory(t) { return new (t || CalendarFooterComponent)(i0.ɵɵdirectiveInject(i1.DateHelperService), i0.ɵɵdirectiveInject(i2.DatePickerService)); };
CalendarFooterComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: CalendarFooterComponent, selectors: [["calendar-footer"]], inputs: { locale: "locale", showToday: "showToday", showNow: "showNow", hasTimePicker: "hasTimePicker", isRange: "isRange", listDate: "listDate", okDisabled: "okDisabled", disabledDate: "disabledDate", extraFooter: "extraFooter", rangeQuickSelector: "rangeQuickSelector" }, outputs: { clickOk: "clickOk", clickCancel: "clickCancel", clickToday: "clickToday" }, exportAs: ["calendarFooter"], features: [i0.ɵɵNgOnChangesFeature], decls: 4, vars: 6, consts: [[3, "class", 4, "ngIf"], ["class", "flex justify-center px-4 py-2 w-full", 4, "ngIf"], ["class", "flex justify-end items-center px-4 py-2 w-full", 4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngTemplateOutlet"], [3, "innerHTML"], [1, "flex", "justify-center", "px-4", "py-2", "w-full"], ["role", "button", 3, "title", "click"], [1, "flex", "justify-end", "items-center", "px-4", "py-2", "w-full"], ["class", "px-4", 4, "ngIf"], [1, "ml-2"], ["tds-button", "", "color", "secondary", 1, "mr-2", 3, "click"], ["tds-button", "", "color", "primary", 3, "disabled", "click"], [1, "px-4"], [1, "flex", "items-center", "text-neutral-1-400"], [1, "flex"], [1, "text-lg", "leading-5", "tdsi-calendar-fill"], [1, "px-2", "text-body-2", "font-regular", "text-center"], [1, "text-body-2", "font-regular"]], template: function CalendarFooterComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div");
        i0.ɵɵtemplate(1, CalendarFooterComponent_div_1_Template, 4, 6, "div", 0);
        i0.ɵɵtemplate(2, CalendarFooterComponent_div_2_Template, 3, 6, "div", 1);
        i0.ɵɵtemplate(3, CalendarFooterComponent_div_3_Template, 7, 2, "div", 2);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵclassMapInterpolate1("", ctx.prefixCls, "-footer w-full  border-0 border-t  border-neutral-2-200 dark:border-d-neutral-2-200");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.extraFooter);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.isRange);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isRange);
    } }, directives: [i3.NgIf, i3.NgSwitch, i3.NgSwitchCase, i3.NgTemplateOutlet, i4.TDSButtonComponent], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(CalendarFooterComponent, [{
        type: Component,
        args: [{ encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, selector: 'calendar-footer', exportAs: 'calendarFooter', template: "<div class=\"{{ prefixCls }}-footer w-full  border-0 border-t  border-neutral-2-200 dark:border-d-neutral-2-200\">\r\n    <div *ngIf=\"extraFooter\" class=\"{{ prefixCls }}-footer-extra\">\r\n        <ng-container [ngSwitch]=\"true\">\r\n            <ng-container *ngSwitchCase=\"isTemplateRef(extraFooter)\">\r\n                <ng-container *ngTemplateOutlet=\"$any(extraFooter)\"></ng-container>\r\n            </ng-container>\r\n            <ng-container *ngSwitchCase=\"isNonEmptyString(extraFooter)\">\r\n                <span [innerHTML]=\"extraFooter\"></span>\r\n            </ng-container>\r\n        </ng-container>\r\n    </div>\r\n    <div class=\"flex justify-center px-4 py-2 w-full\" *ngIf=\"!isRange\">\r\n        <a class=\"{{ prefixCls }}-today-btn {{ isTodayDisabled ? prefixCls + '-today-btn-disabled' : 'text-primary-1' }} text-body-2 font-semibold uppercase\"\r\n            role=\"button\" (click)=\"isTodayDisabled ? null : onClickToday()\" title=\"{{ todayTitle }}\">\r\n            {{ locale.today }}\r\n        </a>\r\n    </div>\r\n    <div class=\"flex justify-end items-center px-4 py-2 w-full\" *ngIf=\"isRange\">\r\n        <div *ngIf=\"!hasTimePicker && (startDate || endDate)\" class=\"px-4\">\r\n            <div class=\"flex items-center text-neutral-1-400\">\r\n                <div class=\"flex\">\r\n                    <span class=\"text-lg leading-5  tdsi-calendar-fill\"></span>\r\n                </div>\r\n                <div class=\"px-2 text-body-2 font-regular text-center\">\r\n                  {{startDate}}\r\n                </div>\r\n                <div class=\"text-body-2 font-regular\">\r\n                    -\r\n                </div>\r\n                <div class=\"px-2  text-body-2 font-regular text-center\">\r\n                    {{endDate}}\r\n                </div>\r\n            </div>\r\n        </div>\r\n        <div class=\"ml-2\">\r\n            <button tds-button color=\"secondary\" class=\"mr-2\" (click)=\"onClickCancel()\">Cancel</button>\r\n            <button tds-button color=\"primary\"   [disabled]=\"okDisabled\"\r\n            (click)=\"okDisabled ? null : clickOk.emit()\">Apply</button>\r\n        </div>\r\n    </div>\r\n</div>" }]
    }], function () { return [{ type: i1.DateHelperService }, { type: i2.DatePickerService }]; }, { locale: [{
            type: Input
        }], showToday: [{
            type: Input
        }], showNow: [{
            type: Input
        }], hasTimePicker: [{
            type: Input
        }], isRange: [{
            type: Input
        }], listDate: [{
            type: Input
        }], okDisabled: [{
            type: Input
        }], disabledDate: [{
            type: Input
        }], extraFooter: [{
            type: Input
        }], rangeQuickSelector: [{
            type: Input
        }], clickOk: [{
            type: Output
        }], clickCancel: [{
            type: Output
        }], clickToday: [{
            type: Output
        }] }); })();
