// tds-ui/esm2020/date-picker/date-picker.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSTimePickerModule } from 'tds-ui/time-picker';
import { CalendarFooterComponent } from './calendar-footer.component';
import { TDSDatePickerComponent } from './date-picker.component';
import { DateRangePopupComponent } from './date-range-popup.component';
import { InnerPopupComponent } from './inner-popup.component';
import { LibPackerModule } from './lib/lib-packer.module';
import { TDSMonthPickerComponent } from './month-picker.component';
import { TDSRangePickerComponent } from './range-picker.component';
import { TDSWeekPickerComponent } from './week-picker.component';
import { TDSYearPickerComponent } from './year-picker.component';
import * as i0 from "@angular/core";
export class TDSDatePickerModule {
}
TDSDatePickerModule.ɵfac = function TDSDatePickerModule_Factory(t) { return new (t || TDSDatePickerModule)(); };
TDSDatePickerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSDatePickerModule });
TDSDatePickerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            FormsModule,
            OverlayModule,
            LibPackerModule,
            TDSOutletModule,
            TDSTimePickerModule,
            TDSButtonModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDatePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    FormsModule,
                    OverlayModule,
                    LibPackerModule,
                    TDSOutletModule,
                    TDSTimePickerModule,
                    TDSButtonModule,
                ],
                exports: [TDSDatePickerComponent,
                    TDSRangePickerComponent,
                    TDSMonthPickerComponent,
                    TDSYearPickerComponent,
                    TDSWeekPickerComponent],
                declarations: [
                    TDSDatePickerComponent,
                    TDSMonthPickerComponent,
                    TDSYearPickerComponent,
                    TDSWeekPickerComponent,
                    TDSRangePickerComponent,
                    CalendarFooterComponent,
                    InnerPopupComponent,
                    DateRangePopupComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSDatePickerModule, { declarations: [TDSDatePickerComponent,
        TDSMonthPickerComponent,
        TDSYearPickerComponent,
        TDSWeekPickerComponent,
        TDSRangePickerComponent,
        CalendarFooterComponent,
        InnerPopupComponent,
        DateRangePopupComponent], imports: [BidiModule,
        CommonModule,
        FormsModule,
        OverlayModule,
        LibPackerModule,
        TDSOutletModule,
        TDSTimePickerModule,
        TDSButtonModule], exports: [TDSDatePickerComponent,
        TDSRangePickerComponent,
        TDSMonthPickerComponent,
        TDSYearPickerComponent,
        TDSWeekPickerComponent] }); })();
