// tds-ui/esm2020/date-picker/month-picker.component.mjs
import { Directive, Host, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./date-picker.component";
// tslint:disable-next-line:directive-class-suffix
export class TDSMonthPickerComponent {
    constructor(datePicker) {
        this.datePicker = datePicker;
        this.datePicker.mode = 'month';
    }
}
TDSMonthPickerComponent.ɵfac = function TDSMonthPickerComponent_Factory(t) { return new (t || TDSMonthPickerComponent)(i0.ɵɵdirectiveInject(i1.TDSDatePickerComponent, 9)); };
TDSMonthPickerComponent.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSMonthPickerComponent, selectors: [["tds-month-picker"]], exportAs: ["tdsMonthPicker"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSMonthPickerComponent, [{
        type: Directive,
        args: [{
                selector: 'tds-month-picker',
                exportAs: 'tdsMonthPicker'
            }]
    }], function () { return [{ type: i1.TDSDatePickerComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }] }]; }, null); })();
