// tds-ui/esm2020/date-picker/range-picker.component.mjs
import { Directive, Host, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./date-picker.component";
// tslint:disable-next-line:directive-class-suffix
export class TDSRangePickerComponent {
    constructor(datePicker) {
        this.datePicker = datePicker;
        this.datePicker.isRange = true;
    }
}
TDSRangePickerComponent.ɵfac = function TDSRangePickerComponent_Factory(t) { return new (t || TDSRangePickerComponent)(i0.ɵɵdirectiveInject(i1.TDSDatePickerComponent, 9)); };
TDSRangePickerComponent.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSRangePickerComponent, selectors: [["tds-range-picker"]], exportAs: ["tdsRangePicker"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSRangePickerComponent, [{
        type: Directive,
        args: [{
                selector: 'tds-range-picker',
                exportAs: 'tdsRangePicker'
            }]
    }], function () { return [{ type: i1.TDSDatePickerComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }] }]; }, null); })();
