// tds-ui/esm2020/date-picker/week-picker.component.mjs
import { Directive, Host, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./date-picker.component";
// tslint:disable-next-line:directive-class-suffix
export class TDSWeekPickerComponent {
    constructor(datePicker) {
        this.datePicker = datePicker;
        this.datePicker.mode = 'week';
    }
}
TDSWeekPickerComponent.ɵfac = function TDSWeekPickerComponent_Factory(t) { return new (t || TDSWeekPickerComponent)(i0.ɵɵdirectiveInject(i1.TDSDatePickerComponent, 9)); };
TDSWeekPickerComponent.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSWeekPickerComponent, selectors: [["tds-week-picker"]], exportAs: ["tdsWeekPicker"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSWeekPickerComponent, [{
        type: Directive,
        args: [{
                selector: 'tds-week-picker',
                exportAs: 'tdsWeekPicker',
            }]
    }], function () { return [{ type: i1.TDSDatePickerComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }] }]; }, null); })();
