// tds-ui/esm2020/date-picker/year-picker.component.mjs
import { Directive, Host, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./date-picker.component";
// tslint:disable-next-line:directive-class-suffix
export class TDSYearPickerComponent {
    constructor(datePicker) {
        this.datePicker = datePicker;
        this.datePicker.mode = 'year';
    }
}
TDSYearPickerComponent.ɵfac = function TDSYearPickerComponent_Factory(t) { return new (t || TDSYearPickerComponent)(i0.ɵɵdirectiveInject(i1.TDSDatePickerComponent, 9)); };
TDSYearPickerComponent.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSYearPickerComponent, selectors: [["tds-year-picker"]], exportAs: ["tdsYearPicker"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSYearPickerComponent, [{
        type: Directive,
        args: [{
                selector: 'tds-year-picker',
                exportAs: 'tdsYearPicker'
            }]
    }], function () { return [{ type: i1.TDSDatePickerComponent, decorators: [{
                type: Optional
            }, {
                type: Host
            }] }]; }, null); })();
