// tds-ui/esm2020/dropdown/dropdown-button.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class TDSDropdownButtonDirective {
    constructor(renderer, 
    // @Host() @Optional() private nzButtonGroupComponent: NzButtonGroupComponent,
    elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
    }
    ngAfterViewInit() {
        const parentElement = this.renderer.parentNode(this.elementRef.nativeElement);
        // if (this.nzButtonGroupComponent && parentElement) {
        //   this.renderer.addClass(parentElement, 'tds-dropdown-button');
        // }
    }
}
TDSDropdownButtonDirective.ɵfac = function TDSDropdownButtonDirective_Factory(t) { return new (t || TDSDropdownButtonDirective)(i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.ElementRef)); };
TDSDropdownButtonDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSDropdownButtonDirective, selectors: [["", "tds-button", "", "tds-dropdown", ""]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDropdownButtonDirective, [{
        type: Directive,
        args: [{
                selector: '[tds-button][tds-dropdown]'
            }]
    }], function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }]; }, null); })();
