// tds-ui/esm2020/dropdown/dropdown.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSNoAnimationModule } from 'tds-ui/core/no-animation';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSOverlayModule } from 'tds-ui/core/overlay';
import { TDSMenuModule } from 'tds-ui/menu';
import { TDSContextMenuServiceModule } from './context-menu.service.module';
import { TDSDropDownADirective } from './dropdown-a.directive';
import { TDSDropdownButtonDirective } from './dropdown-button.directive';
import { TDSDropDownItemDirective } from './dropdown-item.directive';
import { TDSDropdownMenuComponent } from './dropdown-menu.component';
import { TDSDropDownDirective } from './dropdown.directive';
import * as i0 from "@angular/core";
export class TDSDropDownModule {
}
TDSDropDownModule.ɵfac = function TDSDropDownModule_Factory(t) { return new (t || TDSDropDownModule)(); };
TDSDropDownModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSDropDownModule });
TDSDropDownModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            OverlayModule,
            FormsModule,
            TDSButtonModule,
            TDSMenuModule,
            TDSNoAnimationModule,
            PlatformModule,
            TDSOverlayModule,
            TDSContextMenuServiceModule,
            TDSOutletModule
        ], TDSMenuModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDropDownModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BidiModule,
                    CommonModule,
                    OverlayModule,
                    FormsModule,
                    TDSButtonModule,
                    TDSMenuModule,
                    TDSNoAnimationModule,
                    PlatformModule,
                    TDSOverlayModule,
                    TDSContextMenuServiceModule,
                    TDSOutletModule
                ],
                declarations: [TDSDropDownDirective,
                    TDSDropDownADirective,
                    TDSDropdownMenuComponent,
                    TDSDropdownButtonDirective,
                    TDSDropDownItemDirective
                ],
                exports: [TDSMenuModule,
                    TDSDropDownDirective,
                    TDSDropDownADirective,
                    TDSDropdownMenuComponent,
                    TDSDropdownButtonDirective,
                    TDSDropDownItemDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSDropDownModule, { declarations: [TDSDropDownDirective,
        TDSDropDownADirective,
        TDSDropdownMenuComponent,
        TDSDropdownButtonDirective,
        TDSDropDownItemDirective], imports: [BidiModule,
        CommonModule,
        OverlayModule,
        FormsModule,
        TDSButtonModule,
        TDSMenuModule,
        TDSNoAnimationModule,
        PlatformModule,
        TDSOverlayModule,
        TDSContextMenuServiceModule,
        TDSOutletModule], exports: [TDSMenuModule,
        TDSDropDownDirective,
        TDSDropDownADirective,
        TDSDropdownMenuComponent,
        TDSDropdownButtonDirective,
        TDSDropDownItemDirective] }); })();
