// tds-ui/esm2020/dropdown/dropdown-item.directive.mjs
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/menu";
export class TDSDropDownItemDirective {
    constructor(TDSMenuService, cdr) {
        this.TDSMenuService = TDSMenuService;
        this.cdr = cdr;
        this.destroy$ = new Subject();
        this.selected$ = new Subject();
        this.disabled = false;
        this.selected = false;
    }
    /** clear all item selected status except this */
    clickMenuItem(e) {
        if (this.disabled) {
            e.preventDefault();
            e.stopPropagation();
        }
        else {
            this.TDSMenuService.onDescendantMenuItemClick(this);
        }
    }
    ngOnInit() {
    }
    ngAfterContentInit() {
    }
    ngOnChanges(changes) {
        if (changes.disabled) {
            this.cdr.markForCheck();
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSDropDownItemDirective.ɵfac = function TDSDropDownItemDirective_Factory(t) { return new (t || TDSDropDownItemDirective)(i0.ɵɵdirectiveInject(i1.TDSMenuService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSDropDownItemDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSDropDownItemDirective, selectors: [["", "tds-dropdown-item", ""]], hostAttrs: [1, "tds-dropdown-item", "text-neutral-1-900", "dark:text-d-neutral-1-900", "text-body-2", "px-3", "py-1.5", "font-regular", "dark:hover:bg-d-neutral-3-400", "hover:bg-neutral-3-50", "w-full"], hostVars: 2, hostBindings: function TDSDropDownItemDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function TDSDropDownItemDirective_click_HostBindingHandler($event) { return ctx.clickMenuItem($event); });
    } if (rf & 2) {
        i0.ɵɵclassProp("cursor-pointer", !ctx.disabled);
    } }, inputs: { disabled: "disabled", selected: "selected" }, exportAs: ["tdsDropdownItem"], features: [i0.ɵɵNgOnChangesFeature] });
__decorate([
    InputBoolean()
], TDSDropDownItemDirective.prototype, "disabled", void 0);
__decorate([
    InputBoolean()
], TDSDropDownItemDirective.prototype, "selected", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSDropDownItemDirective, [{
        type: Directive,
        args: [{
                selector: '[tds-dropdown-item]',
                exportAs: 'tdsDropdownItem',
                host: {
                    class: "tds-dropdown-item text-neutral-1-900  dark:text-d-neutral-1-900  text-body-2 px-3 py-1.5 font-regular dark:hover:bg-d-neutral-3-400 hover:bg-neutral-3-50  w-full",
                    '[class.cursor-pointer]': "!disabled",
                    '(click)': 'clickMenuItem($event)'
                }
            }]
    }], function () { return [{ type: i1.TDSMenuService }, { type: i0.ChangeDetectorRef }]; }, { disabled: [{
            type: Input
        }], selected: [{
            type: Input
        }] }); })();
