// tds-ui/esm2020/breadcrumb/breadcrumb-item.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "tds-ui/dropdown";
import * as i3 from "tds-ui/core/outlet";
function TDSBreadCrumbItemComponent_ng_container_0_ng_template_2_Template(rf, ctx) { }
function TDSBreadCrumbItemComponent_ng_container_0_span_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "span", 6);
} }
function TDSBreadCrumbItemComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "span", 3);
    i0.ɵɵtemplate(2, TDSBreadCrumbItemComponent_ng_container_0_ng_template_2_Template, 0, 0, "ng-template", 4);
    i0.ɵɵtemplate(3, TDSBreadCrumbItemComponent_ng_container_0_span_3_Template, 1, 0, "span", 5);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    const _r1 = i0.ɵɵreference(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsDropdownMenu", ctx_r0.overlay);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r1);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !!ctx_r0.overlay);
} }
function TDSBreadCrumbItemComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 7);
    i0.ɵɵprojection(1);
    i0.ɵɵelementEnd();
} }
function TDSBreadCrumbItemComponent_span_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "span", 10);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", ctx_r6.separator);
} }
function TDSBreadCrumbItemComponent_span_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 8);
    i0.ɵɵtemplate(1, TDSBreadCrumbItemComponent_span_3_ng_container_1_Template, 2, 1, "ng-container", 9);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r3.separator);
} }
const _c0 = ["*"];
export class TDSBreadCrumbItemComponent {
    constructor() {
        this.separator = '/';
        this.active = false;
    }
}
TDSBreadCrumbItemComponent.ɵfac = function TDSBreadCrumbItemComponent_Factory(t) { return new (t || TDSBreadCrumbItemComponent)(); };
TDSBreadCrumbItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSBreadCrumbItemComponent, selectors: [["tds-breadcrumb-item"]], hostAttrs: [1, "text-caption-1", "font-regular", "text-neutral-1-400", "dark:text-d-neutral-1-400", "flex", "items-center", "hover:text-primary-1", "dark:hover:text-primary-1", "cursor-pointer"], hostVars: 2, hostBindings: function TDSBreadCrumbItemComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("tds-breadcrumitem-active", ctx.active == true);
    } }, inputs: { overlay: "overlay", separator: "separator", active: "active" }, exportAs: ["tdsBreadcrumbItem"], ngContentSelectors: _c0, decls: 4, vars: 3, consts: [[4, "ngIf", "ngIfElse"], ["noMenuTpl", ""], ["class", "tds-breadcrumb-separator px-1 flex text-neutral-1-400 dark:text-d-neutral-1-400", 4, "ngIf"], ["tds-dropdown", "", 1, "tds-breadcrumb-overlay-link", "flex", "items-center", 3, "tdsDropdownMenu"], [3, "ngTemplateOutlet"], ["class", "flex items-center tdsi-arrow-down-fill text-base leading-4 px-1", 4, "ngIf"], [1, "flex", "items-center", "tdsi-arrow-down-fill", "text-base", "leading-4", "px-1"], [1, "tds-breadcrumb-link"], [1, "tds-breadcrumb-separator", "px-1", "flex", "text-neutral-1-400", "dark:text-d-neutral-1-400"], [4, "tdsStringTemplateOutlet"], [1, "text-base", "leading-4", 3, "ngClass"]], template: function TDSBreadCrumbItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtemplate(0, TDSBreadCrumbItemComponent_ng_container_0_Template, 4, 3, "ng-container", 0);
        i0.ɵɵtemplate(1, TDSBreadCrumbItemComponent_ng_template_1_Template, 2, 0, "ng-template", null, 1, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(3, TDSBreadCrumbItemComponent_span_3_Template, 2, 1, "span", 2);
    } if (rf & 2) {
        const _r1 = i0.ɵɵreference(2);
        i0.ɵɵproperty("ngIf", !!ctx.overlay)("ngIfElse", _r1);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.separator);
    } }, directives: [i1.NgIf, i2.TDSDropDownDirective, i1.NgTemplateOutlet, i3.TDSStringTemplateOutletDirective, i1.NgClass], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSBreadCrumbItemComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-breadcrumb-item',
                exportAs: 'tdsBreadcrumbItem',
                preserveWhitespaces: false,
                template: `
    <ng-container *ngIf="!!overlay; else noMenuTpl">
      <span class="tds-breadcrumb-overlay-link flex items-center" tds-dropdown [tdsDropdownMenu]="overlay">
        <ng-template [ngTemplateOutlet]="noMenuTpl"></ng-template>
        <span *ngIf="!!overlay" class='flex items-center tdsi-arrow-down-fill text-base leading-4 px-1'></span>
      </span>
    </ng-container>

    <ng-template #noMenuTpl>
      <span class="tds-breadcrumb-link">
        <ng-content></ng-content>
      </span>
    </ng-template>

    <span class="tds-breadcrumb-separator px-1 flex text-neutral-1-400 dark:text-d-neutral-1-400" *ngIf="separator">
      <ng-container *tdsStringTemplateOutlet="separator">
       <span class='text-base leading-4' [ngClass]="separator"></span>
      </ng-container>
    </span>
  `,
                host: {
                    '[class.tds-breadcrumitem-active]': 'active == true',
                    'class': 'text-caption-1 font-regular text-neutral-1-400 dark:text-d-neutral-1-400 flex items-center hover:text-primary-1 dark:hover:text-primary-1 cursor-pointer'
                }
            }]
    }], function () { return []; }, { overlay: [{
            type: Input
        }], separator: [{
            type: Input
        }], active: [{
            type: Input
        }] }); })();
