// tds-ui/esm2020/breadcrumb/breadcrumb.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSOverlayModule } from 'tds-ui/core/overlay';
import { TDSDropDownModule } from 'tds-ui/dropdown';
// import { TDSDropDownModule } from 'ng-zorro-antd/dropdown';
// import { NzIconModule } from 'ng-zorro-antd/icon';
import { TDSBreadCrumbItemComponent } from './breadcrumb-item.component';
import { TDSBreadCrumbSeparatorComponent } from './breadcrumb-separator.component';
import { TDSBreadCrumbComponent } from './breadcrumb.component';
import * as i0 from "@angular/core";
export class TDSBreadCrumbModule {
}
TDSBreadCrumbModule.ɵfac = function TDSBreadCrumbModule_Factory(t) { return new (t || TDSBreadCrumbModule)(); };
TDSBreadCrumbModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSBreadCrumbModule });
TDSBreadCrumbModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            TDSOutletModule,
            OverlayModule,
            TDSOverlayModule,
            TDSDropDownModule
        ], BidiModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSBreadCrumbModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    TDSOutletModule,
                    OverlayModule,
                    TDSOverlayModule,
                    TDSDropDownModule
                ],
                declarations: [
                    TDSBreadCrumbComponent,
                    TDSBreadCrumbItemComponent,
                    TDSBreadCrumbSeparatorComponent
                ],
                exports: [
                    BidiModule,
                    TDSBreadCrumbComponent,
                    TDSBreadCrumbItemComponent,
                    TDSBreadCrumbSeparatorComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSBreadCrumbModule, { declarations: [TDSBreadCrumbComponent,
        TDSBreadCrumbItemComponent,
        TDSBreadCrumbSeparatorComponent], imports: [CommonModule,
        TDSOutletModule,
        OverlayModule,
        TDSOverlayModule,
        TDSDropDownModule], exports: [BidiModule,
        TDSBreadCrumbComponent,
        TDSBreadCrumbItemComponent,
        TDSBreadCrumbSeparatorComponent] }); })();
