// tds-ui/esm2020/breadcrumb/breadcrumb-separator.component.mjs
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSBreadCrumbSeparatorComponent {
}
TDSBreadCrumbSeparatorComponent.ɵfac = function TDSBreadCrumbSeparatorComponent_Factory(t) { return new (t || TDSBreadCrumbSeparatorComponent)(); };
TDSBreadCrumbSeparatorComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSBreadCrumbSeparatorComponent, selectors: [["tds-breadcrumb-separator"]], exportAs: ["tdsBreadcrumbSeparator"], ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "tds-breadcrumb-separator", "text-neutral-1-400", "dark:text-d-neutral-1-400", "px-1", "flex"]], template: function TDSBreadCrumbSeparatorComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "span", 0);
        i0.ɵɵprojection(1);
        i0.ɵɵelementEnd();
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSBreadCrumbSeparatorComponent, [{
        type: Component,
        args: [{
                selector: 'tds-breadcrumb-separator',
                exportAs: 'tdsBreadcrumbSeparator',
                template: `
    <span class="tds-breadcrumb-separator text-neutral-1-400 dark:text-d-neutral-1-400 px-1 flex">
      <ng-content></ng-content>
    </span>
  `
            }]
    }], null, null); })();
