// tds-ui/esm2020/alert/alert.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSAlertComponent } from './alert.component';
import * as i0 from "@angular/core";
export class TDSAlertModule {
}
TDSAlertModule.ɵfac = function TDSAlertModule_Factory(t) { return new (t || TDSAlertModule)(); };
TDSAlertModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSAlertModule });
TDSAlertModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, TDSOutletModule, TDSButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSAlertModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSAlertComponent],
                exports: [TDSAlertComponent],
                imports: [BidiModule, CommonModule, TDSOutletModule, TDSButtonModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSAlertModule, { declarations: [TDSAlertComponent], imports: [BidiModule, CommonModule, TDSOutletModule, TDSButtonModule], exports: [TDSAlertComponent] }); })();
