// tds-ui/esm2020/empty/empty.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/i18n";
import * as i2 from "@angular/common";
import * as i3 from "tds-ui/core/outlet";
import * as i4 from "./partial/default";
import * as i5 from "./partial/simple";
function TDSEmptyComponent_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "img", 5);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("src", ctx_r5.notFoundImage, i0.ɵɵsanitizeUrl)("alt", ctx_r5.isContentString ? ctx_r5.notFoundContent : "empty");
} }
function TDSEmptyComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TDSEmptyComponent_ng_container_1_ng_container_1_Template, 2, 2, "ng-container", 4);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r0.notFoundImage);
} }
function TDSEmptyComponent_tds_empty_default_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-empty-default");
} }
function TDSEmptyComponent_tds_empty_simple_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-empty-simple");
} }
function TDSEmptyComponent_p_4_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r6.isContentString ? ctx_r6.notFoundContent : ctx_r6.locale["description"], " ");
} }
function TDSEmptyComponent_p_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "p", 6);
    i0.ɵɵtemplate(1, TDSEmptyComponent_p_4_ng_container_1_Template, 2, 1, "ng-container", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r3.notFoundContent);
} }
function TDSEmptyComponent_div_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r7.notFoundFooter, " ");
} }
function TDSEmptyComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 7);
    i0.ɵɵtemplate(1, TDSEmptyComponent_div_5_ng_container_1_Template, 2, 1, "ng-container", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r4.notFoundFooter);
} }
const TDSEmptyDefaultImages = ['default', 'simple'];
export class TDSEmptyComponent {
    constructor(i18n, cdr) {
        this.i18n = i18n;
        this.cdr = cdr;
        this.notFoundImage = 'default';
        this.isContentString = false;
        this.isImageBuildIn = true;
        this.destroy$ = new Subject();
    }
    ngOnChanges(changes) {
        const { notFoundContent, notFoundImage } = changes;
        if (notFoundContent) {
            const content = notFoundContent.currentValue;
            this.isContentString = typeof content === 'string';
        }
        if (notFoundImage) {
            const image = notFoundImage.currentValue || 'default';
            this.isImageBuildIn = TDSEmptyDefaultImages.findIndex(i => i === image) > -1;
        }
    }
    ngOnInit() {
        this.i18n.localeChange.pipe(takeUntil(this.destroy$)).subscribe(() => {
            this.locale = this.i18n.getLocaleData('Empty');
            this.cdr.markForCheck();
        });
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
}
TDSEmptyComponent.ɵfac = function TDSEmptyComponent_Factory(t) { return new (t || TDSEmptyComponent)(i0.ɵɵdirectiveInject(i1.TDSI18nService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSEmptyComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSEmptyComponent, selectors: [["tds-empty"]], hostAttrs: [1, "tds-empty", "flex", "flex-col", "items-center"], inputs: { notFoundImage: "notFoundImage", notFoundContent: "notFoundContent", notFoundFooter: "notFoundFooter" }, exportAs: ["tdsEmpty"], features: [i0.ɵɵNgOnChangesFeature], decls: 6, vars: 5, consts: [[1, "tds-empty-image"], [4, "ngIf"], ["class", "tds-empty-description text-neutral-1-900 text-title-1 font-semibold mt-3", 4, "ngIf"], ["class", "tds-empty-footer mt-3", 4, "ngIf"], [4, "tdsStringTemplateOutlet"], [1, "h-full", 3, "src", "alt"], [1, "tds-empty-description", "text-neutral-1-900", "text-title-1", "font-semibold", "mt-3"], [1, "tds-empty-footer", "mt-3"]], template: function TDSEmptyComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, TDSEmptyComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
        i0.ɵɵtemplate(2, TDSEmptyComponent_tds_empty_default_2_Template, 1, 0, "tds-empty-default", 1);
        i0.ɵɵtemplate(3, TDSEmptyComponent_tds_empty_simple_3_Template, 1, 0, "tds-empty-simple", 1);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(4, TDSEmptyComponent_p_4_Template, 2, 1, "p", 2);
        i0.ɵɵtemplate(5, TDSEmptyComponent_div_5_Template, 2, 1, "div", 3);
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.isImageBuildIn);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isImageBuildIn && ctx.notFoundImage !== "simple");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isImageBuildIn && ctx.notFoundImage === "simple");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.notFoundContent !== null);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.notFoundFooter);
    } }, directives: [i2.NgIf, i3.TDSStringTemplateOutletDirective, i4.TDSEmptyDefaultComponent, i5.TDSEmptySimpleComponent], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptyComponent, [{
        type: Component,
        args: [{
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                selector: 'tds-empty',
                exportAs: 'tdsEmpty',
                template: `
    <div class="tds-empty-image">
      <ng-container *ngIf="!isImageBuildIn">
        <ng-container *tdsStringTemplateOutlet="notFoundImage">
          <img class="h-full" [src]="notFoundImage" [alt]="isContentString ? notFoundContent : 'empty'" />
        </ng-container>
      </ng-container>
      <tds-empty-default *ngIf="isImageBuildIn && notFoundImage !== 'simple'"></tds-empty-default>
      <tds-empty-simple *ngIf="isImageBuildIn && notFoundImage === 'simple'"></tds-empty-simple>
    </div>
    <p class="tds-empty-description text-neutral-1-900 text-title-1 font-semibold mt-3" *ngIf="notFoundContent !== null">
      <ng-container *tdsStringTemplateOutlet="notFoundContent">
        {{ isContentString ? notFoundContent : locale['description'] }}
      </ng-container>
    </p>
    <div class="tds-empty-footer mt-3" *ngIf="notFoundFooter">
      <ng-container *tdsStringTemplateOutlet="notFoundFooter">
        {{ notFoundFooter }}
      </ng-container>
    </div>
  `,
                host: {
                    class: 'tds-empty  flex flex-col items-center'
                }
            }]
    }], function () { return [{ type: i1.TDSI18nService }, { type: i0.ChangeDetectorRef }]; }, { notFoundImage: [{
            type: Input
        }], notFoundContent: [{
            type: Input
        }], notFoundFooter: [{
            type: Input
        }] }); })();
