// tds-ui/esm2020/empty/empty.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSEmbedEmptyComponent } from './embed-empty.component';
import { TDSEmptyComponent } from './empty.component';
import { TDSEmptyDefaultComponent } from './partial/default';
import { TDSEmptySimpleComponent } from './partial/simple';
import * as i0 from "@angular/core";
export class TDSEmptyModule {
}
TDSEmptyModule.ɵfac = function TDSEmptyModule_Factory(t) { return new (t || TDSEmptyModule)(); };
TDSEmptyModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSEmptyModule });
TDSEmptyModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[BidiModule, CommonModule, PortalModule, TDSOutletModule, TDSI18nModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSEmptyModule, [{
        type: NgModule,
        args: [{
                imports: [BidiModule, CommonModule, PortalModule, TDSOutletModule, TDSI18nModule],
                declarations: [TDSEmptyComponent, TDSEmbedEmptyComponent, TDSEmptyDefaultComponent, TDSEmptySimpleComponent],
                exports: [TDSEmptyComponent, TDSEmbedEmptyComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSEmptyModule, { declarations: [TDSEmptyComponent, TDSEmbedEmptyComponent, TDSEmptyDefaultComponent, TDSEmptySimpleComponent], imports: [BidiModule, CommonModule, PortalModule, TDSOutletModule, TDSI18nModule], exports: [TDSEmptyComponent, TDSEmbedEmptyComponent] }); })();
