// tds-ui/esm2020/button/a-link.directive.mjs
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { InputBoolean } from 'tds-ui/shared/utility';
import * as i0 from "@angular/core";
export class TDSLinkDirective {
    constructor() {
        this.disabled = false;
        this.danger = false;
    }
    _haltDisabledEvents(event) {
        // A disabled button shouldn't apply any actions
        if (this.disabled) {
            event.preventDefault();
            event.stopImmediatePropagation();
        }
    }
}
TDSLinkDirective.ɵfac = function TDSLinkDirective_Factory(t) { return new (t || TDSLinkDirective)(); };
TDSLinkDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSLinkDirective, selectors: [["a", "tds-link", ""]], hostAttrs: [1, "text-body-2", "font-semibold"], hostVars: 28, hostBindings: function TDSLinkDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function TDSLinkDirective_click_HostBindingHandler($event) { return ctx._haltDisabledEvents($event); });
    } if (rf & 2) {
        i0.ɵɵclassProp("cursor-pointer", !ctx.disabled)("hover:underline", !ctx.disabled)("active:pointer-events-none", ctx.disabled)("cursor-not-allowed", ctx.disabled)("text-neutral-1-300", ctx.disabled)("dark:text-d-neutral-1-300", ctx.disabled)("text-info-400", !ctx.danger && !ctx.disabled)("dark:text-d-info-400", !ctx.danger && !ctx.disabled)("hover:text-info-500", !ctx.danger && !ctx.disabled)("dark:hover:text-d-info-500", !ctx.danger && !ctx.disabled)("text-error-400", ctx.danger && !ctx.disabled)("dark:text-d-error-400", ctx.danger && !ctx.disabled)("hover:text-error-500", ctx.danger && !ctx.disabled)("dark:hover:text-d-error-500", ctx.danger && !ctx.disabled);
    } }, inputs: { disabled: "disabled", danger: "danger" } });
__decorate([
    InputBoolean()
], TDSLinkDirective.prototype, "disabled", void 0);
__decorate([
    InputBoolean()
], TDSLinkDirective.prototype, "danger", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSLinkDirective, [{
        type: Directive,
        args: [{
                selector: 'a[tds-link]',
                host: {
                    'class': 'text-body-2 font-semibold ',
                    '[class.cursor-pointer]': '!disabled',
                    '[class.hover:underline]': '!disabled',
                    '[class.active:pointer-events-none]': 'disabled',
                    '[class.cursor-not-allowed]': 'disabled',
                    '[class.text-neutral-1-300]': 'disabled',
                    '[class.dark:text-d-neutral-1-300]': 'disabled',
                    '[class.text-info-400]': '!danger && !disabled',
                    '[class.dark:text-d-info-400]': '!danger && !disabled',
                    '[class.hover:text-info-500]': '!danger && !disabled',
                    '[class.dark:hover:text-d-info-500]': '!danger && !disabled',
                    '[class.text-error-400]': 'danger && !disabled',
                    '[class.dark:text-d-error-400]': 'danger && !disabled',
                    '[class.hover:text-error-500]': 'danger && !disabled',
                    '[class.dark:hover:text-d-error-500]': 'danger && !disabled',
                    '(click)': '_haltDisabledEvents($event)',
                }
            }]
    }], function () { return []; }, { disabled: [{
            type: Input
        }], danger: [{
            type: Input
        }] }); })();
