// tds-ui/esm2020/button/button-close.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "tds-ui/core/outlet";
const _c0 = ["tds-button-close", ""];
function TDSButtonCloseComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "i");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵclassMap(ctx_r0.iconTemplate);
} }
export class TDSButtonCloseComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.size = 'xl';
        this.iconTemplate = 'tdsi-close-fill';
        this.cssClass = '';
    }
    ngOnChanges(changes) {
        this.buildCss();
        this.cdr.markForCheck();
    }
    ngOnInit() {
        this.buildCss();
    }
    buildCss() {
        this.cssClass = `${this.getClassCssBySize(this.size)} ${this.defaultClassCss()} `;
    }
    getClassCssBySize(size) {
        let result = '';
        switch (size) {
            case "sm":
                result = 'p-0 h-4';
                break;
            case "md":
                result = 'p-0.5 h-5';
                break;
            case "lg":
                result = 'p-1 h-6';
                break;
            case "xl":
                result = 'p-1.5 h-7';
                break;
        }
        return result;
    }
    defaultClassCss() {
        let result = 'focus:outline-none box-border  whitespace-nowrap  border-0 text-base leading-4 relative text-neutral-1-500 dark:text-d-neutral-1-500 hover:bg-neutral-3-50 dark:hover:bg-d-neutral-3-300 hover:rounded-md inline-flex items-center justify-center';
        return result;
    }
}
TDSButtonCloseComponent.ɵfac = function TDSButtonCloseComponent_Factory(t) { return new (t || TDSButtonCloseComponent)(i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
TDSButtonCloseComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSButtonCloseComponent, selectors: [["button", "tds-button-close", ""], ["a", "tds-button-close", ""]], hostVars: 4, hostBindings: function TDSButtonCloseComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("tabindex", ctx.disabled ? -1 : ctx.tabIndex === null ? null : ctx.tabIndex)("disabled", ctx.disabled || ctx.loading || null);
        i0.ɵɵclassMap(ctx.cssClass);
    } }, inputs: { size: "size", iconTemplate: "iconTemplate" }, exportAs: ["tdsButtonClose"], features: [i0.ɵɵNgOnChangesFeature], attrs: _c0, decls: 1, vars: 1, consts: [[4, "tdsStringTemplateOutlet"]], template: function TDSButtonCloseComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, TDSButtonCloseComponent_ng_container_0_Template, 2, 2, "ng-container", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("tdsStringTemplateOutlet", ctx.iconTemplate);
    } }, directives: [i1.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonCloseComponent, [{
        type: Component,
        args: [{
                selector: 'button[tds-button-close], a[tds-button-close]',
                exportAs: 'tdsButtonClose',
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: `
   <ng-container *tdsStringTemplateOutlet="iconTemplate"><i [class]="iconTemplate"></i></ng-container>  
    `,
                host: {
                    '[class]': 'cssClass',
                    '[attr.tabindex]': 'disabled ? -1 : (tabIndex === null ? null : tabIndex)',
                    '[attr.disabled]': 'disabled || loading || null',
                }
            }]
    }], function () { return [{ type: i0.ChangeDetectorRef }]; }, { size: [{
            type: Input
        }], iconTemplate: [{
            type: Input
        }] }); })();
