// tds-ui/esm2020/button/button.component.mjs
import { ChangeDetectionStrategy, Component, Optional, ViewEncapsulation, } from "@angular/core";
import { TDSButtonBase } from "./button.base";
import * as i0 from "@angular/core";
import * as i1 from "./button-group.component";
import * as i2 from "@angular/common";
import * as i3 from "tds-ui/progress-spinner";
import * as i4 from "tds-ui/core/outlet";
const _c0 = ["tds-button", ""];
function TDSButtonComponent_div_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function TDSButtonComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtemplate(1, TDSButtonComponent_div_0_ng_container_1_Template, 1, 0, "ng-container", 6);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r4 = i0.ɵɵreference(5);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r4);
} }
function TDSButtonComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div");
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r1.loadingCss);
} }
function TDSButtonComponent_ng_template_2_div_0_tds_spinner_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "tds-spinner", 11);
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(3);
    i0.ɵɵproperty("color", ctx_r10.spinerColor);
} }
function TDSButtonComponent_ng_template_2_div_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 8);
    i0.ɵɵtemplate(1, TDSButtonComponent_ng_template_2_div_0_tds_spinner_1_Template, 1, 1, "tds-spinner", 9);
    i0.ɵɵelementStart(2, "div", 10);
    i0.ɵɵtext(3, " Loading... ");
    i0.ɵɵelementEnd()();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r9.loading);
} }
function TDSButtonComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSButtonComponent_ng_template_2_div_0_Template, 4, 1, "div", 7);
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    const _r6 = i0.ɵɵreference(7);
    i0.ɵɵproperty("ngIf", !ctx_r3.loadingTemplate)("ngIfElse", _r6);
} }
function TDSButtonComponent_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵprojection(0);
} }
function TDSButtonComponent_ng_template_6_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r11.loadingTemplate, " ");
} }
function TDSButtonComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtemplate(0, TDSButtonComponent_ng_template_6_ng_container_0_Template, 2, 1, "ng-container", 12);
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext();
    i0.ɵɵproperty("tdsStringTemplateOutlet", ctx_r7.loadingTemplate);
} }
const _c1 = ["*"];
export class TDSButtonComponent extends TDSButtonBase {
    constructor(ngZone, elementRef, cdr, _buttonGroup) {
        super(ngZone, elementRef, cdr, _buttonGroup);
        this.ngZone = ngZone;
        this.elementRef = elementRef;
        this.cdr = cdr;
        this._buttonGroup = _buttonGroup;
    }
}
TDSButtonComponent.ɵfac = function TDSButtonComponent_Factory(t) { return new (t || TDSButtonComponent)(i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSButtonGroupComponent, 8)); };
TDSButtonComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSButtonComponent, selectors: [["button", "tds-button", ""], ["a", "tds-button", ""]], hostVars: 4, hostBindings: function TDSButtonComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵattribute("tabindex", ctx.disabled ? -1 : ctx.tabIndex === null ? null : ctx.tabIndex)("disabled", ctx.disabled || null);
        i0.ɵɵclassMap(ctx.cssClass);
    } }, exportAs: ["tdsButton"], features: [i0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c1, decls: 8, vars: 3, consts: [["class", "flex w-full items-center justify-center", 4, "ngIf", "ngIfElse"], [3, "class", 4, "ngIf"], ["loadingtmpl", ""], ["content", ""], ["CustomLoadingtmpl", ""], [1, "flex", "w-full", "items-center", "justify-center"], [4, "ngTemplateOutlet"], ["class", "flex w-full items-center", 4, "ngIf", "ngIfElse"], [1, "flex", "w-full", "items-center"], ["class", "w-4 h-4 mr-2", 3, "color", 4, "ngIf"], [1, "flex-auto"], [1, "w-4", "h-4", "mr-2", 3, "color"], [4, "tdsStringTemplateOutlet"]], template: function TDSButtonComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtemplate(0, TDSButtonComponent_div_0_Template, 2, 1, "div", 0);
        i0.ɵɵtemplate(1, TDSButtonComponent_div_1_Template, 1, 2, "div", 1);
        i0.ɵɵtemplate(2, TDSButtonComponent_ng_template_2_Template, 1, 2, "ng-template", null, 2, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(4, TDSButtonComponent_ng_template_4_Template, 1, 0, "ng-template", null, 3, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(6, TDSButtonComponent_ng_template_6_Template, 1, 1, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r2 = i0.ɵɵreference(3);
        i0.ɵɵproperty("ngIf", !ctx.loading || ctx.disabled)("ngIfElse", _r2);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.loading && !ctx.disabled);
    } }, directives: [i2.NgIf, i2.NgTemplateOutlet, i3.TDSProgressSpinnerComponent, i4.TDSStringTemplateOutletDirective], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonComponent, [{
        type: Component,
        args: [{ selector: 'button[tds-button], a[tds-button]', exportAs: 'tdsButton', preserveWhitespaces: false, changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, host: {
                    '[class]': 'cssClass',
                    '[attr.tabindex]': 'disabled ? -1 : (tabIndex === null ? null : tabIndex)',
                    '[attr.disabled]': 'disabled || null',
                }, template: "<div class='flex w-full items-center justify-center' *ngIf=\"!loading || disabled; else loadingtmpl\">\r\n    <ng-container *ngTemplateOutlet=\"content\"></ng-container>\r\n</div>\r\n<div [class]=\"loadingCss\" *ngIf=\"loading && !disabled\">\r\n</div>\r\n<ng-template #loadingtmpl>    \r\n    <div class='flex w-full items-center' *ngIf=\"!loadingTemplate; else CustomLoadingtmpl\">    \r\n        <tds-spinner *ngIf=\"loading\" [color]=\"spinerColor\" class=\"w-4 h-4 mr-2\"></tds-spinner>\r\n        <div class='flex-auto'>\r\n            Loading...\r\n        </div>\r\n    </div>\r\n</ng-template>\r\n<ng-template #content>\r\n    <ng-content></ng-content>\r\n</ng-template>\r\n<ng-template #CustomLoadingtmpl>\r\n    <ng-container *tdsStringTemplateOutlet=\"loadingTemplate\">\r\n        {{loadingTemplate}}\r\n    </ng-container>\r\n</ng-template>\r\n" }]
    }], function () { return [{ type: i0.NgZone }, { type: i0.ElementRef }, { type: i0.ChangeDetectorRef }, { type: i1.TDSButtonGroupComponent, decorators: [{
                type: Optional
            }] }]; }, null); })();
