// tds-ui/esm2020/button/button-group.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class TDSButtonGroupComponent {
    constructor() {
    }
    ngOnInit() {
    }
}
TDSButtonGroupComponent.ɵfac = function TDSButtonGroupComponent_Factory(t) { return new (t || TDSButtonGroupComponent)(); };
TDSButtonGroupComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TDSButtonGroupComponent, selectors: [["tds-button-group"]], hostAttrs: [1, "inline-flex"], ngContentSelectors: _c0, decls: 1, vars: 0, template: function TDSButtonGroupComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonGroupComponent, [{
        type: Component,
        args: [{
                selector: 'tds-button-group',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                host: {
                    class: "inline-flex"
                },
                preserveWhitespaces: false,
                template: `
    <ng-content></ng-content>
  `
            }]
    }], function () { return []; }, null); })();
