// tds-ui/esm2020/button/button.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { TDSButtonGroupComponent } from './button-group.component';
import { TDSSpinnerModule } from 'tds-ui/progress-spinner';
import { TDSButtonComponent } from './button.component';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSButtonIconComponent } from './button-icon.component';
import { TDSButtonFlatIconComponent } from './button-flat-icon.component';
import { TDSButtonOutlineIconComponent } from './button-outline-icon.component';
import { TDSButtonOutlineComponent } from './button-outline.component';
import { TDSButtonFlatComponent } from './button-flat.component';
import { TDSButtonCloseComponent } from './button-close';
import { TDSLinkDirective } from './a-link.directive';
import * as i0 from "@angular/core";
export class TDSButtonModule {
}
TDSButtonModule.ɵfac = function TDSButtonModule_Factory(t) { return new (t || TDSButtonModule)(); };
TDSButtonModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSButtonModule });
TDSButtonModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            OverlayModule,
            PortalModule,
            TDSSpinnerModule,
            TDSOutletModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    OverlayModule,
                    PortalModule,
                    TDSSpinnerModule,
                    TDSOutletModule
                ],
                declarations: [
                    TDSButtonGroupComponent,
                    TDSButtonComponent,
                    TDSButtonIconComponent,
                    TDSButtonFlatIconComponent,
                    TDSButtonOutlineIconComponent,
                    TDSButtonOutlineComponent,
                    TDSButtonFlatComponent,
                    TDSButtonCloseComponent,
                    TDSLinkDirective
                ],
                exports: [
                    TDSButtonGroupComponent,
                    TDSButtonComponent,
                    TDSButtonIconComponent,
                    TDSButtonFlatIconComponent,
                    TDSButtonOutlineIconComponent,
                    TDSButtonOutlineComponent,
                    TDSButtonFlatComponent,
                    TDSButtonCloseComponent,
                    TDSLinkDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSButtonModule, { declarations: [TDSButtonGroupComponent,
        TDSButtonComponent,
        TDSButtonIconComponent,
        TDSButtonFlatIconComponent,
        TDSButtonOutlineIconComponent,
        TDSButtonOutlineComponent,
        TDSButtonFlatComponent,
        TDSButtonCloseComponent,
        TDSLinkDirective], imports: [CommonModule,
        OverlayModule,
        PortalModule,
        TDSSpinnerModule,
        TDSOutletModule], exports: [TDSButtonGroupComponent,
        TDSButtonComponent,
        TDSButtonIconComponent,
        TDSButtonFlatIconComponent,
        TDSButtonOutlineIconComponent,
        TDSButtonOutlineComponent,
        TDSButtonFlatComponent,
        TDSButtonCloseComponent,
        TDSLinkDirective] }); })();
