// tds-ui/esm2020/button/button.base.mjs
import { __decorate } from "tslib";
import { Directive, Input, Optional } from "@angular/core";
import { fromEvent, Subject } from "rxjs";
import { takeUntil } from "rxjs/operators";
import { InputBoolean } from "tds-ui/shared/utility";
import * as i0 from "@angular/core";
import * as i1 from "./button-group.component";
export class TDSButtonBase {
    constructor(ngZone, elementRef, cdr, _buttonGroup) {
        this.ngZone = ngZone;
        this.elementRef = elementRef;
        this.cdr = cdr;
        this._buttonGroup = _buttonGroup;
        this.loading = false;
        this.color = 'primary';
        this.size = 'md';
        this.disabled = false;
        this.rounded = 'rounded-md';
        this.cssClass = '';
        this.spinerColor = 'white';
        this.loadingCss = '';
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        this.buildCss();
        this.ngZone.runOutsideAngular(() => {
            // Caretaker note: this event listener could've been added through `host.click` or `HostListener`.
            // The compiler generates the `ɵɵlistener` instruction which wraps the actual listener internally into the
            // function, which runs `markDirty()` before running the actual listener (the decorated class method).
            // Since we're preventing the default behavior and stopping event propagation this doesn't require Angular to run the change detection.
            fromEvent(this.elementRef.nativeElement, 'click', { capture: true })
                .pipe(takeUntil(this.destroy$))
                .subscribe(event => {
                if ((this.disabled && event.target?.tagName === 'A') || this.loading) {
                    event.preventDefault();
                    event.stopImmediatePropagation();
                }
            });
        });
    }
    ngOnChanges(changes) {
        const { size, loading, color, loadingTemplate, disabled } = changes;
        if ((size && !size.firstChange) || (loading && !loading.firstChange) || (color && !color.firstChange) || (disabled && !disabled.firstChange)) {
            this.buildCss();
            this.cdr.markForCheck();
        }
        if (loadingTemplate && !loadingTemplate.firstChange) {
            this.cdr.markForCheck();
        }
    }
    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }
    getClassCssLoading() {
        return `absolute -inset-px opacity-[.35] bg-white cursor-progress dark:bg-d-neutral-3-50 ${this.getClassCssButtonGroup()}`;
    }
    getClassCssDisabled() {
        return 'cursor-not-allowed border text-neutral-1-300 border-neutral-2-200 bg-neutral-3-50 dark:text-d-neutral-1-300 dark:border-d-neutral-2-200 dark:bg-d-neutral-3-200 ';
    }
    getClassCssActive() {
        let result = 'focus:ring focus:outline-none cursor-pointer';
        switch (this.color) {
            case "primary":
                result += ` text-white ring-primary-1 border-primary-1 bg-primary-1 hover:bg-primary-2 focus:bg-primary-1 hover:border-primary-2 focus:border-primary-1 dark:ring-d-primary-1 dark:border-d-primary-1 dark:bg-d-primary-1 dark:hover:bg-d-primary-2 dark:focus:bg-d-primary-1 dark:hover:border-d-primary-2 dark:focus:border-d-primary-1 `;
                break;
            case "secondary":
                result += ` text-neutral-1-900 dark:text-d-neutral-1-900  ring-primary-1 dark:ring-d-primary-1 border-neutral-2-200 dark:border-d-neutral-2-200 bg-white hover:bg-neutral-3-50 focus:bg-white dark:focus:bg-d-neutral-3-300 focus:border-primary-1 dark:focus:border-d-primary-1 dark:bg-d-neutral-3-300 dark:hover:bg-d-neutral-3-400 `;
                break;
            case 'info':
                result += ` text-white ring-info-400 border-info-400 bg-info-400 hover:bg-info-500 focus:bg-info-400 hover:border-info-500 focus:border-info-400  dark:ring-d-info-400 dark:border-d-info-400 dark:bg-d-info-400 dark:hover:bg-d-info-500 dark:focus:bg-d-info-400 dark:hover:border-d-info-500 dark:focus:border-d-info-400`;
                break;
            case 'success':
                result += ` text-white ring-success-400 border-success-400 bg-success-400 hover:bg-success-500 focus:bg-success-400 hover:border-success-500 focus:border-success-400  dark:ring-d-success-400 dark:border-d-success-400 dark:bg-d-success-400 dark:hover:bg-d-success-500 dark:focus:bg-d-success-400 dark:hover:border-d-success-500 dark:focus:border-d-success-400 `;
                break;
            case 'warning':
                result += ` text-white ring-warning-400 border-warning-400 bg-warning-400 hover:bg-warning-500 focus:bg-warning-400 hover:border-warning-500 focus:border-warning-400 dark:ring-d-warning-400 dark:border-d-warning-400 dark:bg-d-warning-400 dark:hover:bg-d-warning-500 dark:focus:bg-d-warning-400 dark:hover:border-d-warning-500 dark:focus:border-d-warning-400 `;
                break;
            case 'error':
                result += ` text-white ring-error-400 border-error-400 bg-error-400 hover:bg-error-500 focus:bg-error-400 hover:border-error-500 focus:border-error-400 dark:ring-d-error-400 dark:border-d-error-400 dark:bg-d-error-400 dark:hover:bg-d-error-500 dark:focus:bg-d-error-400 dark:hover:border-d-error-500 dark:focus:border-d-error-400 `;
                break;
            case 'red':
                result += ' text-white ring-base-red-500 border-base-red-500 bg-base-red-500 hover:bg-base-red-600 focus:bg-base-red-500 hover:border-base-red-600 focus:border-base-red-500 dark:ring-d-base-red-500 dark:border-d-base-red-500 dark:bg-d-base-red-500 dark:hover:bg-d-base-red-600 dark:focus:bg-d-base-red-500 dark:hover:border-d-base-red-600 dark:focus:border-d-base-red-500 ';
                break;
            case 'yellow':
                result += ' text-white ring-base-yellow-500 border-base-yellow-500 bg-base-yellow-500 hover:bg-base-yellow-600 focus:bg-base-yellow-500 hover:border-base-yellow-600 focus:border-base-yellow-500 dark:ring-d-base-yellow-500 dark:border-d-base-yellow-500 dark:bg-d-base-yellow-500 dark:hover:bg-d-base-yellow-600 dark:focus:bg-d-base-yellow-500 dark:hover:border-d-base-yellow-600 dark:focus:border-d-base-yellow-500 ';
                break;
            case 'green':
                result += ' text-white ring-base-green-500 border-base-green-500 bg-base-green-500 hover:bg-base-green-600 focus:bg-base-green-500 hover:border-base-green-600 focus:border-base-green-500 dark:ring-d-base-green-500 dark:border-d-base-green-500 dark:bg-d-base-green-500 dark:hover:bg-d-base-green-600 dark:focus:bg-d-base-green-500 dark:hover:border-d-base-green-600 dark:focus:border-d-base-green-500 ';
                break;
            case 'blue':
                result += ' text-white ring-base-blue-500 border-base-blue-500 bg-base-blue-500 hover:bg-base-blue-600 focus:bg-base-blue-500 hover:border-base-blue-600 focus:border-base-blue-500 dark:ring-d-base-blue-500 dark:border-d-base-blue-500 dark:bg-d-base-blue-500 dark:hover:bg-d-base-blue-600 dark:focus:bg-d-base-blue-500 dark:hover:border-d-base-blue-600 dark:focus:border-d-base-blue-500 ';
                break;
            case 'indigo':
                result += ' text-white ring-base-indigo-500 border-base-indigo-500 bg-base-indigo-500 hover:bg-base-indigo-600 focus:bg-base-indigo-500 hover:border-base-indigo-600 focus:border-base-indigo-500 dark:ring-d-base-indigo-500 dark:border-d-base-indigo-500 dark:bg-d-base-indigo-500 dark:hover:bg-d-base-indigo-600 dark:focus:bg-d-base-indigo-500 dark:hover:border-d-base-indigo-600 dark:focus:border-d-base-indigo-500 ';
                break;
            case 'purple':
                result += ' text-white ring-base-purple-500 border-base-purple-500 bg-base-purple-500 hover:bg-base-purple-600 focus:bg-base-purple-500 hover:border-base-purple-600 focus:border-base-purple-500 dark:ring-d-base-purple-500 dark:border-d-base-purple-500 dark:bg-d-base-purple-500 dark:hover:bg-d-base-purple-600 dark:focus:bg-d-base-purple-500 dark:hover:border-d-base-purple-600 dark:focus:border-d-base-purple-500 ';
                break;
            case 'pink':
                result += ' text-white ring-base-pink-500 border-base-pink-500 bg-base-pink-500 hover:bg-base-pink-600 focus:bg-base-pink-500 hover:border-base-pink-600 focus:border-base-pink-500 dark:ring-d-base-pink-500 dark:border-d-base-pink-500 dark:bg-d-base-pink-500 dark:hover:bg-d-base-pink-600 dark:focus:bg-d-base-pink-500 dark:hover:border-d-base-pink-600 dark:focus:border-d-base-pink-500 ';
                break;
            case 'amber':
                result += ' text-white ring-base-amber-500 border-base-amber-500 bg-base-amber-500 hover:bg-base-amber-600 focus:bg-base-amber-500 hover:border-base-amber-600 focus:border-base-amber-500 dark:ring-d-base-amber-500 dark:border-d-base-amber-500 dark:bg-d-base-amber-500 dark:hover:bg-d-base-amber-600 dark:focus:bg-d-base-amber-500 dark:hover:border-d-base-amber-600 dark:focus:border-d-base-amber-500 ';
                break;
            case 'sky':
                result += ' text-white ring-base-sky-500 border-base-sky-500 bg-base-sky-500 hover:bg-base-sky-600 focus:bg-base-sky-500 hover:border-base-sky-600 focus:border-base-sky-500 dark:ring-d-base-sky-500 dark:border-d-base-sky-500 dark:bg-d-base-sky-500 dark:hover:bg-d-base-sky-600 dark:focus:bg-d-base-sky-500 dark:hover:border-d-base-sky-600 dark:focus:border-d-base-sky-500 ';
                break;
            case 'lime':
                result += ' text-white ring-base-lime-500 border-base-lime-500 bg-base-lime-500 hover:bg-base-lime-600 focus:bg-base-lime-500 hover:border-base-lime-600 focus:border-base-lime-500 dark:ring-d-base-lime-500 dark:border-d-base-lime-500 dark:bg-d-base-lime-500 dark:hover:bg-d-base-lime-600 dark:focus:bg-d-base-lime-500 dark:hover:border-d-base-lime-600 dark:focus:border-d-base-lime-500 ';
                break;
            case 'teal':
                result += ' text-white ring-base-teal-500 border-base-teal-500 bg-base-teal-500 hover:bg-base-teal-600 focus:bg-base-teal-500 hover:border-base-teal-600 focus:border-base-teal-500 dark:ring-d-base-teal-500 dark:border-d-base-teal-500 dark:bg-d-base-teal-500 dark:hover:bg-d-base-teal-600 dark:focus:bg-d-base-teal-500 dark:hover:border-d-base-teal-600 dark:focus:border-d-base-teal-500 ';
                break;
            case 'orange':
                result += ' text-white ring-base-orange-500 border-base-orange-500 bg-base-orange-500 hover:bg-base-orange-600 focus:bg-base-orange-500 hover:border-base-orange-600 focus:border-base-orange-500 dark:ring-d-base-orange-500 dark:border-d-base-orange-500 dark:bg-d-base-orange-500 dark:hover:bg-d-base-orange-600 dark:focus:bg-d-base-orange-500 dark:hover:border-d-base-orange-600 dark:focus:border-d-base-orange-500 ';
                break;
            // custom : dev tự style theo thiết kế
            case 'custom':
                result += '';
                break;
        }
        return result;
    }
    defaultClassCss() {
        let result = 'font-semibold box-border whitespace-nowrap border ring-opacity-20 dark:ring-opacity-20 relative text-body-2';
        return result;
    }
    getClassCssBySize(size) {
        let result = '';
        switch (size) {
            case "sm":
                result = 'py-1 px-2.5  h-sm';
                break;
            case "md":
                result = 'py-1.5 px-3  h-md';
                break;
            case "lg":
                result = 'py-2 px-4 h-lg';
                break;
            case "xl":
                result = 'py-2.5 px-4.5 h-xl';
                break;
        }
        return result;
    }
    getClassCssButtonGroup() {
        let result = '';
        if (this._buttonGroup) {
            result = 'first:rounded-l-md last:rounded-r-md first:border-l last:border-r';
        }
        else {
            result = this.rounded;
        }
        return result;
    }
    buildCss() {
        this.cssClass = this.builClassCssButton();
        this.spinerColor = this.buildSpinerColor();
        this.loadingCss = this.getClassCssLoading();
    }
    builClassCssButton() {
        let result = '';
        if (this.elementRef.nativeElement.tagName === 'A') {
            result = 'inline-block ';
        }
        result += this.defaultClassCss();
        result += ` ${this.getClassCssBySize(this.size)}`;
        result += ` ${this.getClassCssButtonGroup()}`;
        if (this.disabled) {
            result += ` ${this.getClassCssDisabled()}`;
        }
        else {
            // custom : dev tự style theo thiết kế
            if (this.color != 'custom')
                result += ` ${this.getClassCssActive()}`;
        }
        return result;
    }
    buildSpinerColor() {
        let result = 'white';
        if (this.color == 'secondary') {
            result = 'primary';
        }
        return result;
    }
}
TDSButtonBase.ɵfac = function TDSButtonBase_Factory(t) { return new (t || TDSButtonBase)(i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i1.TDSButtonGroupComponent, 8)); };
TDSButtonBase.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: TDSButtonBase, inputs: { loading: "loading", color: "color", size: "size", loadingTemplate: "loadingTemplate", disabled: "disabled", rounded: "rounded" }, features: [i0.ɵɵNgOnChangesFeature] });
__decorate([
    InputBoolean()
], TDSButtonBase.prototype, "loading", void 0);
__decorate([
    InputBoolean()
], TDSButtonBase.prototype, "disabled", void 0);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSButtonBase, [{
        type: Directive
    }], function () { return [{ type: i0.NgZone }, { type: i0.ElementRef }, { type: i0.ChangeDetectorRef }, { type: i1.TDSButtonGroupComponent, decorators: [{
                type: Optional
            }] }]; }, { loading: [{
            type: Input
        }], color: [{
            type: Input
        }], size: [{
            type: Input
        }], loadingTemplate: [{
            type: Input
        }], disabled: [{
            type: Input
        }], rounded: [{
            type: Input
        }] }); })();
