// tds-ui/esm2020/time-picker/time-picker.module.mjs
import { BidiModule } from '@angular/cdk/bidi';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TDSButtonModule } from 'tds-ui/button';
import { TDSOutletModule } from 'tds-ui/core/outlet';
import { TDSI18nModule } from 'tds-ui/i18n';
import { TDSSelectModule } from 'tds-ui/select';
import { TDSTimePickerPanelComponent } from './time-picker-panel.component';
import { TDSTimePickerComponent } from './time-picker.component';
import * as i0 from "@angular/core";
export class TDSTimePickerModule {
}
TDSTimePickerModule.ɵfac = function TDSTimePickerModule_Factory(t) { return new (t || TDSTimePickerModule)(); };
TDSTimePickerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TDSTimePickerModule });
TDSTimePickerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            BidiModule,
            CommonModule,
            FormsModule,
            TDSI18nModule,
            OverlayModule,
            TDSOutletModule,
            TDSButtonModule,
            TDSSelectModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TDSTimePickerModule, [{
        type: NgModule,
        args: [{
                declarations: [TDSTimePickerComponent, TDSTimePickerPanelComponent],
                exports: [TDSTimePickerPanelComponent, TDSTimePickerComponent],
                imports: [
                    BidiModule,
                    CommonModule,
                    FormsModule,
                    TDSI18nModule,
                    OverlayModule,
                    TDSOutletModule,
                    TDSButtonModule,
                    TDSSelectModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TDSTimePickerModule, { declarations: [TDSTimePickerComponent, TDSTimePickerPanelComponent], imports: [BidiModule,
        CommonModule,
        FormsModule,
        TDSI18nModule,
        OverlayModule,
        TDSOutletModule,
        TDSButtonModule,
        TDSSelectModule], exports: [TDSTimePickerPanelComponent, TDSTimePickerComponent] }); })();
